{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cytocraft Analysis Notebook - Reproducibility\n",
    "\n",
    "This notebook contains the downstream analysis code for the cytocraft project.\n",
    "\n",
    "## Data Requirements\n",
    "Before running this notebook, please ensure you have downloaded the required datasets and placed them in the appropriate directories:\n",
    "\n",
    "- `./data/` - Main data directory\n",
    "- `./results/` - Results from cytocraft analysis\n",
    "- `./figures/` - Output directory for generated figures\n",
    "\n",
    "## Dependencies\n",
    "Make sure you have installed the cytocraft package:\n",
    "```bash\n",
    "pip install cytocraft\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "from cytocraft.craft import *\n",
    "from cytocraft.plot import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colormap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zeileis_28 = [\n",
    "    \"#023fa5\",\n",
    "    \"#7d87b9\",\n",
    "    \"#bec1d4\",\n",
    "    \"#d6bcc0\",\n",
    "    \"#bb7784\",\n",
    "    \"#8e063b\",\n",
    "    \"#4a6fe3\",\n",
    "    \"#8595e1\",\n",
    "    \"#b5bbe3\",\n",
    "    \"#e6afb9\",\n",
    "    \"#e07b91\",\n",
    "    \"#d33f6a\",\n",
    "    \"#11c638\",\n",
    "    \"#8dd593\",\n",
    "    \"#c6dec7\",\n",
    "    \"#ead3c6\",\n",
    "    \"#f0b98d\",\n",
    "    \"#ef9708\",\n",
    "    \"#0fcfc0\",\n",
    "    \"#9cded6\",\n",
    "    \"#d5eae7\",\n",
    "    \"#f3e1eb\",\n",
    "    \"#f6c4e1\",\n",
    "    \"#f79cd4\",\n",
    "    # these last ones were added:\n",
    "    \"#7f7f7f\",\n",
    "    \"#c7c7c7\",\n",
    "    \"#1CE6FF\",\n",
    "    \"#336600\",\n",
    "]\n",
    "\n",
    "godsnot_102 = [\n",
    "    #\"#000000\",  # remove the black, as often, we have black colored annotation\n",
    "    #\"#FFFF00\",\n",
    "    \"#1CE6FF\",\n",
    "    \"#FF34FF\",\n",
    "    \"#FF4A46\",\n",
    "    \"#008941\",\n",
    "    \"#006FA6\",\n",
    "    \"#A30059\",\n",
    "    \"#FFDBE5\",\n",
    "    \"#7A4900\",\n",
    "    \"#0000A6\",\n",
    "    \"#63FFAC\",\n",
    "    \"#B79762\",\n",
    "    \"#004D43\",\n",
    "    \"#8FB0FF\",\n",
    "    \"#997D87\",\n",
    "    \"#5A0007\",\n",
    "    \"#809693\",\n",
    "    \"#6A3A4C\",\n",
    "    \"#1B4400\",\n",
    "    \"#4FC601\",\n",
    "    \"#3B5DFF\",\n",
    "    \"#4A3B53\",\n",
    "    \"#FF2F80\",\n",
    "    \"#61615A\",\n",
    "    \"#BA0900\",\n",
    "    \"#6B7900\",\n",
    "    \"#00C2A0\",\n",
    "    \"#FFAA92\",\n",
    "    \"#FF90C9\",\n",
    "    \"#B903AA\",\n",
    "    \"#D16100\",\n",
    "    \"#DDEFFF\",\n",
    "    \"#000035\",\n",
    "    \"#7B4F4B\",\n",
    "    \"#A1C299\",\n",
    "    \"#300018\",\n",
    "    \"#0AA6D8\",\n",
    "    \"#013349\",\n",
    "    \"#00846F\",\n",
    "    \"#372101\",\n",
    "    \"#FFB500\",\n",
    "    \"#C2FFED\",\n",
    "    \"#A079BF\",\n",
    "    \"#CC0744\",\n",
    "    \"#C0B9B2\",\n",
    "    \"#C2FF99\",\n",
    "    \"#001E09\",\n",
    "    \"#00489C\",\n",
    "    \"#6F0062\",\n",
    "    \"#0CBD66\",\n",
    "    \"#EEC3FF\",\n",
    "    \"#456D75\",\n",
    "    \"#B77B68\",\n",
    "    \"#7A87A1\",\n",
    "    \"#788D66\",\n",
    "    \"#885578\",\n",
    "    \"#FAD09F\",\n",
    "    \"#FF8A9A\",\n",
    "    \"#D157A0\",\n",
    "    \"#BEC459\",\n",
    "    \"#456648\",\n",
    "    \"#0086ED\",\n",
    "    \"#886F4C\",\n",
    "    \"#34362D\",\n",
    "    \"#B4A8BD\",\n",
    "    \"#00A6AA\",\n",
    "    \"#452C2C\",\n",
    "    \"#636375\",\n",
    "    \"#A3C8C9\",\n",
    "    \"#FF913F\",\n",
    "    \"#938A81\",\n",
    "    \"#575329\",\n",
    "    \"#00FECF\",\n",
    "    \"#B05B6F\",\n",
    "    \"#8CD0FF\",\n",
    "    \"#3B9700\",\n",
    "    \"#04F757\",\n",
    "    \"#C8A1A1\",\n",
    "    \"#1E6E00\",\n",
    "    \"#7900D7\",\n",
    "    \"#A77500\",\n",
    "    \"#6367A9\",\n",
    "    \"#A05837\",\n",
    "    \"#6B002C\",\n",
    "    \"#772600\",\n",
    "    \"#D790FF\",\n",
    "    \"#9B9700\",\n",
    "    \"#549E79\",\n",
    "    \"#FFF69F\",\n",
    "    \"#201625\",\n",
    "    \"#72418F\",\n",
    "    \"#BC23FF\",\n",
    "    \"#99ADC0\",\n",
    "    \"#3A2465\",\n",
    "    \"#922329\",\n",
    "    \"#5B4534\",\n",
    "    \"#FDE8DC\",\n",
    "    \"#404E55\",\n",
    "    \"#0089A3\",\n",
    "    \"#CB7E98\",\n",
    "    \"#A4E804\",\n",
    "    \"#324E72\",\n",
    "]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Down-stream analysis"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0.1 Dataview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv('./results/Simulation/HMEC/chr22.txt', sep=\"\\t\", comment=\"#\")[[\"x\",\"y\",\"z\"]].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "\n",
    "coords = pd.read_csv('./results/Simulation/HMEC/chr22.txt', sep=\"\\t\", comment=\"#\")[[\"x\",\"y\",\"z\"]].to_numpy()\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "ax.plot_trisurf(coords[:, 0], coords[:, 1], coords[:, 2], linewidth=0.2, antialiased=True, cmap='Blues', alpha=0.5, label='True')\n",
    "ax.view_init(elev=60, azim=-40)\n",
    "ax.axis('off')\n",
    "plt.show()\n",
    "fig.savefig('figures/Simulation_benchmark.png', dpi=300, transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "\n",
    "coords = pd.read_csv('./results/Simulation/HMEC/chr22.txt', sep=\"\\t\", comment=\"#\")[[\"x\",\"y\",\"z\"]].to_numpy()\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111, projection='3d')\n",
    "ax.plot_trisurf(coords[:, 0], coords[:, 1], coords[:, 2], linewidth=0.2, antialiased=True, cmap='Reds', alpha=0.5, label='True')\n",
    "ax.view_init(elev=60, azim=-40)\n",
    "ax.axis('off')\n",
    "plt.show()\n",
    "fig.savefig('figures/Simulation_reconstructed.png', dpi=300, transparent=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('./data/Mice/E16.5_E1S3/E16.5_E1S3_cell_bin.h5ad')\n",
    "adata.obs['x']=adata.obsm['spatial'][:,0]\n",
    "adata.obs['y']=adata.obsm['spatial'][:,1]\n",
    "adata.obsm['meta']=adata.obs[['x','y','annotation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['annotation'] = adata.obs['annotation'].replace({'Dorsal midbrain neuron': 'Dorsal Mb neuron', 'Mid-/hindbrain and spinal cord neuron': 'Mb/Hb/SpC neuron','Forebrain radial glia cell':'Fb radial glia cell','Forebrain neuron':'Fb neuron', 'Spinal cord neuron':'SpC neuron'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi_save=300, figsize=[6,10], fontsize=12, facecolor=(0,0,0,0), transparent=True)\n",
    "sc.pl.scatter(adata,x=\"x\",y=\"y\",color='annotation', size=2, legend_loc = 'right margin', \n",
    "            legend_fontsize='small', show=False, frameon=False, title='')\n",
    "ax = plt.gca()\n",
    "# To remove the background frame\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "# To remove x-y axis\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "ax.axes.get_yaxis().set_visible(False)\n",
    "h, l = ax.get_legend_handles_labels()\n",
    "ax.legend_.remove()\n",
    "\n",
    "ax.legend(h, l,frameon=False,ncols=1,bbox_to_anchor=(1, 0.95),markerscale=10, fontsize=\"13\")\n",
    "plt.savefig('mice_dataview.png', dpi=300, transparent=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### preprocess obs and gems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs = pd.read_csv('./data/external/small_lung_cancer/lung_cellmeta.csv', sep=',')\n",
    "lung_samples,i = np.unique(obs.Run_Tissue_name,return_index=True)\n",
    "lung_samples=lung_samples[np.argsort(i)]\n",
    "print(lung_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_slide_anno={}\n",
    "obs_slide_xy={}\n",
    "obs_slide={}\n",
    "tx_slide={}\n",
    "# process obs\n",
    "for i,sample in enumerate(lung_samples):\n",
    "    obs_slide_anno[sample] = pd.read_csv('./data/NSCLC/'+sample+'.cellmeta.anno.csv', sep=',')\n",
    "    obs_slide_xy[sample] = pd.read_csv('./data/external/small_lung_cancer/'+sample+'/'+sample+'-Flat_files_and_images'+'/'+sample+'_metadata_file.csv', sep=',')\n",
    "    obs_slide_xy[sample]['cell_ID'] ='c_'+str(i+1)+'_'+obs_slide_xy[sample]['fov'].astype(str)+'_'+obs_slide_xy[sample]['cell_ID'].astype(str)\n",
    "    obs_slide[sample] = pd.merge(obs_slide_anno[sample],obs_slide_xy[sample], how='inner', on='cell_ID')\n",
    "    obs_slide[sample].to_csv('./data/NSCLC/'+sample+'.cellmeta.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### draw plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# color map\n",
    "obs = pd.read_csv('./data/NSCLC/lung_cellmeta.csv', sep=',')\n",
    "color_map = {\n",
    "    'B-cell': '#ef9708',\n",
    "    'plasmablast': '#f0b98d',\n",
    "    'NK': '#0fcfc0',\n",
    "    'T CD4 naive': '#e07b91',\n",
    "    'T CD4 memory': '#d33f6a',\n",
    "    'T CD8 naive': '#bb7784',\n",
    "    'T CD8 memory': '#8e063b',\n",
    "    'Treg': '#e6afb9',\n",
    "    'endothelial': '#bec1d4',\n",
    "    'epithelial': '#800080',  # purple\n",
    "    'fibroblast': '#7d87b9',\n",
    "    'macrophage': '#9cded6',\n",
    "    'mast': '#11c638',\n",
    "    'monocyte': '#8dd593',\n",
    "    'neutrophil': '#c6dec7',\n",
    "    'pDC': '#ead3c6',\n",
    "    'mDC': '#d6bcc0',\n",
    "    'tumor 5': '#0000A6',\n",
    "    'tumor 6': '#1CE6FF',\n",
    "    'tumor 9': '#3B5DFF',\n",
    "    'tumor 12': '#023fa5',\n",
    "    'tumor 13': '#4a6fe3',\n",
    "}\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(18, 6)}, font_scale=1)\n",
    "f,(ax1,ax2,ax3) = plt.subplots(1,3, gridspec_kw={'width_ratios':[1,1,1]})\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "\n",
    "obs_ct1 = pd.read_csv('./data/NSCLC/Lung5_Rep1.cellmeta.csv', sep=',')\n",
    "g1=sns.scatterplot(data=obs_ct1, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1.5, ax=ax1)\n",
    "g1.set_ylabel('')\n",
    "g1.set_xlabel('')\n",
    "g1.set_yticks([])\n",
    "g1.set_xticks([])\n",
    "obs_ct2 = pd.read_csv('./data/NSCLC/Lung5_Rep2.cellmeta.csv', sep=',')\n",
    "g2=sns.scatterplot(data=obs_ct2, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1.5, ax=ax2)\n",
    "g2.set_ylabel('')\n",
    "g2.set_xlabel('')\n",
    "g2.set_yticks([])\n",
    "g2.set_xticks([])\n",
    "obs_ct3 = pd.read_csv('./data/NSCLC/Lung5_Rep3.cellmeta.csv', sep=',')\n",
    "g3=sns.scatterplot(data=obs_ct3, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1.5, ax=ax3)\n",
    "g3.set_ylabel('')\n",
    "g3.set_xlabel('')\n",
    "g3.set_yticks([])\n",
    "g3.set_xticks([])\n",
    "\n",
    "# draw shared legend\n",
    "from matplotlib.patches import Patch\n",
    "lut = color_map\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend=ax3.legend(handles, sorted(lut), bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=1, frameon=False, fontsize=13, labelspacing=0.2)\n",
    "#legend.get_frame().set_facecolor('white')\n",
    "\n",
    "# remove other legends\n",
    "ax1.get_legend().remove()\n",
    "ax2.get_legend().remove()\n",
    "\n",
    "f.savefig('figures/Fig4a.pdf', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_types = ['epithelial', 'tumor 5']\n",
    "\n",
    "f, (ax1, ax2, ax3) = plt.subplots(1, 3, gridspec_kw={'width_ratios': [1, 1, 1]})\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "\n",
    "g1 = sns.scatterplot(\n",
    "    data=obs_ct1[obs_ct1['cell_type'].isin(selected_types)],\n",
    "    x=\"CenterX_global_px\", y=\"CenterY_global_px\",\n",
    "    hue=\"cell_type\", palette=color_map, s=1.5, ax=ax1\n",
    ")\n",
    "g1.set_ylabel('')\n",
    "g1.set_xlabel('')\n",
    "g1.set_yticks([])\n",
    "g1.set_xticks([])\n",
    "\n",
    "g2 = sns.scatterplot(\n",
    "    data=obs_ct2[obs_ct2['cell_type'].isin(selected_types)],\n",
    "    x=\"CenterX_global_px\", y=\"CenterY_global_px\",\n",
    "    hue=\"cell_type\", palette=color_map, s=1.5, ax=ax2\n",
    ")\n",
    "g2.set_ylabel('')\n",
    "g2.set_xlabel('')\n",
    "g2.set_yticks([])\n",
    "g2.set_xticks([])\n",
    "\n",
    "g3 = sns.scatterplot(\n",
    "    data=obs_ct3[obs_ct3['cell_type'].isin(selected_types)],\n",
    "    x=\"CenterX_global_px\", y=\"CenterY_global_px\",\n",
    "    hue=\"cell_type\", palette=color_map, s=1.5, ax=ax3\n",
    ")\n",
    "g3.set_ylabel('')\n",
    "g3.set_xlabel('')\n",
    "g3.set_yticks([])\n",
    "g3.set_xticks([])\n",
    "\n",
    "from matplotlib.patches import Patch\n",
    "handles = [Patch(facecolor=color_map[name]) for name in selected_types]\n",
    "ax3.legend(handles, selected_types, bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=1, frameon=False, fontsize=13, labelspacing=0.2)\n",
    "\n",
    "ax1.get_legend().remove()\n",
    "ax2.get_legend().remove()\n",
    "\n",
    "f.savefig('figures/Fig4a_selected.pdf', bbox_inches='tight')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs = pd.read_csv('./data/NSCLC/lung_cellmeta.csv', sep=',')\n",
    "color_map = dict(zip(np.unique(obs.cell_type), godsnot_102[0:np.unique(obs.cell_type).size]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung5_Rep1.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung5_Rep2.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung5_Rep3.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung6.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung9_Rep1.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung9_Rep2.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung12.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_ct = pd.read_csv('./data/NSCLC/Lung13.cellmeta.csv', sep=',')\n",
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(6, 6)}, font_scale=1)\n",
    "sns.scatterplot(data=obs_ct, x=\"CenterX_global_px\", y=\"CenterY_global_px\", hue=\"cell_type\", palette=color_map, s=1)\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=2, frameon=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Liver Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### preprocess gems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs = pd.read_csv('./data/external/liver/liver_cellmeta.csv', sep=',')\n",
    "liver_samples,i = np.unique(obs.Run_Tissue_name,return_index=True)\n",
    "liver_samples=liver_samples[np.argsort(i)]\n",
    "print(liver_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "obs_slide_anno={}\n",
    "obs_slide_xy={}\n",
    "obs_slide={}\n",
    "tx_slide={}\n",
    "# process obs\n",
    "for i,sample in enumerate(liver_samples):\n",
    "    print(sample)\n",
    "    tx_slide[sample] = pd.read_csv('./data/Liver/'+sample+'.raw.gem.csv', sep=' ')\n",
    "    tx_slide[sample]['cell_ID'] = 'c_'+str(i+1)+'_'+tx_slide[sample]['fov'].astype(int).astype(str)+'_'+tx_slide[sample]['cell_ID'].astype(str)\n",
    "    tx_slide[sample] = tx_slide[sample][[\"cell_ID\",\"x\",\"y\",\"target\"]]\n",
    "    tx_slide[sample][\"MIDCount\"] = tx_slide[sample].groupby([\"cell_ID\",\"x\",\"y\",\"target\"]).transform(\"size\")\n",
    "    tx_slide[sample].rename(columns={\"cell_ID\": \"cell\", \"x\": \"x\", \"y\": \"y\", \"target\":\"geneID\"}, inplace=True)\n",
    "    tx_slide[sample].to_csv('./data/Liver/'+sample+'.gem.csv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### draw plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# color map\n",
    "obs = pd.read_csv('./data/external/liver/liver_cellmeta.csv', sep=',')\n",
    "types=np.unique(obs.cellType)\n",
    "\n",
    "color_map = {\n",
    "'Antibody.secreting.B.cells': '#8dd593',\n",
    "'Mature.B.cells': '#11c638',\n",
    "'CD3+.alpha.beta.T.cells': '#c6dec7',\n",
    "'gamma.delta.T.cells.1': '#9cded6',\n",
    "'NK.like.cells': '#f3e1eb',\n",
    "'Central.venous.LSECs': '#bec1d4',\n",
    "'Cholangiocytes': '#8595e1',\n",
    "'Erthyroid.cells': '#d5eae7',\n",
    "'Hep': '#8e063b',\n",
    "'Hep.1': '#d6bcc0',\n",
    "'Hep.3': '#e6afb9',\n",
    "'Hep.4': '#bb7784',\n",
    "'Hep.5': '#e07b91',\n",
    "'Hep.6': '#d33f6a',\n",
    "'Inflammatory.macrophages': '#ef9708',\n",
    "'Non.inflammatory.macrophages': '#f0b98d',\n",
    "'Periportal.LSECs': '#7d87b9',\n",
    "'Portal.endothelial.cells': '#ead3c6',\n",
    "'Stellate.cells': '#0fcfc0',\n",
    "'tumor_1': '#4a6fe3',\n",
    "'tumor_2': '#023fa5',\n",
    "'NotDet': '#b5bbe3'}\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(16, 8)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "f,(ax1,ax2) = plt.subplots(1,2, gridspec_kw={'width_ratios':[1,1]})\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "\n",
    "g1 = sns.scatterplot(data=obs[obs.Run_Tissue_name=='NormalLiver'], x=\"x_slide_mm\", y=\"y_slide_mm\", hue=\"cellType\", palette=color_map, s=1.0, ax=ax1)\n",
    "g1.set_ylabel('')\n",
    "g1.set_xlabel('')\n",
    "g1.set_yticks([])\n",
    "g1.set_xticks([])\n",
    "obs_ct2 = pd.read_csv('./data/NSCLC/Lung5_Rep2.cellmeta.csv', sep=',')\n",
    "g2 = sns.scatterplot(data=obs[obs.Run_Tissue_name=='CancerousLiver'], x=\"x_slide_mm\", y=\"y_slide_mm\", hue=\"cellType\", palette=color_map, s=1.0, ax=ax2)\n",
    "g2.set_ylabel('')\n",
    "g2.set_xlabel('')\n",
    "g2.set_yticks([])\n",
    "g2.set_xticks([])\n",
    "\n",
    "# draw shared legend\n",
    "from matplotlib.patches import Patch\n",
    "lut = color_map\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend=ax2.legend(handles, sorted(lut), title='Cell type', bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0., ncol=1, frameon=False, fontsize=9.6)\n",
    "#legend.get_frame().set_facecolor('white')\n",
    "\n",
    "# remove other legends\n",
    "ax1.get_legend().remove()\n",
    "\n",
    "# show\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0.2 Confview - gene forest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genechr_mice=gene_chr(species='Mice')\n",
    "D = {}\n",
    "samples = list(Xs.keys())\n",
    "for sample in samples:\n",
    "    X = Xs[sample]  \n",
    "    D[sample] = gene_gene_distace_matrix(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=[]\n",
    "for g in GeneLists['Cardiomyocyte_D84F']:\n",
    "    labels.append(genechr_mice[g])\n",
    "cmap=dict(zip(np.unique(labels), zeileis_28))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_network(D['Cardiomyocyte_D84F'], labels=labels, cmap=cmap, \n",
    "            csep='\\n', corder=0, N_neighbor=1, html='GeneConf_Cardiomyocyte_D84F.html', \n",
    "            width=1400, height=1000,edge_scale=0.1, edge_adjust=+1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lungs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and 'Lung5' in chip: \n",
    "        print(sample)\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build dict\n",
    "genechr=gene_chr(species='Human')\n",
    "# Build distance matrix\n",
    "D = gene_gene_distance_matrix(Xs['Lung5_Rep3 macrophage'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GeneLists['Lung5_Rep3 macrophage']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# labels\n",
    "labels=[]\n",
    "for g in GeneLists['Lung5_Rep3 macrophage']:\n",
    "    try:\n",
    "        labels.append(genechr[g])\n",
    "    except KeyError:\n",
    "        labels.append('Unknown')\n",
    "# cmap\n",
    "cmap=dict(zip(np.unique(labels), zeileis_28))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_network(D, labels=labels, cmap=cmap, \n",
    "            csep='\\n', corder=0, N_neighbor=1, html='GeneConf_Lung5_Rep3_macrophage.html', \n",
    "            width=1400, height=1000,edge_scale=0.1, edge_adjust=+1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.1 Transcription Center - single gene compared with shared location"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def draw_scatter_with_labels_skip_nan(x, y, label, width=10, height=8):\n",
    "    from adjustText import adjust_text\n",
    "    plt.figure(figsize=(width, height)) # width and height are in inches\n",
    "    # Convert x and y to pandas Series objects\n",
    "    x = pd.Series(x)\n",
    "    y = pd.Series(y)\n",
    "    # Drop any rows that contain nan values in either x or y\n",
    "    X, Y = x.dropna(), y.dropna()\n",
    "    # Create a scatter plot with x and y values\n",
    "    plt.scatter(X, Y)\n",
    "\n",
    "    # Loop through the x and y values and add labels for each point\n",
    "    texts=[]\n",
    "    for i in range(len(x)):\n",
    "        if not np.isnan(x[i]) or np.isnan(y[i]):\n",
    "            # Use plt.annotate() to add the labels near the points\n",
    "            texts.append(plt.text(x[i], y[i],label[i]))\n",
    "    adjust_text(texts, force_points=0.2, force_text=0.2,\n",
    "            expand_points=(1, 1), expand_text=(1, 1),\n",
    "            arrowprops=dict(arrowstyle=\"-\", color='black', lw=0.5))\n",
    "    # Show the plot\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inpath = \"./data/E16.5_E1S3/SS200000108BR_A3A4_scgem.Cardiomyocyte.csv\"\n",
    "gem = pd.read_csv(inpath, sep=\"\\t\", comment=\"#\")\n",
    "X = load_data(\"D:\\pysfm\\Results\\SS200000108BR_A3A4_scgem.Cardiomyocyte\\CLAK\\SS200000108BR_A3A4_scgem.Cardiomyocyte_CLAK_ChromosomeStructure.pkl\")\n",
    "RM = load_data(\"D:\\pysfm\\Results\\SS200000108BR_A3A4_scgem.Cardiomyocyte\\CLAK\\SS200000108BR_A3A4_scgem.Cardiomyocyte_CLAK_RotationMatrixes.pkl\")\n",
    "W = load_data(\"D:\\pysfm\\Results\\SS200000108BR_A3A4_scgem.Cardiomyocyte\\CLAK\\SS200000108BR_A3A4_scgem.Cardiomyocyte_CLAK_W.pkl\")\n",
    "GeneList = load_data(\"D:\\pysfm\\Results\\SS200000108BR_A3A4_scgem.Cardiomyocyte\\CLAK\\SS200000108BR_A3A4_scgem.Cardiomyocyte_CLAK_GeneList.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nan_idx=np.isnan(W[0,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "~nan_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_scatter_with_labels_skip_nan(W[0,~nan_idx],W[1,~nan_idx],label=GeneList[~nan_idx], width=10,height=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RM.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C=np.dot(X,RM[0])\n",
    "draw_scatter_with_labels_skip_nan(C[~nan_idx,0],C[~nan_idx,1],label=GeneList[~nan_idx], width=10,height=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "gdist={}\n",
    "for g in tqdm(range(X.shape[0]),desc=\"measuring variance for each gene\"):\n",
    "    gdist[g]=[]\n",
    "    for c in range(RM.shape[0]):\n",
    "        if not np.isnan(W[c*2:c*2+2,g]).any():\n",
    "            x, y = W[c*2:c*2+2,g]\n",
    "            C = np.dot(X,RM[c])\n",
    "            x_pre, y_pre = (C[g,0],C[g,1])\n",
    "            d = math.sqrt((x_pre - x)**2 + (y_pre - y)**2)\n",
    "            gdist[g].append(d)\n",
    "\n",
    "vars=[]\n",
    "for i in gdist.keys():\n",
    "    vars.append(np.var(gdist[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame([GeneList,vars],index=[\"gene\",\"vars\"]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.sort_values([\"vars\"], ascending=False)[0:100].gene.to_csv(\"Var_tail100gene_Card.txt\",index=False,header=False)\n",
    "df.sort_values([\"vars\"], ascending=True)[0:100].gene.to_csv(\"Var_top100gene_Card.txt\",index=False,header=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.set(rc={'figure.figsize':(4, 10)})\n",
    "# Plot a bar chart\n",
    "top=df.sort_values([\"vars\"], ascending=True)[0:100]\n",
    "bar_plot=sns.barplot(x='vars',y='gene',data=top,color=\"black\",orient=\"h\",)\n",
    "# Add labels and title\n",
    "plt.xlabel('Variance')\n",
    "plt.ylabel('Gene')\n",
    "plt.title('The Top 100 Genes with the Highest Variance in Cardiomyocytes')\n",
    "plt.xticks(fontsize=8) \n",
    "plt.yticks(fontsize=8) \n",
    "# Show the plot\n",
    "plt.show()\n",
    "# create a line plot\n",
    "#fig = bar_plot.get_figure() # get the figure object\n",
    "#fig.savefig('./figures/barplot_umap_orientations.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.set(rc={'figure.figsize':(4, 10)})\n",
    "# Plot a bar chart\n",
    "bar_plot=sns.barplot(x='gene',y='vars',data=df,color=\"black\")\n",
    "# Add labels and title\n",
    "plt.xlabel('Variance')\n",
    "plt.ylabel('Gene')\n",
    "plt.title('The Top 100 Genes with the Highest Variance in Cardiomyocytes')\n",
    "plt.xticks(fontsize=8) \n",
    "plt.yticks(fontsize=8) \n",
    "# Show the plot\n",
    "plt.show()\n",
    "# create a line plot\n",
    "#fig = bar_plot.get_figure() # get the figure object\n",
    "#fig.savefig('./figures/barplot_umap_orientations.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Axolotls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "gems={}\n",
    "Xs={}\n",
    "RMs={}\n",
    "Ws={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "ctypes=[]\n",
    "sample='15DPI_1'\n",
    "data_dir = os.path.join(\"./data/Axolotls/regeneration/\",sample)\n",
    "result_dir = os.path.join(\"D:/pysfm/Results/Axolotls/\",sample)\n",
    "for file in os.listdir(data_dir):\n",
    "    if file.endswith(\".csv\"): \n",
    "        ctype = file.split('.')[1]\n",
    "        ctypes.append(ctype)\n",
    "        gems[ctype] = pd.read_csv(os.path.join(data_dir, file), sep=\"\\t\", comment=\"#\")\n",
    "        target_dir = os.path.join(os.path.join(result_dir, sample+'.'+ctype), os.listdir(os.path.join(result_dir, sample+'.'+ctype))[0])\n",
    "        for f in os.listdir(target_dir):\n",
    "            if f.endswith(\"_ChromosomeStructure.pkl\"):\n",
    "                Xs[ctype] = load_data(os.path.join(target_dir, f))\n",
    "            elif f.endswith(\"_RotationMatrixes.pkl\"):\n",
    "                RMs[ctype] = load_data(os.path.join(target_dir, f))\n",
    "            elif f.endswith(\"_W.pkl\"):\n",
    "                Ws[ctype] = load_data(os.path.join(target_dir, f))\n",
    "            elif f.endswith(\"_GeneList.pkl\"):\n",
    "                GeneLists[ctype] = load_data(os.path.join(target_dir, f))\n",
    "            elif f.endswith(\"_CellList.pkl\"):\n",
    "                CellLists[ctype] = load_data(os.path.join(target_dir, f))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "GeneFull=reduce(np.union1d, (GeneLists.values()))\n",
    "GdistMatrix={}\n",
    "for ctype in ctypes:\n",
    "    try:\n",
    "        X = Xs[ctype]\n",
    "        RM = RMs[ctype]\n",
    "        W = Ws[ctype]\n",
    "        Glist = GeneLists[ctype]\n",
    "    except KeyError:\n",
    "        print(ctype)\n",
    "        continue\n",
    "    GdistMatrix[ctype] = np.zeros((GeneFull.shape[0],RM.shape[0]))\n",
    "    for gene in tqdm(GeneFull,desc=\"measuring variance for each gene\"):\n",
    "        if gene in Glist:\n",
    "            g, = np.where(GeneFull == gene)\n",
    "            h, = np.where(Glist == gene)\n",
    "            for c in range(RM.shape[0]):\n",
    "                if not np.isnan(W[c*2:c*2+2,h]).any():\n",
    "                    x, y = W[c*2:c*2+2,h]\n",
    "                    C = np.dot(X,RM[c])\n",
    "                    x_pre, y_pre = (C[h,0],C[h,1])\n",
    "                    d = math.sqrt((x_pre - x)**2 + (y_pre - y)**2)\n",
    "                    GdistMatrix[ctype][g,c]=d\n",
    "                else:\n",
    "                    GdistMatrix[ctype][g,c]=np.nan\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GdistMatrix['CMPN'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(GdistMatrix.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gem_15DPI_1.cell.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1 = sc.read_h5ad(\"./data/Axolotls/15DPI_1.h5ad\")\n",
    "adata_15DPI_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1[adata_15DPI_1.obs['Annotation'].isin(['CMPN'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1[adata_15DPI_1.obs['Annotation'].isin(list(GdistMatrix.keys()))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1_CMPN = adata_15DPI_1[adata_15DPI_1.obs['Annotation'].isin(['CMPN'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CellLists['CMPN']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n=0\n",
    "for i in adata_15DPI_1_CMPN.obs.cell_id.values:\n",
    "    if i not in CellLists['CMPN'].values:\n",
    "        print(i)\n",
    "        n+=1\n",
    "print(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(adata_15DPI_1_CMPN.X.toarray(),index=adata_15DPI_1_CMPN.obs['cell_id']).to_csv('adata_15DPI_1_CMPN.X.csv')\n",
    "adata_15DPI_1_CMPN.obs[['cell_id','x','y']].set_index(\"cell_id\").rename({'x': 'row','y': 'col'},axis=1).to_csv(\"adata_15DPI_1_CMPN.spatial.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GdistMatrix.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat((CellLists['CMPN'],CellLists['dpEX'],CellLists['mpEX'],CellLists['MSN'],CellLists['reaEGC'],CellLists['scgnIN'],CellLists['VLMC'],CellLists['wntEGC']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Cs={}\n",
    "for ct in GdistMatrix.keys():\n",
    "    Cs[ct]=gems[ct].groupby('cell')[['x', 'y']].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat(Cs.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat(Cs.values()).reindex(cellids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat(Cs.values()).reindex(cellids).to_csv('SpatialCoords_Axolotls_15DPI_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cellids=pd.concat((CellLists['CMPN'],CellLists['dpEX'],CellLists['mpEX'],CellLists['MSN'],CellLists['reaEGC'],CellLists['scgnIN'],CellLists['VLMC'],CellLists['wntEGC']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(np.hstack(GdistMatrix.values()),index=GeneFull,columns=cellids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(np.hstack(GdistMatrix.values()),index=GeneFull,columns=cellids).to_csv('ErrorDistances_Axolotls_15DPI_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(file=\"ErrorDistances_Axolotls_15DPI_1\",arr=np.hstack(GdistMatrix.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.hstack(GdistMatrix.values()).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GeneFull.size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1.obs['x']=adata_15DPI_1.obsm['spatial'][:,0]\n",
    "adata_15DPI_1.obs['y']=adata_15DPI_1.obsm['spatial'][:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1.obs[['cell_id','x','y']].set_index('cell_id').loc[cellids.values.tolist()].to_csv('SpatialCoords_Axolotls_15DPI_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_moran_max_dist = pd.read_csv(\"D:/pysfm/spatialpattern/0807/morani_dis.csv\", index_col=0)\n",
    "Gene_index=[]\n",
    "for i in df_moran_max_dist.index[0:20]:\n",
    "    Gene_index.append(adata_15DPI_1.var['Gene'][adata_15DPI_1.var['Gene']==i].index.values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from matplotlib.pyplot import rc_context\n",
    "#with rc_context({'figure.figsize': (5, 7)}):\n",
    "df_moran_max_dist = pd.read_csv(\"D:/pysfm/spatialpattern/0807/morani_dis.csv\", index_col=0)\n",
    "sc.pl.spatial(adata_15DPI_1, color=Gene_index, spot_size=50, ncols=5, cmap='viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(adata_15DPI_1, color=adata_15DPI_1.var['Gene'][adata_15DPI_1.var['Gene']==\"AMEX60DD000708\"].index.values[0], spot_size=50, ncols=5, cmap='viridis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(file=\"SpatialCoords_Axolotls_15DPI_1\",arr=np.hstack(Cs.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# variances for genes\n",
    "X = Xs['CMPN']\n",
    "RM = RMs['CMPN']\n",
    "W = Ws['CMPN']\n",
    "import math\n",
    "gdist={}\n",
    "for g in tqdm(range(X.shape[0]),desc=\"measuring variance for each gene\"):\n",
    "    gdist[g]=[]\n",
    "    for c in range(RM.shape[0]):\n",
    "        if not np.isnan(W[c*2:c*2+2,g]).any():\n",
    "            x, y = W[c*2:c*2+2,g]\n",
    "            C = np.dot(X,RM[c])\n",
    "            x_pre, y_pre = (C[g,0],C[g,1])\n",
    "            d = math.sqrt((x_pre - x)**2 + (y_pre - y)**2)\n",
    "            gdist[g].append(d)\n",
    "\n",
    "vars=[]\n",
    "for i in gdist.keys():\n",
    "    vars.append(np.var(gdist[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame([GeneLists['CMPN'],vars],index=[\"gene\",\"vars\"]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "sns.set(rc={'figure.figsize':(10, 4)})\n",
    "# Plot a bar chart\n",
    "bar_plot=sns.barplot(x='gene',y='vars',data=df,color=\"black\")\n",
    "# Add labels and title\n",
    "plt.xlabel('Gene')\n",
    "plt.ylabel('Variance')\n",
    "plt.title('Variance by each gene in CMPN')\n",
    "plt.xticks(fontsize=4) \n",
    "plt.yticks(fontsize=8) \n",
    "# Show the plot\n",
    "plt.show()\n",
    "# create a line plot\n",
    "#fig = bar_plot.get_figure() # get the figure object\n",
    "#fig.savefig('./figures/barplot_umap_orientations.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 Transcription Center - common gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check completeness\n",
    "import os\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in os.listdir(result_path) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    if 'adata.h5ad' not in files:\n",
    "        print(sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.stats import ttest_ind\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "        \n",
    "result_path=\"./results/Mice/\"\n",
    "\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        #print(sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "print(\"Common Gene Number: \" + str(len(intersected_values)))\n",
    "\n",
    "# housekeeping \n",
    "housekeep = pd.read_csv(\"./data/Housekeeping_TranscriptsMouse.csv\",sep=';').Genes.unique()\n",
    "intersect_housekeep = np.intersect1d(intersected_values, housekeep)\n",
    "print(\"Common Housekeeping Gene Number: \" + str(len(intersect_housekeep)))\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersect_housekeep) for a in GeneLists.values()]\n",
    "\n",
    "# build matrix\n",
    "N = len(Xs.keys())\n",
    "D1 = []\n",
    "for i in range(int(N/3)):\n",
    "    X1 = list(Xs.values())[3*i]\n",
    "    X2 = list(Xs.values())[3*i+1]\n",
    "    d1, _, _ = numpy_svd_rmsd_rot(normalizeX(X1[boolean_arrays[3*i]]), normalizeX(X2[boolean_arrays[3*i+1]]))\n",
    "    X2 = list(Xs.values())[3*i+2]\n",
    "    d2, _, _ = numpy_svd_rmsd_rot(normalizeX(X1[boolean_arrays[3*i]]), normalizeX(X2[boolean_arrays[3*i+2]]))\n",
    "    X1 = list(Xs.values())[3*i+1]\n",
    "    d3, _, _ = numpy_svd_rmsd_rot(normalizeX(X1[boolean_arrays[3*i+1]]), normalizeX(X2[boolean_arrays[3*i+2]]))\n",
    "    D1.append(d1)\n",
    "    D1.append(d2)\n",
    "    D1.append(d3)\n",
    "\n",
    "# common\n",
    "intersect_functional = np.setdiff1d(intersected_values, housekeep)\n",
    "\n",
    "x = 0\n",
    "y = 0\n",
    "for k in range(1):\n",
    "    intersect_functional_sub = np.random.choice(intersect_functional, size=len(intersect_housekeep), replace = False)\n",
    "    # Using in1d and nonzero\n",
    "    boolean_arrays = [np.in1d(a, intersect_functional_sub) for a in GeneLists.values()]\n",
    "\n",
    "    # build matrix\n",
    "    D2 = []\n",
    "    for i in range(int(N/3)):\n",
    "        X1 = list(Xs.values())[3*i]\n",
    "        X2 = list(Xs.values())[3*i+1]\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalizeX(X1[boolean_arrays[3*i]]), normalizeX(X2[boolean_arrays[3*i+1]]))\n",
    "        X2 = list(Xs.values())[3*i+2]\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(normalizeX(X1[boolean_arrays[3*i]]), normalizeX(X2[boolean_arrays[3*i+2]]))\n",
    "        X1 = list(Xs.values())[3*i+1]\n",
    "        d3, _, _ = numpy_svd_rmsd_rot(normalizeX(X1[boolean_arrays[3*i+1]]), normalizeX(X2[boolean_arrays[3*i+2]]))\n",
    "        D2.append(d1)\n",
    "        D2.append(d2)\n",
    "        D2.append(d3)\n",
    "\n",
    "    result = ttest_ind(D1, D2, alternative='greater')\n",
    "    if result[1] < 0.05:\n",
    "        x += 1\n",
    "print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect_functional"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":150,'figure.figsize':(3, 3)}, font_scale=1)\n",
    "ax = sns.boxplot([D1,D2],palette=['#A288C7','#F4AE68'], fill=True, gap=0.2, saturation=1)\n",
    "ax.set_xticklabels([\"Housekeeping\\nGenes\",\"Commonly\\nExpressed\\nGenes\"])\n",
    "ax.set_ylabel(\"RMSD\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        #print(sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "            CellLists[sample] = adata.obs.index\n",
    "\n",
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "print(\"Common Gene Number: \" + str(len(intersected_values)))\n",
    "# housekeeping \n",
    "housekeep = pd.read_csv(\"./data/Human_Mouse_Common.csv\",sep=';')\n",
    "intersect_housekeep = np.intersect1d(intersected_values, housekeep.Mouse)\n",
    "print(\"Common Housekeeping Gene Number: \" + str(len(intersect_housekeep)))\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersect_housekeep) for a in GeneLists.values()]\n",
    "\n",
    "N = len(Xs.keys())\n",
    "D = np.zeros((N,N))\n",
    "n = 0\n",
    "for i in Xs.keys():\n",
    "    m=0\n",
    "    for j in Xs.keys():\n",
    "        d, _, _ = numpy_svd_rmsd_rot(Xs[i][boolean_arrays[n]]/np.linalg.norm(Xs[i][boolean_arrays[n]]),\n",
    "                                     Xs[j][boolean_arrays[m]]/np.linalg.norm(Xs[j][boolean_arrays[m]]))\n",
    "        D[n,m] = d\n",
    "        D[m,n] = d\n",
    "        m+=1\n",
    "    n+=1\n",
    "    \n",
    "sns.clustermap(D, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis_r\", figsize=(18, 18))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "RMs={}\n",
    "Ws={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "ctypes=[]\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        print(sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            RMs[sample] = adata.obsm[\"Rotation\"]\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "            CellLists[sample] = adata.obs.index\n",
    "\n",
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "print(\"Common Gene Number: \" + str(len(intersected_values)))\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersected_values) for a in GeneLists.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = len(Xs.keys())\n",
    "D = np.zeros((N,N))\n",
    "n = 0\n",
    "for i in Xs.keys():\n",
    "    m=0\n",
    "    for j in Xs.keys():\n",
    "        d, _, _ = numpy_svd_rmsd_rot(Xs[i][boolean_arrays[n]]/np.linalg.norm(Xs[i][boolean_arrays[n]]),\n",
    "                                     Xs[j][boolean_arrays[m]]/np.linalg.norm(Xs[j][boolean_arrays[m]]))\n",
    "        D[n,m] = d\n",
    "        D[m,n] = d\n",
    "        m+=1\n",
    "    n+=1\n",
    "    \n",
    "sns.clustermap(D, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis_r\", figsize=(18, 18))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Axolotls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### all sample celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "gems={}\n",
    "Xs={}\n",
    "RMs={}\n",
    "Ws={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "ctypes=[]\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[1].split('_')[0]\n",
    "    sample = chip+'_'+ctype\n",
    "    if 'adata.h5ad' in files: \n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        #RMs[sample] = adata.obsm[\"Rotation\"]\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "        #CellLists[sample] = adata.obs.index\n",
    "\n",
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "print(\"Common Gene Number: \" + str(len(intersected_values)))\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersected_values) for a in GeneLists.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = len(Xs.keys())\n",
    "D = np.zeros((N,N))\n",
    "n = 0\n",
    "for i in Xs.keys():\n",
    "    m=0\n",
    "    for j in Xs.keys():\n",
    "        d, _, _ = numpy_svd_rmsd_rot(Xs[i][boolean_arrays[n]]/np.linalg.norm(Xs[i][boolean_arrays[n]]),\n",
    "                                     Xs[j][boolean_arrays[m]]/np.linalg.norm(Xs[j][boolean_arrays[m]]))\n",
    "        D[n,m] = d\n",
    "        D[m,n] = d\n",
    "        m+=1\n",
    "    n+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(15, 15)},font_scale=0.16)\n",
    "sns.clustermap(D, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CMPN along timeslots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'CMPN' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            #RMs[sample] = adata.obsm[\"Rotation\"]\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "            #CellLists[sample] = adata.obs.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersected_values) for a in GeneLists.values()]\n",
    "\n",
    "N = len(Xs.keys())\n",
    "D = np.zeros((N,N))\n",
    "n = 0\n",
    "for i in Xs.keys():\n",
    "    m=0\n",
    "    for j in Xs.keys():\n",
    "        d, _, _ = numpy_svd_rmsd_rot(Xs[i][boolean_arrays[n]]/np.linalg.norm(Xs[i][boolean_arrays[n]]),\n",
    "                                     Xs[j][boolean_arrays[m]]/np.linalg.norm(Xs[j][boolean_arrays[m]]))\n",
    "        D[n,m] = d\n",
    "        D[m,n] = d\n",
    "        m+=1\n",
    "    n+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(8, 8)},font_scale=1)\n",
    "sns.clustermap(D, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis_r\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Brain Cancer "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "RMs={}\n",
    "Ws={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "ctypes=[]\n",
    "\n",
    "result_path=\"./results/Cancer/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[4].split('_')[0]\n",
    "    sample = chip+'_'+ctype\n",
    "    if 'adata.h5ad' in files: \n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        RMs[sample] = adata.obsm[\"Rotation\"]\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "        CellLists[sample] = adata.obs.index\n",
    "\n",
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "print(\"Common Gene Number: \" + str(len(intersected_values)))\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersected_values) for a in GeneLists.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "N = len(Xs.keys())\n",
    "D = np.zeros((N,N))\n",
    "n = 0\n",
    "for i in Xs.keys():\n",
    "    m=0\n",
    "    for j in Xs.keys():\n",
    "        d, _, _ = numpy_svd_rmsd_rot(Xs[i][boolean_arrays[n]]/np.linalg.norm(Xs[i][boolean_arrays[n]]),\n",
    "                                     Xs[j][boolean_arrays[m]]/np.linalg.norm(Xs[j][boolean_arrays[m]]))\n",
    "        D[n,m] = d\n",
    "        D[m,n] = d\n",
    "        m+=1\n",
    "    n+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "mpl.rcParams['figure.dpi'] = 300\n",
    "sns.set(font_scale=0.5)\n",
    "sns.clustermap(D, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis_r\", figsize=(6, 6))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3 Transcription Center - shared gene in every two types"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### body"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kabsch_numpy(P, Q):\n",
    "    \"\"\"\n",
    "    Computes the optimal rotation and translation to align two sets of points (P -> Q),\n",
    "    and their RMSD.\n",
    "\n",
    "    :param P: A Nx3 matrix of points\n",
    "    :param Q: A Nx3 matrix of points\n",
    "    :return: A tuple containing the optimal rotation matrix, the optimal\n",
    "             translation vector, and the RMSD.\n",
    "    \"\"\"\n",
    "    P = np.array(P)\n",
    "    Q = np.array(Q)\n",
    "    assert P.shape == Q.shape, \"Matrix dimensions must match\"\n",
    "\n",
    "    # Compute centroids\n",
    "    centroid_P = np.mean(P, axis=0)\n",
    "    centroid_Q = np.mean(Q, axis=0)\n",
    "\n",
    "    # Optimal translation\n",
    "    t = centroid_Q - centroid_P\n",
    "\n",
    "    # Center the points\n",
    "    p = P - centroid_P\n",
    "    q = Q - centroid_Q\n",
    "\n",
    "    # Compute the covariance matrix\n",
    "    H = np.dot(p.T, q)\n",
    "\n",
    "    # SVD\n",
    "    U, S, Vt = np.linalg.svd(H)\n",
    "\n",
    "    # Validate right-handed coordinate system\n",
    "    is_reflection = np.linalg.det(np.dot(Vt.T, U.T)) < 0.0\n",
    "    if is_reflection:\n",
    "        Vt[-1, :] *= -1.0\n",
    "\n",
    "    # Optimal rotation\n",
    "    R = np.dot(Vt.T, U.T)\n",
    "\n",
    "    # RMSD\n",
    "    rmsd = np.sqrt(np.sum(np.square(np.dot(p, R.T) - q)) / P.shape[0])\n",
    "\n",
    "    return R, t, rmsd, is_reflection\n",
    "    \n",
    "def RMSD_distance_matrix_new(\n",
    "    Confs, GeneLists, order=None, ngene=100, compare_method=\"pair\", norm_method=None\n",
    "):\n",
    "    # check if Confs and GeneLists have the same keys\n",
    "    if order == None:\n",
    "        keys = list(Confs.keys())\n",
    "    else:\n",
    "        keys = order\n",
    "    for i in keys:\n",
    "        if not i in GeneLists.keys():\n",
    "            print(f\"Warning: {i} in Confs but not in GeneLists\")\n",
    "            return\n",
    "\n",
    "    # calculate the distance matrix\n",
    "    N = len(keys)\n",
    "    DM = np.zeros((N, N))\n",
    "    if compare_method == \"complete\":\n",
    "        from functools import reduce\n",
    "\n",
    "        intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "        intersected_values = intersected_values[:ngene]\n",
    "        if len(intersected_values) < ngene:\n",
    "            print(\n",
    "                f\"Warning: {len(intersected_values)} common genes between samples are less than {ngene}\"\n",
    "            )\n",
    "    from tqdm import tqdm\n",
    "\n",
    "    for n, key_n in enumerate(tqdm(keys)):\n",
    "        for m, key_m in enumerate(keys[: n + 1]):\n",
    "            if compare_method == \"pair\":\n",
    "                intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])\n",
    "                intersected_values = intersected_values[:ngene]\n",
    "                if len(intersected_values) < ngene:\n",
    "                    print(\n",
    "                        f\"Warning: {len(intersected_values)} common genes between {key_n} and {key_m} are less than {ngene}\"\n",
    "                    )\n",
    "            boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "            boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "            Conf_n = Confs[key_n][boolean_arrays_n]\n",
    "            Conf_m = Confs[key_m][boolean_arrays_m]\n",
    "            if norm_method:\n",
    "                Conf_n = normalizeF(Conf_n, method=norm_method)\n",
    "                Conf_m = normalizeF(Conf_m, method=norm_method)\n",
    "            rmsd1 = kabsch_numpy(Conf_n, Conf_m)[2]\n",
    "            rmsd2 = kabsch_numpy(mirror(Conf_n), Conf_m)[2]\n",
    "            DM[n, m] = DM[m, n] = min(rmsd1, rmsd2)\n",
    "    return DM\n",
    "\n",
    "def plot_RMSD_heatmap_group(\n",
    "    D,\n",
    "    labels,\n",
    "    cmap=\"viridis\",\n",
    "    method=\"average\",\n",
    "    cthresh=4.8,\n",
    "    font_scale=1,\n",
    "    legend=True,\n",
    "    cbbox=(0.07, 0.84, 0.04, 0.12),\n",
    "    lgbox=(0.86, 0.96, 0.01, 0.01),\n",
    "    ticks=[0, 0.5, 1, 1.5],\n",
    "    figsize=(8, 8),\n",
    "    dpi=300,\n",
    "    legend_ncol=2,\n",
    "    mask=None,\n",
    "):\n",
    "    import matplotlib.pyplot as plt\n",
    "\n",
    "    clustergrid = sns.clustermap(\n",
    "        D, xticklabels=labels, yticklabels=labels, method=method\n",
    "    )\n",
    "    plt.close()\n",
    "\n",
    "    # Get clusters with threshold\n",
    "    linkage = clustergrid.dendrogram_row.linkage\n",
    "    threshold = cthresh\n",
    "    clusters = fcluster(linkage, threshold, criterion=\"distance\")\n",
    "\n",
    "    # Create a color map for clusters\n",
    "    n_clusters = len(np.unique(clusters))\n",
    "    color_map = sns.color_palette(\"rainbow\", n_clusters)\n",
    "\n",
    "    # Assign colors to clusters\n",
    "    cluster_colors = [color_map[c - 1] for c in clusters]\n",
    "\n",
    "    # Draw plot\n",
    "    sns.set(rc={\"figure.dpi\": dpi, \"figure.figsize\": figsize}, font_scale=font_scale)\n",
    "\n",
    "    ax = sns.clustermap(\n",
    "        D,\n",
    "        xticklabels=labels,\n",
    "        yticklabels=labels,\n",
    "        cmap=cmap,\n",
    "        method=method,\n",
    "        row_colors=[cluster_colors],\n",
    "        col_colors=[cluster_colors],\n",
    "        cbar_kws={\"ticks\": ticks},\n",
    "        mask=mask,\n",
    "    )\n",
    "    ax.ax_cbar.set_title(\"RMSD\")\n",
    "    ax.ax_cbar.set_position(cbbox)\n",
    "\n",
    "    if legend:\n",
    "        # Add legend for row color bar\n",
    "        from matplotlib.patches import Patch\n",
    "\n",
    "        lut = dict(zip(clusters, cluster_colors))\n",
    "        handles = [Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "        legend = plt.legend(\n",
    "            handles,\n",
    "            sorted(lut),\n",
    "            title=\"Groups\",\n",
    "            bbox_to_anchor=lgbox,\n",
    "            bbox_transform=plt.gcf().transFigure,\n",
    "            loc=\"upper right\",\n",
    "            ncol=legend_ncol,\n",
    "        )\n",
    "        legend.get_frame().set_facecolor(\"white\")\n",
    "\n",
    "    return ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "#result_path=\"./results/Mice_forfigs/\"\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "            \n",
    "labels=['Cardiomyocyte',\n",
    " 'Chondrocyte',\n",
    " 'Choroid plexus',\n",
    " 'Dorsal Mb neuron',\n",
    " 'Endothelial cell',\n",
    " 'Epithelial cell',\n",
    " 'Erythrocyte',\n",
    " 'Facial fibroblast',\n",
    " 'Fibroblast',\n",
    " 'Fb neuron',\n",
    " 'Fb radial glia cell',\n",
    " 'Ganglion',\n",
    " 'Hepatocyte',\n",
    " 'Immune cell',\n",
    " 'Keratinocyte',\n",
    " 'Limb fibroblast',\n",
    " 'Macrophage',\n",
    " 'Meninges cell',\n",
    " 'Mb/Hb/SpC neuron',\n",
    " 'Myoblast',\n",
    " 'Olfactory epithelial cell',\n",
    " 'Radial glia cell',\n",
    " 'Smooth muscle cell',\n",
    " 'SpC neuron',\n",
    " 'Thalamus neuron']\n",
    "\n",
    "DM_mouse = RMSD_distance_matrix_new(Xs, GeneLists, compare_method='complete', ngene=20, norm_method=None)\n",
    "\n",
    "from cytocraft.plot import plot_RMSD_heatmap\n",
    "plot_RMSD_heatmap_group(DM_mouse, labels, method='ward', cthresh=2, font_scale=1.4, cbbox=(0.06, .83, .04, .11),\n",
    "                lgbox=(0.90, 0.95, .01, .01), ticks=[0,0.5,1,1.5], figsize=(4,4), dpi=300, legend_ncol=2)\n",
    "plt.savefig('mice_clustermap.png', dpi=300, transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.patches import Patch\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "labels=list(Xs.keys())\n",
    "# Draw heatmap\n",
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (8,7)}, font_scale=1)\n",
    "\n",
    "ax = sns.heatmap(\n",
    "    DM_mouse,\n",
    "    xticklabels=labels,\n",
    "    yticklabels=labels,\n",
    "    cmap=\"viridis\",\n",
    "    cbar_kws={\"ticks\": [0, 0.5, 1, 1.5]},\n",
    "    mask=None,\n",
    ")\n",
    "\n",
    "plt.savefig('mice_heatmap.png', dpi=300, transparent=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "#result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "labels=['Cardiomyocyte',\n",
    " '',\n",
    " '',\n",
    " 'Chondrocyte',\n",
    " '',\n",
    " '',\n",
    " '',\n",
    " 'Choroid plexus',\n",
    " '',\n",
    " 'Dorsal Mb neuron',\n",
    " '',\n",
    " '',\n",
    " 'Endothelial cell',\n",
    " '',\n",
    " '',\n",
    " 'Epithelial cell',\n",
    " '',\n",
    " '',\n",
    " '',\n",
    " 'Erythrocyte',\n",
    " '',\n",
    " 'Facial fibroblast',\n",
    " '',\n",
    " '',\n",
    " 'Fibroblast',\n",
    " '',\n",
    " '',\n",
    " '',\n",
    " 'Fb neuron',\n",
    " '',\n",
    " 'Fb radial glia cell',\n",
    " '',\n",
    " '',\n",
    " 'Ganglion',\n",
    " '',\n",
    " '',\n",
    " 'Hepatocyte',\n",
    " '',\n",
    " '',\n",
    " 'Immune cell',\n",
    " '',\n",
    " '',\n",
    " 'Keratinocyte',\n",
    " '',\n",
    " '',\n",
    " '',\n",
    " 'Limb fibroblast',\n",
    " '',\n",
    " '',\n",
    " 'Macrophage',\n",
    " '',\n",
    " '',\n",
    " 'Meninges cell',\n",
    " '',\n",
    " '',\n",
    " 'Mb/Hb/SpC neuron',\n",
    " '',\n",
    " 'Myoblast',\n",
    " '',\n",
    " '',\n",
    " 'Olfactory epithelial cell',\n",
    " '',\n",
    " '',\n",
    " 'Radial glia cell',\n",
    " '',\n",
    " '',\n",
    " '',\n",
    " 'Smooth muscle cell',\n",
    " '',\n",
    " 'SpC neuron',\n",
    " '',\n",
    " '',\n",
    " '',\n",
    " 'Thalamus neuron',\n",
    " '']\n",
    "DM_mouse = RMSD_distance_matrix_new(Xs, GeneLists, compare_method='complete', ngene=50, norm_method=None)\n",
    "from cytocraft.plot import plot_RMSD_heatmap\n",
    "plot_RMSD_heatmap_group(DM_mouse, labels, method='ward', cthresh=4.5, font_scale=1.4, cbbox=(0.06, .83, .04, .11),\n",
    "                lgbox=(0.94, 0.96, .01, .01), ticks=[0,0.5,1,1.5], figsize=(4,4), dpi=300, legend_ncol=2)\n",
    "plt.savefig('mice_clustermap.png', dpi=300, transparent=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 3.housekeeping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "housekeep = pd.read_csv(\"./data/Human_Mouse_Common.csv\",sep=';')\n",
    "key=list(Xs.keys())[0]\n",
    "GeneList = GeneLists[key]\n",
    "intersect_housekeep = np.intersect1d(GeneList, housekeep.Mouse)\n",
    "boolean_arrays_n = np.in1d(GeneList, intersect_housekeep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import *\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "Nspecific = Ncommon = Nhousekeep = 100     \n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        #print(sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "\n",
    "# housekeeping genes and other functional genes\n",
    "housekeep = pd.read_csv(\"./data/Housekeeping_TranscriptsMouse.csv\",sep=';').Genes.unique()\n",
    "housekeep_genes_each_ct = []\n",
    "non_housekeep_genes_each_ct = []\n",
    "samples = list(Xs.keys())\n",
    "\n",
    "for i in range(int(len(samples)/3)):\n",
    "    for j in range(3):\n",
    "        GeneList = GeneLists[samples[(i*3)+j]]\n",
    "        intersect_housekeep = np.intersect1d(GeneList, housekeep)\n",
    "        intersect_functional = np.setdiff1d(GeneList, housekeep)\n",
    "        housekeep_genes_each_ct.append(intersect_housekeep)\n",
    "        non_housekeep_genes_each_ct.append(intersect_functional)\n",
    "# common genes\n",
    "intersected_genes = reduce(np.intersect1d, GeneLists.values())\n",
    "intersect_common = np.setdiff1d(intersected_genes, housekeep)[:Ncommon]\n",
    "\n",
    "\n",
    "# build matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_M_hk = np.zeros((N,N))\n",
    "D_M_sp = np.zeros((N,N))\n",
    "D_M_cm = np.zeros((N,N))\n",
    "D_df = pd.DataFrame(columns=['sample1', 'sample2', 'distance', 'genetype', 'source'])\n",
    "\n",
    "for k in range(10):\n",
    "    for n, key_n in enumerate(keys):\n",
    "        for m, key_m in enumerate(keys[:n+1]):\n",
    "            hk_intersected_values = np.intersect1d(housekeep_genes_each_ct[n], housekeep_genes_each_ct[m])[:Nspecific]\n",
    "            # Nhousekeep =  len(hk_intersected_values)\n",
    "            if Nhousekeep <= Ncommon:\n",
    "                fc_intersected_values = np.random.choice(np.intersect1d(non_housekeep_genes_each_ct[n], non_housekeep_genes_each_ct[m]), size=Nhousekeep, replace = False)\n",
    "                cm_intersected_values = np.random.choice(intersect_common, size=Nhousekeep, replace = False)\n",
    "            else:\n",
    "                fc_intersected_values = np.random.choice(np.intersect1d(non_housekeep_genes_each_ct[n], non_housekeep_genes_each_ct[m]), size=Ncommon, replace = False)\n",
    "                cm_intersected_values = np.random.choice(intersect_common, size=Ncommon, replace = False)\n",
    "            # housekeep\n",
    "            hk_boolean_arrays_n = np.in1d(GeneLists[key_n], hk_intersected_values)\n",
    "            hk_boolean_arrays_m = np.in1d(GeneLists[key_m], hk_intersected_values)\n",
    "            hk_normalized_x_n = Xs[key_n][hk_boolean_arrays_n]\n",
    "            hk_normalized_x_m = Xs[key_m][hk_boolean_arrays_m]\n",
    "            d1 = min(kabsch_numpy(hk_normalized_x_n, hk_normalized_x_m)[2],kabsch_numpy(hk_normalized_x_n, mirror(hk_normalized_x_m))[2])\n",
    "\n",
    "\n",
    "            # specific\n",
    "            fc_boolean_arrays_n = np.in1d(GeneLists[key_n], fc_intersected_values)\n",
    "            fc_boolean_arrays_m = np.in1d(GeneLists[key_m], fc_intersected_values)\n",
    "            fc_normalized_x_n = Xs[key_n][fc_boolean_arrays_n]\n",
    "            fc_normalized_x_m = Xs[key_m][fc_boolean_arrays_m]\n",
    "            d2 = min(kabsch_numpy(fc_normalized_x_n, fc_normalized_x_m)[2],kabsch_numpy(fc_normalized_x_n, mirror(fc_normalized_x_m))[2])\n",
    "\n",
    "            # common\n",
    "            cm_boolean_arrays_n = np.in1d(GeneLists[key_n], cm_intersected_values)\n",
    "            cm_boolean_arrays_m = np.in1d(GeneLists[key_m], cm_intersected_values)\n",
    "            cm_normalized_x_n = Xs[key_n][cm_boolean_arrays_n]\n",
    "            cm_normalized_x_m = Xs[key_m][cm_boolean_arrays_m]\n",
    "            d3 = min(kabsch_numpy(cm_normalized_x_n, cm_normalized_x_m)[2],kabsch_numpy(cm_normalized_x_n, mirror(cm_normalized_x_m))[2])\n",
    "\n",
    "            # define source\n",
    "            D_M_hk[n,m] = D_M_hk[m,n] = d1\n",
    "            D_M_sp[n,m] = D_M_sp[m,n] = d2\n",
    "            D_M_cm[n,m] = D_M_cm[m,n] = d3\n",
    "            if not m == n:\n",
    "                if key_n[:-5] == key_m[:-5]:\n",
    "                    source = 'Between Replicates'\n",
    "                else:\n",
    "                    source = 'Between Celltypes'\n",
    "            else:\n",
    "                continue\n",
    "            # write\n",
    "            s1 = pd.DataFrame([{'sample1':key_n,'sample2':key_m,'distance':d1,'genetype':'Housekeeping Genes', 'source':source}])\n",
    "            s2 = pd.DataFrame([{'sample1':key_n,'sample2':key_m,'distance':d2,'genetype':'Specific Genes', 'source':source}])\n",
    "            s3 = pd.DataFrame([{'sample1':key_n,'sample2':key_m,'distance':d3,'genetype':'Common Genes', 'source':source}])\n",
    "            D_df = D_df.append(s1, ignore_index=True)\n",
    "            D_df = D_df.append(s2, ignore_index=True)\n",
    "            D_df = D_df.append(s3, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statannot import add_stat_annotation\n",
    "D_sub = D_df[D_df['source']=='Between Celltypes']\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(2, 3)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "ax = sns.boxplot(data=D_sub, x='genetype', y='distance', palette=['#FDA06E','#FDCCA8','#9B95C6'], width=0.8, saturation=1, showfliers=False, order=['Specific Genes', 'Housekeeping Genes', 'Common Genes'])\n",
    "boxpairs = [\n",
    "    ('Housekeeping Genes', 'Specific Genes'),\n",
    "    ('Housekeeping Genes', 'Common Genes'),\n",
    "    ('Common Genes', 'Specific Genes'),\n",
    "]\n",
    "add_stat_annotation(ax, data=D_sub, x='genetype', y='distance', box_pairs=boxpairs, loc='outside', test=\"Mann-Whitney\", text_format=\"star\", verbose=2)\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_xticklabels(labels=['Cell Type Specific Gene', 'Housekeeping Gene', 'Ubiquitously Expressed Gene'],rotation=90)\n",
    "ax.set_ylabel(\"RMSD\")\n",
    "ax.legend(title='', loc='best', \n",
    "        fontsize='10', title_fontsize='10', \n",
    "        bbox_to_anchor=(0.8, 1.16), frameon=False\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statannot import add_stat_annotation\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(3, 6)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "ax = sns.boxplot(data=D_df, x='source', y='distance', hue='genetype', palette=['#FDA06E','#FDCCA8','#9B95C6'],\n",
    "                width=0.8, saturation=1, showfliers=False, hue_order=['Housekeeping Genes', 'Common Genes', 'Specific Genes'])\n",
    "boxpairs = [\n",
    "    (('Between Replicates', 'Housekeeping Genes'), ('Between Replicates', 'Specific Genes')),\n",
    "    (('Between Replicates', 'Housekeeping Genes'), ('Between Replicates', 'Common Genes')),\n",
    "    (('Between Replicates', 'Common Genes'), ('Between Replicates', 'Specific Genes')),\n",
    "    (('Between Celltypes', 'Housekeeping Genes'), ('Between Celltypes', 'Specific Genes')),\n",
    "    (('Between Celltypes', 'Housekeeping Genes'), ('Between Celltypes', 'Common Genes')),\n",
    "    (('Between Celltypes', 'Common Genes'), ('Between Celltypes', 'Specific Genes')),\n",
    "]\n",
    "add_stat_annotation(ax, data=D_df, x='source', y='distance', hue='genetype', hue_order=['Housekeeping Genes', 'Common Genes', 'Specific Genes'], box_pairs=boxpairs, loc='inside', test=\"Mann-Whitney\", text_format=\"star\", verbose=2)\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_xticklabels(labels=['Between Replicates','Between Celltypes'],rotation=45)\n",
    "ax.set_ylabel(\"RMSD\")\n",
    "ax.legend(title='', loc='best', \n",
    "        fontsize='10', title_fontsize='10', \n",
    "        bbox_to_anchor=(0.8, 1.16), frameon=False\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(15, 5)}, font_scale=1.1)\n",
    "f,(ax1,ax2,ax3,axcb) = plt.subplots(1,4, gridspec_kw={'width_ratios':[1,1,1,0.05]})\n",
    "plt.subplots_adjust(wspace=0.05)\n",
    "ax1.get_shared_y_axes().join(ax2,ax3)\n",
    "dmax = np.nanmax([np.nanmax(D_M_hk),np.nanmax(D_M_cm),np.nanmax(D_M_sp)])\n",
    "\n",
    "labels = ['', 'Cardiomyocyte', '', '', 'Chondrocyte', '', '', 'Choroid plexus', '', '', 'Dorsal midbrain neuron', '', '', 'Endothelial cell', '', \n",
    "'', 'Epithelial cell', '', '', 'Erythrocyte', '', '', 'Facial fibroblast', '', '', 'Fibroblast', '', '', 'Forebrain neuron', '', \n",
    "'', 'Forebrain radial glia cell', '', '', 'Ganglion', '', '', 'Hepatocyte', '', '', 'Immune cell', '', '', 'Keratinocyte', '', \n",
    "'', 'Limb fibroblast', '', '', 'Macrophage', '', '', 'Meninges cell', '', '', 'Mid-hindbrain and spinal cord neuron', '', '', 'Myoblast', '', \n",
    "'', 'Olfactory epithelial cell', '', '', 'Radial glia cell', '', '', 'Smooth muscle cell', '', '', 'Spinal cord neuron', '', '', 'Thalamus neuron', '']\n",
    "\n",
    "g1 = sns.heatmap(D_M_hk,xticklabels=labels, yticklabels=labels, cmap=\"viridis\", vmax = dmax, cbar=False, ax=ax1)\n",
    "g1.set_ylabel('')\n",
    "g1.set_xlabel('Housekeeping Genes', size=20, weight='bold', color='#FDA06E')\n",
    "g2 = sns.heatmap(D_M_cm,xticklabels=labels, yticklabels=labels, cmap=\"viridis\", vmax = dmax, cbar=False, ax=ax2)\n",
    "g2.set_ylabel('')\n",
    "g2.set_xlabel('Common Genes', size=20, weight='bold', color='#FDCCA8')\n",
    "g2.set_yticks([])\n",
    "g3 = sns.heatmap(D_M_sp,xticklabels=labels, yticklabels=labels, cmap=\"viridis\", vmax = dmax, ax=ax3, cbar_ax=axcb, cbar_kws={'location':'right'})\n",
    "g3.set_ylabel('')\n",
    "g3.set_xlabel('Specific Genes', size=20, weight='bold', color='#9B95C6')\n",
    "g3.set_yticks([])\n",
    "\n",
    "axcb.set_xlabel('RMSD', rotation=0, va=\"bottom\", loc=\"center\", size=11, labelpad=20)\n",
    "# may be needed to rotate the ticklabels correctly:\n",
    "for ax in [g1,g2,g3]:\n",
    "    tl = ax.get_xticklabels()\n",
    "    ax.set_xticklabels(tl, rotation=90)\n",
    "    tly = ax.get_yticklabels()\n",
    "    ax.set_yticklabels(tly, rotation=0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        print(sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = Xs[key_n][boolean_arrays_n]/np.linalg.norm(Xs[key_n][boolean_arrays_n])\n",
    "        normalized_x_m = Xs[key_m][boolean_arrays_m]/np.linalg.norm(Xs[key_m][boolean_arrays_m])\n",
    "        d, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        D[n,m] = D[m,n] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "clustergrid = sns.clustermap(D, xticklabels=list(Xs.keys()), yticklabels=list(Xs.keys()), method='complete')\n",
    "plt.close()\n",
    "\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 2.8\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "\n",
    "# Create a color map for clusters\n",
    "n_clusters = len(np.unique(clusters))\n",
    "color_map = sns.color_palette(\"rainbow\", n_clusters)\n",
    "\n",
    "# Assign colors to clusters\n",
    "cluster_colors = [color_map[c-1] for c in clusters]\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(18, 18)},font_scale=0.5)\n",
    "sns.clustermap(D, xticklabels=list(Xs.keys()), yticklabels=list(Xs.keys()), cmap=\"viridis\", method='complete', row_colors=[cluster_colors], col_colors=[cluster_colors])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Axolotls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "ctypes={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[1].split('_')[0]\n",
    "    if ctype in ctypes.keys():\n",
    "        ctypes[ctype]+=1\n",
    "    else:\n",
    "        ctypes[ctype]=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ctypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files : \n",
    "        print(sample)\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_axo_dev = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in tqdm(enumerate(keys)):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D_axo_dev[n,m] = D_axo_dev[m,n] = min(d1,d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyvis.network import Network\n",
    "from IPython.display import HTML\n",
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "random.seed(997)\n",
    "np.random.seed(997)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_lung.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i, label=br_keys[i])\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_lung[i,:]\n",
    "    min_indices = np.argpartition(arr, 3)[:3]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,int(j),{'weight':float(1/arr[j]*100-60)}))\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "\n",
    "# Draw pyvis\n",
    "net = Network(width=1400, height=1000, notebook=True, cdn_resources='remote')\n",
    "net.from_nx(G)\n",
    "\n",
    "# set node color based on label\n",
    "for node in net.nodes:\n",
    "    node['group'] = node['label'].split(\"\\n\")[1]\n",
    "    node['color'] = color_map_ct[node['group']]\n",
    "\n",
    "# show\n",
    "net.show_buttons()\n",
    "net.set_options(\"\"\"\n",
    "const options = {\n",
    "  \"configure\": {\n",
    "    \"enabled\": true,\n",
    "    \"filter\": [\"nodes\",\"edges\",\"physics\"]\n",
    "  },\n",
    "  \"nodes\": {\n",
    "    \"borderWidth\": 3,\n",
    "    \"opacity\": 1,\n",
    "    \"font\": {\n",
    "      \"size\": 15,\n",
    "      \"strokeWidth\": 5\n",
    "    },\n",
    "    \"size\": 0\n",
    "  },\n",
    "  \"edges\": {\n",
    "    \"color\": {\n",
    "      \"opacity\": 0.7\n",
    "    },\n",
    "    \"selfReferenceSize\": 0,\n",
    "    \"selfReference\": {\n",
    "      \"size\": 0,\n",
    "      \"angle\": 0.7853981633974483\n",
    "    },\n",
    "    \"smooth\": {\n",
    "      \"forceDirection\": \"vertical\"\n",
    "    }\n",
    "  },\n",
    "  \"physics\": {\n",
    "    \"minVelocity\": 0.75\n",
    "  }\n",
    "}\n",
    "\"\"\")\n",
    "\n",
    "net.show('networkplot_Axo_neighbor2.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### VLMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[1].split('_')[0]\n",
    "    sample = chip+'_'+ctype\n",
    "    if 'adata.h5ad' in files and ctype=='VLMC':\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "keys = ['Stage44_VLMC',\n",
    "        'Stage54_VLMC',\n",
    "        'Stage57_VLMC',\n",
    "        'Control_Juv_VLMC',\n",
    "        'Adult_VLMC',\n",
    "        'Meta_VLMC']\n",
    "N = len(keys)\n",
    "print(N)\n",
    "D_VLMC_dev = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D_VLMC_dev[n,m] = D_VLMC_dev[m,n] = min(d1,d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels_dev = ['St.44',\n",
    "        'St.54',\n",
    "        'St.57',\n",
    "        'Juv.',\n",
    "        'Adult',\n",
    "        'Meta.']\n",
    "sns.set(rc={\"figure.dpi\":600,'figure.figsize':(1.8, 1.5)},font_scale=1)\n",
    "f,(ax,axcb) = plt.subplots(1,2, gridspec_kw={'width_ratios':[1,0.1]})\n",
    "g = sns.heatmap(D_VLMC_dev, xticklabels=labels_dev, yticklabels=labels_dev, cmap=\"viridis\", cbar_ax=axcb, ax=ax)\n",
    "g.set(title='Development - VLMC')\n",
    "axcb.set_ylabel('NRMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "keys = [\n",
    "'2DPI_1_VLMC',\n",
    "'2DPI_2_VLMC',\n",
    "'2DPI_3_VLMC',\n",
    "'5DPI_1_VLMC',\n",
    "'5DPI_2_VLMC',\n",
    "'5DPI_3_VLMC',\n",
    "'10DPI_1_VLMC',\n",
    "'10DPI_2_VLMC',\n",
    "'10DPI_3_VLMC',\n",
    "'15DPI_1_VLMC',\n",
    "'15DPI_2_VLMC',\n",
    "'15DPI_3_VLMC',\n",
    "'15DPI_4_VLMC',\n",
    "'20DPI_1_VLMC',\n",
    "'20DPI_2_VLMC',\n",
    "'20DPI_3_VLMC',\n",
    "'30DPI_VLMC',\n",
    "'60DPI_VLMC',\n",
    "'Control_Juv_VLMC',\n",
    "]\n",
    "\n",
    "D_VLMC_reg = RMSD_distance_matrix(Xs, GeneLists, keys=keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels_reg = [\n",
    "'2 DPI-1',\n",
    "'2 DPI-2',\n",
    "'2 DPI-3',\n",
    "'5 DPI-1',\n",
    "'5 DPI-2',\n",
    "'5 DPI-3',\n",
    "'10 DPI-1',\n",
    "'10 DPI-2',\n",
    "'10 DPI-3',\n",
    "'15 DPI-1',\n",
    "'15 DPI-2',\n",
    "'15 DPI-3',\n",
    "'15 DPI-4',\n",
    "'20 DPI-1',\n",
    "'20 DPI-2',\n",
    "'20 DPI-3',\n",
    "'30 DPI',\n",
    "'60 DPI',\n",
    "'Control',\n",
    "]\n",
    "sns.set(rc={\"figure.dpi\":600,'figure.figsize':(4, 3.2)},font_scale=1)\n",
    "f,(ax,axcb) = plt.subplots(1,2, gridspec_kw={'width_ratios':[1,0.1]})\n",
    "g = sns.heatmap(D_VLMC_reg, xticklabels=labels_reg, yticklabels=labels_reg, cmap=\"viridis\", cbar_ax=axcb, ax=ax)\n",
    "g.set(title='Regeneration - VLMC')\n",
    "axcb.set_ylabel('NRMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Graph Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "random.seed(994)\n",
    "np.random.seed(994)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_VLMC_dev.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i)\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "# known edges:\n",
    "#for i,j in [(3,4),(4,5),(5,1),(1,0),(0,2)]:\n",
    "#    edges.append((i,j,{'weight':1/D_VLMC_dev[i,j]}))\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_VLMC_dev[i,:]\n",
    "    min_indices = np.argpartition(arr, 3)[:3]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,j,{'weight':1/arr[j]}))\n",
    "\n",
    "#G.add_weighted_edges_from(edges)\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "labels = nx.get_edge_attributes(G,'weight')\n",
    "final = dict()\n",
    "for key in labels:\n",
    "    final[key] = round(labels[key], 2)\n",
    "\n",
    "widths = nx.get_edge_attributes(G,'weight')\n",
    "widths.update((key, value * 6) for key, value in widths.items())\n",
    "nodelist = G.nodes()\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "\n",
    "#pos = nx.shell_layout(G)\n",
    "pos = nx.spring_layout(G, k=0.5, iterations=100)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p\n",
    "\n",
    "# edges\n",
    "# edge_trace setup\n",
    "edge_trace = []\n",
    "for edge in G.edges():\n",
    "    weight = G.edges[edge]['weight'] * 150 - 95\n",
    "    edge_trace.append(go.Scatter(\n",
    "       x=[],\n",
    "       y=[],\n",
    "       line=dict(width=weight, color='#888'),\n",
    "       hoverinfo='none',\n",
    "       mode='lines'))\n",
    "\n",
    "for trace, edge in zip(edge_trace, G.edges()):\n",
    "   x0, y0 = G.nodes[edge[0]]['pos']\n",
    "   x1, y1 = G.nodes[edge[1]]['pos']\n",
    "   trace['x'] += tuple([x0, x1, None])\n",
    "   trace['y'] += tuple([y0, y1, None])\n",
    "\n",
    "# nodes\n",
    "node_trace = go.Scatter(\n",
    "    x=[],\n",
    "    y=[],\n",
    "    text=[],\n",
    "    mode='markers+text',\n",
    "    hoverinfo='text',\n",
    "    marker=dict(\n",
    "        showscale=True,\n",
    "        colorscale='YlOrRd',\n",
    "        reversescale=False,\n",
    "        color=[],\n",
    "        size=50,\n",
    "        colorbar=dict(\n",
    "            thickness=7,\n",
    "            title='Node Connections',\n",
    "            xanchor='left',\n",
    "            titleside='right',\n",
    "            tickmode='array',\n",
    "            tickvals=[1, 2, 3, 4, 5],\n",
    "        ),\n",
    "        line=dict(width=0)))\n",
    "\n",
    "for node in G.nodes():\n",
    "    x, y = G.nodes[node]['pos']\n",
    "    node_trace['x'] += tuple([x])\n",
    "    node_trace['y'] += tuple([y])\n",
    "\n",
    "node_adjacencies = []\n",
    "for node, adjacencies in enumerate(G.adjacency()):\n",
    "    node_adjacencies.append(len(adjacencies[1]))\n",
    "    \n",
    "node_trace.marker.color = node_adjacencies +[4,5]+[1]\n",
    "node_trace.text = labels_dev\n",
    "\n",
    "fig = go.Figure(data = edge_trace+ [node_trace],\n",
    "            layout=go.Layout(\n",
    "                title=\"\",\n",
    "                titlefont=dict(size=16),\n",
    "                showlegend=False,\n",
    "                hovermode='closest',\n",
    "                margin=dict(b=0, l=0, r=0, t=0),\n",
    "                plot_bgcolor='#fff',\n",
    "                annotations=[dict(\n",
    "                    text=\"Development - VLMC\",\n",
    "                    showarrow=False,\n",
    "                    xref=\"paper\", yref=\"paper\",\n",
    "                    x=0.98, y=0.98, font=dict(size=16))],\n",
    "                xaxis=dict(showgrid=False, zeroline=False,\n",
    "                        showticklabels=False, mirror=True),\n",
    "                yaxis=dict(showgrid=False, zeroline=False,\n",
    "                showticklabels=False, mirror=True)))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=400,\n",
    "    height=300)\n",
    "\n",
    "fig.show()\n",
    "#save a figure of 300dpi, with 1.5 inches, and  height 0.75inches\n",
    "pio.write_image(fig, \"networkplot_dev.png\", width=1.3*300, height=0.9*300, scale=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "\n",
    "seed = 31\n",
    "random.seed(seed)\n",
    "np.random.seed(seed)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_VLMC_reg.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i)\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "# known edges:\n",
    "for i,j in [(0,1),(0,2),(1,2),(3,4),(3,5),(4,5),(6,7),(6,8),(7,8),(9,10),(9,11),(9,12),(10,11),(10,12),(11,12),(13,14),(13,15),(14,15)]:\n",
    "    edges.append((i,j,{'weight':1/D_VLMC_reg[i,j]}))\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_VLMC_reg[i,:]\n",
    "    min_indices = np.argpartition(arr, 4)[:4]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,j,{'weight':1/arr[j]}))\n",
    "\n",
    "#G.add_weighted_edges_from(edges)\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "labels = nx.get_edge_attributes(G,'weight')\n",
    "final = dict()\n",
    "for key in labels:\n",
    "    final[key] = round(labels[key], 2)\n",
    "\n",
    "widths = nx.get_edge_attributes(G,'weight')\n",
    "widths.update((key, value * 6) for key, value in widths.items())\n",
    "nodelist = G.nodes()\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "\n",
    "#pos = nx.shell_layout(G)\n",
    "pos = nx.spring_layout(G, k=0.5, iterations=100)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p\n",
    "\n",
    "# edges\n",
    "# edge_trace setup\n",
    "edge_trace = []\n",
    "for edge in G.edges():\n",
    "    weight = G.edges[edge]['weight'] * 150 - 96\n",
    "    edge_trace.append(go.Scatter(\n",
    "       x=[],\n",
    "       y=[],\n",
    "       line=dict(width=weight, color='#888', shape='linear',simplify=False),\n",
    "       hoverinfo='none',\n",
    "       mode='lines'))\n",
    "\n",
    "for trace, edge in zip(edge_trace, G.edges()):\n",
    "   x0, y0 = G.nodes[edge[0]]['pos']\n",
    "   x1, y1 = G.nodes[edge[1]]['pos']\n",
    "   trace['x'] += tuple([x0, x1, None])\n",
    "   trace['y'] += tuple([y0, y1, None])\n",
    "\n",
    "# nodes\n",
    "node_trace = go.Scatter(\n",
    "    x=[],\n",
    "    y=[],\n",
    "    text=[],\n",
    "    mode='markers+text',\n",
    "    hoverinfo='text',\n",
    "    marker=dict(\n",
    "        showscale=True,\n",
    "        colorscale='YlOrRd',\n",
    "        reversescale=False,\n",
    "        color=[],\n",
    "        size=52,\n",
    "        colorbar=dict(\n",
    "            thickness=10,\n",
    "            title='Node Connections',\n",
    "            xanchor='left',\n",
    "            titleside='right',\n",
    "            tickmode='array',\n",
    "            dtick=1\n",
    "        ),\n",
    "        line=dict(width=0)))\n",
    "\n",
    "for node in G.nodes():\n",
    "    x, y = G.nodes[node]['pos']\n",
    "    node_trace['x'] += tuple([x])\n",
    "    node_trace['y'] += tuple([y])\n",
    "\n",
    "node_adjacencies = []\n",
    "for node, adjacencies in enumerate(G.adjacency()):\n",
    "    node_adjacencies.append(len(adjacencies[1]))\n",
    "    \n",
    "node_trace.marker.color = node_adjacencies\n",
    "node_trace.text = labels_reg\n",
    "\n",
    "fig = go.Figure(data = edge_trace+ [node_trace],\n",
    "            layout=go.Layout(\n",
    "                title=\"\",\n",
    "                titlefont=dict(size=16),\n",
    "                showlegend=False,\n",
    "                hovermode='closest',\n",
    "                margin=dict(b=0, l=0, r=0, t=0),\n",
    "                plot_bgcolor='#fff',\n",
    "                annotations=[dict(\n",
    "                    text=\"Regeneration - VLMC\",\n",
    "                    showarrow=False,\n",
    "                    xref=\"paper\", yref=\"paper\",\n",
    "                    x=0.98, y=0.98, font=dict(size=16))],\n",
    "                xaxis=dict(showgrid=False, zeroline=False,\n",
    "                        showticklabels=False, mirror=True),\n",
    "                yaxis=dict(showgrid=False, zeroline=False,\n",
    "                showticklabels=False, mirror=True)))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=850,\n",
    "    height=650)\n",
    "\n",
    "fig.show()\n",
    "#save a figure of 300dpi, with 1.5 inches, and  height 0.75inches\n",
    "pio.write_image(fig, \"networkplot_reg.png\", width=2.5*300, height=2.0*300, scale=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyvis.network import Network\n",
    "from IPython.display import HTML\n",
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "random.seed(997)\n",
    "np.random.seed(997)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_lung.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i, label=br_keys[i])\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_lung[i,:]\n",
    "    min_indices = np.argpartition(arr, 4)[:4]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,int(j),{'weight':float(1/arr[j]*100-60)}))\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "\n",
    "# Draw pyvis\n",
    "net = Network(width=1400, height=1000, notebook=True, cdn_resources='remote')\n",
    "net.from_nx(G)\n",
    "\n",
    "# set node color based on label\n",
    "for node in net.nodes:\n",
    "    node['group'] = node['label'].split(\"\\n\")[1]\n",
    "    node['color'] = color_map_ct[node['group']]\n",
    "\n",
    "# show\n",
    "net.show_buttons()\n",
    "net.set_options(\"\"\"\n",
    "const options = {\n",
    "  \"configure\": {\n",
    "    \"enabled\": true,\n",
    "    \"filter\": [\"nodes\",\"edges\",\"physics\"]\n",
    "  },\n",
    "  \"nodes\": {\n",
    "    \"borderWidth\": 3,\n",
    "    \"opacity\": 1,\n",
    "    \"font\": {\n",
    "      \"size\": 15,\n",
    "      \"strokeWidth\": 5\n",
    "    },\n",
    "    \"size\": 0\n",
    "  },\n",
    "  \"edges\": {\n",
    "    \"color\": {\n",
    "      \"opacity\": 0.7\n",
    "    },\n",
    "    \"selfReferenceSize\": 0,\n",
    "    \"selfReference\": {\n",
    "      \"size\": 0,\n",
    "      \"angle\": 0.7853981633974483\n",
    "    },\n",
    "    \"smooth\": {\n",
    "      \"forceDirection\": \"vertical\"\n",
    "    }\n",
    "  },\n",
    "  \"physics\": {\n",
    "    \"minVelocity\": 0.75\n",
    "  }\n",
    "}\n",
    "\"\"\")\n",
    "\n",
    "net.show('networkplot_Axo_VLMC_neighbor3.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels_reg = [\n",
    "'2 DPI-1',\n",
    "'2 DPI-2',\n",
    "'2 DPI-3',\n",
    "'5 DPI-1',\n",
    "'5 DPI-2',\n",
    "'5 DPI-3',\n",
    "'10 DPI-1',\n",
    "'10 DPI-2',\n",
    "'10 DPI-3',\n",
    "'15 DPI-1',\n",
    "'15 DPI-2',\n",
    "'15 DPI-3',\n",
    "'15 DPI-4',\n",
    "'20 DPI-1',\n",
    "'20 DPI-2',\n",
    "'20 DPI-3',\n",
    "'30 DPI',\n",
    "'60 DPI',\n",
    "'Control',\n",
    "]\n",
    "color_map_ct =  dict(zip(['2','5','10','15','20','30','60','Control'], get_cmap('RdYlGn',8)))\n",
    "\n",
    "plot_network(D=D_VLMC_reg, labels=labels_reg, cmap=color_map_ct, N_neighbor=3, csep=' ', corder=0, html='networkplot_Axo_VLMC_neighbor3.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### all samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[1].split('_')[0]\n",
    "    sample = chip+'_'+ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "print(N)\n",
    "D = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D[n,m] = D[m,n] = min(d1,d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":600,'figure.figsize':(15, 15)},font_scale=0.16)\n",
    "sns.heatmap(D, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### development"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_dev = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        D_dev[n,m] = D_dev[m,n] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":600,'figure.figsize':(12, 12)},font_scale=0.7)\n",
    "sns.clustermap(D_dev, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis_r\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### regeneration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "#RMs={}\n",
    "#Ws={}\n",
    "GeneLists={}\n",
    "#CellLists={}\n",
    "#ctypes=[]\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isdigit() or 'Control_Juv' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            #RMs[sample] = adata.obsm[\"Rotation\"]\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "            #CellLists[sample] = adata.obs.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_reg = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in tqdm(enumerate(keys)):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        D_reg[n,m] = D_reg[m,n] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":600,'figure.figsize':(12, 12)},font_scale=0.7)\n",
    "sns.clustermap(D_reg, xticklabels=list(Xs.keys()),yticklabels=list(Xs.keys()), cmap=\"viridis_r\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Brain Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FP200000493BR_E4 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Cancer/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'celltype2' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = ctype\n",
    "        if 'adata.h5ad' in files and chip =='FP200000493BR_E4': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_cancer_E4 = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in tqdm(enumerate(keys)):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D_cancer_E4[n,m] = D_cancer_E4[m,n] = min(d1,d2)\n",
    "\n",
    "status=['Cancer','Normal','Normal','Cancer','Cancer','Cancer','Cancer']\n",
    "        \n",
    "\n",
    "status_num={}\n",
    "for i,s in enumerate(np.unique(status)):\n",
    "    status_num[s]=i\n",
    "\n",
    "# Create a color map for status\n",
    "n_status = len(np.unique(status))\n",
    "color_map = sns.color_palette(\"Set1\", n_status)\n",
    "status_colors = [color_map[status_num[s]] for s in status]\n",
    "\n",
    "row_colors = pd.DataFrame({'status':status_colors})\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 1)},font_scale=1)\n",
    "clustergrid = sns.clustermap(pd.DataFrame(D_cancer_E4), xticklabels=keys, yticklabels=keys, method='average', cmap=\"coolwarm_r\", vmin=1.3, row_colors=row_colors)\n",
    "\n",
    "# color bar\n",
    "from matplotlib.patches import Patch\n",
    "lut = dict(zip(status, status_colors))\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend = plt.legend(handles, sorted(lut), title='status', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=1,)\n",
    "legend.get_frame().set_facecolor('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "random.seed(994)\n",
    "np.random.seed(994)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_E4.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i)\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "# known edges:\n",
    "#for i,j in [(3,4),(4,5),(5,1),(1,0),(0,2)]:\n",
    "#    edges.append((i,j,{'weight':1/D_cancer_E4[i,j]}))\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_E4[i,:]\n",
    "    min_indices = np.argpartition(arr, 3)[:3]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,j,{'weight':1/arr[j]}))\n",
    "\n",
    "#G.add_weighted_edges_from(edges)\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "labels = nx.get_edge_attributes(G,'weight')\n",
    "final = dict()\n",
    "for key in labels:\n",
    "    final[key] = round(labels[key], 2)\n",
    "\n",
    "widths = nx.get_edge_attributes(G,'weight')\n",
    "widths.update((key, value * 6) for key, value in widths.items())\n",
    "nodelist = G.nodes()\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "\n",
    "#pos = nx.shell_layout(G)\n",
    "pos = nx.spring_layout(G, k=0.5, iterations=100)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p\n",
    "\n",
    "# edges\n",
    "# edge_trace setup\n",
    "edge_trace = []\n",
    "for edge in G.edges():\n",
    "    weight = G.edges[edge]['weight'] * 150 - 90\n",
    "    edge_trace.append(go.Scatter(\n",
    "       x=[],\n",
    "       y=[],\n",
    "       line=dict(width=weight, color='#888'),\n",
    "       hoverinfo='none',\n",
    "       mode='lines'))\n",
    "\n",
    "for trace, edge in zip(edge_trace, G.edges()):\n",
    "   x0, y0 = G.nodes[edge[0]]['pos']\n",
    "   x1, y1 = G.nodes[edge[1]]['pos']\n",
    "   trace['x'] += tuple([x0, x1, None])\n",
    "   trace['y'] += tuple([y0, y1, None])\n",
    "\n",
    "# nodes\n",
    "node_trace = go.Scatter(\n",
    "    x=[],\n",
    "    y=[],\n",
    "    text=[],\n",
    "    mode='markers+text',\n",
    "    hoverinfo='text',\n",
    "    marker=dict(\n",
    "        showscale=True,\n",
    "        colorscale='YlOrRd',\n",
    "        reversescale=False,\n",
    "        color=[],\n",
    "        size=30,\n",
    "        colorbar=dict(\n",
    "            thickness=7,\n",
    "            title='Node Connections',\n",
    "            xanchor='left',\n",
    "            titleside='right',\n",
    "            tickmode='array',\n",
    "            tickvals=[1, 2, 3, 4, 5],\n",
    "        ),\n",
    "        line=dict(width=0)))\n",
    "\n",
    "for node in G.nodes():\n",
    "    x, y = G.nodes[node]['pos']\n",
    "    node_trace['x'] += tuple([x])\n",
    "    node_trace['y'] += tuple([y])\n",
    "\n",
    "node_adjacencies = []\n",
    "for node, adjacencies in enumerate(G.adjacency()):\n",
    "    node_adjacencies.append(len(adjacencies[1]))\n",
    "    \n",
    "node_trace.marker.color = node_adjacencies #+[4,5]+[1]\n",
    "node_trace.text = keys\n",
    "\n",
    "fig = go.Figure(data = edge_trace + [node_trace],\n",
    "            layout=go.Layout(\n",
    "                title=\"\",\n",
    "                titlefont=dict(size=16),\n",
    "                showlegend=False,\n",
    "                hovermode='closest',\n",
    "                margin=dict(b=0, l=0, r=0, t=0),\n",
    "                plot_bgcolor='#fff',\n",
    "                annotations=[dict(\n",
    "                    text=\"Brain Cancer - FP200000493BR_E4\",\n",
    "                    showarrow=False,\n",
    "                    xref=\"paper\", yref=\"paper\",\n",
    "                    x=0.98, y=0.98, font=dict(size=16))],\n",
    "                xaxis=dict(showgrid=False, zeroline=False,\n",
    "                        showticklabels=False, mirror=True),\n",
    "                yaxis=dict(showgrid=False, zeroline=False,\n",
    "                showticklabels=False, mirror=True)))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=800,\n",
    "    height=600)\n",
    "\n",
    "fig.show()\n",
    "#save figure\n",
    "pio.write_image(fig, \"./figures/networkplot_Cancer_E4.png\", width=3.0*300, height=3.0*300, scale=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FP200000615BR_A4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Cancer/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'celltype2' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = ctype\n",
    "        if 'adata.h5ad' in files and chip == 'FP200000615BR_A4': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_cancer_A4 = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in tqdm(enumerate(keys)):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D_cancer_A4[n,m] = D_cancer_A4[m,n] = min(d1,d2)\n",
    "\n",
    "status=['Normal','Cancer','Normal','Normal','Cancer','Cancer','Cancer','Cancer']\n",
    "        \n",
    "\n",
    "status_num={}\n",
    "for i,s in enumerate(np.unique(status)):\n",
    "    status_num[s]=i\n",
    "\n",
    "# Create a color map for status\n",
    "n_status = len(np.unique(status))\n",
    "color_map = sns.color_palette(\"Set1\", n_status)\n",
    "status_colors = [color_map[status_num[s]] for s in status]\n",
    "\n",
    "row_colors = pd.DataFrame({'status':status_colors})\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 1)},font_scale=1)\n",
    "clustergrid = sns.clustermap(pd.DataFrame(D_cancer_A4), xticklabels=keys, yticklabels=keys, method='average', cmap=\"coolwarm_r\", vmin=1.3, row_colors=row_colors)\n",
    "\n",
    "# color bar\n",
    "from matplotlib.patches import Patch\n",
    "lut = dict(zip(status, status_colors))\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend = plt.legend(handles, sorted(lut), title='status', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=1,)\n",
    "legend.get_frame().set_facecolor('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "random.seed(992)\n",
    "np.random.seed(992)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_A4.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i)\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "# known edges:\n",
    "#for i,j in [(3,4),(4,5),(5,1),(1,0),(0,2)]:\n",
    "#    edges.append((i,j,{'weight':1/D_cancer_A4[i,j]}))\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_A4[i,:]\n",
    "    min_indices = np.argpartition(arr, 3)[:3]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,j,{'weight':1/arr[j]}))\n",
    "\n",
    "#G.add_weighted_edges_from(edges)\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "labels = nx.get_edge_attributes(G,'weight')\n",
    "final = dict()\n",
    "for key in labels:\n",
    "    final[key] = round(labels[key], 2)\n",
    "\n",
    "widths = nx.get_edge_attributes(G,'weight')\n",
    "widths.update((key, value * 6) for key, value in widths.items())\n",
    "nodelist = G.nodes()\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "\n",
    "#pos = nx.shell_layout(G)\n",
    "pos = nx.spring_layout(G, k=0.5, iterations=100)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p\n",
    "\n",
    "# edges\n",
    "# edge_trace setup\n",
    "edge_trace = []\n",
    "for edge in G.edges():\n",
    "    weight = G.edges[edge]['weight'] * 150 - 90\n",
    "    edge_trace.append(go.Scatter(\n",
    "       x=[],\n",
    "       y=[],\n",
    "       line=dict(width=weight, color='#888'),\n",
    "       hoverinfo='none',\n",
    "       mode='lines'))\n",
    "\n",
    "for trace, edge in zip(edge_trace, G.edges()):\n",
    "   x0, y0 = G.nodes[edge[0]]['pos']\n",
    "   x1, y1 = G.nodes[edge[1]]['pos']\n",
    "   trace['x'] += tuple([x0, x1, None])\n",
    "   trace['y'] += tuple([y0, y1, None])\n",
    "\n",
    "# nodes\n",
    "node_trace = go.Scatter(\n",
    "    x=[],\n",
    "    y=[],\n",
    "    text=[],\n",
    "    mode='markers+text',\n",
    "    hoverinfo='text',\n",
    "    marker=dict(\n",
    "        showscale=True,\n",
    "        colorscale='YlOrRd',\n",
    "        reversescale=False,\n",
    "        color=[],\n",
    "        size=30,\n",
    "        colorbar=dict(\n",
    "            thickness=7,\n",
    "            title='Node Connections',\n",
    "            xanchor='left',\n",
    "            titleside='right',\n",
    "            tickmode='array',\n",
    "            tickvals=[1, 2, 3, 4, 5],\n",
    "        ),\n",
    "        line=dict(width=0)))\n",
    "\n",
    "for node in G.nodes():\n",
    "    x, y = G.nodes[node]['pos']\n",
    "    node_trace['x'] += tuple([x])\n",
    "    node_trace['y'] += tuple([y])\n",
    "\n",
    "node_adjacencies = []\n",
    "for node, adjacencies in enumerate(G.adjacency()):\n",
    "    node_adjacencies.append(len(adjacencies[1]))\n",
    "    \n",
    "node_trace.marker.color = node_adjacencies #+[4,5]+[1]\n",
    "node_trace.text = keys\n",
    "\n",
    "fig = go.Figure(data = edge_trace+ [node_trace],\n",
    "            layout=go.Layout(\n",
    "                title=\"\",\n",
    "                titlefont=dict(size=16),\n",
    "                showlegend=False,\n",
    "                hovermode='closest',\n",
    "                margin=dict(b=0, l=0, r=0, t=0),\n",
    "                plot_bgcolor='#fff',\n",
    "                annotations=[dict(\n",
    "                    text=\"Brain Cancer - FP200000615BR_A4\",\n",
    "                    showarrow=False,\n",
    "                    xref=\"paper\", yref=\"paper\",\n",
    "                    x=0.98, y=0.98, font=dict(size=16))],\n",
    "                xaxis=dict(showgrid=False, zeroline=False,\n",
    "                        showticklabels=False, mirror=True),\n",
    "                yaxis=dict(showgrid=False, zeroline=False,\n",
    "                showticklabels=False, mirror=True)))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=800,\n",
    "    height=600)\n",
    "\n",
    "fig.show()\n",
    "#save figure\n",
    "pio.write_image(fig, \"./figures/networkplot_Cancer_A4.png\", width=3.0*300, height=3.0*300, scale=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FP200000615BR_F3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Cancer/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'celltype2' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = ctype\n",
    "        if 'adata.h5ad' in files and chip == 'FP200000615BR_F3': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_cancer_F3 = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in tqdm(enumerate(keys)):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D_cancer_F3[n,m] = D_cancer_F3[m,n] = min(d1,d2)\n",
    "\n",
    "status=['Cancer','Normal','Normal','Cancer','Cancer','Cancer','Cancer']\n",
    "        \n",
    "\n",
    "status_num={}\n",
    "for i,s in enumerate(np.unique(status)):\n",
    "    status_num[s]=i\n",
    "\n",
    "# Create a color map for status\n",
    "n_status = len(np.unique(status))\n",
    "color_map = sns.color_palette(\"Set1\", n_status)\n",
    "status_colors = [color_map[status_num[s]] for s in status]\n",
    "\n",
    "row_colors = pd.DataFrame({'status':status_colors})\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 1)},font_scale=1)\n",
    "clustergrid = sns.clustermap(pd.DataFrame(D_cancer_F3), xticklabels=keys, yticklabels=keys, method='average', cmap=\"coolwarm_r\", vmin=1.3, row_colors=row_colors)\n",
    "\n",
    "# color bar\n",
    "from matplotlib.patches import Patch\n",
    "lut = dict(zip(status, status_colors))\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend = plt.legend(handles, sorted(lut), title='status', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=1,)\n",
    "legend.get_frame().set_facecolor('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "random.seed(994)\n",
    "np.random.seed(994)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_F3.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i)\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "# known edges:\n",
    "#for i,j in [(3,4),(4,5),(5,1),(1,0),(0,2)]:\n",
    "#    edges.append((i,j,{'weight':1/D_cancer_F3[i,j]}))\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_F3[i,:]\n",
    "    min_indices = np.argpartition(arr, 3)[:3]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,j,{'weight':1/arr[j]}))\n",
    "\n",
    "#G.add_weighted_edges_from(edges)\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "labels = nx.get_edge_attributes(G,'weight')\n",
    "final = dict()\n",
    "for key in labels:\n",
    "    final[key] = round(labels[key], 2)\n",
    "\n",
    "widths = nx.get_edge_attributes(G,'weight')\n",
    "widths.update((key, value * 6) for key, value in widths.items())\n",
    "nodelist = G.nodes()\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "\n",
    "#pos = nx.shell_layout(G)\n",
    "pos = nx.spring_layout(G, k=0.5, iterations=100)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p\n",
    "\n",
    "# edges\n",
    "# edge_trace setup\n",
    "edge_trace = []\n",
    "for edge in G.edges():\n",
    "    weight = G.edges[edge]['weight'] * 150 - 90\n",
    "    edge_trace.append(go.Scatter(\n",
    "       x=[],\n",
    "       y=[],\n",
    "       line=dict(width=weight, color='#888'),\n",
    "       hoverinfo='none',\n",
    "       mode='lines'))\n",
    "\n",
    "for trace, edge in zip(edge_trace, G.edges()):\n",
    "   x0, y0 = G.nodes[edge[0]]['pos']\n",
    "   x1, y1 = G.nodes[edge[1]]['pos']\n",
    "   trace['x'] += tuple([x0, x1, None])\n",
    "   trace['y'] += tuple([y0, y1, None])\n",
    "\n",
    "# nodes\n",
    "node_trace = go.Scatter(\n",
    "    x=[],\n",
    "    y=[],\n",
    "    text=[],\n",
    "    mode='markers+text',\n",
    "    hoverinfo='text',\n",
    "    marker=dict(\n",
    "        showscale=True,\n",
    "        colorscale='YlOrRd',\n",
    "        reversescale=False,\n",
    "        color=[],\n",
    "        size=30,\n",
    "        colorbar=dict(\n",
    "            thickness=7,\n",
    "            title='Node Connections',\n",
    "            xanchor='left',\n",
    "            titleside='right',\n",
    "            tickmode='array',\n",
    "            tickvals=[1, 2, 3, 4, 5],\n",
    "        ),\n",
    "        line=dict(width=0)))\n",
    "\n",
    "for node in G.nodes():\n",
    "    x, y = G.nodes[node]['pos']\n",
    "    node_trace['x'] += tuple([x])\n",
    "    node_trace['y'] += tuple([y])\n",
    "\n",
    "node_adjacencies = []\n",
    "for node, adjacencies in enumerate(G.adjacency()):\n",
    "    node_adjacencies.append(len(adjacencies[1]))\n",
    "    \n",
    "node_trace.marker.color = node_adjacencies #+[4,5]+[1]\n",
    "node_trace.text = keys\n",
    "\n",
    "fig = go.Figure(data = edge_trace+ [node_trace],\n",
    "            layout=go.Layout(\n",
    "                title=\"\",\n",
    "                titlefont=dict(size=16),\n",
    "                showlegend=False,\n",
    "                hovermode='closest',\n",
    "                margin=dict(b=0, l=0, r=0, t=0),\n",
    "                plot_bgcolor='#fff',\n",
    "                annotations=[dict(\n",
    "                    text=\"Brain Cancer - FP200000615BR_F3\",\n",
    "                    showarrow=False,\n",
    "                    xref=\"paper\", yref=\"paper\",\n",
    "                    x=0.98, y=0.98, font=dict(size=16))],\n",
    "                xaxis=dict(showgrid=False, zeroline=False,\n",
    "                        showticklabels=False, mirror=True),\n",
    "                yaxis=dict(showgrid=False, zeroline=False,\n",
    "                showticklabels=False, mirror=True)))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=800,\n",
    "    height=600)\n",
    "\n",
    "fig.show()\n",
    "#save figure\n",
    "pio.write_image(fig, \"./figures/networkplot_Cancer_F3.png\", width=3.0*300, height=3.0*300, scale=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cytocraft.plot import *\n",
    "def plot_RMSD_heatmap(\n",
    "    D,\n",
    "    labels,\n",
    "    cmap=\"viridis\",\n",
    "    method=\"average\",\n",
    "    cthresh=4.8,\n",
    "    font_scale=1,\n",
    "    cbbox=(0.07, 0.84, 0.04, 0.12),\n",
    "    lgbox=(0.86, 0.96, 0.01, 0.01),\n",
    "    figsize=(8,8),\n",
    "    dpi=300,\n",
    "):\n",
    "    clustergrid = sns.clustermap(\n",
    "        D, xticklabels=labels, yticklabels=labels, method=method\n",
    "    )\n",
    "    plt.close()\n",
    "\n",
    "    # Get clusters with threshold\n",
    "    linkage = clustergrid.dendrogram_row.linkage\n",
    "    threshold = cthresh\n",
    "    clusters = fcluster(linkage, threshold, criterion=\"distance\")\n",
    "\n",
    "    # Create a color map for clusters\n",
    "    n_clusters = len(np.unique(clusters))\n",
    "    color_map = sns.color_palette(\"rainbow\", n_clusters)\n",
    "\n",
    "    # Assign colors to clusters\n",
    "    cluster_colors = [color_map[c - 1] for c in clusters]\n",
    "\n",
    "    # Draw plot\n",
    "    sns.set(rc={\"figure.dpi\": dpi, \"figure.figsize\": figsize}, font_scale=font_scale)\n",
    "\n",
    "    ax = sns.clustermap(\n",
    "        D,\n",
    "        xticklabels=labels,\n",
    "        yticklabels=labels,\n",
    "        cmap=cmap,\n",
    "        method=method,\n",
    "        row_colors=[cluster_colors],\n",
    "        col_colors=[cluster_colors],\n",
    "        cbar_kws={\"ticks\": [0, 0.5, 1, 1.5]},\n",
    "    )\n",
    "    ax.ax_cbar.set_title(\"NRMSD\")\n",
    "    ax.ax_cbar.set_position(cbbox)\n",
    "\n",
    "    # Add legend for row color bar\n",
    "    from matplotlib.patches import Patch\n",
    "\n",
    "    lut = dict(zip(clusters, cluster_colors))\n",
    "    handles = [Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "    legend = plt.legend(\n",
    "        handles,\n",
    "        sorted(lut),\n",
    "        title=\"Clusters\",\n",
    "        bbox_to_anchor=lgbox,\n",
    "        bbox_transform=plt.gcf().transFigure,\n",
    "        loc=\"upper right\",\n",
    "        ncol=2,\n",
    "    )\n",
    "    legend.get_frame().set_facecolor(\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files: \n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "    else:\n",
    "        print(sample)\n",
    "\n",
    "D_cancer_lung = RMSD_distance_matrix(Xs, GeneLists, keys=list(Xs.keys()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_RMSD_heatmap(D_cancer_lung, labels=list(Xs.keys()), cmap='viridis', method='ward', cthresh = 4.8, dpi=300, figsize=(10,10), font_scale=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=['Lung12\\nB-cell',\n",
    "'Lung12\\nNK',\n",
    "'Lung12\\nT CD4 memory',\n",
    "'Lung12\\nT CD4 naive',\n",
    "'Lung12\\nT CD8 naive',\n",
    "'Lung12\\nendothelial',\n",
    "'Lung12\\nepithelial',\n",
    "'Lung12\\nfibroblast',\n",
    "'Lung12\\nmDC',\n",
    "'Lung12\\nmacrophage',\n",
    "'Lung12\\nmast',\n",
    "'Lung12\\nmonocyte',\n",
    "'Lung12\\nneutrophil',\n",
    "'Lung12\\npDC',\n",
    "'Lung12\\nplasmablast',\n",
    "'Lung12\\ntumor 12',\n",
    "'Lung12\\ntumor 5',\n",
    "'Lung13\\nB-cell',\n",
    "'Lung13\\nNK',\n",
    "'Lung13\\nT CD4 memory',\n",
    "'Lung13\\nT CD4 naive',\n",
    "'Lung13\\nT CD8 memory',\n",
    "'Lung13\\nT CD8 naive',\n",
    "'Lung13\\nTreg',\n",
    "'Lung13\\nendothelial',\n",
    "'Lung13\\nepithelial',\n",
    "'Lung13\\nfibroblast',\n",
    "'Lung13\\nmDC',\n",
    "'Lung13\\nmacrophage',\n",
    "'Lung13\\nmast',\n",
    "'Lung13\\nmonocyte',\n",
    "'Lung13\\nneutrophil',\n",
    "'Lung13\\npDC',\n",
    "'Lung13\\nplasmablast',\n",
    "'Lung13\\ntumor 13',\n",
    "'Lung13\\ntumor 6',\n",
    "'Lung13\\ntumor 9',\n",
    "'Lung5 Rep1\\nB-cell',\n",
    "'Lung5 Rep1\\nNK',\n",
    "'Lung5 Rep1\\nT CD4 memory',\n",
    "'Lung5 Rep1\\nT CD4 naive',\n",
    "'Lung5 Rep1\\nT CD8 memory',\n",
    "'Lung5 Rep1\\nT CD8 naive',\n",
    "'Lung5 Rep1\\nTreg',\n",
    "'Lung5 Rep1\\nendothelial',\n",
    "'Lung5 Rep1\\nepithelial',\n",
    "'Lung5 Rep1\\nfibroblast',\n",
    "'Lung5 Rep1\\nmDC',\n",
    "'Lung5 Rep1\\nmacrophage',\n",
    "'Lung5 Rep1\\nmast',\n",
    "'Lung5 Rep1\\nmonocyte',\n",
    "'Lung5 Rep1\\nneutrophil',\n",
    "'Lung5 Rep1\\npDC',\n",
    "'Lung5 Rep1\\nplasmablast',\n",
    "'Lung5 Rep1\\ntumor 12',\n",
    "'Lung5 Rep1\\ntumor 5',\n",
    "'Lung5 Rep2\\nB-cell',\n",
    "'Lung5 Rep2\\nNK',\n",
    "'Lung5 Rep2\\nT CD4 memory',\n",
    "'Lung5 Rep2\\nT CD4 naive',\n",
    "'Lung5 Rep2\\nT CD8 memory',\n",
    "'Lung5 Rep2\\nT CD8 naive',\n",
    "'Lung5 Rep2\\nTreg',\n",
    "'Lung5 Rep2\\nendothelial',\n",
    "'Lung5 Rep2\\nepithelial',\n",
    "'Lung5 Rep2\\nfibroblast',\n",
    "'Lung5 Rep2\\nmDC',\n",
    "'Lung5 Rep2\\nmacrophage',\n",
    "'Lung5 Rep2\\nmast',\n",
    "'Lung5 Rep2\\nmonocyte',\n",
    "'Lung5 Rep2\\nneutrophil',\n",
    "'Lung5 Rep2\\npDC',\n",
    "'Lung5 Rep2\\nplasmablast',\n",
    "'Lung5 Rep2\\ntumor 12',\n",
    "'Lung5 Rep2\\ntumor 5',\n",
    "'Lung5 Rep3\\nB-cell',\n",
    "'Lung5 Rep3\\nNK',\n",
    "'Lung5 Rep3\\nT CD4 memory',\n",
    "'Lung5 Rep3\\nT CD4 naive',\n",
    "'Lung5 Rep3\\nT CD8 memory',\n",
    "'Lung5 Rep3\\nT CD8 naive',\n",
    "'Lung5 Rep3\\nTreg',\n",
    "'Lung5 Rep3\\nendothelial',\n",
    "'Lung5 Rep3\\nepithelial',\n",
    "'Lung5 Rep3\\nfibroblast',\n",
    "'Lung5 Rep3\\nmDC',\n",
    "'Lung5 Rep3\\nmacrophage',\n",
    "'Lung5 Rep3\\nmast',\n",
    "'Lung5 Rep3\\nmonocyte',\n",
    "'Lung5 Rep3\\nneutrophil',\n",
    "'Lung5 Rep3\\npDC',\n",
    "'Lung5 Rep3\\nplasmablast',\n",
    "'Lung5 Rep3\\ntumor 12',\n",
    "'Lung5 Rep3\\ntumor 13',\n",
    "'Lung5 Rep3\\ntumor 5',\n",
    "'Lung6\\nB-cell',\n",
    "'Lung6\\nNK',\n",
    "'Lung6\\nT CD4 memory',\n",
    "'Lung6\\nT CD4 naive',\n",
    "'Lung6\\nT CD8 memory',\n",
    "'Lung6\\nT CD8 naive',\n",
    "'Lung6\\nTreg',\n",
    "'Lung6\\nendothelial',\n",
    "'Lung6\\nepithelial',\n",
    "'Lung6\\nfibroblast',\n",
    "'Lung6\\nmDC',\n",
    "'Lung6\\nmacrophage',\n",
    "'Lung6\\nmast',\n",
    "'Lung6\\nmonocyte',\n",
    "'Lung6\\nneutrophil',\n",
    "'Lung6\\npDC',\n",
    "'Lung6\\nplasmablast',\n",
    "'Lung6\\ntumor 12',\n",
    "'Lung6\\ntumor 13',\n",
    "'Lung6\\ntumor 5',\n",
    "'Lung6\\ntumor 6',\n",
    "'Lung6\\ntumor 9',\n",
    "'Lung9 Rep1\\nB-cell',\n",
    "'Lung9 Rep1\\nNK',\n",
    "'Lung9 Rep1\\nT CD4 memory',\n",
    "'Lung9 Rep1\\nT CD4 naive',\n",
    "'Lung9 Rep1\\nT CD8 memory',\n",
    "'Lung9 Rep1\\nT CD8 naive',\n",
    "'Lung9 Rep1\\nTreg',\n",
    "'Lung9 Rep1\\nendothelial',\n",
    "'Lung9 Rep1\\nepithelial',\n",
    "'Lung9 Rep1\\nfibroblast',\n",
    "'Lung9 Rep1\\nmDC',\n",
    "'Lung9 Rep1\\nmacrophage',\n",
    "'Lung9 Rep1\\nmast',\n",
    "'Lung9 Rep1\\nmonocyte',\n",
    "'Lung9 Rep1\\nneutrophil',\n",
    "'Lung9 Rep1\\npDC',\n",
    "'Lung9 Rep1\\nplasmablast',\n",
    "'Lung9 Rep1\\ntumor 12',\n",
    "'Lung9 Rep1\\ntumor 13',\n",
    "'Lung9 Rep1\\ntumor 5',\n",
    "'Lung9 Rep1\\ntumor 6',\n",
    "'Lung9 Rep1\\ntumor 9',\n",
    "'Lung9 Rep2\\nB-cell',\n",
    "'Lung9 Rep2\\nNK',\n",
    "'Lung9 Rep2\\nT CD4 memory',\n",
    "'Lung9 Rep2\\nT CD4 naive',\n",
    "'Lung9 Rep2\\nT CD8 memory',\n",
    "'Lung9 Rep2\\nT CD8 naive',\n",
    "'Lung9 Rep2\\nTreg',\n",
    "'Lung9 Rep2\\nendothelial',\n",
    "'Lung9 Rep2\\nepithelial',\n",
    "'Lung9 Rep2\\nfibroblast',\n",
    "'Lung9 Rep2\\nmDC',\n",
    "'Lung9 Rep2\\nmacrophage',\n",
    "'Lung9 Rep2\\nmast',\n",
    "'Lung9 Rep2\\nmonocyte',\n",
    "'Lung9 Rep2\\nneutrophil',\n",
    "'Lung9 Rep2\\npDC',\n",
    "'Lung9 Rep2\\nplasmablast',\n",
    "'Lung9 Rep2\\ntumor 12',\n",
    "'Lung9 Rep2\\ntumor 13',\n",
    "'Lung9 Rep2\\ntumor 5',\n",
    "'Lung9 Rep2\\ntumor 6',\n",
    "'Lung9 Rep2\\ntumor 9']\n",
    "\n",
    "color_map = {\n",
    "'B-cell': '#ef9708',\n",
    "'plasmablast': '#f0b98d',\n",
    "'NK': '#0fcfc0',\n",
    "'T CD4 naive': '#e07b91',\n",
    "'T CD4 memory': '#d33f6a',\n",
    "'T CD8 naive': '#bb7784',\n",
    "'T CD8 memory': '#8e063b',\n",
    "'Treg': '#e6afb9',\n",
    "'endothelial': '#bec1d4',\n",
    "'epithelial': '#b5bbe3',\n",
    "'fibroblast': '#7d87b9',\n",
    "'macrophage': '#9cded6',\n",
    "'mast': '#11c638',\n",
    "'monocyte': '#8dd593',\n",
    "'neutrophil': '#c6dec7',\n",
    "'pDC': '#ead3c6',\n",
    "'mDC': '#d6bcc0',\n",
    "'tumor 5': '#0000A6',\n",
    "'tumor 6': '#1CE6FF',\n",
    "'tumor 9': '#3B5DFF',\n",
    "'tumor 12': '#023fa5',\n",
    "'tumor 13': '#4a6fe3',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_network(D_cancer_lung, labels=labels, cmap = color_map, csep='\\n', corder=1, N_neighbor=1, html='Lung_all_neighbor1.html', width=1400, height=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyvis.network import Network\n",
    "from IPython.display import HTML\n",
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_lung.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i, label=labels[i])\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_lung[i,:]\n",
    "    min_indices = np.argpartition(arr, int(2))[:int(2)]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,int(j),{'weight':float(1/arr[j]*100-60)}))\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "\n",
    "# Draw pyvis\n",
    "net = Network(width=2000, height=1000, notebook=True, cdn_resources='remote')\n",
    "net.from_nx(G)\n",
    "\n",
    "# set node color based on label\n",
    "for node in net.nodes:\n",
    "    node['group'] = node['label'].split('\\n')[1]\n",
    "    node['color'] = color_map[node['group']]\n",
    "\n",
    "# show\n",
    "net.show_buttons()\n",
    "net.set_options(\"\"\"\n",
    "    const options = {\n",
    "    \"configure\": {\n",
    "        \"enabled\": true,\n",
    "        \"filter\": [\"nodes\",\"edges\",\"physics\"]\n",
    "    },\n",
    "    \"nodes\": {\n",
    "        \"borderWidth\": 3,\n",
    "        \"opacity\": 1,\n",
    "        \"font\": {\n",
    "        \"size\": 15,\n",
    "        \"strokeWidth\": 5\n",
    "        },\n",
    "        \"size\": 0\n",
    "    },\n",
    "    \"edges\": {\n",
    "        \"color\": {\n",
    "        \"opacity\": 0.7\n",
    "        },\n",
    "        \"selfReferenceSize\": 0,\n",
    "        \"selfReference\": {\n",
    "        \"size\": 0,\n",
    "        \"angle\": 0.7853981633974483\n",
    "        },\n",
    "        \"smooth\": {\n",
    "        \"forceDirection\": \"vertical\"\n",
    "        }\n",
    "    },\n",
    "    \"physics\": {\n",
    "        \"minVelocity\": 0.75\n",
    "    }\n",
    "    }\n",
    "\"\"\")\n",
    "\n",
    "net.show('Lung_all_neighbor2.html')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Lung5 replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung_old/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and 'Lung5' in chip: \n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        if len(adata.uns['X'].index)>150:\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "        else:\n",
    "            print(sample)\n",
    "\n",
    "\n",
    "# Build distance matrix\n",
    "D_cancer_lung5 = RMSD_distance_matrix(Xs, GeneLists, keys=list(Xs.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys=[\n",
    " 'Rep1 B-cell',\n",
    " 'Rep1 NK',\n",
    " 'Rep1 T CD4 memory',\n",
    " 'Rep1 T CD4 naive',\n",
    " 'Rep1 T CD8 memory',\n",
    " 'Rep1 T CD8 naive',\n",
    " 'Rep1 Treg',\n",
    " 'Rep1 endothelial',\n",
    " 'Rep1 epithelial',\n",
    " 'Rep1 fibroblast',\n",
    " 'Rep1 mDC',\n",
    " 'Rep1 macrophage',\n",
    " 'Rep1 mast',\n",
    " 'Rep1 monocyte',\n",
    " 'Rep1 neutrophil',\n",
    " 'Rep1 pDC',\n",
    " 'Rep1 plasmablast',\n",
    " 'Rep1 tumor 12',\n",
    " 'Rep1 tumor 5',\n",
    " 'Rep2 B-cell',\n",
    " 'Rep2 NK',\n",
    " 'Rep2 T CD4 memory',\n",
    " 'Rep2 T CD4 naive',\n",
    " 'Rep2 T CD8 memory',\n",
    " 'Rep2 T CD8 naive',\n",
    " 'Rep2 Treg',\n",
    " 'Rep2 endothelial',\n",
    " 'Rep2 epithelial',\n",
    " 'Rep2 fibroblast',\n",
    " 'Rep2 mDC',\n",
    " 'Rep2 macrophage',\n",
    " 'Rep2 mast',\n",
    " 'Rep2 monocyte',\n",
    " 'Rep2 neutrophil',\n",
    " 'Rep2 pDC',\n",
    " 'Rep2 plasmablast',\n",
    " 'Rep2 tumor 12',\n",
    " 'Rep2 tumor 5',\n",
    " 'Rep3 B-cell',\n",
    " 'Rep3 NK',\n",
    " 'Rep3 T CD4 memory',\n",
    " 'Rep3 T CD4 naive',\n",
    " 'Rep3 T CD8 memory',\n",
    " 'Rep3 T CD8 naive',\n",
    " 'Rep3 Treg',\n",
    " 'Rep3 endothelial',\n",
    " 'Rep3 epithelial',\n",
    " 'Rep3 fibroblast',\n",
    " 'Rep3 mDC',\n",
    " 'Rep3 macrophage',\n",
    " 'Rep3 mast',\n",
    " 'Rep3 monocyte',\n",
    " 'Rep3 neutrophil',\n",
    " 'Rep3 pDC',\n",
    " 'Rep3 plasmablast',\n",
    " 'Rep3 tumor 5']\n",
    "br_keys=[\n",
    " 'Rep1\\nB-cell',\n",
    " 'Rep1\\nNK',\n",
    " 'Rep1\\nT CD4 memory',\n",
    " 'Rep1\\nT CD4 naive',\n",
    " 'Rep1\\nT CD8 memory',\n",
    " 'Rep1\\nT CD8 naive',\n",
    " 'Rep1\\nTreg',\n",
    " 'Rep1\\nendothelial',\n",
    " 'Rep1\\nepithelial',\n",
    " 'Rep1\\nfibroblast',\n",
    " 'Rep1\\nmDC',\n",
    " 'Rep1\\nmacrophage',\n",
    " 'Rep1\\nmast',\n",
    " 'Rep1\\nmonocyte',\n",
    " 'Rep1\\nneutrophil',\n",
    " 'Rep1\\npDC',\n",
    " 'Rep1\\nplasmablast',\n",
    " 'Rep1\\ntumor 12',\n",
    " 'Rep1\\ntumor 5',\n",
    " 'Rep2\\nB-cell',\n",
    " 'Rep2\\nNK',\n",
    " 'Rep2\\nT CD4 memory',\n",
    " 'Rep2\\nT CD4 naive',\n",
    " 'Rep2\\nT CD8 memory',\n",
    " 'Rep2\\nT CD8 naive',\n",
    " 'Rep2\\nTreg',\n",
    " 'Rep2\\nendothelial',\n",
    " 'Rep2\\nepithelial',\n",
    " 'Rep2\\nfibroblast',\n",
    " 'Rep2\\nmDC',\n",
    " 'Rep2\\nmacrophage',\n",
    " 'Rep2\\nmast',\n",
    " 'Rep2\\nmonocyte',\n",
    " 'Rep2\\nneutrophil',\n",
    " 'Rep2\\npDC',\n",
    " 'Rep2\\nplasmablast',\n",
    " 'Rep2\\ntumor 12',\n",
    " 'Rep2\\ntumor 5',\n",
    " 'Rep3\\nB-cell',\n",
    " 'Rep3\\nNK',\n",
    " 'Rep3\\nT CD4 memory',\n",
    " 'Rep3\\nT CD4 naive',\n",
    " 'Rep3\\nT CD8 memory',\n",
    " 'Rep3\\nT CD8 naive',\n",
    " 'Rep3\\nTreg',\n",
    " 'Rep3\\nendothelial',\n",
    " 'Rep3\\nepithelial',\n",
    " 'Rep3\\nfibroblast',\n",
    " 'Rep3\\nmDC',\n",
    " 'Rep3\\nmacrophage',\n",
    " 'Rep3\\nmast',\n",
    " 'Rep3\\nmonocyte',\n",
    " 'Rep3\\nneutrophil',\n",
    " 'Rep3\\npDC',\n",
    " 'Rep3\\nplasmablast',\n",
    " 'Rep3\\ntumor 5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "clustergrid = sns.clustermap(D_cancer_lung5, xticklabels=keys, yticklabels=keys, method='complete')\n",
    "plt.close()\n",
    "\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 2.4\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "\n",
    "# Create a color map for clusters\n",
    "n_clusters = len(np.unique(clusters))\n",
    "color_map = sns.color_palette(\"rainbow\", n_clusters)\n",
    "\n",
    "# Assign colors to clusters\n",
    "cluster_colors = [color_map[c-1] for c in clusters]\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(4, 4)},font_scale=0.75)\n",
    "ax = sns.clustermap(\n",
    "                pd.DataFrame(D_cancer_lung5), xticklabels=keys, yticklabels=keys, \n",
    "                method='complete', cmap=\"viridis\", \n",
    "                #vmin=0.3, \n",
    "                #row_colors=row_colors,\n",
    "                row_colors=[cluster_colors], col_colors=[cluster_colors], cbar_kws={'ticks': [0,0.5,1,1.5]}\n",
    "                )\n",
    "ax.ax_cbar.set_title(\"NRMSD\")\n",
    "ax.ax_cbar.set_position((.10, .83, .04, .12))\n",
    "\n",
    "# color bar\n",
    "from matplotlib.patches import Patch\n",
    "lut = dict(zip(clusters, cluster_colors))\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend=plt.legend(handles, sorted(lut), title='Clusters', bbox_to_anchor=(0.98, 0.95, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=2,)\n",
    "legend.get_frame().set_facecolor('white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cts = []\n",
    "for c in br_keys:\n",
    "    cts.append(c.split('\\n')[1])\n",
    "cts_unique=np.unique(cts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map_ct =  {\n",
    "    'B-cell': '#ef9708',\n",
    "    'plasmablast': '#f0b98d',\n",
    "    'NK': '#0fcfc0',\n",
    "    'T CD4 naive': '#e07b91',\n",
    "    'T CD4 memory': '#d33f6a',\n",
    "    'T CD8 naive': '#bb7784',\n",
    "    'T CD8 memory': '#8e063b',\n",
    "    'Treg': '#e6afb9',\n",
    "    'endothelial': '#bec1d4',\n",
    "    'epithelial': '#b5bbe3',\n",
    "    'fibroblast': '#7d87b9',\n",
    "    'macrophage': '#9cded6',\n",
    "    'mast': '#11c638',\n",
    "    'monocyte': '#8dd593',\n",
    "    'neutrophil': '#c6dec7',\n",
    "    'pDC': '#ead3c6',\n",
    "    'mDC': '#d6bcc0',\n",
    "    'tumor 5': '#0000A6',\n",
    "    'tumor 12': '#023fa5',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyvis.network import Network\n",
    "from IPython.display import HTML\n",
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido\n",
    "random.seed(997)\n",
    "np.random.seed(997)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_cancer_lung5.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i, label=br_keys[i])\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_cancer_lung5[i,:]\n",
    "    min_indices = np.argpartition(arr, 2)[:2]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,int(j),{'weight':float(1/arr[j]*100-60)}))\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "\n",
    "# Draw pyvis\n",
    "net = Network(width=1400, height=1000, notebook=True, cdn_resources='remote')\n",
    "net.from_nx(G)\n",
    "\n",
    "# set node color based on label\n",
    "for node in net.nodes:\n",
    "    node['group'] = node['label'].split(\"\\n\")[1]\n",
    "    node['color'] = color_map_ct[node['group']]\n",
    "\n",
    "# show\n",
    "net.show_buttons()\n",
    "net.set_options(\"\"\"\n",
    "const options = {\n",
    "  \"configure\": {\n",
    "    \"enabled\": true,\n",
    "    \"filter\": [\"nodes\",\"edges\",\"physics\"]\n",
    "  },\n",
    "  \"nodes\": {\n",
    "    \"borderWidth\": 3,\n",
    "    \"opacity\": 1,\n",
    "    \"font\": {\n",
    "      \"size\": 15,\n",
    "      \"strokeWidth\": 5\n",
    "    },\n",
    "    \"size\": 0\n",
    "  },\n",
    "  \"edges\": {\n",
    "    \"color\": {\n",
    "      \"opacity\": 0.7\n",
    "    },\n",
    "    \"selfReferenceSize\": 0,\n",
    "    \"selfReference\": {\n",
    "      \"size\": 0,\n",
    "      \"angle\": 0.7853981633974483\n",
    "    },\n",
    "    \"smooth\": {\n",
    "      \"forceDirection\": \"vertical\"\n",
    "    }\n",
    "  },\n",
    "  \"physics\": {\n",
    "    \"minVelocity\": 0.75\n",
    "  }\n",
    "}\n",
    "\"\"\")\n",
    "\n",
    "net.show('networkplot_Lung5s_neighbor1.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Lung tumor cells"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### all tumors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and 'tumor' in ctype: \n",
    "        print(sample)\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "D_cancer_lungtumors = RMSD_distance_matrix(Xs, GeneLists, keys=list(Xs.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_RMSD_heatmap(D_cancer_lungtumors, labels=list(Xs.keys()), cmap='viridis', method='ward', cthresh = 2.5, font_scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=[\n",
    "'Lung12\\ntumor 12',\n",
    "'Lung12\\ntumor 5',\n",
    "'Lung13\\ntumor 13',\n",
    "'Lung13\\ntumor 6',\n",
    "'Lung13\\ntumor 9',\n",
    "'Lung5 Rep1\\ntumor 12',\n",
    "'Lung5 Rep1\\ntumor 5',\n",
    "'Lung5 Rep2\\ntumor 12',\n",
    "'Lung5 Rep2\\ntumor 5',\n",
    "'Lung5 Rep3\\ntumor 12',\n",
    "'Lung5 Rep3\\ntumor 13',\n",
    "'Lung5 Rep3\\ntumor 5',\n",
    "'Lung6\\ntumor 12',\n",
    "'Lung6\\ntumor 13',\n",
    "'Lung6\\ntumor 5',\n",
    "'Lung6\\ntumor 6',\n",
    "'Lung6\\ntumor 9',\n",
    "'Lung9 Rep1\\ntumor 12',\n",
    "'Lung9 Rep1\\ntumor 13',\n",
    "'Lung9 Rep1\\ntumor 5',\n",
    "'Lung9 Rep1\\ntumor 6',\n",
    "'Lung9 Rep1\\ntumor 9',\n",
    "'Lung9 Rep2\\ntumor 12',\n",
    "'Lung9 Rep2\\ntumor 13',\n",
    "'Lung9 Rep2\\ntumor 5',\n",
    "'Lung9 Rep2\\ntumor 6',\n",
    "'Lung9 Rep2\\ntumor 9']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cmap('Blues', ncolor=10)\n",
    "['#f7fbff',\n",
    " '#e3eef9',\n",
    " '#d0e2f2',\n",
    " '#b8d5ea',\n",
    " '#97c6df',\n",
    " '#6fb0d7',\n",
    " '#4e9acb',\n",
    " '#3282be',\n",
    " '#1a68ae',\n",
    " '#084f99']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map_ctype = {\n",
    "'tumor 5': '#d0e2f2',\n",
    "'tumor 6':  '#97c6df',\n",
    "'tumor 9': '#6fb0d7',\n",
    "'tumor 12': '#3282be',\n",
    "'tumor 13': '#084f99',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map_sample = {\n",
    "'Lung5 Rep1': '#d0e2f2',\n",
    "'Lung5 Rep2': '#d0e2f2',\n",
    "'Lung5 Rep3': '#d0e2f2',\n",
    "'Lung6': '#97c6df',\n",
    "'Lung9 Rep1': '#6fb0d7',\n",
    "'Lung9 Rep2': '#6fb0d7',\n",
    "'Lung12': '#3282be',\n",
    "'Lung13': '#084f99',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_network(D_cancer_lungtumors, labels=labels, solver='forceAtlas2Based', cmap = color_map_sample, csep='\\n', corder=0, \n",
    "        N_neighbor=1, html='Lung_tumors_neighbor1.html', width=1400, height=1000, edge_scale = 20, edge_adjust=-10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### main tumors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and sample in ['Lung5_Rep1 tumor_5', 'Lung5_Rep2 tumor_5', 'Lung5_Rep3 tumor_5','Lung6 tumor_6','Lung9_Rep1 tumor_9','Lung9_Rep2 tumor_9','Lung12 tumor_12','Lung13 tumor_13',]: \n",
    "        print(sample)\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "D_cancer_main_lungtumors = RMSD_distance_matrix(Xs, GeneLists, keys=list(Xs.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_RMSD_heatmap(D_cancer_main_lungtumors, labels=list(Xs.keys()), cmap='viridis', method='ward', cthresh = 2.5, font_scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=[\n",
    "'Lung12\\ntumor 12',\n",
    "'Lung13\\ntumor 13',\n",
    "'Lung5 Rep1\\ntumor 5',\n",
    "'Lung5 Rep2\\ntumor 5',\n",
    "'Lung5 Rep3\\ntumor 5',\n",
    "'Lung6\\ntumor 6',\n",
    "'Lung9 Rep1\\ntumor 9',\n",
    "'Lung9 Rep2\\ntumor 9']\n",
    "color_map_ctype = {\n",
    "'tumor 5': '#d0e2f2',\n",
    "'tumor 6':  '#97c6df',\n",
    "'tumor 9': '#6fb0d7',\n",
    "'tumor 12': '#3282be',\n",
    "'tumor 13': '#084f99',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_network(D_cancer_main_lungtumors, labels=labels, solver='forceAtlas2Based', cmap = color_map_ctype, csep='\\n', corder=1, \n",
    "        N_neighbor=1, html='Lung_main_tumors_neighbor1.html', width=1400, height=1000, edge_scale = 20, edge_adjust=-10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Liver Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Liver/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\"cellType\")\n",
    "    chip = info[0].split(\".\")[0]\n",
    "    ctype = info[1][1:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files: \n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "    else:\n",
    "        print(sample)\n",
    " \n",
    "D_cancer_liver = RMSD_distance_matrix(Xs, GeneLists, keys=list(Xs.keys()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_RMSD_heatmap(D_cancer_liver, labels=list(Xs.keys()), cmap='viridis', method='ward', cthresh = 2.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map = {\n",
    "'Antibody secreting B cells': '#8dd593',\n",
    "'Mature B cells': '#11c638',\n",
    "'CD3+ alpha beta T cells': '#c6dec7',\n",
    "'gamma delta T cells': '#9cded6',\n",
    "'NK like cells': '#f3e1eb',\n",
    "'Central venous LSECs': '#bec1d4',\n",
    "'Cholangiocytes': '#8595e1',\n",
    "'Erthyroid cells': '#d5eae7',\n",
    "'Hep': '#8e063b',\n",
    "'Hep 1': '#d6bcc0',\n",
    "'Hep 3': '#e6afb9',\n",
    "'Hep 4': '#bb7784',\n",
    "'Hep 5': '#e07b91',\n",
    "'Hep 6': '#d33f6a',\n",
    "'Inflammatory macrophages': '#ef9708',\n",
    "'Non inflammatory macrophages': '#f0b98d',\n",
    "'Periportal LSECs': '#7d87b9',\n",
    "'Portal endothelial cells': '#ead3c6',\n",
    "'Stellate cells': '#0fcfc0',\n",
    "'tumor_1': '#4a6fe3',\n",
    "'tumor_2': '#023fa5',\n",
    "'NotDet': '#b5bbe3'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=[\n",
    "'CancerousLiver\\nAntibody secreting B cells',\n",
    "'CancerousLiver\\nCD3+ alpha beta T cells',\n",
    "'CancerousLiver\\nCentral venous LSECs',\n",
    "'CancerousLiver\\nCholangiocytes',\n",
    "'CancerousLiver\\nErthyroid cells',\n",
    "'CancerousLiver\\nHep',\n",
    "'CancerousLiver\\nInflammatory macrophages',\n",
    "'CancerousLiver\\nMature B cells',\n",
    "'CancerousLiver\\nNK like cells',\n",
    "'CancerousLiver\\nNon inflammatory macrophages',\n",
    "'CancerousLiver\\nPeriportal LSECs',\n",
    "'CancerousLiver\\nPortal endothelial cells',\n",
    "'CancerousLiver\\nStellate cells',\n",
    "'CancerousLiver\\ngamma delta T cells',\n",
    "'NormalLiver\\nAntibody secreting B cells',\n",
    "'NormalLiver\\nCD3+ alpha beta T cells',\n",
    "'NormalLiver\\nCentral venous LSECs',\n",
    "'NormalLiver\\nCholangiocytes',\n",
    "'NormalLiver\\nErthyroid cells',\n",
    "'NormalLiver\\nHep 1',\n",
    "'NormalLiver\\nHep 3',\n",
    "'NormalLiver\\nHep 6',\n",
    "'NormalLiver\\nInflammatory macrophages',\n",
    "'NormalLiver\\nMature B cells',\n",
    "'NormalLiver\\nNK like cells',\n",
    "'NormalLiver\\nNon inflammatory macrophages',\n",
    "'NormalLiver\\nPeriportal LSECs',\n",
    "'NormalLiver\\nPortal endothelial cells',\n",
    "'NormalLiver\\nStellate cells',\n",
    "'NormalLiver\\ngamma delta T cells']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_network(D_cancer_liver, labels=labels, solver='forceAtlas2Based', cmap = color_map, csep='\\n', corder=1, N_neighbor=2, html='Liver_neighbor2.html', width=1400, height=1000, edge_scale = 20, edge_adjust=-10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata=sc.read_h5ad(\"./data/Mice/E16.5_E1S3/E16.5_E1S3_cell_bin.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ct in adata.obs['annotation'].drop_duplicates().values[14:]:\n",
    "    adata_ = adata[adata.obs['annotation'].isin([ct])]\n",
    "    adata_.obs['row'] = adata_.obsm['spatial'][:,0]\n",
    "    adata_.obs['row'] = adata_.obsm['spatial'][:,0]\n",
    "    adata_.obs['col'] = adata_.obsm['spatial'][:,1]\n",
    "\n",
    "    pd.DataFrame(adata_.X.toarray(),index=adata_.obs['cell_id']).to_csv(\"adata_\"+filename(ct)+\".X.csv\")\n",
    "    adata_.obs[['cell_id','row','col']].set_index(\"cell_id\").to_csv(\"adata_\"+filename(ct)+\".spatial.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib.pyplot import rc_context\n",
    "import os\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "path = './BayesSpace/results/Mice/'\n",
    "adata = sc.read_h5ad(\"./data/Mice/E16.5_E1S3_cell_bin.h5ad\")\n",
    "for f in os.listdir(path):\n",
    "    if f.endswith(\"4_bayes_label.csv\"):\n",
    "        ct = f.replace(\"_4_bayes_label.csv\",\"\")\n",
    "        print(ct)\n",
    "        result = pd.read_csv(path+'/'+f, index_col=0)\n",
    "        result.reset_index(names=['cell_id'],inplace=True)\n",
    "        adata_ = adata[adata.obs['annotation'].isin([ct.replace('_',' ')])]\n",
    "        adata_.obs = pd.merge(adata_.obs, result, on='cell_id')\n",
    "        #adata_.obs['spatial.cluster'] = adata_.obs['spatial.cluster'].astype(str)\n",
    "        #adata_.obs['spatial.cluster'].value_counts()\n",
    "        #with rc_context({'figure.figsize': (6, 6)}):\n",
    "        #    sc.pl.scatter(adata_,x=\"row\",y=\"col\",color='spatial.cluster',size=5, color_map='gist_ncar', title=ct)\n",
    "        sample = \"E16.5_E1S3\"\n",
    "        gem = pd.read_csv(\"./data/Mice/SS200000108BR_A3A4_scgem.\"+ct+\".csv\", sep=\"\\t\", comment=\"#\")\n",
    "        anno = adata_.obs[[\"cell_id\",\"spatial.cluster\"]]\n",
    "        for cluster in adata_.obs[\"spatial.cluster\"].unique():\n",
    "            gem_subset=keep_rows_with_values_in_list(gem, 'cell', anno[anno[\"spatial.cluster\"]==cluster].cell_id.values)\n",
    "            gem_subset.to_csv(\"./BayesSpace/data/Mice/gem/%s.%s.\"%(sample,ct)+str(cluster)+\".csv\",sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Axo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1_CMPN.obs = pd.merge(adata_15DPI_1_CMPN.obs, result, on='cell_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1_CMPN.obs['spatial.cluster']=adata_15DPI_1_CMPN.obs['spatial.cluster'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_15DPI_1_CMPN.obs['spatial.cluster'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with rc_context({'figure.figsize': (6, 6)}):\n",
    "    sc.pl.scatter(adata_15DPI_1_CMPN,x=\"x\",y=\"y\",color='spatial.cluster',size=30, color_map='gist_ncar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample=\"15DPI_1\"\n",
    "gem = pd.read_csv(\"./data/Axolotls/regeneration/15DPI_1/15DPI_1.CMPN.csv\", sep=\"\\t\", comment=\"#\")\n",
    "anno = adata_15DPI_1_CMPN.obs[[\"cell_id\",\"spatial.cluster\"]]\n",
    "for ct in adata_15DPI_1_CMPN.obs[\"spatial.cluster\"].unique():\n",
    "        gem_subset=keep_rows_with_values_in_list(gem, 'cell', anno[anno[\"spatial.cluster\"]==ct].cell_id.values)\n",
    "        gem_subset.to_csv(\"./data/Axolotls/regeneration/%s/%s.%s.\"%(sample,sample,\"CMPN\")+filename(ct)+\".csv\",sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compare Multi-gene Structure btwn clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "RMs={}\n",
    "Ws={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "ctypes=[]\n",
    "\n",
    "result_path=\"./BayesSpace/results/Mice/\"\n",
    "samples = [ name for name in os.listdir(result_path) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    print(sample)\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    for TID in os.listdir(result_dir):\n",
    "        target_dir = os.path.join(result_dir,TID)\n",
    "        files = os.listdir(os.path.join(target_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(target_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            RMs[sample] = adata.obsm[\"Rotation\"]\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "            CellLists[sample] = adata.obs.index\n",
    "\n",
    "intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "print(\"Common Gene Number: \" + str(len(intersected_values)))\n",
    "# Using in1d and nonzero\n",
    "boolean_arrays = [np.in1d(a, intersected_values) for a in GeneLists.values()]\n",
    "\n",
    "N = len(samples)\n",
    "D = np.zeros((N,N))\n",
    "n = 0\n",
    "for i in samples:\n",
    "    m=0\n",
    "    for j in samples:\n",
    "        d, _, _ = numpy_svd_rmsd_rot(Xs[i][boolean_arrays[n]]/np.linalg.norm(Xs[i][boolean_arrays[n]]),\n",
    "                                     Xs[j][boolean_arrays[m]]/np.linalg.norm(Xs[j][boolean_arrays[m]]))\n",
    "        D[n,m] = d\n",
    "        D[m,n] = d\n",
    "        m+=1\n",
    "    n+=1\n",
    "\n",
    "sns.set(rc={'figure.figsize':(6, 5)})\n",
    "sns.heatmap(D, xticklabels=samples,yticklabels=samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.Gene Clustering on spatial distance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### clustering function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Leiden algorithm\n",
    "def RunLeiden(adj_matrix, resolution=0.1):\n",
    "    from igraph import Graph\n",
    "    import leidenalg\n",
    "    import numpy as np\n",
    "    \n",
    "    G = Graph.Weighted_Adjacency(adj_matrix)\n",
    "    part = leidenalg.find_partition(G, leidenalg.ModularityVertexPartition)\n",
    "    # part = leidenalg.CPMVertexPartition(G, resolution_parameter=resolution)\n",
    "    labels = np.zeros(adj_matrix.shape[0]).astype(int)\n",
    "    for i in range(len(part)):\n",
    "        labels[part[i]] = i\n",
    "\n",
    "    return list(labels)\n",
    "\n",
    "def leiden(mat, use_weights=False, resolution=1, iterations=-1, seed=1):\n",
    "    import leidenalg\n",
    "    g = get_igraph_from_adjacency(mat, directed=True)\n",
    "    weights = None\n",
    "    if use_weights:\n",
    "        weights = np.array(g.es[\"weight\"]).astype(np.float64)\n",
    "    part = leidenalg.find_partition(\n",
    "        g, leidenalg.RBConfigurationVertexPartition, \n",
    "        resolution_parameter = resolution, weights = weights, \n",
    "        n_iterations = iterations, \n",
    "        seed=seed*1000\n",
    "    )\n",
    "    groups = np.array(part.membership)\n",
    "        # adata.obs['louvain'] = pd.Categorical(\n",
    "        #       values=groups.astype('U'),\n",
    "        #       categories=natsorted(np.unique(groups).astype('U')),\n",
    "        # )\n",
    "    return groups, part.quality()\n",
    "\n",
    "def get_igraph_from_adjacency(adjacency, directed=None):\n",
    "    \"\"\"Get igraph graph from adjacency matrix.\"\"\"\n",
    "    import igraph as ig\n",
    "\n",
    "    sources, targets = adjacency.nonzero()\n",
    "    weights = adjacency[sources, targets]\n",
    "    if isinstance(weights, np.matrix):\n",
    "        weights = weights.A1\n",
    "    g = ig.Graph(directed=directed)\n",
    "    g.add_vertices(adjacency.shape[0])  # this adds adjacency.shape[0] vertices\n",
    "    g.add_edges(list(zip(sources, targets)))\n",
    "    try:\n",
    "        g.es['weight'] = weights\n",
    "    except KeyError:\n",
    "        pass\n",
    "    if g.vcount() != adjacency.shape[0]:\n",
    "        logg.warning(\n",
    "            f'The constructed graph has only {g.vcount()} nodes. '\n",
    "            'Your adjacency matrix contained redundant nodes.'\n",
    "        )\n",
    "    return g\n",
    "\n",
    "\n",
    "def knn_neighbors(X, k):\n",
    "    if k != 0:\n",
    "        knn_m = np.zeros((X.shape[0], X.shape[0]))\n",
    "        for i in range(X.shape[0]):\n",
    "            ids = np.argpartition(X[i], -k)[-k:]\n",
    "            top_set = set(X[i, ids])\n",
    "            if len(top_set) == 1:\n",
    "                b = X[i] == top_set.pop()\n",
    "                ids = []\n",
    "                offset = 1\n",
    "                left = True\n",
    "                while (len(ids) < k) and (offset < X.shape[0]):\n",
    "                # while len(ids) < k:\n",
    "                    if left:\n",
    "                        idx = i + offset\n",
    "                    else:\n",
    "                        idx = i - offset\n",
    "                    if idx < 0 or idx > len(b)-1:\n",
    "                        offset += 1\n",
    "                        left = not left\n",
    "                        continue\n",
    "                    if b[idx]:\n",
    "                        ids.append(idx)\n",
    "                    offset += 1\n",
    "                    left = not left\n",
    "            knn_m[i, ids] = 1\n",
    "\n",
    "        knn_m = (knn_m + knn_m.T)/2\n",
    "        knn_m[np.nonzero(knn_m)] = 1\n",
    "    else:\n",
    "        knn_m = X\n",
    "    return knn_m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "RMs={}\n",
    "Ws={}\n",
    "GeneLists={}\n",
    "CellLists={}\n",
    "ctypes=[]\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in os.listdir(result_path) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        print(sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Xs.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.metrics.pairwise import rbf_kernel\n",
    "aff = rbf_kernel(Xs['Immune_cell_E43X'],)\n",
    "# knn\n",
    "aff_knn = knn_neighbors(aff, k=5)\n",
    "# run leiden\n",
    "labels = np.array(RunLeiden(aff_knn, resolution=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster count (optimal)\n",
    "np.unique(labels,return_counts=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check cluster 0 (optimal)\n",
    "Xs['Immune_cell_E43X'].index[labels==0].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test\n",
    "# matplotlib.cm.get_cmap('tab20b',20)(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "colors1 = plt.cm.tab20b( (4./3*np.arange(20*3/4)).astype(int) )\n",
    "colors2 = plt.cm.tab20c( (4./3*np.arange(20*3/4)).astype(int) )\n",
    "colors = np.vstack((colors1, colors2))\n",
    "[matplotlib.colors.rgb2hex(c) for c in colors]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib widget\n",
    "import matplotlib.pyplot as plt\n",
    "fig = plt.figure()\n",
    "fig.set_figwidth(15)\n",
    "fig.set_figheight(15)\n",
    "ax = fig.add_subplot(111, projection=\"3d\")\n",
    "#ax.view_init(elev=90, azim=-90, roll=0)\n",
    "#ax.set_aspect(\"auto\")\n",
    "#colormap = ['#FF0000', '#00FF00', '#0000FF', '#FFFF00', '#00FFFF', '#FF00FF', '#C0C0C0', '#808080']\n",
    "X = Xs['Immune_cell_E43X']\n",
    "\n",
    "scatter = ax.scatter(X.iloc[:,0],X.iloc[:,1],X.iloc[:,2], c=labels, cmap = 'tab20')\n",
    "legend = ax.legend(*[scatter.legend_elements(num=None)[0],np.unique(labels)], \n",
    "                    title=\"Legend\", loc='upper left')\n",
    "ax.add_artist(legend)\n",
    "\n",
    "ax.set_xlabel(\"x\")\n",
    "ax.set_ylabel(\"y\")\n",
    "ax.set_zlabel(\"z\")\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "import numpy as np\n",
    "\n",
    "t = np.linspace(-1, 1.2, 500)\n",
    "x = (t**3) + (0.3 * np.random.randn(500))\n",
    "y = (t**6) + (0.3 * np.random.randn(500))\n",
    "z = (t**2) + (0.3 * np.random.randn(500))\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "fig.add_scatter3d(x = X.iloc[:,0], y = X.iloc[:,1], z = X.iloc[:,2],\n",
    "        mode = 'markers',\n",
    "        marker = dict(color = labels, size = 4, colorscale='rainbow')\n",
    "    )\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize = True,\n",
    "    xaxis = dict(\n",
    "        zeroline = False,\n",
    "        showgrid = False,\n",
    "        showline=True, linecolor='black',\n",
    "    ),\n",
    "    yaxis = dict(\n",
    "        zeroline = False,\n",
    "        showgrid = False ),)\n",
    "\n",
    "# xaxis.backgroundcolor is used to set background color\n",
    "fig.update_layout(scene = dict(\n",
    "                    xaxis = dict(\n",
    "                         backgroundcolor=\"rgb(200, 200, 230)\",\n",
    "                         gridcolor=\"white\",\n",
    "                         showbackground=True,\n",
    "                         zerolinecolor=\"white\",),\n",
    "                    yaxis = dict(\n",
    "                        backgroundcolor=\"rgb(230, 200,230)\",\n",
    "                        gridcolor=\"white\",\n",
    "                        showbackground=True,\n",
    "                        zerolinecolor=\"white\"),\n",
    "                    zaxis = dict(\n",
    "                        backgroundcolor=\"rgb(230, 230,200)\",\n",
    "                        gridcolor=\"white\",\n",
    "                        showbackground=True,\n",
    "                        zerolinecolor=\"white\",),),\n",
    "                    margin=dict(\n",
    "                    r=10, l=10,\n",
    "                    b=10, t=10),\n",
    "                    height = 1000,\n",
    "                    width = 1000,\n",
    "                    showlegend = False\n",
    "                                )\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### show certain clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_num(num):\n",
    "  if num in [1, 3, 6, 12]:\n",
    "    return num # \n",
    "  else:\n",
    "    return 0\n",
    "labels_adjust = [check_num(x) for x in labels]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib widget\n",
    "import matplotlib.pyplot as plt\n",
    "fig = plt.figure()\n",
    "fig.set_figwidth(15)\n",
    "fig.set_figheight(15)\n",
    "ax = fig.add_subplot(111, projection=\"3d\")\n",
    "#ax.view_init(elev=90, azim=-90, roll=0)\n",
    "#ax.set_aspect(\"auto\")\n",
    "#colormap = ['#FF0000', '#00FF00', '#0000FF', '#FFFF00', '#00FFFF', '#FF00FF', '#C0C0C0', '#808080']\n",
    "X=adata.uns['X']\n",
    "\n",
    "scatter1=ax.scatter(X.iloc[:,0],X.iloc[:,1],X.iloc[:,2], c=labels_adjust, cmap = 'rainbow')\n",
    "#legend1 = ax.legend(*[scatter1.legend_elements()[0],[\"0\",\"1\",\"2\",\"3\",\"4\",\"5\",\"6\",\"7\",\"8\",\"9\",\"10\",\"11\",\"12\",\"13\",\"14\"]], \n",
    "#                    title=\"Legend\", loc='upper left')\n",
    "legend1 = ax.legend(*[scatter1.legend_elements()[0],[\"others\",\"1\",\"4\",\"9\",\"11\"]], \n",
    "title=\"Legend\", loc='upper left')\n",
    "ax.add_artist(legend1)\n",
    "\n",
    "ax.set_xlabel(\"x\")\n",
    "ax.set_ylabel(\"y\")\n",
    "ax.set_zlabel(\"z\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GO enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "# Enricr API\n",
    "fig = plt.figure()\n",
    "fig.set_figwidth(18)\n",
    "fig.set_figheight(24)\n",
    "for i in np.unique(labels):\n",
    "    try:\n",
    "        enr_up = gp.enrichr(list(adata.uns['X'].index[labels==i]),\n",
    "                            gene_sets='GO_Biological_Process_2023',\n",
    "                            outdir=None)\n",
    "\n",
    "        # trim (go:...)\n",
    "        enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "\n",
    "        # dotplot\n",
    "        fig.axes.append(gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(i), cmap = plt.cm.autumn_r))\n",
    "    except ValueError:\n",
    "        continue\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.Meaning of physical distance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1 physical distance and regulatory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "pathways = pd.read_csv(\"./pathways.tsv\", sep='\\t', index_col=0)\n",
    "# list all lr pairs\n",
    "pairs = pathways[pathways.species==\"Mouse\"][['src','dest']]\n",
    "# list all genes in pathway\n",
    "#GeneLR = pairs_in_pathway.src.unique().tolist() + pairs_in_pathway.dest.unique().tolist()\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = list(adata.uns['X'].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import copy\n",
    "import itertools \n",
    "samples = list(Xs.keys())\n",
    "E = pd.DataFrame(columns=['sample','gene1', 'gene2', 'distance', 'source'])\n",
    "pair_ct={}\n",
    "for sample in samples:\n",
    "    # Pairs in celltype\n",
    "    GeneList = GeneLists[sample]\n",
    "    pair_ct[sample] = copy.deepcopy(pairs)\n",
    "    delete_idx = [idx for idx, row in pair_ct[sample].iterrows() if row[0] not in GeneList or row[1] not in GeneList]\n",
    "    pair_ct[sample].drop(delete_idx, inplace=True)\n",
    "    del delete_idx\n",
    "    print(sample+' '+str(len(pair_ct[sample])))\n",
    "    # Physical Distance\n",
    "    # regulatory pairs\n",
    "    gene1 = Xs[sample].loc[pair_ct[sample].src].to_numpy() # convert to numpy array\n",
    "    gene2 = Xs[sample].loc[pair_ct[sample].dest].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':pair_ct[sample].src,'gene2':pair_ct[sample].dest,'distance':dist,'source':'rgl'})\n",
    "    E = pd.concat([E, s1], ignore_index=True)\n",
    "    # all pairs\n",
    "    all_pairs = np.array(list(itertools.combinations(GeneList, 2))) # convert to numpy array\n",
    "    gene1 = Xs[sample].loc[all_pairs[:,0]].to_numpy() # convert to numpy array\n",
    "    gene2 = Xs[sample].loc[all_pairs[:,1]].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':all_pairs[:,0],'gene2':all_pairs[:,1],'distance':dist,'source':'all'})\n",
    "    E = pd.concat([E, s1], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_pairs = []\n",
    "for s in samples:\n",
    "    ls_pairs.append(((s, 'rgl'), (s, 'all')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statannot import add_stat_annotation\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(40, 5)},font_scale=2)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "ax = sns.boxplot(data=E, x='sample', y='distance', hue='source', showfliers=False, width=0.7, saturation=1)\n",
    "add_stat_annotation(ax, data=E, x='sample', y='distance', hue='source', box_pairs=ls_pairs, loc='outside', test=\"Mann-Whitney\", text_format=\"star\", verbose=2)\n",
    "\n",
    "#ax.set_xticklabels([\"Regulatory Pairs\",\"All Pairs\"])\n",
    "#ax.legend(title=' ', loc='upper right', labels=['regulatory','all'])\n",
    "\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"Euclidean Distance\", fontsize=16)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xticklabels(['Cardiomyocyte rep1', 'Cardiomyocyte rep2', 'Cardiomyocyte rep3', 'Chondrocyte rep1', 'Chondrocyte rep2', 'Chondrocyte rep3',\n",
    " 'Choroid plexus rep1', 'Choroid plexus rep2', 'Choroid plexus rep3', 'Dorsal midbrain neuron rep1', 'Dorsal midbrain neuron rep2', 'Dorsal midbrain neuron rep3', \n",
    " 'Endothelial cell rep1', 'Endothelial cell rep2', 'Endothelial cell rep3', 'Epithelial cell rep1', 'Epithelial cell rep2', 'Epithelial cell rep3',\n",
    " 'Erythrocyte rep1', 'Erythrocyte rep2', 'Erythrocyte rep3', 'Facial fibroblast rep1', 'Facial fibroblast rep2', 'Facial fibroblast rep3',\n",
    " 'Fibroblast rep1', 'Fibroblast rep2', 'Fibroblast rep3', 'Forebrain neuron rep1', 'Forebrain neuron rep2', 'Forebrain neuron rep3', \n",
    " 'Forebrain radial glia cell rep1', 'Forebrain radial glia cell rep2', 'Forebrain radial glia cell rep3', 'Ganglion rep1', 'Ganglion rep2', 'Ganglion rep3', \n",
    " 'Hepatocyte rep1', 'Hepatocyte rep2', 'Hepatocyte rep3', 'Immune cell rep1', 'Immune cell rep2', 'Immune cell rep3', 'Keratinocyte rep1', 'Keratinocyte rep2', 'Keratinocyte rep3', \n",
    " 'Limb fibroblast rep1', 'Limb fibroblast rep2', 'Limb fibroblast rep3', 'Macrophage rep1', 'Macrophage rep2', 'Macrophage rep3', 'Meninges cell rep1', 'Meninges cell rep2', 'Meninges cell rep3', \n",
    " 'Mid-hindbrain and spinal cord neuron rep1', 'Mid-hindbrain and spinal cord neuron rep2', 'Mid-hindbrain and spinal cord neuron rep3', 'Myoblast rep1', 'Myoblast rep2', 'Myoblast rep3', \n",
    " 'Olfactory epithelial cell rep1', 'Olfactory epithelial cell rep2', 'Olfactory epithelial cell rep3', 'Radial glia cell rep1', 'Radial glia cell rep2', 'Radial glia cell rep3', \n",
    " 'Smooth muscle cell rep1', 'Smooth muscle cell rep2', 'Smooth muscle cell rep3', 'Spinal cord neuron rep1', 'Spinal cord neuron rep2', 'Spinal cord neuron rep3', \n",
    " 'Thalamus neuron rep1', 'Thalamus neuron rep2', 'Thalamus neuron rep3'], rotation=90, fontsize=16)\n",
    "\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "new_labels = ['Regulatory Gene Pairs', 'All Gene Pairs']\n",
    "plt.legend(handles, new_labels, fontsize=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2 Fig.3i.phisical distance and chromosome distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "pathways = pd.read_csv(\"./pathways.tsv\", sep='\\t', index_col=0)\n",
    "# list all lr pairs\n",
    "pairs = pathways[pathways.species==\"Mouse\"][['src','dest']]\n",
    "# list all genes in pathway\n",
    "#GeneLR = pairs_in_pathway.src.unique().tolist() + pairs_in_pathway.dest.unique().tolist()\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Mice_onesample/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in tqdm(samples):\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            try:\n",
    "                Xs[sample] = adata.uns['X']\n",
    "                GeneLists[sample] = list(adata.uns['X'].index)\n",
    "                print(sample)\n",
    "            except KeyError:\n",
    "                continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv(\"./main/src/cytocraft/gtf/mice_genes.gtf\", sep='\\t', comment='#', header=None)\n",
    "gtf = gtf.rename(columns=lambda x: 'col_' + str(x))\n",
    "# get position of genes\n",
    "gene_chr={}\n",
    "gene_position={}\n",
    "for _, row in gtf.query(\"col_2 == 'gene'\").iterrows():\n",
    "    key = row['col_8'].split(';')[2].split()[1].strip('\"')\n",
    "    gene_chr[key] = row['col_0']\n",
    "    gene_position[key] = int(np.mean([row['col_3'],row['col_4']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import itertools\n",
    "samples = list(Xs.keys())\n",
    "F = pd.DataFrame(columns=['sample','gene1', 'gene2', 'distance', 'source'])\n",
    "pair_ct={}\n",
    "for sample in tqdm(samples):\n",
    "    # All Pairs in celltype\n",
    "    GeneList = GeneLists[sample]\n",
    "    all_pairs = np.array(list(itertools.combinations(GeneList, 2))) # convert to numpy array\n",
    "    gene1 = Xs[sample].loc[all_pairs[:,0]].to_numpy() # convert to numpy array\n",
    "    gene2 = Xs[sample].loc[all_pairs[:,1]].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    same_chrom = [gene_chr[g1] == gene_chr[g2] for g1, g2 in all_pairs]\n",
    "    # Assign a different value to the 'source' column based on the condition\n",
    "    source = []\n",
    "    for g1, g2, s, d in zip(all_pairs[:,0], all_pairs[:,1], same_chrom, dist):\n",
    "        chr_dist = abs(gene_position[g1] - gene_position[g2]) # calculate the chromosomal distance\n",
    "        if s and chr_dist < 1e5: # check if same chromosome and distance less than 1m\n",
    "            source.append('close')\n",
    "        elif s and chr_dist >= 1e5: # check if same chromosome and distance greater than or equal to 1m\n",
    "            source.append('far')\n",
    "        else: # check if different chromosome\n",
    "            source.append('different')\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':all_pairs[:,0],'gene2':all_pairs[:,1],'distance':dist,'source':source})\n",
    "    F = pd.concat([F, s1], ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_pairs = []\n",
    "for s in samples:\n",
    "    ls_pairs.append(((s, 'close'), (s, 'far')))\n",
    "    ls_pairs.append(((s, 'close'), (s, 'different')))                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statannotations,matplotlib\n",
    "sns.__version__,statannotations.__version__,matplotlib.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statannot import add_stat_annotation\n",
    "from matplotlib import pyplot as plt\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(16, 5)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "ax = sns.boxplot(data=F, x='sample', y='distance', hue='source', hue_order=['close', 'far', 'different'], showfliers=False, \n",
    "    width=0.7, saturation=1,  palette=['#4a6fe3','#b5bbe3','#e07b91'])\n",
    "\n",
    "add_stat_annotation(ax, data=F, x='sample', y='distance', hue='source', hue_order=['close', 'far', 'different'], box_pairs=ls_pairs, loc='outside', test=\"Mann-Whitney\", text_format=\"star\", verbose=2, comparisons_correction='bonferroni')\n",
    "#ax.set_xticklabels([\"Regulatory Pairs\",\"All Pairs\"])\n",
    "#plt.legend(title=' ', loc='upper right', labels=['regulatory','all'])\n",
    "ax.set_xlabel(\"\")\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_ylabel(\"Euclidean Distance\", fontsize=16)\n",
    "ax.set_xticklabels(['Cardiomyocyte',\n",
    "'Chondrocyte',\n",
    "'Choroid plexus',\n",
    "'Dorsal Mb neuron',\n",
    "'Endothelial cell',\n",
    "'Epithelial cell',\n",
    "'Erythrocyte',\n",
    "'Facial fibroblast',\n",
    "'Fibroblast',\n",
    "'Fb neuron',\n",
    "'Fb radial glia cell',\n",
    "'Ganglion',\n",
    "'Hepatocyte',\n",
    "'Immune cell',\n",
    "'Keratinocyte',\n",
    "'Limb fibroblast',\n",
    "'Macrophage',\n",
    "'Meninges cell',\n",
    "'Mb/Hb/SpC neuron',\n",
    "'Myoblast',\n",
    "'Olfactory epithelial cell',\n",
    "'Radial glia cell',\n",
    "'Smooth muscle cell',\n",
    "'SpC neuron',\n",
    "'Thalamus neuron'], rotation=90, fontsize=16)\n",
    "\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "new_labels = ['Close Same Chromosome Gene Pairs', 'Distant Same Chromosome Gene Pairs', 'Different Chromosome Gene Pairs']\n",
    "plt.legend(handles, new_labels, fontsize=14)\n",
    "plt.savefig(\"Fig3i.pdf\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Forebrain_radial_glia_cell / Nonsense-Mediated Decay (NMD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ir_in_pathway = pathways[pathways.species==\"Mouse\"][pathways[pathways.species==\"Mouse\"].pathway==\"Nonsense-Mediated Decay (NMD)\"]\n",
    "import scanpy as sc\n",
    "adata={}\n",
    "adata['Forebrain_radial_glia_cell']=sc.read_h5ad(\"./results/Mice/SS200000108BR_A3A4_scgem.Radial_glia_cell_HU50/adata.h5ad\")\n",
    "X = adata['Forebrain_radial_glia_cell'].uns['X']\n",
    "GeneList = X.index.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### count most common pathways for cell type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "Lpath=[]\n",
    "for idx,row in pathways.iterrows():\n",
    "    if row[0] in GeneList and row[1] in GeneList:\n",
    "        Lpath.append(row[2])\n",
    "counter = Counter(Lpath)\n",
    "sorted_list = counter.most_common()\n",
    "print(sorted_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### generating KG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset X condtion on if gene in pathway\n",
    "for g in X.index.values:\n",
    "    if g not in GeneLR:\n",
    "        X.drop([g],inplace=True)\n",
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# knowledge graph matrix\n",
    "KG = np.zeros((X.shape[0],X.shape[0]))\n",
    "distances_1 = []\n",
    "distances_0 = []\n",
    "for _,row in ir_in_pathway.iterrows():\n",
    "    l = row[0]\n",
    "    r = row[1]\n",
    "    try:\n",
    "        l_idx = np.where(X.index.values == l)\n",
    "        r_idx = np.where(X.index.values == r)\n",
    "        KG[l_idx,r_idx] = 1\n",
    "        KG[r_idx,l_idx] = 1\n",
    "    except ValueError:\n",
    "        continue\n",
    "KG.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### analyse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# knn matrix\n",
    "from sklearn.metrics.pairwise import rbf_kernel\n",
    "aff = rbf_kernel(X,)\n",
    "aff_knn = knn_neighbors(aff, k=10)\n",
    "np.fill_diagonal(aff_knn,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy\n",
    "# corr\n",
    "A = np.array([aff_knn.flatten(), KG.flatten()])\n",
    "idx = np.argwhere(np.all(A[..., :] == 0, axis=0))\n",
    "A2 = np.delete(A, idx, axis=1)\n",
    "print(\"Pearson:\" + str(np.corrcoef(A2[0],A2[1])[0][1]))\n",
    "print(\"Spearman:\" + str(scipy.stats.spearmanr(A2[0],A2[1])[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = np.array([aff_knn.flatten(),KG.flatten()])\n",
    "idx = np.argwhere(KG.flatten() == 0)\n",
    "A2 = np.delete(A, idx, axis=1)\n",
    "A2\n",
    "print(\"Hit Score:\" + str(np.sum(A2[0]/np.sum(A2[1]))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = np.array([aff_knn.flatten(),KG.flatten()])\n",
    "idx = np.argwhere(aff_knn.flatten() == 0)\n",
    "A3 = np.delete(A, idx, axis=1)\n",
    "A3\n",
    "print(\"Hit Score:\" + str(np.sum(A3[1])/np.sum(A3[0])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### average distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distances_0=[]\n",
    "distances_1=[]\n",
    "for (x,y), value in np.ndenumerate(KG):\n",
    "    if x>y:\n",
    "        if value==0:\n",
    "            distances_0.append(np.linalg.norm(X.iloc[x]-X.iloc[y]))\n",
    "        elif value==1:\n",
    "            distances_1.append(np.linalg.norm(X.iloc[x]-X.iloc[y]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(distances_0)+len(distances_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import ttest_ind\n",
    "ttest_ind(distances_0, distances_1, alternative='greater')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ss.t.cdf(t_stat, len(distances_0) + len(distances_1) - 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ss.t.sf(t_stat, len(distances_0) + len(distances_1) - 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "fig, ax = plt.subplots(facecolor='white', figsize=(4, 5))\n",
    "ax =sns.boxplot(data=normalized_df,x='Metric',y='value', hue='Category', palette={'SVI':'#80b1d3','Excluded': '#fb8072'}, width=0.8, hue_order=['SVI', 'Excluded'])\n",
    "pairs = []\n",
    "for i in metrics:\n",
    "    pairs.append( ((i, 'SVI'), (i, 'Excluded')))\n",
    "annot = Annotator(ax, pairs, data=normalized_df, x='Metric',y='value', hue='Category', hue_order=['SVI', 'Excluded'])\n",
    "annot.configure(test='Mann-Whitney-gt',comparisons_correction=\"BH\", correction_format=\"replace\")\n",
    "annot.apply_test()\n",
    "annot.annotate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(facecolor='white', figsize=(4, 5))\n",
    "\n",
    "# rectangular box plot\n",
    "bplot = ax.boxplot([distances_0,distances_1],\n",
    "                     vert=True,  # vertical box alignment\n",
    "                     patch_artist=True,  # fill with color\n",
    "                     labels=['0','1'] )  # will be used to label x-ticks\n",
    "#ax.set_title('Rectangular box plot')\n",
    "\n",
    "# fill with colors\n",
    "colors = ['pink','lightblue']\n",
    "\n",
    "for patch, color in zip(bplot['boxes'], colors):\n",
    "    patch.set_facecolor(color)\n",
    "\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "#ax.set_xlabel()\n",
    "plt.xticks(rotation=90)\n",
    "ax.set_ylabel('Distance')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=pd.DataFrame(data={'0':pd.Series(distances_0),'1':pd.Series(distances_1)})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.3 Volcano Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from collections import Counter\n",
    "from copy import deepcopy\n",
    "from scipy.stats import ttest_ind\n",
    "from statistics import mean\n",
    "\n",
    "# load data\n",
    "pathways = pd.read_csv(\"./pathways.tsv\", sep='\\t', index_col=0)\n",
    "pathway = pathways[pathways.species==\"Mouse\"]\n",
    "adata={}\n",
    "adata['Forebrain_radial_glia_cell']=sc.read_h5ad(\"./results/Mice/SS200000108BR_A3A4_scgem.Radial_glia_cell_9B0R/adata.h5ad\")\n",
    "X = adata['Forebrain_radial_glia_cell'].uns['X']\n",
    "GeneList = X.index.values\n",
    "\n",
    "# sort pathway\n",
    "Lpath=[]\n",
    "for idx,row in pathways.iterrows():\n",
    "    if row[0] in GeneList and row[1] in GeneList:\n",
    "        Lpath.append(row[2])\n",
    "counter = Counter(Lpath)\n",
    "sorted_list = counter.most_common()\n",
    "\n",
    "pvalue={}\n",
    "FC={}\n",
    "distances_0={}\n",
    "distances_1={}\n",
    "for i,v in tqdm(sorted_list):\n",
    "    if v >= 20:\n",
    "        # list all lr pairs in pathway i \n",
    "        ir_in_pathway = pathway[pathway.pathway==i]\n",
    "        # list all genes in pathway\n",
    "        GeneLR = ir_in_pathway.src.unique().tolist() + ir_in_pathway.dest.unique().tolist()\n",
    "        # subset X condtion on if gene in pathway\n",
    "        X_subset = deepcopy(X)\n",
    "        X_subset = X_subset[X_subset.index.isin(GeneLR)]\n",
    "        # knowledge graph matrix\n",
    "        KG = np.zeros((X_subset.shape[0],X_subset.shape[0]))\n",
    "        distances_1[i] = []\n",
    "        distances_0[i] = []\n",
    "        for _,row in ir_in_pathway.iterrows():\n",
    "            l, r = row[:2]\n",
    "            l_idx = np.isin(X_subset.index.values, l).nonzero()[0]\n",
    "            r_idx = np.isin(X_subset.index.values, r).nonzero()[0]\n",
    "            if l_idx.size > 0 and r_idx.size > 0:\n",
    "                KG[l_idx, r_idx] = KG[r_idx, l_idx] = 1\n",
    "        distances_0[i] = [np.linalg.norm(X_subset.iloc[x]-X_subset.iloc[y]) for (x,y), value in np.ndenumerate(KG) if x>y and value==0]\n",
    "        distances_1[i] = [np.linalg.norm(X_subset.iloc[x]-X_subset.iloc[y]) for (x,y), value in np.ndenumerate(KG) if x>y and value==1]\n",
    "        pvalue[i] = -np.log10(ttest_ind(distances_0[i], distances_1[i], alternative='greater')[1])\n",
    "        try:\n",
    "            FC[i] = mean(distances_0[i]) / mean(distances_1[i])\n",
    "        except ZeroDivisionError:\n",
    "            FC[i] = np.inf # assign a large value to indicate infinite fold change\n",
    "        else:\n",
    "            pass # do nothing if no error occurs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# draw\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.pyplot import figure\n",
    "from adjustText import adjust_text\n",
    "fig, ax = plt.subplots(figsize=(5, 4), dpi=300)\n",
    "\n",
    "# Convert dictionary to lists for scatter plot\n",
    "x = list(pvalue.values())\n",
    "y = list(FC.values())\n",
    "\n",
    "plt.scatter(x, y, s=5, c=\"blue\", alpha=0.5)\n",
    "\n",
    "# Add labels for dots\n",
    "text_labels = []\n",
    "for i, label in enumerate(pvalue.keys()):\n",
    "    text_labels.append(plt.text(x[i], y[i], label, fontsize=3))\n",
    "\n",
    "ax.autoscale_view()\n",
    "adjust_text(text_labels,arrowprops=dict(arrowstyle=\"-\", color='k', lw=0.5),)\n",
    "\n",
    "plt.xlabel(\"-log10(p-value)\")\n",
    "plt.ylabel(\"fold change\")\n",
    "plt.title(\"LR pair vs. non-LR pair\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Set up the figure and axis\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# Extract the data for plotting\n",
    "fold_changes = [entry['fold_change'] for entry in data]\n",
    "pathways = [entry['pathway'] for entry in data]\n",
    "p_values = [entry['p_value'] for entry in data]\n",
    "\n",
    "# Assign colors based on p-values\n",
    "colors = ['red' if p_value < 0.05 else 'blue' for p_value in p_values]\n",
    "\n",
    "# Plot the dots\n",
    "plt.scatter(fold_changes, pathways, c=colors)\n",
    "\n",
    "# Set labels and title\n",
    "plt.xlabel('Fold Change')\n",
    "plt.ylabel('Pathway')\n",
    "plt.title('Dot Plot')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.Discance between TC and CC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "adata={}\n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "Distance_data = []\n",
    "for ct in list(adata.keys()):\n",
    "    for i in range(adata[ct].n_obs):\n",
    "        vec_cc = np.nanmean(adata[ct].uns['W'][i*2:i*2+2,:],axis=1)\n",
    "        vec_tc = np.dot(adata[ct].uns['X'],adata[ct].obsm['Rotation'][i]).mean(axis=0)[0:2]\n",
    "        d = np.linalg.norm(vec_tc - vec_cc)\n",
    "        n_tc = len(adata[ct].uns['X'])\n",
    "        Distance_data.append({'cellid': i, 'sample': ct, 'Bias': d, 'ct':ct[:-5], 'number of transcription centers':n_tc})\n",
    "\n",
    "Distance = pd.DataFrame(Distance_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Distance['ct'] = Distance['ct'].replace({'Cardiomyocyte': 'Cardiomyocyte', 'Chondrocyte': 'Chondrocyte', 'Choroid_plexus': 'Choroid plexus', \n",
    "'Dorsal_midbrain_neuron': 'Dorsal Mb neuron', 'Endothelial_cell': 'Endothelial cell', \n",
    "'Epithelial_cell': 'Epithelial cell', 'Erythrocyte': 'Erythrocyte', 'Facial_fibroblast': 'Facial fibroblast', \n",
    "'Fibroblast': 'Fibroblast', 'Forebrain_neuron': 'Fb neuron', 'Forebrain_radial_glia_cell': 'Fb radial glia cell', \n",
    "'Ganglion': 'Ganglion', 'Hepatocyte': 'Hepatocyte', 'Immune_cell': 'Immune cell', 'Keratinocyte': 'Keratinocyte', \n",
    "'Limb_fibroblast': 'Limb fibroblast', 'Macrophage': 'Macrophage', 'Meninges_cell': 'Meninges cell', 'Mid-hindbrain_and_spinal_cord_neuron': 'Mb/Hb/SpC neuron', \n",
    "'Myoblast': 'Myoblast', 'Olfactory_epithelial_cell': 'Olfactory epithelial cell', 'Radial_glia_cell': 'Radial glia cell', \n",
    "'Smooth_muscle_cell': 'Smooth muscle cell', 'Spinal_cord_neuron': 'SpC neuron', 'Thalamus_neuron': 'Thalamus neuron'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(10, 4)}, font_scale=1)\n",
    "ax=sns.barplot(data=Distance,x='ct',y='number of transcription centers', color='#728FBF')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha=\"right\")\n",
    "ax.set_xlabel(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig3e distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "palette={'Limb fibroblast': '#809693',\n",
    " 'Keratinocyte': '#A30059',\n",
    " 'Endothelial cell': '#006FA6',\n",
    " 'Facial fibroblast': '#63FFAC',\n",
    " 'Meninges cell': '#4FC601',\n",
    " 'Mb/Hb/SpC neuron': '#3B5DFF',\n",
    " 'SpC neuron': '#FFAA92',\n",
    " 'Myoblast': '#FF2F80',\n",
    " 'Choroid plexus': '#FF34FF',\n",
    " 'Cardiomyocyte': '#8B0000',\n",
    " 'Chondrocyte': '#1CE6FF',\n",
    " 'Dorsal Mb neuron': '#FF92CC',\n",
    " 'Fibroblast': '#FF4A46',\n",
    " 'Thalamus neuron': '#B79762',\n",
    " 'Erythrocyte': '#FFE4E1',\n",
    " 'Smooth muscle cell': '#00C2A0',\n",
    " 'Macrophage': '#1B4400',\n",
    " 'Fb neuron': '#004D43',\n",
    " 'Hepatocyte': '#997D87',\n",
    " 'Ganglion': '#800080',\n",
    " 'Olfactory epithelial cell': '#008941',\n",
    " 'Immune cell': '#5A0007',\n",
    " 'Radial glia cell': '#6B7900',\n",
    " 'Epithelial cell': '#0000A6',\n",
    " 'Fb radial glia cell': '#8FB0FF'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import skew\n",
    "\n",
    "# Set the plot style and size\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(7.5, 4)}, font_scale=0.9, style='ticks')\n",
    "\n",
    "# Create the histogram plot\n",
    "ax = sns.histplot(data=Distance, x='Bias', hue='ct', multiple=\"layer\", palette=palette, legend=True)\n",
    "\n",
    "# Calculate and display the negative skewness coefficient for each cell type\n",
    "skewness_values = Distance.groupby('ct')['Bias'].apply(lambda x: skew(x))\n",
    "\n",
    "# Update the legend labels to include skewness values\n",
    "legend_labels = [f'{cell_type} ({skewness:.2f})' for cell_type, skewness in skewness_values.items()]\n",
    "\n",
    "# Put the legend on top\n",
    "lut = palette\n",
    "from matplotlib.patches import Patch\n",
    "handles = [Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "plt.legend(handles, legend_labels, loc=\"upper center\", bbox_to_anchor=(0.43, 1.60), fancybox=True, ncol=3, fontsize=9)\n",
    "\n",
    "# Plot the axhlines\n",
    "for y in range(5000, 30000, 5000):\n",
    "    ax.axhline(y, color=\"gray\", dashes=(4, 5), zorder=0)\n",
    "\n",
    "# Set the labels\n",
    "ax.set_xlabel(\"Bias between Cell Centroid and Nucleus Centroid\")\n",
    "\n",
    "# Show the plot\n",
    "plt.savefig('Fig3e.pdf', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(4, 2)}, font_scale=1)\n",
    "g = sns.FacetGrid(data=Distance, col='ct', col_wrap=5, height=2, aspect=1.5)\n",
    "g.map(sns.histplot, 'Bias')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig3d sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors as mcolors\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(6, 5)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "# Calculate the mean Distance for each ct\n",
    "mean_distance = Distance.groupby('ct')['Bias'].median()\n",
    "# Sort the ct based on the mean Distance\n",
    "order = mean_distance.sort_values(ascending=False).index\n",
    "\n",
    "# IQR\n",
    "q1 = np.percentile(Distance.Bias, 25)\n",
    "q3 = np.percentile(Distance.Bias, 75)\n",
    "\n",
    "colorlst=[\n",
    " '#809693',\n",
    " '#A30059',\n",
    " '#006FA6',\n",
    " '#63FFAC',\n",
    " '#4FC601',\n",
    " '#3B5DFF',\n",
    " '#FFAA92',\n",
    " '#FF2F80',\n",
    " '#FF34FF',\n",
    " '#8B0000',\n",
    " '#1CE6FF',\n",
    " '#FF92CC',\n",
    " '#FF4A46',\n",
    " '#B79762',\n",
    " '#FFE4E1',\n",
    " '#00C2A0',\n",
    " '#1B4400',\n",
    " '#004D43',\n",
    " '#997D87',\n",
    " '#800080',\n",
    " '#008941',\n",
    " '#5A0007',\n",
    " '#6B7900',\n",
    " '#0000A6',\n",
    " '#8FB0FF']\n",
    "\n",
    "# rectangular box plot\n",
    "ax = sns.boxplot(data=Distance, x='ct', y='Bias', showfliers = False, width=0.5, saturation=1, \n",
    "    order = order, palette=colorlst, linewidth=0.8\n",
    "    )\n",
    "\n",
    "# barplot\n",
    "iqr = Distance.groupby('ct')['Bias'].agg([lambda x: x.quantile(0.25), lambda x: x.quantile(0.75)])\n",
    "iqr[\"iqr\"] = iqr[\"<lambda_0>\"] - iqr[\"<lambda_1>\"]\n",
    "sns.barplot(x=iqr.index, y=\"iqr\", data=iqr, color=\"black\", order = order, palette=colorlst)\n",
    "#ax.set_title('Rectangular box plot')\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "#ax.set_ylabel('IQR                   Bias between CC and NC   ')\n",
    "ax.set_ylabel('     Nucleus Size                      Cell Size                                ')\n",
    "ax.axhline(q3, color=\"darkred\", dashes=(4, 5))\n",
    "ax.axhline(q1, color=\"darkblue\", dashes=(4, 5))\n",
    "ax.axhline(0, color=\".3\")\n",
    "plt.savefig(\"Fig3d.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "adata={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and 'Lung5' in chip: \n",
    "        print(sample)\n",
    "        adata[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "\n",
    "Distance_data = []\n",
    "for ct in tqdm(list(adata.keys())):\n",
    "    for i in range(adata[ct].n_obs):\n",
    "        vec_cc = np.nanmean(adata[ct].uns['W'][i*2:i*2+2,:],axis=1)\n",
    "        vec_tc = np.dot(adata[ct].uns['X'],adata[ct].obsm['Rotation'][i]).mean(axis=0)[0:2]\n",
    "        d = np.linalg.norm(vec_tc - vec_cc)\n",
    "        Distance_data.append({'cellid': i, 'sample': ct, 'distance': d, 'ct':ct[:-5]})\n",
    "\n",
    "Distance = pd.DataFrame(Distance_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print([i[5:]+' '+i[:5].strip() for i in [\n",
    "    'Rep1 B-cell', 'Rep2 B-cell', 'Rep3 B-cell', 'Rep1 plasmablast', 'Rep2 plasmablast', 'Rep3 plasmablast', \n",
    "    'Rep1 NK', 'Rep2 NK', 'Rep3 NK', 'Rep1 T CD4 naive', 'Rep2 T CD4 naive', 'Rep3 T CD4 naive',\n",
    "    'Rep1 T CD4 memory', 'Rep2 T CD4 memory', 'Rep3 T CD4 memory', 'Rep1 T CD8 naive', 'Rep2 T CD8 naive', 'Rep3 T CD8 naive', \n",
    "    'Rep1 T CD8 memory', 'Rep2 T CD8 memory', 'Rep3 T CD8 memory', 'Rep1 Treg', 'Rep2 Treg', 'Rep3 Treg', \n",
    "    'Rep1 endothelial', 'Rep2 endothelial', 'Rep3 endothelial', 'Rep1 epithelial', 'Rep2 epithelial', 'Rep3 epithelial', \n",
    "    'Rep1 fibroblast', 'Rep2 fibroblast', 'Rep3 fibroblast', 'Rep1 macrophage', 'Rep2 macrophage', 'Rep3 macrophage', \n",
    "    'Rep1 mast', 'Rep2 mast', 'Rep3 mast', 'Rep1 monocyte', 'Rep2 monocyte', 'Rep3 monocyte',\n",
    "    'Rep1 neutrophil', 'Rep2 neutrophil', 'Rep3 neutrophil', 'Rep1 pDC', 'Rep2 pDC', 'Rep3 pDC', 'Rep1 mDC', 'Rep2 mDC', 'Rep3 mDC', \n",
    "    'Rep1 tumor 5', 'Rep2 tumor 5', 'Rep3 tumor 5', 'Rep1 tumor 12', 'Rep2 tumor 12', 'Rep3 tumor 12', 'Rep3 tumor 13',\n",
    "    ]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colorlst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors as mcolors\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(15, 6)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "color_map_ct =  {\n",
    "    'B-cell': '#ef9708',\n",
    "    'plasmablast': '#f0b98d',\n",
    "    'NK': '#0fcfc0',\n",
    "    'T CD4 naive': '#e07b91',\n",
    "    'T CD4 memory': '#d33f6a',\n",
    "    'T CD8 naive': '#bb7784',\n",
    "    'T CD8 memory': '#8e063b',\n",
    "    'Treg': '#e6afb9',\n",
    "    'endothelial': '#bec1d4',\n",
    "    'epithelial': '#b5bbe3',\n",
    "    'fibroblast': '#7d87b9',\n",
    "    'macrophage': '#9cded6',\n",
    "    'mast': '#11c638',\n",
    "    'monocyte': '#8dd593',\n",
    "    'neutrophil': '#c6dec7',\n",
    "    'pDC': '#ead3c6',\n",
    "    'mDC': '#d6bcc0',\n",
    "    'tumor 5': '#0000A6',\n",
    "    'tumor 12': '#023fa5',\n",
    "    'tumor 13': '#4a6fe3',\n",
    "}\n",
    "# order\n",
    "order = ['Lung5_Rep1 B-cell', 'Lung5_Rep2 B-cell', 'Lung5_Rep3 B-cell', 'Lung5_Rep1 plasmablast', 'Lung5_Rep2 plasmablast', 'Lung5_Rep3 plasmablast', \n",
    "         'Lung5_Rep1 NK', 'Lung5_Rep2 NK', 'Lung5_Rep3 NK', 'Lung5_Rep1 T_CD4_naive', 'Lung5_Rep2 T_CD4_naive', 'Lung5_Rep3 T_CD4_naive',\n",
    "         'Lung5_Rep1 T_CD4_memory', 'Lung5_Rep2 T_CD4_memory', 'Lung5_Rep3 T_CD4_memory', 'Lung5_Rep1 T_CD8_naive', 'Lung5_Rep2 T_CD8_naive', 'Lung5_Rep3 T_CD8_naive', \n",
    "         'Lung5_Rep1 T_CD8_memory', 'Lung5_Rep2 T_CD8_memory', 'Lung5_Rep3 T_CD8_memory', 'Lung5_Rep1 Treg', 'Lung5_Rep2 Treg', 'Lung5_Rep3 Treg', \n",
    "         'Lung5_Rep1 endothelial', 'Lung5_Rep2 endothelial', 'Lung5_Rep3 endothelial', 'Lung5_Rep1 epithelial', 'Lung5_Rep2 epithelial', 'Lung5_Rep3 epithelial', \n",
    "         'Lung5_Rep1 fibroblast', 'Lung5_Rep2 fibroblast', 'Lung5_Rep3 fibroblast', 'Lung5_Rep1 macrophage', 'Lung5_Rep2 macrophage', 'Lung5_Rep3 macrophage', \n",
    "         'Lung5_Rep1 mast', 'Lung5_Rep2 mast', 'Lung5_Rep3 mast', 'Lung5_Rep1 monocyte', 'Lung5_Rep2 monocyte', 'Lung5_Rep3 monocyte',\n",
    "         'Lung5_Rep1 neutrophil', 'Lung5_Rep2 neutrophil', 'Lung5_Rep3 neutrophil', 'Lung5_Rep1 pDC', 'Lung5_Rep2 pDC', 'Lung5_Rep3 pDC', 'Lung5_Rep1 mDC', 'Lung5_Rep2 mDC', 'Lung5_Rep3 mDC', \n",
    "         'Lung5_Rep1 tumor_5', 'Lung5_Rep2 tumor_5', 'Lung5_Rep3 tumor_5', 'Lung5_Rep1 tumor_12', 'Lung5_Rep2 tumor_12', 'Lung5_Rep3 tumor_12']\n",
    "\n",
    "# colors\n",
    "## Define the color map\n",
    "#colorlst=[]\n",
    "#cmap = plt.get_cmap(\"Spectral\")\n",
    "## Get the number of colors in the color map\n",
    "#ncolors=25\n",
    "#rgb_values=[]\n",
    "#for i in range(ncolors):\n",
    "#    rgb_values.append(cmap(i*10))\n",
    "# Print the RGB values\n",
    "colorlst=[]\n",
    "for o in ['B-cell Rep1', 'B-cell Rep2', 'B-cell Rep3', 'plasmablast Rep1', 'plasmablast Rep2', 'plasmablast Rep3', 'NK Rep1', 'NK Rep2', 'NK Rep3', 'T CD4 naive Rep1', 'T CD4 naive Rep2', 'T CD4 naive Rep3', 'T CD4 memory Rep1', 'T CD4 memory Rep2', 'T CD4 memory Rep3', 'T CD8 naive Rep1', 'T CD8 naive Rep2', 'T CD8 naive Rep3', 'T CD8 memory Rep1', 'T CD8 memory Rep2', 'T CD8 memory Rep3', 'Treg Rep1', 'Treg Rep2', 'Treg Rep3', 'endothelial Rep1', 'endothelial Rep2', 'endothelial Rep3', 'epithelial Rep1', 'epithelial Rep2', 'epithelial Rep3', 'fibroblast Rep1', 'fibroblast Rep2', 'fibroblast Rep3', 'macrophage Rep1', 'macrophage Rep2', 'macrophage Rep3', 'mast Rep1', 'mast Rep2', 'mast Rep3', 'monocyte Rep1', 'monocyte Rep2', 'monocyte Rep3', 'neutrophil Rep1', 'neutrophil Rep2', 'neutrophil Rep3', 'pDC Rep1', 'pDC Rep2', 'pDC Rep3', 'mDC Rep1', 'mDC Rep2', 'mDC Rep3', 'tumor 5 Rep1', 'tumor 5 Rep2', 'tumor 5 Rep3', 'tumor 12 Rep1', 'tumor 12 Rep2', 'tumor 12 Rep3']:\n",
    "    colorlst.append(color_map_ct[o[:-5]])\n",
    "\n",
    "# rectangular box plot\n",
    "ax = sns.boxplot(data=Distance, x='sample', y='distance', showfliers = False, width=0.5, saturation=1, order=order,palette=colorlst\n",
    "    )\n",
    "\n",
    "# barplot\n",
    "iqr = Distance.groupby('sample')['distance'].agg([lambda x: x.quantile(0.25), lambda x: x.quantile(0.75)])\n",
    "iqr[\"iqr\"] = iqr[\"<lambda_0>\"] - iqr[\"<lambda_1>\"]\n",
    "sns.barplot(x=iqr.index, y=\"iqr\", data=iqr, color=\"black\", order=order, palette=colorlst)\n",
    "#ax.set_title('Rectangular box plot')\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(['B-cell Rep1', 'B-cell Rep2', 'B-cell Rep3', 'plasmablast Rep1', 'plasmablast Rep2', 'plasmablast Rep3', 'NK Rep1', 'NK Rep2', 'NK Rep3', 'T CD4 naive Rep1', 'T CD4 naive Rep2', 'T CD4 naive Rep3', 'T CD4 memory Rep1', 'T CD4 memory Rep2', 'T CD4 memory Rep3', 'T CD8 naive Rep1', 'T CD8 naive Rep2', 'T CD8 naive Rep3', 'T CD8 memory Rep1', 'T CD8 memory Rep2', 'T CD8 memory Rep3', 'Treg Rep1', 'Treg Rep2', 'Treg Rep3', 'endothelial Rep1', 'endothelial Rep2', 'endothelial Rep3', 'epithelial Rep1', 'epithelial Rep2', 'epithelial Rep3', 'fibroblast Rep1', 'fibroblast Rep2', 'fibroblast Rep3', 'macrophage Rep1', 'macrophage Rep2', 'macrophage Rep3', 'mast Rep1', 'mast Rep2', 'mast Rep3', 'monocyte Rep1', 'monocyte Rep2', 'monocyte Rep3', 'neutrophil Rep1', 'neutrophil Rep2', 'neutrophil Rep3', 'pDC Rep1', 'pDC Rep2', 'pDC Rep3', 'mDC Rep1', 'mDC Rep2', 'mDC Rep3', 'tumor 5 Rep1', 'tumor 5 Rep2', 'tumor 5 Rep3', 'tumor 12 Rep1', 'tumor 12 Rep2', 'tumor 12 Rep3'], rotation=90)\n",
    "ax.set_ylabel('IQR                Distance between CC and TC   ')\n",
    "#ax.axhline(0.08, color=\"darkred\", dashes=(4, 5))\n",
    "#ax.axhline(0.02, color=\"darkblue\", dashes=(4, 5))\n",
    "ax.axhline(0, color=\".3\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Axolotls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "adata={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and sample[0].isalpha(): \n",
    "        adata[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "\n",
    "Distance_data = []\n",
    "for ct in tqdm(list(adata.keys())):\n",
    "    for i in range(adata[ct].n_obs):\n",
    "        vec_cc = np.nanmean(adata[ct].uns['W'][i*2:i*2+2,:],axis=1)\n",
    "        vec_tc = np.dot(adata[ct].uns['X'],adata[ct].obsm['Rotation'][i]).mean(axis=0)[0:2]\n",
    "        d = np.linalg.norm(vec_tc - vec_cc)\n",
    "        Distance_data.append({'cellid': i, 'sample': ct, 'distance': d, 'ct':ct[:-5]})\n",
    "\n",
    "Distance = pd.DataFrame(Distance_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors as mcolors\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(15, 6)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "color_map_ct =  {\n",
    "    'B-cell': '#ef9708',\n",
    "    'plasmablast': '#f0b98d',\n",
    "    'NK': '#0fcfc0',\n",
    "    'T CD4 naive': '#e07b91',\n",
    "    'T CD4 memory': '#d33f6a',\n",
    "    'T CD8 naive': '#bb7784',\n",
    "    'T CD8 memory': '#8e063b',\n",
    "    'Treg': '#e6afb9',\n",
    "    'endothelial': '#bec1d4',\n",
    "    'epithelial': '#b5bbe3',\n",
    "    'fibroblast': '#7d87b9',\n",
    "    'macrophage': '#9cded6',\n",
    "    'mast': '#11c638',\n",
    "    'monocyte': '#8dd593',\n",
    "    'neutrophil': '#c6dec7',\n",
    "    'pDC': '#ead3c6',\n",
    "    'mDC': '#d6bcc0',\n",
    "    'tumor 5': '#0000A6',\n",
    "    'tumor 12': '#023fa5',\n",
    "    'tumor 13': '#4a6fe3',\n",
    "}\n",
    "# order\n",
    "order = ['Lung5_Rep1 B-cell', 'Lung5_Rep2 B-cell', 'Lung5_Rep3 B-cell', 'Lung5_Rep1 plasmablast', 'Lung5_Rep2 plasmablast', 'Lung5_Rep3 plasmablast', \n",
    "         'Lung5_Rep1 NK', 'Lung5_Rep2 NK', 'Lung5_Rep3 NK', 'Lung5_Rep1 T_CD4_naive', 'Lung5_Rep2 T_CD4_naive', 'Lung5_Rep3 T_CD4_naive',\n",
    "         'Lung5_Rep1 T_CD4_memory', 'Lung5_Rep2 T_CD4_memory', 'Lung5_Rep3 T_CD4_memory', 'Lung5_Rep1 T_CD8_naive', 'Lung5_Rep2 T_CD8_naive', 'Lung5_Rep3 T_CD8_naive', \n",
    "         'Lung5_Rep1 T_CD8_memory', 'Lung5_Rep2 T_CD8_memory', 'Lung5_Rep3 T_CD8_memory', 'Lung5_Rep1 Treg', 'Lung5_Rep2 Treg', 'Lung5_Rep3 Treg', \n",
    "         'Lung5_Rep1 endothelial', 'Lung5_Rep2 endothelial', 'Lung5_Rep3 endothelial', 'Lung5_Rep1 epithelial', 'Lung5_Rep2 epithelial', 'Lung5_Rep3 epithelial', \n",
    "         'Lung5_Rep1 fibroblast', 'Lung5_Rep2 fibroblast', 'Lung5_Rep3 fibroblast', 'Lung5_Rep1 macrophage', 'Lung5_Rep2 macrophage', 'Lung5_Rep3 macrophage', \n",
    "         'Lung5_Rep1 mast', 'Lung5_Rep2 mast', 'Lung5_Rep3 mast', 'Lung5_Rep1 monocyte', 'Lung5_Rep2 monocyte', 'Lung5_Rep3 monocyte',\n",
    "         'Lung5_Rep1 neutrophil', 'Lung5_Rep2 neutrophil', 'Lung5_Rep3 neutrophil', 'Lung5_Rep1 pDC', 'Lung5_Rep2 pDC', 'Lung5_Rep3 pDC', 'Lung5_Rep1 mDC', 'Lung5_Rep2 mDC', 'Lung5_Rep3 mDC', \n",
    "         'Lung5_Rep1 tumor_5', 'Lung5_Rep2 tumor_5', 'Lung5_Rep3 tumor_5', 'Lung5_Rep1 tumor_12', 'Lung5_Rep2 tumor_12', 'Lung5_Rep3 tumor_12']\n",
    "\n",
    "# colors\n",
    "## Define the color map\n",
    "#colorlst=[]\n",
    "#cmap = plt.get_cmap(\"Spectral\")\n",
    "## Get the number of colors in the color map\n",
    "#ncolors=25\n",
    "#rgb_values=[]\n",
    "#for i in range(ncolors):\n",
    "#    rgb_values.append(cmap(i*10))\n",
    "# Print the RGB values\n",
    "colorlst=[]\n",
    "for o in ['B-cell Rep1', 'B-cell Rep2', 'B-cell Rep3', 'plasmablast Rep1', 'plasmablast Rep2', 'plasmablast Rep3', 'NK Rep1', 'NK Rep2', 'NK Rep3', 'T CD4 naive Rep1', 'T CD4 naive Rep2', 'T CD4 naive Rep3', 'T CD4 memory Rep1', 'T CD4 memory Rep2', 'T CD4 memory Rep3', 'T CD8 naive Rep1', 'T CD8 naive Rep2', 'T CD8 naive Rep3', 'T CD8 memory Rep1', 'T CD8 memory Rep2', 'T CD8 memory Rep3', 'Treg Rep1', 'Treg Rep2', 'Treg Rep3', 'endothelial Rep1', 'endothelial Rep2', 'endothelial Rep3', 'epithelial Rep1', 'epithelial Rep2', 'epithelial Rep3', 'fibroblast Rep1', 'fibroblast Rep2', 'fibroblast Rep3', 'macrophage Rep1', 'macrophage Rep2', 'macrophage Rep3', 'mast Rep1', 'mast Rep2', 'mast Rep3', 'monocyte Rep1', 'monocyte Rep2', 'monocyte Rep3', 'neutrophil Rep1', 'neutrophil Rep2', 'neutrophil Rep3', 'pDC Rep1', 'pDC Rep2', 'pDC Rep3', 'mDC Rep1', 'mDC Rep2', 'mDC Rep3', 'tumor 5 Rep1', 'tumor 5 Rep2', 'tumor 5 Rep3', 'tumor 12 Rep1', 'tumor 12 Rep2', 'tumor 12 Rep3']:\n",
    "    colorlst.append(color_map_ct[o[:-5]])\n",
    "\n",
    "# rectangular box plot\n",
    "ax = sns.boxplot(data=Distance, x='sample', y='distance', showfliers = False, width=0.5, saturation=1, order=order,palette=colorlst\n",
    "    )\n",
    "\n",
    "# barplot\n",
    "iqr = Distance.groupby('sample')['distance'].agg([lambda x: x.quantile(0.25), lambda x: x.quantile(0.75)])\n",
    "iqr[\"iqr\"] = iqr[\"<lambda_0>\"] - iqr[\"<lambda_1>\"]\n",
    "sns.barplot(x=iqr.index, y=\"iqr\", data=iqr, color=\"black\", order=order, palette=colorlst)\n",
    "#ax.set_title('Rectangular box plot')\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(['B-cell Rep1', 'B-cell Rep2', 'B-cell Rep3', 'plasmablast Rep1', 'plasmablast Rep2', 'plasmablast Rep3', 'NK Rep1', 'NK Rep2', 'NK Rep3', 'T CD4 naive Rep1', 'T CD4 naive Rep2', 'T CD4 naive Rep3', 'T CD4 memory Rep1', 'T CD4 memory Rep2', 'T CD4 memory Rep3', 'T CD8 naive Rep1', 'T CD8 naive Rep2', 'T CD8 naive Rep3', 'T CD8 memory Rep1', 'T CD8 memory Rep2', 'T CD8 memory Rep3', 'Treg Rep1', 'Treg Rep2', 'Treg Rep3', 'endothelial Rep1', 'endothelial Rep2', 'endothelial Rep3', 'epithelial Rep1', 'epithelial Rep2', 'epithelial Rep3', 'fibroblast Rep1', 'fibroblast Rep2', 'fibroblast Rep3', 'macrophage Rep1', 'macrophage Rep2', 'macrophage Rep3', 'mast Rep1', 'mast Rep2', 'mast Rep3', 'monocyte Rep1', 'monocyte Rep2', 'monocyte Rep3', 'neutrophil Rep1', 'neutrophil Rep2', 'neutrophil Rep3', 'pDC Rep1', 'pDC Rep2', 'pDC Rep3', 'mDC Rep1', 'mDC Rep2', 'mDC Rep3', 'tumor 5 Rep1', 'tumor 5 Rep2', 'tumor 5 Rep3', 'tumor 12 Rep1', 'tumor 12 Rep2', 'tumor 12 Rep3'], rotation=90)\n",
    "ax.set_ylabel('IQR                Distance between CC and TC   ')\n",
    "#ax.axhline(0.08, color=\"darkred\", dashes=(4, 5))\n",
    "#ax.axhline(0.02, color=\"darkblue\", dashes=(4, 5))\n",
    "ax.axhline(0, color=\".3\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6.sub-conformation by clustering on variance matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Axolotls - development - VLMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='VLMC': \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[sample] = adata\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(lambda x, y: x.intersection(y), GeneLists.values())\n",
    "M = intersected_genes.size\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build gene-gene distance matrix\n",
    "keys = ['Stage44_VLMC',\n",
    "        'Stage54_VLMC',\n",
    "        'Stage57_VLMC',\n",
    "        'Control_Juv_VLMC',\n",
    "        'Adult_VLMC',\n",
    "        'Meta_VLMC']\n",
    "N = len(keys)\n",
    "GG_dev = np.zeros((N,M,M))\n",
    "for n in tqdm(range(N)):\n",
    "    index_k_intersection = [list(GeneLists[keys[n]]).index(gene) for gene in intersected_genes]\n",
    "    for i, idx_n in enumerate(index_k_intersection):\n",
    "        for j, idx_m in enumerate(index_k_intersection[:i+1]):\n",
    "            g1_cord = Xs[keys[n]].iloc[idx_n,:].to_numpy()\n",
    "            g2_cord = Xs[keys[n]].iloc[idx_m,:].to_numpy()\n",
    "            d = np.linalg.norm(g1_cord - g2_cord)\n",
    "            GG_dev[n,i,j] = GG_dev[n,j,i] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "V_dev = np.zeros((M,M))\n",
    "for i in range(M):\n",
    "    for j in range(M):\n",
    "        v = statistics.variance(GG_dev[:,i,j])\n",
    "        V_dev[i,j] = V_dev[j,i] = v"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Translate gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "gtf = pd.read_csv(\"./main/gtf/src/cytocraft/gtf/AmexT_v47-AmexG_v6.0-DD.gene.gtf\", sep='\\t',header=None)\n",
    "df = gtf[gtf.iloc[:,2]==\"gene\"].iloc[:,8].str.split(';',expand=True)\n",
    "\n",
    "GD_Axo = {}\n",
    "for idx, row in df.iterrows():\n",
    "    k = row[0].split(\" \")[1].strip(\"\\\"\")\n",
    "    if \"hs\" in row[1]:\n",
    "        names = row[1].strip(\"gene_name \").strip(\"\\\"\").split(\"|\")\n",
    "        for n in names:\n",
    "            if \"hs\" in n:\n",
    "                v = n.split(\" \")[0]\n",
    "    else:\n",
    "        v = row[1].strip().split(\" \")[1].strip(\"\\\"\")\n",
    "    GD_Axo[k]=v \n",
    "    \n",
    "intersected_symbols=[]\n",
    "for i in intersected_genes:\n",
    "    intersected_symbols.append(GD_Axo[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Hierarchy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hierarchy\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.patches as patches\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "clustergrid = sns.clustermap(V_dev, xticklabels=intersected_genes, yticklabels=intersected_genes, method='average')\n",
    "plt.close()\n",
    "\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 35\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "\n",
    "# Create a color map for clusters\n",
    "n_clusters = len(np.unique(clusters))\n",
    "color_map = sns.color_palette(\"Set1\", n_clusters)\n",
    "\n",
    "# Assign colors to clusters\n",
    "cluster_colors = [color_map[c-1] for c in clusters]\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(30, 30)},font_scale=1)\n",
    "\n",
    "ax = sns.clustermap(V_dev, xticklabels=[], yticklabels=[], cmap=\"viridis\", method='average', row_colors=[cluster_colors], col_colors=[cluster_colors], cbar_kws={'ticks': [0,15,30,45]}, vmax=40)\n",
    "ax.ax_cbar.set_title(\"Var\")\n",
    "ax.ax_cbar.set_position((0.10, .80, .05, .12))\n",
    "# Add legend for row color bar\n",
    "#from matplotlib.patches import Patch\n",
    "#lut = dict(zip(clusters, cluster_colors))\n",
    "#handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "#legend=plt.legend(handles, sorted(lut), title='Clusters', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=2,)\n",
    "#legend.get_frame().set_facecolor('white')\n",
    "\n",
    "# add patches\n",
    "row_order = ax.dendrogram_row.reordered_ind\n",
    "# start\n",
    "x = y = 0\n",
    "# width\n",
    "w = h = 0\n",
    "formal_c = None\n",
    "count = 1\n",
    "for c in np.roll(clusters[row_order],-1):\n",
    "    if c == formal_c:\n",
    "        formal_c = c\n",
    "        count += 1\n",
    "    else:\n",
    "        w = h = count\n",
    "        ax.ax_heatmap.add_patch(patches.Rectangle((x, y), w, h, fill=False, edgecolor='crimson', lw=0.5, clip_on=False))\n",
    "        x += count\n",
    "        y = x\n",
    "        formal_c = c\n",
    "        count = 1\n",
    "\n",
    "#ax.ax_heatmap.tick_params(length=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_values, counts = np.unique(clusters, return_counts=True)\n",
    "for value, count in zip(unique_values, counts):\n",
    "    if count>=10:\n",
    "        print(f\"{value}: {count}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### UMAP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP \n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "# Get clusters with threshold\n",
    "clusters = np.array(umap_leiden(V_dev, resolution=1, n_neighbors=3, min_dist=0.1, plot_umap=True), dtype=int)\n",
    "\n",
    "# reorder\n",
    "df = pd.DataFrame(V_dev)\n",
    "df['cluster'] = clusters\n",
    "df_sorted = df.sort_values(by=['cluster'])\n",
    "sorted_indices = df_sorted.index\n",
    "V_dev_sorted = V_dev[sorted_indices,:][:,sorted_indices]\n",
    "\n",
    "# Create a color map for clusters\n",
    "n_clusters = len(np.unique(clusters))\n",
    "color_map = sns.color_palette(\"rainbow\", n_clusters)\n",
    "cluster_colors = [color_map[c-1] for c in df_sorted.cluster]\n",
    "\n",
    "# Draw \n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(30, 25)},font_scale=1)\n",
    "\n",
    "ax=sns.heatmap(V_dev_sorted, xticklabels=intersected_genes[sorted_indices], yticklabels=intersected_genes[sorted_indices], cmap=\"viridis\", \n",
    "                #row_colors=[cluster_colors], col_colors=[cluster_colors], \n",
    "                cbar_kws={'ticks': [0,15,30,45]})\n",
    "ax.set_xticklabels([])\n",
    "ax.set_yticklabels([])\n",
    "#ax.ax_cbar.set_title(\"Var\")\n",
    "#ax.ax_cbar.set_position((0.07, .84, .04, .12))\n",
    "ax.tick_params(axis='y', which='major', pad=20, length=0)\n",
    "for i, color in enumerate(cluster_colors):\n",
    "    ax.add_patch(plt.Rectangle(xy=(-0.05, i), width=0.05, height=1, color=color, lw=0,\n",
    "                               transform=ax.get_yaxis_transform(), clip_on=False))\n",
    "\n",
    "# Add legend for row color bar\n",
    "from matplotlib.patches import Patch\n",
    "lut = dict(zip(clusters, cluster_colors))\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend=plt.legend(handles, sorted(lut), title='Clusters', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=2,)\n",
    "legend.get_frame().set_facecolor('white')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exam conservativeness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hierarchy\n",
    "subXs={}\n",
    "for sample in keys:\n",
    "    subXs[sample] = Xs[sample].loc[intersected_genes[clusters==8]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['Stage44_VLMC',\n",
    "        'Stage54_VLMC',\n",
    "        'Stage57_VLMC',\n",
    "        'Control_Juv_VLMC',\n",
    "        'Adult_VLMC',\n",
    "        'Meta_VLMC']\n",
    "# Hierarchy\n",
    "for value, count in zip(unique_values, counts):\n",
    "        if count >= 5:\n",
    "                subXs={}\n",
    "                for sample in keys:\n",
    "                        subXs[sample] = Xs[sample].loc[intersected_genes[clusters==value]]\n",
    "\n",
    "                # Build distance matrix\n",
    "                D_VLMC_dev_subconf = np.zeros((N,N))\n",
    "                for n, key_n in enumerate(keys):\n",
    "                        for m, key_m in enumerate(keys[:n+1]):\n",
    "                                normalized_x_n = normalizeX(subXs[key_n], method='mean')\n",
    "                                normalized_x_m = normalizeX(subXs[key_m], method='mean')\n",
    "                                d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "                                d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "                                D_VLMC_dev_subconf[n,m] = D_VLMC_dev_subconf[m,n] = min(d1,d2)\n",
    "\n",
    "                dmax = np.nanmax([np.nanmax(D_VLMC_dev_subconf),np.nanmax(D_VLMC_dev)])\n",
    "                labels_dev = ['St.44',\n",
    "                        'St.54',\n",
    "                        'St.57',\n",
    "                        'Juv.',\n",
    "                        'Adult',\n",
    "                        'Meta.']\n",
    "                sns.set(rc={\"figure.dpi\":600,'figure.figsize':(3.3, 1.5)},font_scale=1)\n",
    "                f,(ax1,ax2,axcb) = plt.subplots(1,3, gridspec_kw={'width_ratios':[1,1,0.1]})\n",
    "                g1 = sns.heatmap(D_VLMC_dev_subconf, xticklabels=labels_dev, yticklabels=labels_dev, vmax=dmax, cmap=\"viridis\", cbar=False, ax=ax1)\n",
    "                g2 = sns.heatmap(D_VLMC_dev, xticklabels=labels_dev, yticklabels=labels_dev, vmax=dmax, cmap=\"viridis\", cbar_ax=axcb, ax=ax2)\n",
    "                g2.set_yticks([])\n",
    "                g1.set(title='Sub-Conf'+str(value))\n",
    "                g2.set(title='Whole-Conf')\n",
    "                axcb.set_ylabel('NRMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 3X3 heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['Stage44_VLMC',\n",
    "        'Stage54_VLMC',\n",
    "        'Stage57_VLMC',\n",
    "        'Control_Juv_VLMC',\n",
    "        'Adult_VLMC',\n",
    "        'Meta_VLMC']\n",
    "labels_dev = ['St.44',\n",
    "            'St.54',\n",
    "            'St.57',\n",
    "            'Juv.',\n",
    "            'Adult',\n",
    "            'Meta.']\n",
    "# UMAP\n",
    "M = len(labels.cat.categories)\n",
    "# Build distance matrix\n",
    "D_VLMC_dev_subconf = np.zeros((M,N,N))\n",
    "for idx,label in enumerate(labels.cat.categories):\n",
    "    subXs={}\n",
    "    for sample in keys:\n",
    "        subXs[sample] = Xs[sample].loc[intersected_genes[labels==label]]\n",
    "    for n, key_n in enumerate(keys):\n",
    "        for m, key_m in enumerate(keys[:n+1]):\n",
    "            normalized_x_n = normalizeX(subXs[key_n], method='mean')\n",
    "            normalized_x_m = normalizeX(subXs[key_m], method='mean')\n",
    "            d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "            d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "            D_VLMC_dev_subconf[idx,n,m] = D_VLMC_dev_subconf[idx,m,n] = min(d1,d2)\n",
    "\n",
    "dmax = np.nanmax(D_VLMC_dev_subconf)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1.5, 1.2)},font_scale=1)\n",
    "gw = sns.heatmap(D_VLMC_dev, xticklabels=labels_dev, yticklabels=labels_dev, vmax=dmax, cmap=\"viridis\")\n",
    "gw.set(title='Development-VLMC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 1X2 heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(4.5, 4.5)},font_scale=1)\n",
    "f,axs = plt.subplots(3,3)\n",
    "\n",
    "for i in range(3):\n",
    "    for j in range(3):\n",
    "        g = sns.heatmap(D_VLMC_dev_subconf[i*3+j], xticklabels=labels_dev, yticklabels=labels_dev, vmin=0, vmax=dmax, cmap=\"viridis\", cbar=False, ax=axs[i,j])\n",
    "        #g2 = sns.heatmap(D_VLMC_dev, xticklabels=labels_dev, yticklabels=labels_dev, vmax=dmax, cmap=\"viridis\", cbar_ax=axcb, ax=ax2)\n",
    "        if j>0:\n",
    "            g.set_yticks([])\n",
    "        if i<2:\n",
    "            g.set_xticks([])\n",
    "        g.set(title='Sub-Conf '+str(i*3+j))\n",
    "        #axcb.set_ylabel('NRMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### barplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(4.5, 4.5)},font_scale=1)\n",
    "\n",
    "bar_data = np.concatenate([np.delete(D_VLMC_dev_subconf[i], dia).flatten() for i in range(M)])\n",
    "sns.boxplot(x = np.array(['sub-conf 0'] * (N * N) + ['sub-conf 1'] * (N * N) + ['sub-conf 2'] * (N * N) +\n",
    "                        ['sub-conf 3'] * (N * N) + ['sub-conf 4'] * (N * N) + ['sub-conf 5'] * (N * N) +\n",
    "                        ['sub-conf 6'] * (N * N) + ['sub-conf 7'] * (N * N) + ['sub-conf 8'] * (N * N)), y=bar_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### GO "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "gp.get_library_name(organism='Fish')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO on umap results\n",
    "enr = pd.DataFrame(columns=['Cluster','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "\n",
    "# Enricr API\n",
    "for value, count in zip(unique_values, counts):\n",
    "        if count >= 10:\n",
    "            try:\n",
    "                genelist = np.array(intersected_symbols)[clusters==value].tolist()\n",
    "                filterlist = list(filter(lambda x: 'AMEX60DD' not in x, genelist))\n",
    "                for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "                    enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                                        organism='Human',\n",
    "                                        gene_sets=gs,\n",
    "                                        outdir=None, \n",
    "                                        cutoff=0.05,\n",
    "                                        no_plot= True)\n",
    "\n",
    "                    # trim (go:...)\n",
    "                    enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "                    enr_up.res2d['Cluster'] = \"sub-conformation \"+str(value)\n",
    "                    enr = pd.concat([enr, enr_up.res2d[enr_up.res2d['Adjusted P-value']<0.05]], ignore_index=True)\n",
    "                    # dotplot\n",
    "                    # gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(value), size=10, cmap = plt.cm.viridis_r)\n",
    "            except ValueError or FileNotFoundError or TypeError as e:\n",
    "                print(value)\n",
    "                print(e)\n",
    "                continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enr.to_csv(\"AXO_dev_VLMC_subconformation_hierachy_clustering_GO&KEGG_p0.05.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GO top5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enr_top5 = pd.DataFrame(columns=['Cluster','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "    enr_gs = enr[enr[\"Gene_set\"]==gs]\n",
    "    for i,cluster in enumerate(np.unique(enr_BP.Cluster)):\n",
    "        enr_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "        enr_top5 = pd.concat([enr_top5, enr_gs_cluster], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enr_top5.to_csv(\"AXO_dev_VLMC_subconformation_hierachy_clustering_GO&KEGG_top5.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### GO plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1.35, 10)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "f,axs = plt.subplots(8,1)\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.cm.get_cmap('viridis') # create a color map object\n",
    "enr_top5_gs = enr_top5[enr_top5[\"Gene_set\"]=='GO_Biological_Process_2023']\n",
    "\n",
    "for i,cluster in enumerate(np.unique(enr_top5_gs.Cluster)):\n",
    "    enr_top5_cluster = enr_top5_gs[enr_top5_gs[\"Cluster\"]==cluster]\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=0.05)\n",
    "    colors = [cmap(norm(value)) for value in enr_top5_cluster[\"Adjusted P-value\"]]\n",
    "    g = sns.barplot(data=enr_top5_cluster, y=enr_top5_cluster[\"Term\"], x=\"Combined Score\", fill=\"Adjusted P-value\", ax=axs[i], palette=colors)\n",
    "    g.set_ylabel(\"\")\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='Adjusted P-value') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-4, 0.24, 0.9, 0.05]) # xmin, ymin, dx, and dy\n",
    "plt.show()\n",
    "f.savefig('AXO_dev_VLMC_GO_subconformations.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Highest Var Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = np.array(intersected_symbols)[clusters==value].tolist()\n",
    "filterlist = list(filter(lambda x: 'AMEX60DD' not in x, genelist))\n",
    "enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                    organism='Human',\n",
    "                    #gene_sets='GO_Molecular_Function_2023',\n",
    "                    gene_sets='GO_Biological_Process_2023',\n",
    "                    outdir=None, \n",
    "                    cutoff=0.05,\n",
    "                    no_plot= True)\n",
    "\n",
    "# trim (go:...)\n",
    "enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(intersected_symbols)[82 in clusters]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung Cancer - all types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[4][:-5]\n",
    "    if ctype not in adatas.keys():\n",
    "        adatas[ctype]={}\n",
    "        Xs[ctype]={}\n",
    "        GeneLists[ctype]={}\n",
    "    #sample = chip+'_'+ctype\n",
    "    if 'adata.h5ad' in files and 'Lung5' in chip: \n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        adatas[ctype][chip] = adata\n",
    "        Xs[ctype][chip] = adata.uns['X']\n",
    "        GeneLists[ctype][chip] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for ct in Xs.keys():\n",
    "    print(ct)\n",
    "    keys = list(Xs[ct].keys())\n",
    "    intersected_genes = reduce(lambda x, y: x.intersection(y), GeneLists[ct].values())\n",
    "    M = intersected_genes.size\n",
    "    print(M)\n",
    "\n",
    "    # Build gene-gene distance matrix\n",
    "    N = len(keys)\n",
    "    GG_lung = np.zeros((N,M,M))\n",
    "    for n in tqdm(range(N)):\n",
    "        index_k_intersection = [list(GeneLists[ct][keys[n]]).index(gene) for gene in intersected_genes]\n",
    "        for i, idx_n in enumerate(index_k_intersection):\n",
    "            for j, idx_m in enumerate(index_k_intersection[:i+1]):\n",
    "                g1_cord = Xs[ct][keys[n]].iloc[idx_n,:].to_numpy()\n",
    "                g2_cord = Xs[ct][keys[n]].iloc[idx_m,:].to_numpy()\n",
    "                d = np.linalg.norm(g1_cord - g2_cord)\n",
    "                GG_lung[n,i,j] = GG_lung[n,j,i] = d\n",
    "\n",
    "    import statistics\n",
    "    V_lung = np.zeros((M,M))\n",
    "    for i in range(M):\n",
    "        for j in range(M):\n",
    "            v = statistics.variance(GG_lung[:,i,j])\n",
    "            V_lung[i,j] = V_lung[j,i] = v\n",
    "\n",
    "    # Draw\n",
    "    # hierarchy\n",
    "    import seaborn as sns\n",
    "    import pandas as pd\n",
    "    import matplotlib.patches as patches\n",
    "    from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "    clustergrid = sns.clustermap(V_lung, xticklabels=intersected_genes, yticklabels=intersected_genes, method='average')\n",
    "    plt.close()\n",
    "\n",
    "    # Get clusters with threshold\n",
    "    linkage = clustergrid.dendrogram_row.linkage\n",
    "    threshold = 35\n",
    "    clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "\n",
    "    # Create a color map for clusters\n",
    "    n_clusters = len(np.unique(clusters))\n",
    "    color_map = sns.color_palette(\"Set1\", n_clusters)\n",
    "\n",
    "    # Assign colors to clusters\n",
    "    cluster_colors = [color_map[c-1] for c in clusters]\n",
    "    sns.set(rc={\"figure.dpi\":300,'figure.figsize':(30, 30)},font_scale=1)\n",
    "\n",
    "    ax = sns.clustermap(V_lung, xticklabels=[], yticklabels=[], cmap=\"viridis\", method='average', row_colors=[cluster_colors], col_colors=[cluster_colors], cbar_kws={'ticks': [0,15,30,45]}, vmax=40)\n",
    "    ax.ax_cbar.set_title(\"Var\")\n",
    "    ax.ax_cbar.set_position((0.10, .80, .05, .12))\n",
    "    # Add legend for row color bar\n",
    "    #from matplotlib.patches import Patch\n",
    "    #lut = dict(zip(clusters, cluster_colors))\n",
    "    #handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "    #legend=plt.legend(handles, sorted(lut), title='Clusters', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=2,)\n",
    "    #legend.get_frame().set_facecolor('white')\n",
    "\n",
    "    # add patches\n",
    "    row_order = ax.dendrogram_row.reordered_ind\n",
    "    # start\n",
    "    x = y = 0\n",
    "    # width\n",
    "    w = h = 0\n",
    "    formal_c = None\n",
    "    count = 1\n",
    "    for c in np.roll(clusters[row_order],-1):\n",
    "        if c == formal_c:\n",
    "            formal_c = c\n",
    "            count += 1\n",
    "        else:\n",
    "            w = h = count\n",
    "            ax.ax_heatmap.add_patch(patches.Rectangle((x, y), w, h, fill=False, edgecolor='crimson', lw=0.5, clip_on=False))\n",
    "            x += count\n",
    "            y = x\n",
    "            formal_c = c\n",
    "            count = 1\n",
    "\n",
    "    #ax.ax_heatmap.tick_params(length=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7.Conversiveness Across Species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "print(samples[::3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "adatas={}\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "species=[]\n",
    "# mice\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples[::3]:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    sample = sample.split('.')[1]\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    brain_ct = [\"Choroid_plexus_1RNO\",\"Dorsal_midbrain_neuron_3DXO\",\"Forebrain_neuron_2U9E\",\"Forebrain_radial_glia_cell_CZGW\",\n",
    "    \"Meninges_cell_NGKE\",\"Mid-hindbrain_and_spinal_cord_neuron_23UD\",\"Olfactory_epithelial_cell_1AXJ\",\"Radial_glia_cell_9B0R\",\"Thalamus_neuron_BDOD\"]\n",
    "    if 'adata.h5ad' in files and sample in brain_ct: \n",
    "        sample='Mice ' + sample\n",
    "        print(sample)\n",
    "        adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        species.append('Mice')\n",
    "\n",
    "# axolotls\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[1].split('_')[0]\n",
    "    sample = 'Axolotls ' +chip+'_'+ctype\n",
    "    if 'adata.h5ad' in files and chip=='Control_Juv':\n",
    "        print(sample)\n",
    "        adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        species.append('Axolotls')\n",
    "\n",
    "# cancer\n",
    "result_path=\"./results/Cancer/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'celltype2' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = 'Human ' + ctype+' '+chip\n",
    "        print(sample)\n",
    "        if 'adata.h5ad' in files: \n",
    "            adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            species.append('Human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mouse gene to human gene\n",
    "HOM_MouseHuman = pd.read_csv('./data/HMD_HumanPhenotype.rpt', sep='\\t', header=None)\n",
    "HOM_MouseHuman = HOM_MouseHuman.set_index(2)\n",
    "HOM_MouseHuman = HOM_MouseHuman[~HOM_MouseHuman.index.duplicated()]\n",
    "D_HOM_MouseHuman = pd.Series(HOM_MouseHuman[0]).to_dict()\n",
    "\n",
    "# axolotls gene to human gene\n",
    "gtf=pd.read_csv(\"./gtf/AmexT_v47-AmexG_v6.0-DD.gtf\",sep='\\t',header=None)\n",
    "df=gtf[gtf.iloc[:,2]==\"gene\"].iloc[:,8].str.split(';',expand=True)\n",
    "\n",
    "D_HOM_AxolotlsHuman = {}\n",
    "for idx, row in df.iterrows():\n",
    "    k = row[0].split(\" \")[1].strip(\"\\\"\")\n",
    "    if \"hs\" in row[1]:\n",
    "        names = row[1].strip(\"gene_name \").strip(\"\\\"\").split(\"|\")\n",
    "        for n in names:\n",
    "            if \"hs\" in n:\n",
    "                v = n.split(\" \")[0]\n",
    "    else:\n",
    "        v = row[1].strip().split(\" \")[1].strip(\"\\\"\")\n",
    "    D_HOM_AxolotlsHuman[k]=v "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = list(adatas.keys())\n",
    "for sample in keys:\n",
    "    Xs[sample] = adatas[sample].uns['X']\n",
    "    GeneLists[sample] = list(adatas[sample].uns['X'].index)\n",
    "    if 'Mice' in sample:\n",
    "        GeneLists[sample] = [D_HOM_MouseHuman.get(x, x) for x in GeneLists[sample]]\n",
    "    elif 'Axolotls' in sample:\n",
    "        GeneLists[sample] = [D_HOM_AxolotlsHuman.get(x, x) for x in GeneLists[sample]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "N = len(keys)\n",
    "DM_Species = np.zeros((N,N))\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        index_n = []\n",
    "        index_m = []\n",
    "        for i,v1 in enumerate(intersected_values):\n",
    "            for j,v2 in enumerate(GeneLists[key_n]):\n",
    "                if v1==v2:\n",
    "                    index_n.append(j)\n",
    "                    break\n",
    "        for i,v1 in enumerate(intersected_values):\n",
    "            for j,v2 in enumerate(GeneLists[key_m]):\n",
    "                if v1==v2:\n",
    "                    index_m.append(j)\n",
    "                    break\n",
    "        #index_n.append(np.where(GeneLists[key_n] == g)[0][0])\n",
    "        #index_m.append(np.where(GeneLists[key_m] == g)[0][0])\n",
    "        normalized_x_n = normalizeX(Xs[key_n].iloc[index_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m].iloc[index_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        DM_Species[n,m] = DM_Species[m,n] = min(d1,d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_num={}\n",
    "for i,s in enumerate(np.unique(species)):\n",
    "    species_num[s]=i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a color map for species\n",
    "n_species = len(np.unique(species))\n",
    "color_map = sns.color_palette(\"Set1\", n_species)\n",
    "species_colors = [color_map[species_num[s]] for s in species]\n",
    "\n",
    "row_colors = pd.DataFrame({'species':species_colors})\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(30, 30)},font_scale=0.6)\n",
    "clustergrid = sns.clustermap(pd.DataFrame(DM_Species), xticklabels=keys, yticklabels=keys, method='average', cmap=\"coolwarm_r\", vmin=1.2, row_colors=row_colors)\n",
    "\n",
    "# color bar\n",
    "from matplotlib.patches import Patch\n",
    "lut = dict(zip(species, species_colors))\n",
    "handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "legend = plt.legend(handles, sorted(lut), title='species', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=1,)\n",
    "legend.get_frame().set_facecolor('white')\n",
    "#plt.gca().add_artist(legend)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fibroblast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata_Fb={}\n",
    "\n",
    "# Cancer celltype\n",
    "#adata_Fb['Cancer1'] = sc.read_h5ad(\"./results/Cancer/FP200000493BR_D4.cellbin.gem.celltype.Fibroblasts_240L/adata.h5ad\")\n",
    "#adata_Fb['Cancer2'] = sc.read_h5ad(\"./results/Cancer/FP200000493BR_E3.cellbin.gem.celltype.Fibroblasts_VXWH/adata.h5ad\")\n",
    "#adata_Fb['Cancer3'] = sc.read_h5ad(\"./results/Cancer/FP200000493BR_E4.cellbin.gem.celltype.Fibroblasts_4W9P/adata.h5ad\")\n",
    "#adata_Fb['Cancer4'] = sc.read_h5ad(\"./results/Cancer/FP200000615BR_A4.cellbin.gem.celltype.Fibroblasts_SZ6J/adata.h5ad\")\n",
    "#adata_Fb['Cancer5'] = sc.read_h5ad(\"./results/Cancer/FP200000615BR_F3.cellbin.gem.celltype.Fibroblasts_Q28N/adata.h5ad\")\n",
    "\n",
    "# Cancer celltype2\n",
    "adata_Fb['Cancer_rep1'] = sc.read_h5ad(\"./results/Cancer/FP200000493BR_D4.cellbin.gem.celltype2.Fibroblasts_J7M7/adata.h5ad\")\n",
    "adata_Fb['Cancer_rep2'] = sc.read_h5ad(\"./results/Cancer/FP200000493BR_E3.cellbin.gem.celltype2.Fibroblasts_SY1T/adata.h5ad\")\n",
    "adata_Fb['Cancer_rep3'] = sc.read_h5ad(\"./results/Cancer/FP200000493BR_E4.cellbin.gem.celltype2.Fibroblasts_JKV0/adata.h5ad\")\n",
    "adata_Fb['Cancer_rep4'] = sc.read_h5ad(\"./results/Cancer/FP200000615BR_A4.cellbin.gem.celltype2.Fibroblasts_PDX2/adata.h5ad\")\n",
    "adata_Fb['Cancer_rep5'] = sc.read_h5ad(\"./results/Cancer/FP200000615BR_F3.cellbin.gem.celltype2.Fibroblasts_2U5L/adata.h5ad\")\n",
    "\n",
    "# Mice\n",
    "adata_Fb['Mice1'] = sc.read_h5ad(\"./results/Mice/SS200000108BR_A3A4_scgem.Fibroblast_DPOL/adata.h5ad\")\n",
    "adata_Fb['Mice2'] = sc.read_h5ad(\"./results/Mice/SS200000108BR_A3A4_scgem.Fibroblast_RU6F/adata.h5ad\")\n",
    "adata_Fb['Mice3'] = sc.read_h5ad(\"./results/Mice/SS200000108BR_A3A4_scgem.Fibroblast_UH87/adata.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HOM_MouseHuman = pd.read_csv('./data/HMD_HumanPhenotype.rpt', sep='\\t', header=None)\n",
    "HOM_MouseHuman = HOM_MouseHuman.set_index(2)\n",
    "HOM_MouseHuman = HOM_MouseHuman[~HOM_MouseHuman.index.duplicated()]\n",
    "D_HOM_MouseHuman = pd.Series(HOM_MouseHuman[0]).to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = list(adata_Fb.keys())\n",
    "for sample in keys:\n",
    "    Xs[sample] = adata_Fb[sample].uns['X']\n",
    "    GeneLists[sample] = list(adata_Fb[sample].uns['X'].index)\n",
    "    if 'Mice' in sample:\n",
    "        GeneLists[sample] = [D_HOM_MouseHuman.get(x, x) for x in GeneLists[sample]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build distance matrix\n",
    "N = len(keys)\n",
    "DM_Species = np.zeros((N,N))\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])\n",
    "        index_n = []\n",
    "        index_m = []\n",
    "        for i,v1 in enumerate(intersected_values):\n",
    "            for j,v2 in enumerate(GeneLists[key_n]):\n",
    "                if v1==v2:\n",
    "                    index_n.append(j)\n",
    "                    break\n",
    "        for i,v1 in enumerate(intersected_values):\n",
    "            for j,v2 in enumerate(GeneLists[key_m]):\n",
    "                if v1==v2:\n",
    "                    index_m.append(j)\n",
    "                    break\n",
    "        #index_n.append(np.where(GeneLists[key_n] == g)[0][0])\n",
    "        #index_m.append(np.where(GeneLists[key_m] == g)[0][0])\n",
    "        normalized_x_n = normalizeX(Xs[key_n].iloc[index_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m].iloc[index_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        DM_Species[n,m] = DM_Species[m,n] = min(d1,d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(2.0, 1.6)},font_scale=1)\n",
    "sns.heatmap(DM_Species, xticklabels=keys, yticklabels=keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "clustergrid = sns.clustermap(DM_Species, xticklabels=keys, yticklabels=keys, method='average')\n",
    "#plt.close()\n",
    "#\n",
    "## Get clusters with threshold\n",
    "#linkage = clustergrid.dendrogram_row.linkage\n",
    "#threshold = 4.8\n",
    "#clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "#\n",
    "## Create a color map for clusters\n",
    "#n_clusters = len(np.unique(clusters))\n",
    "#color_map = sns.color_palette(\"rainbow\", n_clusters)\n",
    "#\n",
    "## Assign colors to clusters\n",
    "#cluster_colors = [color_map[c-1] for c in clusters]\n",
    "#sns.set(rc={\"figure.dpi\":300,'figure.figsize':(8, 8)},font_scale=1)\n",
    "#\n",
    "#ax=sns.clustermap(DM_mouse, xticklabels=labels, yticklabels=labels, cmap=\"viridis\", method='ward', row_colors=[cluster_colors], col_colors=[cluster_colors], cbar_kws={'ticks': [0,0.5,1,1.5]})\n",
    "#ax.ax_cbar.set_title(\"NRMSD\")\n",
    "#ax.ax_cbar.set_position((0.07, .84, .04, .12))\n",
    "#\n",
    "## Add legend for row color bar\n",
    "#from matplotlib.patches import Patch\n",
    "#lut = dict(zip(clusters, cluster_colors))\n",
    "#handles=[Patch(facecolor=lut[name]) for name in sorted(lut)]\n",
    "#legend=plt.legend(handles, sorted(lut), title='Clusters', bbox_to_anchor=(0.86, 0.96, .01, .01), bbox_transform=plt.gcf().transFigure, loc='upper right', ncol=2,)\n",
    "#legend.get_frame().set_facecolor('white')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Microglia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8.Pattern of Distance over Time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='CP': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(np.intersect1d, GeneLists.values())\n",
    "boolean_arrays = [np.in1d(a, intersected_genes) for a in GeneLists.values()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "D_pattern = pd.DataFrame()\n",
    "order=0\n",
    "for s, df in Xs.items():\n",
    "    print(s)\n",
    "    distances = pdist(df[boolean_arrays[order]].values, metric='euclidean')\n",
    "    df_dist = pd.DataFrame(squareform(distances), index=intersected_genes, columns=intersected_genes).stack()\n",
    "    df_dist = df_dist.reset_index()\n",
    "    df_dist.columns = ['gene1', 'gene2', 'distance']\n",
    "    inf = s.split('_')\n",
    "    df_dist['stage'] = inf[0]\n",
    "    df_dist['celltype'] = inf[1]\n",
    "    df_dist['genepair'] = df_dist['gene1'] + \",\" + df_dist['gene2']\n",
    "    D_pattern = D_pattern.append(df_dist)\n",
    "    order+=1\n",
    "\n",
    "D_pattern.reset_index(drop=True, inplace=True)\n",
    "matrix_pattern = D_pattern.pivot_table(index='genepair', columns='stage', values='distance')\n",
    "matrix_pattern.to_csv('Axo_dev_CP_pattern.mx.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Mfuzz_CP=pd.read_csv('Mfuzz_membership_AXO_dev_CP.xls', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Mfuzz_CP_sub=Mfuzz_CP[Mfuzz_CP['max'] > 0.7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gs_clusters={}\n",
    "for i in np.unique(Mfuzz_VLMC['Cluster']):\n",
    "    Mfuzz_CP_sub_cluster = Mfuzz_CP_sub[Mfuzz_CP['Cluster'] > i]\n",
    "    gs_clusters[i] = Mfuzz_CP_sub_cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno = pd.read_xlsx(\"./Summary_Gene_Annotation_0707.xls\", sep='\\t',header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate gene\n",
    "anno = pd.read_csv(\"./Summary_Gene_Annotation_0707.xls\", sep='\\t',header=None)\n",
    "df = gtf[gtf.iloc[:,2]==\"gene\"].iloc[:,8].str.split(';',expand=True)\n",
    "\n",
    "GD_Axo = {}\n",
    "for idx, row in df.iterrows():\n",
    "    k = row[0].split(\" \")[1].strip(\"\\\"\")\n",
    "    if \"hs\" in row[1]:\n",
    "        names = row[1].strip(\"gene_name \").strip(\"\\\"\").split(\"|\")\n",
    "        for n in names:\n",
    "            if \"hs\" in n:\n",
    "                v = n.split(\" \")[0]\n",
    "    else:\n",
    "        v = row[1].strip().split(\" \")[1].strip(\"\\\"\")\n",
    "    GD_Axo[k]=v \n",
    "    \n",
    "intersected_symbols=[]\n",
    "for i in intersected_genes:\n",
    "    intersected_symbols.append(GD_Axo[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO on umap results\n",
    "import gseapy as gp\n",
    "enr = pd.DataFrame(columns=['Cluster','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "\n",
    "# Enricr API\n",
    "for value, count in zip(unique_values, counts):\n",
    "        if count >= 10:\n",
    "            try:\n",
    "                genelist = np.array(intersected_symbols)[clusters==value].tolist()\n",
    "                filterlist = list(filter(lambda x: 'AMEX60DD' not in x, genelist))\n",
    "                for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "                    enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                                        organism='Human',\n",
    "                                        gene_sets=gs,\n",
    "                                        outdir=None, \n",
    "                                        cutoff=0.05,\n",
    "                                        no_plot= True)\n",
    "\n",
    "                    # trim (go:...)\n",
    "                    enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "                    enr_up.res2d['Cluster'] = \"sub-conformation \"+str(value)\n",
    "                    enr = pd.concat([enr, enr_up.res2d[enr_up.res2d['Adjusted P-value']<0.05]], ignore_index=True)\n",
    "                    # dotplot\n",
    "                    # gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(value), size=10, cmap = plt.cm.viridis_r)\n",
    "            except ValueError or FileNotFoundError or TypeError as e:\n",
    "                print(value)\n",
    "                print(e)\n",
    "                continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VLMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='VLMC': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(np.intersect1d, GeneLists.values())\n",
    "boolean_arrays = [np.in1d(a, intersected_genes) for a in GeneLists.values()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D_pattern = pd.DataFrame()\n",
    "order=0\n",
    "for s, df in Xs.items():\n",
    "    print(s)\n",
    "    distances = pdist(df[boolean_arrays[order]].values, metric='euclidean')\n",
    "    df_dist = pd.DataFrame(squareform(distances), index=intersected_genes, columns=intersected_genes).stack()\n",
    "    df_dist = df_dist.reset_index()\n",
    "    df_dist.columns = ['gene1', 'gene2', 'distance']\n",
    "    inf = s.split('_')\n",
    "    df_dist['stage'] = inf[0]\n",
    "    df_dist['celltype'] = inf[1]\n",
    "    df_dist['genepair'] = df_dist['gene1'] + \",\" + df_dist['gene2']\n",
    "    D_pattern = D_pattern.append(df_dist)\n",
    "    order+=1\n",
    "\n",
    "D_pattern.reset_index(drop=True, inplace=True)\n",
    "matrix_pattern = D_pattern.pivot_table(index='genepair', columns='stage', values='distance')\n",
    "matrix_pattern.to_csv('Axo_dev_VLMC_pattern.mx.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Mfuzz_VLMC=pd.read_csv('Mfuzz_membership_AXO_dev_VLMC.xls', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Running time and Memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from datetime import datetime\n",
    "\n",
    "def time_to_seconds(time_str):\n",
    "    try:\n",
    "        if len(time_str.split(\":\")) == 3:\n",
    "            # h:mm:ss format\n",
    "            h, m, s = map(int, time_str.split(':'))\n",
    "            total_hours = h + m/60 + s/3600\n",
    "        else:\n",
    "            # m:ss format\n",
    "            t = datetime.strptime(time_str, \"%M:%S.%f\")\n",
    "            total_hours = t.microsecond/(3600*1000000) + t.second/3600 + t.minute/60 \n",
    "        return total_hours\n",
    "    except ValueError:\n",
    "        print(\"Invalid time format\")\n",
    "        return None\n",
    "\n",
    "def read_log_file(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if re.search(r'Command being timed: ', line):\n",
    "                datapath = line.split(\":\")[1].strip().split()[2]\n",
    "                size = os.stat(datapath).st_size/(1024 * 1024)\n",
    "            elif re.search(r'Elapsed \\(wall clock\\) time \\(h:mm:ss or m:ss\\): ', line):\n",
    "                time = time_to_seconds(line.strip().lstrip('Elapsed (wall clock) time (h:mm:ss or m:ss): '))\n",
    "            elif re.search(r'Maximum resident set size \\(kbytes\\): ', line):\n",
    "                memory = int(line.strip().lstrip('Maximum resident set size (kbytes): '))/(1024 * 1024)\n",
    "    log_df = pd.DataFrame([{'size (MB)': size, 'time (h)': time, 'memory (GB)': memory}])\n",
    "    return log_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "log_path = './*.time.log'\n",
    "logs = glob.glob(log_path)\n",
    "log_df = pd.DataFrame(columns=['size (MB)','time (h)', 'memory (GB)'])\n",
    "for l in logs:\n",
    "    try:\n",
    "        df = read_log_file(l)\n",
    "        log_df = pd.concat([log_df,df], ignore_index=True)\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(4, 3)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='deep')\n",
    "ax1 = sns.lineplot(data=log_df, x='size (MB)', y='time (h)', color='r', label='time')\n",
    "ax1.set_xlabel('input data size (MB)')\n",
    "ax1.set_ylabel('Running Time (h)')\n",
    "ax2 = ax1.twinx()\n",
    "sns.lineplot(data=log_df, x='size (MB)', y='memory (GB)', ax=ax2 ,label='memory')\n",
    "ax2.set_ylabel('Max Memory (GB)')\n",
    "# Combine the legends\n",
    "lines, labels = ax1.get_legend_handles_labels()\n",
    "lines2, labels2 = ax2.get_legend_handles_labels()\n",
    "ax1.get_legend().remove()\n",
    "ax2.legend(lines + lines2, labels + labels2, loc=0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(4, 3)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='deep')\n",
    "\n",
    "fig, ax1 = plt.subplots()\n",
    "\n",
    "sns.regplot(data=log_df, x='size (MB)', y='time (h)', ax=ax1, label='time')\n",
    "ax1.set_xlabel('input data size (MB)')\n",
    "ax1.set_ylabel('Running Time (h)')\n",
    "\n",
    "ax2 = ax1.twinx()\n",
    "sns.regplot(data=log_df, x='size (MB)', y='memory (GB)', ax=ax2, color='r', label='memory')\n",
    "ax2.set_ylabel('Max Memory (GB)')\n",
    "\n",
    "# Combine the legends\n",
    "lines, labels = ax1.get_legend_handles_labels()\n",
    "lines2, labels2 = ax2.get_legend_handles_labels()\n",
    "ax2.legend(lines + lines2, labels + labels2, loc=0)\n",
    " \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# setting the plot dimension\n",
    "fig, axs = plt.subplots(1,2, figsize=(10,4), dpi=600)\n",
    "\n",
    "# iterating through each Subplot\n",
    "for i, arg in enumerate(args):\n",
    "        # group by arg, calculate sum of 'mirror', and count of rows (not including NaNs in 'mirror')\n",
    "        grouped = df.groupby([arg]).agg(\n",
    "                mirror_sum=('MIRROR', 'sum'),\n",
    "                count=('MIRROR', 'count')\n",
    "        )\n",
    "\n",
    "        # calculate 'mirror percent'\n",
    "        grouped['mirror_percent'] = grouped['mirror_sum'] / grouped['count']\n",
    "        df_mirror = pd.DataFrame([grouped['mirror_sum'],grouped['count'],grouped['mirror_percent']])\n",
    "        \n",
    "        # creating a strip plot for the subplot\n",
    "        sns.barplot(x=df_mirror.T.index.values, y=df_mirror.T.mirror_percent.values, ax=axs[i], palette='Blues')\n",
    "        \n",
    "        # setting title\n",
    "        #axs[i].set_title(f'Capture Rate: {capturerate*100}% | Drop Rate: {droprate*100}%')\n",
    "        #axs[i].set_xticklabels([\"None\",\"Low\",\"High\"])\n",
    "        axs[i].set_xlabel(arg)\n",
    "        axs[i].set_ylabel(\"Mirror Rate\")\n",
    "        axs[i].set_ylim(0.4,0.6)\n",
    "        #axs[i].legend(loc='upper left', title=\"Cell Number / Resolution\", ncol=3)\n",
    "        \n",
    "# setting the layout\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10.Compare Platforms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "df=pd.DataFrame(columns=['platform', 'Detected Cell Number', 'Detected Gene Number', 'Reconstructed Gene Number', 'Reconstruction Rate'])\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            df_sample=pd.DataFrame({'platform':'stereo-seq', 'Detected Cell Number':adata.n_obs, 'Detected Gene Number':adata.n_vars, 'Reconstructed Gene Number': len(adata.uns['X']), 'Reconstruction Rate': len(adata.uns['X'])/adata.n_vars}, index=[0])\n",
    "            df = pd.concat([df,df_sample], ignore_index=True)\n",
    "            #Xs[sample] = adata.uns['X']\n",
    "            #GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        df_sample=pd.DataFrame({'platform':'CosMx', 'Detected Cell Number':adata.n_obs, 'Detected Gene Number':adata.n_vars, 'Reconstructed Gene Number': len(adata.uns['X']), 'Reconstruction Rate': len(adata.uns['X'])/adata.n_vars}, index=[0])\n",
    "        df = pd.concat([df,df_sample], ignore_index=True)\n",
    "\n",
    "result_path=\"./results/merfish_ileum/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        df_sample=pd.DataFrame({'platform':'MERFISH', 'Detected Cell Number':adata.n_obs, 'Detected Gene Number':adata.n_vars, 'Reconstructed Gene Number': len(adata.uns['F']), 'Reconstruction Rate': len(adata.uns['F'])/adata.n_vars}, index=[0])\n",
    "        df = pd.concat([df,df_sample], ignore_index=True)\n",
    "\n",
    "result_path=\"./results/Xenium_BreastCancer/\"\n",
    "#samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in ['Sample2', 'Sample1R2','Sample1R1']:\n",
    "    sample_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(sample_dir))\n",
    "    for file in files:\n",
    "        result_dir = os.path.join(sample_dir,file)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        ctype = file.split(\".\")[-1][:-5]\n",
    "        key = sample+' '+ctype\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            df_sample=pd.DataFrame({'platform':'Xenium', 'Detected Cell Number':adata.n_obs, 'Detected Gene Number':adata.n_vars, 'Reconstructed Gene Number': len(adata.uns['F']), 'Reconstruction Rate': len(adata.uns['F'])/adata.n_vars}, index=[0])\n",
    "            df = pd.concat([df,df_sample], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sns.set(style=\"whitegrid\", context=\"notebook\", font_scale=1.1, rc={\"figure.figsize\": (12,4), \"figure.dpi\": 300})\n",
    "palette = sns.color_palette(\"Set2\")\n",
    "#args=['Detected Cell Number', 'Detected Gene Number', 'Reconstructed Gene Number', 'Reconstruction Rate']\n",
    "args=['Detected Cell Number', 'Detected Gene Number', 'Reconstructed Gene Number']\n",
    "# setting the plot dimension\n",
    "fig, axs = plt.subplots(nrows=1, ncols=3)\n",
    "\n",
    "# converting the 2D axs array to a 1D array\n",
    "axs = axs.ravel()\n",
    "\n",
    "# iterating through each Subplot\n",
    "for i, arg in enumerate(args):\n",
    "        # creating a strip plot for the subplot\n",
    "        sns.boxplot(x=df.platform, y=df[arg], ax=axs[i], showfliers = False, palette=palette, width=0.5)\n",
    "\n",
    "        # setting title\n",
    "        axs[i].set_title(arg)\n",
    "        axs[i].set_xticklabels([\"stereo-seq\",\"CosMx\",\"MERFISH\",\"Xenium\"])\n",
    "        axs[i].set_xlabel('')\n",
    "        axs[i].set_ylabel('')\n",
    "        #axs[i].set_ylim(0.4,0.6)\n",
    "        #axs[i].legend(frameon=True, fontsize=12, loc='upper left')   \n",
    "# setting the layout\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "\n",
    "args=['Detected Cell Number', 'Detected Gene Number', 'Reconstructed Gene Number', 'Reconstruction Rate']\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "df=pd.DataFrame(columns=['platform', 'arg', 'value'])\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            df_1 = pd.DataFrame({'platform':'stereo-seq', 'arg':'Detected Cell Number','value': adata.n_obs}, index=[0])\n",
    "            df_2 = pd.DataFrame({'platform':'stereo-seq', 'arg':'Detected Gene Number','value': adata.n_vars}, index=[0])\n",
    "            df_3 = pd.DataFrame({'platform':'stereo-seq', 'arg':'Reconstructed Gene Number','value': len(adata.uns['X'])}, index=[0])\n",
    "            df_4 = pd.DataFrame({'platform':'stereo-seq', 'arg':'Reconstruction Rate','value': len(adata.uns['X'])/adata.n_vars}, index=[0])\n",
    "            df = pd.concat([df,df_1,df_2,df_3,df_4], ignore_index=True)\n",
    "            #Xs[sample] = adata.uns['X']\n",
    "            #GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        df_1 = pd.DataFrame({'platform':'CosMx', 'arg':'Detected Cell Number','value': adata.n_obs}, index=[0])\n",
    "        df_2 = pd.DataFrame({'platform':'CosMx', 'arg':'Detected Gene Number','value': adata.n_vars}, index=[0])\n",
    "        df_3 = pd.DataFrame({'platform':'CosMx', 'arg':'Reconstructed Gene Number','value': len(adata.uns['X'])}, index=[0])\n",
    "        df_4 = pd.DataFrame({'platform':'CosMx', 'arg':'Reconstruction Rate','value': len(adata.uns['X'])/adata.n_vars}, index=[0])\n",
    "        df = pd.concat([df,df_1,df_2,df_3,df_4], ignore_index=True)\n",
    "\n",
    "result_path=\"./results/merfish_ileum/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        df_1 = pd.DataFrame({'platform':'MERFISH', 'arg':'Detected Cell Number','value': adata.n_obs}, index=[0])\n",
    "        df_2 = pd.DataFrame({'platform':'MERFISH', 'arg':'Detected Gene Number','value': adata.n_vars}, index=[0])\n",
    "        df_3 = pd.DataFrame({'platform':'MERFISH', 'arg':'Reconstructed Gene Number','value': len(adata.uns['F'])}, index=[0])\n",
    "        df_4 = pd.DataFrame({'platform':'MERFISH', 'arg':'Reconstruction Rate','value': len(adata.uns['F'])/adata.n_vars}, index=[0])\n",
    "        df = pd.concat([df,df_1,df_2,df_3,df_4], ignore_index=True)\n",
    "\n",
    "result_path=\"./results/Xenium_BreastCancer/\"\n",
    "#samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in ['Sample2', 'Sample1R2','Sample1R1']:\n",
    "    sample_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(sample_dir))\n",
    "    for file in files:\n",
    "        result_dir = os.path.join(sample_dir,file)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        ctype = file.split(\".\")[-1][:-5]\n",
    "        key = sample+' '+ctype\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            df_1 = pd.DataFrame({'platform':'Xenium', 'arg':'Detected Cell Number','value': adata.n_obs}, index=[0])\n",
    "            df_2 = pd.DataFrame({'platform':'Xenium', 'arg':'Detected Gene Number','value': adata.n_vars}, index=[0])\n",
    "            df_3 = pd.DataFrame({'platform':'Xenium', 'arg':'Reconstructed Gene Number','value': len(adata.uns['F'])}, index=[0])\n",
    "            df_4 = pd.DataFrame({'platform':'Xenium', 'arg':'Reconstruction Rate','value': len(adata.uns['F'])/adata.n_vars}, index=[0])\n",
    "            df = pd.concat([df,df_1,df_2,df_3,df_4], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sns.set(style=\"whitegrid\", context=\"notebook\", font_scale=1, rc={\"figure.dpi\":300,'figure.figsize':(9, 3)})\n",
    "palette = sns.color_palette(\"Set2\")\n",
    "args=['Detected Cell Number', 'Detected Gene Number', 'Reconstructed Gene Number', 'Reconstruction Rate']\n",
    "\n",
    "# creating a strip plot for the subplot\n",
    "g = sns.boxplot(data=df, x='arg', y='value', hue='platform', showfliers = False, palette=palette, width=0.5)\n",
    "\n",
    "# setting title\n",
    "#g.set_title(arg)\n",
    "#g.set_xticklabels([\"stereo-seq\",\"CosMx\",\"MERFISH\",\"Xenium\"])\n",
    "g.set_xlabel('')\n",
    "g.set_ylabel('')\n",
    "#g.set_ylim(0.4,0.6)\n",
    "#g.legend(frameon=True, fontsize=12, loc='upper left')   \n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "sns.set(style=\"whitegrid\", context=\"notebook\")\n",
    "palette = sns.color_palette(\"Set2\")\n",
    "args=['Detected Cell Number', 'Detected Gene Number', 'Reconstructed Gene Number', 'Reconstruction Rate']\n",
    "# setting the plot dimension\n",
    "fig, axs = plt.subplots(nrows=1, ncols=4, figsize=(16,3), dpi=300)\n",
    "\n",
    "# converting the 2D axs array to a 1D array\n",
    "axs = axs.ravel()\n",
    "\n",
    "# iterating through each Subplot\n",
    "for i, arg in enumerate(args):\n",
    "        # creating a strip plot for the subplot\n",
    "        sns.boxplot(x=df.platform, y=df[arg], ax=axs[i], showfliers = False, palette=palette, width=0.5)\n",
    "\n",
    "        # setting title\n",
    "        axs[i].set_title(arg)\n",
    "        axs[i].set_xticklabels([\"stereo-seq\",\"CosMx\",\"MERFISH\",\"Xenium\"])\n",
    "        axs[i].set_xlabel('')\n",
    "        axs[i].set_ylabel('')\n",
    "        #axs[i].set_ylim(0.4,0.6)\n",
    "        #axs[i].legend(frameon=True, fontsize=12, loc='upper left')   \n",
    "# setting the layout\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parent conformations (split unlikely cells)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "adatas={}\n",
    "\n",
    "#result_path=\"./results/Mice/\"\n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from cytocraft.craft import mean_radius\n",
    "# Preallocate a list to store the data\n",
    "data = []\n",
    "\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.concatenate((adatas[sample].uns['W'][::2, :], adatas[sample].uns['W'][1::2, :]), axis=0)\n",
    "    Z = Z.reshape(2, -1).T.reshape(ngene, ncell, 2)\n",
    "    XR = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'])[:, :, :2]\n",
    "    \n",
    "    # Move the loop over cells into a single, vectorized operation.\n",
    "    isnan_mask = ~np.isnan(Z).any(axis=2)\n",
    "    xr = np.where(isnan_mask[:,:,None], XR, np.nan)\n",
    "    z = np.where(isnan_mask[:,:,None], Z, np.nan)\n",
    "    err = np.nansum(np.linalg.norm(xr-z, axis=2), axis=0) / (ngene*mean_radius(adatas[sample].uns['X']))\n",
    "    \n",
    "    # Extend the data list instead of repeatedly concatenating DataFrames.\n",
    "    data.extend([(sample, e) for e in err])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['configuration', 'error'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "g = sns.FacetGrid(diffs, col=\"configuration\", col_wrap=5, sharex=False)\n",
    "g.map(plt.hist, \"error\", alpha=.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs={}\n",
    "for sample in list(adatas.keys()):\n",
    "    ngene = int(adatas[sample].uns['W'].shape[0]/2)\n",
    "    ncell = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.concatenate((adatas[sample].uns['W'][::2,:], adatas[sample].uns['W'][1::2,:]), axis=0)\n",
    "    Z = Z.reshape(2, -1).T.reshape(ncell,ngene,2)\n",
    "    XR=np.dot(adatas[sample].uns['X'],adatas[sample].obsm['Rotation'])[:,:,:2]\n",
    "    diffs[sample]=[]\n",
    "    for c in range(XR.shape[1]):\n",
    "        xr = XR[:,c,:][~np.isnan(Z[:,c,:]).any(axis=1)]\n",
    "        z = Z[:,c,:][~np.isnan(Z[:,c,:]).any(axis=1)]\n",
    "        diffs[sample].append(np.sum([np.linalg.norm(x) for x in xr-z]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming your dictionary is named data_dict\n",
    "fig, axs = plt.subplots(25, 3, figsize=(15,30))  # You can set the figure size as per your convenience\n",
    "\n",
    "axs = axs.ravel()  # To flatten the axis array\n",
    "\n",
    "for i, (key, val) in enumerate(diffs.items()):\n",
    "    axs[i].hist(val, bins=20, alpha=0.5, color='g')  # You can set the number of bins as per your requirement\n",
    "    axs[i].set_title(key)\n",
    "\n",
    "plt.tight_layout()  # To prevent overlapping of subplots\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "gmm = GaussianMixture(n_components=2)\n",
    "X=np.array(diffs['Epithelial_cell_1G3J']).reshape(-1,1)\n",
    "gmm.fit(X)\n",
    "# retrieve the parameters\n",
    "weights = gmm.weights_\n",
    "means = gmm.means_\n",
    "covariances = gmm.covariances_\n",
    "\n",
    "print('weights: ', weights)\n",
    "print('means: ', means)\n",
    "print('covariances ', covariances)\n",
    "predictions = gmm.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "celllist=adatas['Epithelial_cell_1G3J'].obs_names.values[np.array(diffs['Epithelial_cell_1G3J']) > 1781]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(celllist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cytocraft\n",
    "gem = cytocraft.craft.read_gem_as_csv('./data/Mice/E16.5_E1S3/SS200000108BR_A3A4_scgem.Epithelial_cell.csv',sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gem['MIDCount']=gem.MIDCount.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gem.loc[gem['CellID'].isin(celllist)].to_csv('./results/Mice/SS200000108BR_A3A4_scgem.Epithelial_cell_1G3J/SS200000108BR_A3A4_scgem.Epithelial_cell.unlike.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### test convergence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata1 = adatas['Epithelial_cell_1G3J']\n",
    "adata2 = sc.read_h5ad('./results/Mice/SS200000108BR_A3A4_scgem.Epithelial_cell_1G3J/SS200000108BR_A3A4_scgem.Epithelial_cell.unlike_DKR3/adata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersected_values = np.intersect1d(adata1.uns['X'].index.values, adata2.uns['F'].index.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "boolean_arrays_n = np.in1d(adata1.uns['X'].index.values, intersected_values)\n",
    "boolean_arrays_m = np.in1d(adata2.uns['F'].index.values, intersected_values)\n",
    "X_n = adata1.uns['X'].values[boolean_arrays_n]\n",
    "X_m = adata2.uns['F'].values[boolean_arrays_m]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rmsd = cytocraft.craft.numpy_svd_rmsd_rot(X_n, X_m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rmsd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lungs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and 'Lung5' in chip: \n",
    "        adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs={}\n",
    "for sample in list(adatas.keys()):\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0]/2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.concatenate((adatas[sample].uns['W'][::2,:], adatas[sample].uns['W'][1::2,:]), axis=0)\n",
    "    Z = Z.reshape(2, -1).T.reshape(ngene,ncell,2)\n",
    "    XR=np.dot(adatas[sample].uns['X'],adatas[sample].obsm['Rotation'])[:,:,:2]\n",
    "    diffs[sample]=[]\n",
    "    for c in range(XR.shape[1]):\n",
    "        xr = XR[:,c,:][~np.isnan(Z[:,c,:]).any(axis=1)]\n",
    "        z = Z[:,c,:][~np.isnan(Z[:,c,:]).any(axis=1)]\n",
    "        diffs[sample].append(np.sum([np.linalg.norm(x) for x in xr-z]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Assuming your dictionary is named data_dict\n",
    "fig, axs = plt.subplots(20, 3, figsize=(15,30))  # You can set the figure size as per your convenience\n",
    "\n",
    "axs = axs.ravel()  # To flatten the axis array\n",
    "\n",
    "for i, (key, val) in enumerate(diffs.items()):\n",
    "    axs[i].hist(val, bins=20, alpha=0.5, color='g')  # You can set the number of bins as per your requirement\n",
    "    axs[i].set_title(key)\n",
    "\n",
    "plt.tight_layout()  # To prevent overlapping of subplots\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### GMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.mixture import GaussianMixture\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "gmm = GaussianMixture(n_components=2)\n",
    "X=np.array(diffs['Lung5_Rep3 tumor_5']).reshape(-1,1)\n",
    "gmm.fit(X)\n",
    "# retrieve the parameters\n",
    "weights = gmm.weights_\n",
    "means = gmm.means_\n",
    "covariances = gmm.covariances_\n",
    "\n",
    "print('weights: ', weights)\n",
    "print('means: ', means)\n",
    "print('covariances ', covariances)\n",
    "predictions = gmm.predict(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas['Lung5_Rep3 tumor_5'].n_obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11. Error evaluation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Cyrocraft with randomly selected cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cytocraft.craft as cc\n",
    "gem_allcells = cc.read_gem_as_csv('./data/Mice/E16.5_E1S3/SS200000108BR_A3A4_scgem.csv',sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import cytocraft.craft as cc\n",
    "\n",
    "def process_iteration(i):\n",
    "    # Randomly select 10000 cells\n",
    "    random_cells = gem_allcells['CellID'].drop_duplicates().sample(n=10000, random_state=i)\n",
    "    subset_gem_allcells = gem_allcells[gem_allcells['CellID'].isin(random_cells)]\n",
    "    subset_gem_allcells.to_csv(f'./results/Mice_rand10k/data/SS200000108BR_A3A4_scgem.rand1k_{i}.csv', sep='\\t', index=False)\n",
    "\n",
    "    # Run craft\n",
    "    #gem_path = f'./results/Mice_rand1k/data/SS200000108BR_A3A4_scgem.rand1k_{i}.csv'\n",
    "    #cc.run_craft(gem_path, species='Mice', seed=42, outpath=f'./results/Mice_rand1k/rand1k_{i}/', sn=f\"rand1k_{i}\", threshold_for_rmsd=0.25, n_anchor=10)\n",
    "\n",
    "# Use ThreadPoolExecutor to parallelize the process\n",
    "with ThreadPoolExecutor(max_workers=10) as executor:\n",
    "    executor.map(process_iteration, range(1, 101))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "\n",
    "log_files = glob.glob('./results/Mice_rand1k/rand1k_{i}/*.log')\n",
    "\n",
    "for log_file in log_files:\n",
    "    with open(log_file, 'r') as file:\n",
    "        log_content = file.read()\n",
    "        print(log_content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# randomly select 1000 cells\n",
    "random_cells = gem_allcells['CellID'].drop_duplicates().sample(n=1000, random_state=42)\n",
    "subset_gem_allcells = gem_allcells[gem_allcells['CellID'].isin(random_cells)]\n",
    "subset_gem_allcells.to_csv('./data/Mice/E16.5_E1S3/SS200000108BR_A3A4_scgem.rand1k_1.csv', sep='\\t', index=False)\n",
    "\n",
    "# run craft\n",
    "gem_path='./data/Mice/E16.5_E1S3/SS200000108BR_A3A4_scgem.rand1k.csv'\n",
    "gem = cc.read_gem_as_csv(gem_path, sep='\\t')\n",
    "adata = cc.read_gem_as_adata(gem_path, sep='\\t', SN='rand1k')\n",
    "GeneUIDs = cc.get_GeneUID(gem)\n",
    "adata = cc.craft(subset_gem_allcells, adata, species='Mice', seed=42, outpath='./results/Mice_rand1k_1/', samplename=\"rand1k\", thresh_rmsd=0.25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "\n",
    "# optimized\n",
    "ncell = adata.uns['Z'].shape[0] // 2\n",
    "ngene = adata.uns['Z'].shape[1]\n",
    "Z = np.zeros((ncell * 2, ngene))\n",
    "W = adata.uns['Z']\n",
    "for c in range(ncell):\n",
    "    XY = np.dot(adata.uns['F'], adata.obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "    # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "    Z[c * 2] = XY[:, 0]\n",
    "    Z[c * 2 + 1] = XY[:, 1]\n",
    "mse = np.nanmean((W - Z) ** 2)\n",
    "data.append([\"rand1k\",'Optimized', mse]) \n",
    "# shuffled\n",
    "mse_shuffled=[]\n",
    "for _ in range(100):\n",
    "    Z_shuffled = W.copy()\n",
    "    np.random.shuffle(Z_shuffled)\n",
    "    mse_shuffled.append(np.nanmean((W - Z_shuffled) ** 2))\n",
    "mean_mse_shuffled = np.mean(mse_shuffled)\n",
    "data.append([\"rand1k\",'Random', mean_mse_shuffled])\n",
    "# random from all samples\n",
    "#mse_random_all=[]\n",
    "#for _ in range(100):\n",
    "#    all_cells_selected = all_cells[:,np.isin(list(all_genes), list(adata.uns['X'].index))]\n",
    "#    cell_index = np.random.choice(np.arange(all_cells.shape[0] // 2), ncell, replace=False)\n",
    "#    Z_random = all_cells_selected[np.concatenate([cell_index*2, cell_index*2 + 1])]\n",
    "#    mse_random_all.append(np.nanmean((W - Z_random) ** 2))\n",
    "#mean_mse_random_all = np.mean(mse_random_all)\n",
    "#data.append([\"rand1k\",'Random from all cell types', mean_mse_random_all])\n",
    "\n",
    "\n",
    "# Build the final DataXrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'source', 'error'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MSE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Multi sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Mice_onesample/\"\n",
    "#result_path=\"./results/Mice_forfigs/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#all_genes = set().union(*[adata.uns['X'].index for adata in adatas.values()])\n",
    "#all_cells = []\n",
    "#for adata in adatas.values():\n",
    "#    W = adata.uns['W']\n",
    "#    W_full = np.full((W.shape[0], len(all_genes)), np.nan)\n",
    "#    gene_indices = [list(all_genes).index(gene) for gene in adata.uns['X'].index if gene in all_genes]\n",
    "#    W_full[:, gene_indices] = W\n",
    "#    all_cells.append(W_full)\n",
    "#all_cells = np.concatenate(all_cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(adatas.keys()):\n",
    "    # optimized\n",
    "    ncell = adatas[sample].uns['W'].shape[0] // 2\n",
    "    ngene = adatas[sample].uns['W'].shape[1]\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    W = adatas[sample].uns['W']\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    mse = np.nanmean((W - Z) ** 2)\n",
    "    data.append([sample[:-5],'Optimized', mse]) \n",
    "    # shuffled\n",
    "    mse_shuffled=[]\n",
    "    for _ in range(100):\n",
    "        Z_shuffled = W.copy()\n",
    "        np.random.shuffle(Z_shuffled)\n",
    "        mse_shuffled.append(np.nanmean((W - Z_shuffled) ** 2))\n",
    "    mean_mse_shuffled = np.mean(mse_shuffled)\n",
    "    data.append([sample[:-5],'Random', mean_mse_shuffled])\n",
    "    # random from all samples\n",
    "    #mse_random_all=[]\n",
    "    #for _ in range(100):\n",
    "    #    all_cells_selected = all_cells[:,np.isin(list(all_genes), list(adatas[sample].uns['X'].index))]\n",
    "    #    cell_index = np.random.choice(np.arange(all_cells.shape[0] // 2), ncell, replace=False)\n",
    "    #    Z_random = all_cells_selected[np.concatenate([cell_index*2, cell_index*2 + 1])]\n",
    "    #    mse_random_all.append(np.nanmean((W - Z_random) ** 2))\n",
    "    #mean_mse_random_all = np.mean(mse_random_all)\n",
    "    #data.append([sample[:-5],'Random from all cell types', mean_mse_random_all])\n",
    "\n",
    "\n",
    "# Build the final DataXrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'source', 'error'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs.to_csv('./results/Mice_onesample/MSE_diffs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs = pd.read_csv('./results/Mice_onesample/MSE_diffs.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import ttest_ind\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# Separate the errors by source\n",
    "optimized_errors = diffs[diffs['source'] == 'Optimized']['error']\n",
    "random_errors = diffs[diffs['source'] == 'Random']['error']\n",
    "\n",
    "# Perform the t-test\n",
    "t_stat, p_value = ttest_ind(optimized_errors, random_errors)\n",
    "\n",
    "# Adjust the p-value for multiple comparisons\n",
    "adjusted_p_value = multipletests(p_value, method='bonferroni')[1][0]\n",
    "\n",
    "print(f\"T-statistic: {t_stat}, P-value: {p_value}, Adjusted P-value: {adjusted_p_value}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs['sample'] = diffs['sample'].replace({'Cardiomyocyte': 'Cardiomyocyte', 'Chondrocyte': 'Chondrocyte', 'Choroid_plexus': 'Choroid plexus', \n",
    "'Dorsal_midbrain_neuron': 'Dorsal Mb neuron', 'Endothelial_cell': 'Endothelial cell', \n",
    "'Epithelial_cell': 'Epithelial cell', 'Erythrocyte': 'Erythrocyte', 'Facial_fibroblast': 'Facial fibroblast', \n",
    "'Fibroblast': 'Fibroblast', 'Forebrain_neuron': 'Fb neuron', 'Forebrain_radial_glia_cell': 'Fb radial glia cell', \n",
    "'Ganglion': 'Ganglion', 'Hepatocyte': 'Hepatocyte', 'Immune_cell': 'Immune cell', 'Keratinocyte': 'Keratinocyte', \n",
    "'Limb_fibroblast': 'Limb fibroblast', 'Macrophage': 'Macrophage', 'Meninges_cell': 'Meninges cell', 'Mid-hindbrain_and_spinal_cord_neuron': 'Mb/Hb/SpC neuron', \n",
    "'Myoblast': 'Myoblast', 'Olfactory_epithelial_cell': 'Olfactory epithelial cell', 'Radial_glia_cell': 'Radial glia cell', \n",
    "'Smooth_muscle_cell': 'Smooth muscle cell', 'Spinal_cord_neuron': 'SpC neuron', 'Thalamus_neuron': 'Thalamus neuron'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs = diffs.drop_duplicates(subset=['sample', 'source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Calculate the mean optimized MSE for each sample\n",
    "mean_optimized_mse = diffs[diffs['source'] == 'Optimized'].groupby('sample')['error'].mean()\n",
    "\n",
    "# Sort the samples based on the mean optimized MSE in increasing order\n",
    "sorted_samples = mean_optimized_mse.sort_values().index\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(8, 1.5)}, font_scale=1)\n",
    "sns.set_theme(style='ticks', palette='pastel')\n",
    "\n",
    "# Create the bar plot with the specified order for the 'source' and sorted x-axis labels\n",
    "ax = sns.barplot(data=diffs, x='sample', y='error', hue='source', hue_order=['Random', 'Optimized'], palette='Paired', order=sorted_samples)\n",
    "\n",
    "# Set the x labels to be centered\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=90, ha='center')\n",
    "ax.set_ylabel('MSE')\n",
    "ax.set_ylim(min(diffs.error)-10, max(diffs.error)+10)\n",
    "ax.set_title('')\n",
    "\n",
    "handles, labels = plt.gca().get_legend_handles_labels()\n",
    "order = [1, 0]\n",
    "plt.legend([handles[idx] for idx in order], ['Optimized','Random'])\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "\n",
    "adatas={}\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(adatas.keys()):\n",
    "    # optimized\n",
    "    sn, celltype = sample.split(' ')\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    mse = np.nanmean((adatas[sample].uns['W'] - Z) ** 2)\n",
    "    data.append([sn, celltype, 'Optimized', mse]) \n",
    "    # random\n",
    "    mse_random=[]\n",
    "    for _ in range(10):\n",
    "        Z_shuffled = adatas[sample].uns['W'].copy()\n",
    "        np.random.shuffle(Z_shuffled)\n",
    "        mse_random.append(np.nanmean((adatas[sample].uns['W'] - Z_shuffled) ** 2))\n",
    "    mean_mse_random = np.mean(mse_random)\n",
    "    data.append([sn, celltype, 'Random', mean_mse_random])\n",
    "\n",
    "# Build the final DataXrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'celltype', 'source', 'error'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs['celltype'] = diffs['celltype'].str.replace('_', ' ')\n",
    "diffs['sample'] = diffs['sample'].str.replace('_', ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(8, 4)}, font_scale=1)\n",
    "sns.set_theme(style='ticks', palette='pastel')\n",
    "\n",
    "# Define the order of the stages\n",
    "stage_order = ['Lung5 Rep1', 'Lung5 Rep2', 'Lung5 Rep3', 'Lung6', 'Lung9 Rep1', 'Lung9 Rep2', 'Lung12', 'Lung13']\n",
    "\n",
    "# Create a FacetGrid for each sample\n",
    "g = sns.FacetGrid(diffs, col=\"sample\", col_wrap=4, sharey=False, aspect=1.5, col_order=stage_order)\n",
    "\n",
    "# Map the swarmplot to each facet\n",
    "g.map_dataframe(sns.swarmplot, x='celltype', y='error', hue='source', palette='Paired')\n",
    "\n",
    "# Set the x labels to be centered and rotated\n",
    "for ax in g.axes.flat:\n",
    "    for label in ax.get_xticklabels():\n",
    "        label.set_rotation(90)\n",
    "        label.set_ha('center')\n",
    "\n",
    "# Set the title\n",
    "g.fig.suptitle('Human NSCLC Dataset', y=1.02)\n",
    "\n",
    "# Set the y-axis label\n",
    "g.set_axis_labels(y_var='MSE')\n",
    "\n",
    "# Adjust the legend\n",
    "g.add_legend(title='')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Axolotls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+' '+ctype\n",
    "        if 'adata.h5ad' in files:\n",
    "                adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(adatas.keys()):\n",
    "    # optimized\n",
    "    sn, celltype = sample.split(' ')\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    mse = np.nanmean((adatas[sample].uns['W'] - Z) ** 2)\n",
    "    data.append([sn, celltype, 'Optimized', mse]) \n",
    "    # random\n",
    "    mse_random=[]\n",
    "    for _ in range(10):\n",
    "        Z_shuffled = adatas[sample].uns['W'].copy()\n",
    "        np.random.shuffle(Z_shuffled)\n",
    "        mse_random.append(np.nanmean((adatas[sample].uns['W'] - Z_shuffled) ** 2))\n",
    "    mean_mse_random = np.mean(mse_random)\n",
    "    data.append([sn, celltype, 'Random', mean_mse_random])\n",
    "\n",
    "# Build the final DataXrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'celltype', 'source', 'error'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(8, 4)}, font_scale=1)\n",
    "sns.set_theme(style='ticks', palette='pastel')\n",
    "\n",
    "# Define the order of the stages\n",
    "stage_order = ['Stage44', 'Stage54', 'Stage57', 'Control_Juv', 'Adult', 'Meta']\n",
    "\n",
    "# Create a FacetGrid for each sample with the specified order\n",
    "g = sns.FacetGrid(diffs, col=\"sample\", col_wrap=3, sharey=False, aspect=2, col_order=stage_order)\n",
    "\n",
    "# Map the swarmplot to each facet\n",
    "g.map_dataframe(sns.swarmplot, x='celltype', y='error', hue='source', palette='Paired')\n",
    "\n",
    "# Set the x labels to be centered and rotated\n",
    "for ax in g.axes.flat:\n",
    "    for label in ax.get_xticklabels():\n",
    "        label.set_rotation(90)\n",
    "        label.set_ha('center')\n",
    "\n",
    "# Set the title\n",
    "g.fig.suptitle('Axolotl Development Dataset', y=1.02)\n",
    "\n",
    "# Set the y-axis label\n",
    "g.set_axis_labels(y_var='MSE')\n",
    "\n",
    "# Adjust the legend\n",
    "g.add_legend(title='')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### One sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "adatas={}\n",
    "\n",
    "#result_path=\"./results/Mice_forfigs/\"\n",
    "result_path=\"/./results/Mice/SS200000108BR_A3A4_scgem.Keratinocyte_QL9M\"\n",
    "files = sorted(os.listdir(result_path))\n",
    "for file in files:\n",
    "    if file.endswith(\".h5ad\") and 'loop' in file:\n",
    "        file_name = file.split('_')[1]\n",
    "        adatas[file_name] = sc.read_h5ad(os.path.join(result_path, file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    loop = sample\n",
    "    ncell = int(adatas[sample].uns['Z'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['Z'].shape[1])\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['F'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z).reshape(-1, ngene*2), axis=1)\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z), axis=0)\n",
    "    mse = np.nanmean((adatas[sample].uns['Z'] - Z) ** 2)\n",
    "    #E = adatas[sample].uns['Z'] - Z\n",
    "    #E = E.reshape((ncell, 2, ngene))[:,:,:10]\n",
    "    #norms = np.linalg.norm(E, axis=1)\n",
    "    #error = norms.flatten()\n",
    "    data.append([loop, mse])  # Ensure each element is a list or tuple\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['loop', 'error'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _ in range(10):\n",
    "    Z_shuffled = adatas[sample].uns['Z'].copy()\n",
    "    np.random.shuffle(Z_shuffled)\n",
    "    mse_random.append(np.nanmean((adatas[sample].uns['Z'] - Z_shuffled) ** 2))\n",
    "mean_mse_random = np.mean(mse_random)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.nanmax(adatas[sample].uns['Z'] - Z_shuffled), np.nanmin(adatas[sample].uns['Z'] - Z_shuffled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.nanmax(adatas[sample].uns['Z']), np.nanmin(adatas[sample].uns['Z'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.nanmax(Z), np.nanmin(Z)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(18, 4)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "# colors\n",
    "# Define the color map\n",
    "#colorlst=[]\n",
    "#cmap = plt.get_cmap(\"rainbow\")\n",
    "## Get the number of colors in the color map\n",
    "#ncolors=len(np.unique(diffs['loop']))\n",
    "#rgb_values=[]\n",
    "#for i in range(ncolors):\n",
    "#    rgb_values.append(cmap(i*30))\n",
    "## Print the RGB values\n",
    "#for i, rgb in enumerate(rgb_values):\n",
    "#    colorlst.append(mcolors.rgb2hex(rgb))\n",
    "\n",
    "# Order the x-axis labels from 0 to the largest number\n",
    "diffs['loop'] = diffs['loop'].astype(int)\n",
    "diffs = diffs.sort_values('loop')\n",
    "\n",
    "ax = sns.boxplot(data=diffs, x='loop', y='error', showfliers=False, width=0.5, saturation=1)\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=0, ha='center', rotation_mode='anchor')\n",
    "ax.set_ylabel('MSE')\n",
    "ax.set_title('Mice Epithelial_cell')\n",
    "#plt.legend(loc='upper left', ncol=2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### spatial distance error"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Mice "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Mice_forfigs/\"\n",
    "#result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "samples={}\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    if sample[:-5] not in list(samples.keys()): \n",
    "        rep='Rep1'\n",
    "        samples[sample[:-5]]=1\n",
    "    elif sample[:-5] in list(samples.keys()) and samples[sample[:-5]]==1:\n",
    "        rep='Rep2'\n",
    "        samples[sample[:-5]]+=1\n",
    "    elif sample[:-5] in list(samples.keys()) and samples[sample[:-5]]==2:\n",
    "        rep='Rep3'\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z).reshape(-1, ngene*2), axis=1)\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z), axis=0)\n",
    "    E = adatas[sample].uns['W'] - Z\n",
    "    E = E.reshape((ncell, 2, ngene))[:,:,:10]\n",
    "    norms = np.linalg.norm(E, axis=1)\n",
    "    error = norms.flatten()\n",
    "    data.extend([(sample[:-5], rep, e*0.5) for e in error])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['celltype', 'configuration', 'error'])\n",
    "\n",
    "# rectangular box plot\n",
    "# group_means=diffs.groupby(['celltype'])['error'].median().sort_values(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs['celltype'] = diffs['celltype'].replace({'Cardiomyocyte': 'Cardiomyocyte', 'Chondrocyte': 'Chondrocyte', 'Choroid_plexus': 'Choroid plexus', \n",
    "'Dorsal_midbrain_neuron': 'Dorsal Mb neuron', 'Endothelial_cell': 'Endothelial cell', \n",
    "'Epithelial_cell': 'Epithelial cell', 'Erythrocyte': 'Erythrocyte', 'Facial_fibroblast': 'Facial fibroblast', \n",
    "'Fibroblast': 'Fibroblast', 'Forebrain_neuron': 'Fb neuron', 'Forebrain_radial_glia_cell': 'Fb radial glia cell', \n",
    "'Ganglion': 'Ganglion', 'Hepatocyte': 'Hepatocyte', 'Immune_cell': 'Immune cell', 'Keratinocyte': 'Keratinocyte', \n",
    "'Limb_fibroblast': 'Limb fibroblast', 'Macrophage': 'Macrophage', 'Meninges_cell': 'Meninges cell', 'Mid-hindbrain_and_spinal_cord_neuron': 'Mb/Hb/SpC neuron', \n",
    "'Myoblast': 'Myoblast', 'Olfactory_epithelial_cell': 'Olfactory epithelial cell', 'Radial_glia_cell': 'Radial glia cell', \n",
    "'Smooth_muscle_cell': 'Smooth muscle cell', 'Spinal_cord_neuron': 'SpC neuron', 'Thalamus_neuron': 'Thalamus neuron'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors as mcolors\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(18, 4)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "# Calculate the mean Distance for each ct\n",
    "mean_distance = diffs.groupby('celltype')['error'].median()\n",
    "# Sort the ct based on the mean Distance\n",
    "order = mean_distance.sort_values(ascending=False).index\n",
    "\n",
    "# colors\n",
    "# Define the color map\n",
    "colorlst=[]\n",
    "cmap = plt.get_cmap(\"viridis_r\")\n",
    "# Get the number of colors in the color map\n",
    "ncolors=3\n",
    "rgb_values=[]\n",
    "for i in range(ncolors):\n",
    "    rgb_values.append(cmap(i*10+10))\n",
    "# Print the RGB values\n",
    "for i, rgb in enumerate(rgb_values):\n",
    "    colorlst.append(mcolors.rgb2hex(rgb))\n",
    "\n",
    "palette =  dict(zip(np.unique(diffs.configuration),colorlst))\n",
    "\n",
    "ax = sns.boxplot(data=diffs, x='celltype', y='error', showfliers = False, width=0.5, saturation=1, order=order)\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right', rotation_mode='anchor')\n",
    "ax.set_ylabel('Spatial Discrepancy of Top 10 Genes (m)')\n",
    "ax.set_title('Mice Embryo')\n",
    "plt.legend(loc='upper left', ncol=1, title='')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "\n",
    "adatas={}\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files:\n",
    "        adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    chip, ct = sample.split()\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z).reshape(-1, ngene*2), axis=1)\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z), axis=0)\n",
    "    E = adatas[sample].uns['W'] - Z\n",
    "    E = E.reshape((ncell, 2, ngene))[:,:,:10]\n",
    "    norms = np.linalg.norm(E, axis=1)\n",
    "    error = norms.flatten()\n",
    "    data.extend([(chip, ct, e*0.5) for e in error])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'configuration', 'error'])\n",
    "\n",
    "# rectangular box plot\n",
    "group_means=diffs.groupby(['configuration'])['error'].median().sort_values(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(18, 4)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "# colors\n",
    "# Define the color map\n",
    "colorlst=[]\n",
    "cmap = plt.get_cmap(\"rainbow\")\n",
    "# Get the number of colors in the color map\n",
    "ncolors=len(np.unique(diffs['sample']))\n",
    "rgb_values=[]\n",
    "for i in range(ncolors):\n",
    "    rgb_values.append(cmap(i*30))\n",
    "# Print the RGB values\n",
    "for i, rgb in enumerate(rgb_values):\n",
    "    colorlst.append(mcolors.rgb2hex(rgb))\n",
    "\n",
    "ax = sns.boxplot(data=diffs, x='configuration', y='error', showfliers = False, width=0.5, saturation=1, palette=colorlst, \n",
    "    hue='sample', hue_order=['Lung5_Rep1', 'Lung5_Rep2', 'Lung5_Rep3', 'Lung6', 'Lung9_Rep1', 'Lung9_Rep2', 'Lung12', 'Lung13'],\n",
    "    order=['B-cell', 'NK', 'T_CD4_naive', 'T_CD4_memory', 'T_CD8_naive', 'T_CD8_memory', 'Treg', 'endothelial', 'epithelial', 'fibroblast',\n",
    "       'mDC', 'macrophage', 'mast', 'monocyte', 'neutrophil', 'pDC', 'plasmablast', 'tumor_5', 'tumor_6', 'tumor_9', 'tumor_12', 'tumor_13'])\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right', rotation_mode='anchor')\n",
    "ax.set_ylabel('Spatial Discrepancy of Top 10 Genes (m)')\n",
    "ax.set_title('NSCLC')\n",
    "plt.legend(loc='upper left', ncol=2)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Axolotls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+' '+ctype\n",
    "        if 'adata.h5ad' in files:\n",
    "                adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(adatas.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    chip, ct = sample.split()\n",
    "    ncell = int(adatas[sample].uns['W'].shape[0] / 2)\n",
    "    ngene = int(adatas[sample].uns['W'].shape[1])\n",
    "    Z = np.zeros((ncell * 2, ngene))\n",
    "    for c in range(ncell):\n",
    "        XY = np.dot(adatas[sample].uns['X'], adatas[sample].obsm['Rotation'][c, :, :])[:, 0:2]\n",
    "        # XY = np.dot(randRM[c, :, :], simX.T).T[:, 0:2]\n",
    "        Z[c * 2] = XY[:, 0]\n",
    "        Z[c * 2 + 1] = XY[:, 1]\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z).reshape(-1, ngene*2), axis=1)\n",
    "    # error = np.nansum(np.abs(adatas[sample].uns['W']-Z), axis=0)\n",
    "    E = adatas[sample].uns['W'] - Z\n",
    "    E = E.reshape((ncell, 2, ngene))[:,:,:10]\n",
    "    norms = np.linalg.norm(E, axis=1)\n",
    "    error = norms.flatten()\n",
    "    data.extend([(chip, ct, e*0.5) for e in error])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'configuration', 'error'])\n",
    "\n",
    "# rectangular box plot\n",
    "#group_means=diffs.groupby(['configuration'])['error'].median().sort_values(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors as mcolors\n",
    "sns.set(rc={\"figure.dpi\":300, 'figure.figsize':(18, 4)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "# colors\n",
    "# Define the color map\n",
    "colorlst=[]\n",
    "cmap = plt.get_cmap(\"rainbow\")\n",
    "# Get the number of colors in the color map\n",
    "ncolors=6\n",
    "rgb_values=[]\n",
    "for i in range(ncolors):\n",
    "    rgb_values.append(cmap(i*40))\n",
    "# Print the RGB values\n",
    "for i, rgb in enumerate(rgb_values):\n",
    "    colorlst.append(mcolors.rgb2hex(rgb))\n",
    "\n",
    "ax = sns.boxplot(data=diffs, x='configuration', y='error', showfliers = False, width=0.5, saturation=1, palette=colorlst, \n",
    "    order=['CMPN', 'CP', 'Immature', 'MCG', 'MSN', 'Oligo', 'Unknown', 'VLMC',\n",
    "            'cckIN', 'dEGC', 'dNBL1', 'dNBL2', 'dNBL3', 'dNBL4', 'dNBL5',\n",
    "            'dpEX', 'mpEX', 'mpIN', 'nptxEX', 'npyIN', 'ntng1IN', 'ribEGC',\n",
    "            'scgnIN', 'sfrpEGC', 'sstIN', 'tlNBL', 'wntEGC'], \n",
    "    hue='sample', hue_order=['Stage44', 'Stage54', 'Stage57','Control_Juv', 'Adult', 'Meta'])\n",
    "\n",
    "# adding horizontal grid lines\n",
    "ax.yaxis.grid(True)\n",
    "ax.set_xlabel('')\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right', rotation_mode='anchor')\n",
    "ax.set_ylabel('Spatial Discrepancy of Top 10 Genes (m)')\n",
    "ax.set_title('Axolotls Telencephalon')\n",
    "plt.legend(loc='upper left', title='')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Expression Similarity vs. Configuration Similarity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scipy.cluster.hierarchy as sch\n",
    "\n",
    "\n",
    "def distance_matrix_from_dendrogram(Exp, method=\"average\", metric=\"euclidean\"):\n",
    "    linkage = sch.linkage(Exp, method=method, metric=metric)\n",
    "    dendrogram = sch.dendrogram(linkage, no_plot=True)\n",
    "    num_leaves = len(dendrogram['leaves'])\n",
    "    distance_matrix = np.zeros((num_leaves, num_leaves))\n",
    "\n",
    "    for i, leaf_i in enumerate(dendrogram['leaves']):\n",
    "        for j, leaf_j in enumerate(dendrogram['leaves']):\n",
    "            if i != j:\n",
    "                distance_matrix[i, j] = dendrogram['dcoord'][j, 2] - dendrogram['dcoord'][i, 2]\n",
    "\n",
    "    return distance_matrix\n",
    "\n",
    "def expression_similarity(adatas, ngene, method=\"pearson\", compare_method=\"pair\"):\n",
    "    GeneLists = {}\n",
    "    for ct, adata in adatas.items():\n",
    "        if \"F\" in adata.uns.keys():\n",
    "            GeneLists[ct] = adata.uns[\"F\"].index\n",
    "        elif \"X\" in adata.uns.keys():\n",
    "            GeneLists[ct] = adata.uns[\"X\"].index\n",
    "        else:\n",
    "            print(f\"Please run Cytocraft for {ct} first\")\n",
    "    if method == \"pearson\":\n",
    "        similarity_matrix = np.zeros((len(adatas), len(adatas)))\n",
    "        for i, sample1 in enumerate(adatas.keys()):\n",
    "            for j, sample2 in enumerate(adatas.keys()):\n",
    "                if compare_method == \"pair\":\n",
    "                    intersected_values = np.intersect1d(\n",
    "                        GeneLists[sample1], GeneLists[sample2]\n",
    "                    )\n",
    "                    # print(\"number of common gene: \" + str(len(intersected_values)))\n",
    "                    intersected_values = intersected_values[:ngene]\n",
    "                elif compare_method == \"complete\":\n",
    "                    intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "                if len(intersected_values) < ngene:\n",
    "                    print(\n",
    "                        f\"Warning: {len(intersected_values)} common genes between {sample1} and {sample2} are less than {ngene}\"\n",
    "                    )\n",
    "                boolean_arrays_n = np.in1d(GeneLists[sample1], intersected_values)\n",
    "                boolean_arrays_m = np.in1d(GeneLists[sample2], intersected_values)\n",
    "                average_expression1 = np.mean(\n",
    "                    adatas[sample1].X[:, boolean_arrays_n], axis=0\n",
    "                )\n",
    "                average_expression2 = np.mean(\n",
    "                    adatas[sample2].X[:, boolean_arrays_m], axis=0\n",
    "                )\n",
    "                similarity = np.corrcoef(average_expression1, average_expression2)[0, 1]\n",
    "                similarity_matrix[i, j] = similarity\n",
    "        return similarity_matrix\n",
    "    # elif method == \"dendrogram\":\n",
    "    #    Exp = np.zeros((len(adatas), ngene))\n",
    "    #    for i, sample in enumerate(adatas.keys()):\n",
    "    #        if not compare_method == \"complate\":\n",
    "    #            print(\n",
    "    #                \"only 'complete' compare method is supported if using 'dendrogram' similarity\"\n",
    "    #            )\n",
    "    #        intersected_values = reduce(np.intersect1d, GeneLists.values())[:ngene]\n",
    "    #        if len(intersected_values) < ngene:\n",
    "    #            print(\n",
    "    #                f\"Warning: {len(intersected_values)} common genes between samples are less than {ngene}\"\n",
    "    #            )\n",
    "    #        boolean_arrays = np.in1d(GeneLists[sample], intersected_values)\n",
    "    #        Exp[i] = np.mean(adatas[sample].X[:, boolean_arrays], axis=0)\n",
    "    #    similarity_matrix = dendrogram_similarity(\n",
    "    #        Exp, method=\"average\", metric=\"euclidean\"\n",
    "    #    )\n",
    "    #    return similarity_matrix\n",
    "    else:\n",
    "        print(\"method not supported\")\n",
    "        return\n",
    "\n",
    "def dendrogram_distance(Exp, method=\"average\", metric=\"euclidean\"):\n",
    "    linkage = sch.linkage(Exp, method=method, metric=metric)\n",
    "    dendrogram = sch.dendrogram(linkage, no_plot=True)\n",
    "    num_leaves = len(dendrogram['leaves'])\n",
    "    distance_matrix = np.zeros((num_leaves, num_leaves))\n",
    "\n",
    "    for i, leaf_i in enumerate(dendrogram['leaves']):\n",
    "        for j, leaf_j in enumerate(dendrogram['leaves']):\n",
    "            if i != j:\n",
    "                print(len(dendrogram['dcoord']))\n",
    "                distance_matrix[i, j] = tuple(dendrogram['dcoord'])[j][2] - tuple(dendrogram['dcoord'])[i][2]\n",
    "\n",
    "    return distance_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "adatas={}\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Lung5' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = chip+' '+ctype\n",
    "        if 'adata.h5ad' in files : \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            if len(adata.uns['X'].index) > 150:\n",
    "                adatas[sample] = adata\n",
    "                Xs[sample] = adata.uns['X']\n",
    "                GeneLists[sample] = adata.uns['X'].index\n",
    "                print(sample)\n",
    "                print(adata.n_obs)\n",
    "#        else:\n",
    "#            print(sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### conf similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "topgene = 20\n",
    "Conf_DstMx_lung5 = RMSD_distance_matrix(adatas, ngene=topgene)\n",
    "Conf_SimMx_lung5 = np.divide((np.max(Conf_DstMx_lung5) - Conf_DstMx_lung5), np.max(Conf_DstMx_lung5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### exp similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#from cytocraft.craft import expression_similarity\n",
    "Pearson_Exp_SimMx_lung5 = expression_similarity(adatas, ngene=topgene, method='pearson', compare_method=\"pair\")\n",
    "#Dendrogram_Exp_SimMx_lung5 = expression_similarity_celltype(adatas, ngene=topgene, method='dendrogram', compare_method=\"complete\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add labels for each dot\n",
    "labels=list(Xs.keys())\n",
    "Conf_SimMx_lung5[labels.index('Lung5_Rep2 endothelial'),labels.index('Lung5_Rep3 tumor_5')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Pearson_Exp_SimMx_lung5[labels.index('Lung5_Rep2 endothelial'),labels.index('Lung5_Rep3 tumor_5')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "np.fill_diagonal(Pearson_Exp_SimMx_lung5, np.nan)\n",
    "np.fill_diagonal(Conf_SimMx_lung5, np.nan)\n",
    "\n",
    "\n",
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (20,20)})\n",
    "sns.set_theme(style='white', palette='pastel', font_scale=0.5)\n",
    "\n",
    "# Create the scatter plot\n",
    "g = sns.scatterplot(x=Conf_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), y=Pearson_Exp_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), s=0.5)\n",
    "\n",
    "g.set_xlim(0,1)\n",
    "g.set_ylim(0.15,1)\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Configuration Similarity')\n",
    "plt.ylabel('Expression Correlation')\n",
    "\n",
    "\n",
    "Lidx=np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)\n",
    "for (idx_x, idx_y) in zip(Lidx[0],Lidx[1]):\n",
    "    label1 = labels[idx_x]\n",
    "    label2 = labels[idx_y]\n",
    "    plt.annotate(f'{label1} {label2}', (Conf_SimMx_lung5[idx_x, idx_y], Pearson_Exp_SimMx_lung5[idx_x, idx_y]), fontsize=4)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "np.fill_diagonal(Pearson_Exp_SimMx_lung5, np.nan)\n",
    "np.fill_diagonal(Conf_SimMx_lung5, np.nan)\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (2,2)})\n",
    "sns.set_theme(style='white', palette='pastel', font_scale=0.5)\n",
    "\n",
    "# Create the scatter plot\n",
    "g = sns.kdeplot(x=Conf_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), y=Pearson_Exp_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), fill=True, cmap='Spectral_r', cbar=True)\n",
    "\n",
    "g.set_xlim(-0.1,1)\n",
    "g.set_ylim(-0.5,1.1)\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Configuration Similarity')\n",
    "plt.ylabel('Expression Correlation')\n",
    "plt.title('CosMx NSCLC Dataset\\n(Top 20 genes)')\n",
    "\n",
    "\n",
    "# Show the legend\n",
    "# plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### BC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "adatas={}\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Xenium_BreastCancer/\"\n",
    "#samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in ['Sample2', 'Sample1R2','Sample1R1']:\n",
    "    sample_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(sample_dir))\n",
    "    for file in files:\n",
    "        result_dir = os.path.join(sample_dir,file)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        ctype = file.split(\".\")[-1][:-5]\n",
    "        key = sample+' '+ctype\n",
    "        if 'adata.h5ad' in files: \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[key] = adata\n",
    "            Xs[key] = adata.uns['F']\n",
    "            GeneLists[key] = adata.uns['F'].index\n",
    "        else:\n",
    "            print(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### conf similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topgene = 20\n",
    "Conf_DstMx_lung5 = RMSD_distance_matrix(adatas, ngene=topgene)\n",
    "Conf_SimMx_lung5 = np.divide((np.max(Conf_DstMx_lung5) - Conf_DstMx_lung5), np.max(Conf_DstMx_lung5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### exp correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Pearson_Exp_SimMx_lung5 = expression_similarity(adatas, ngene=topgene, method='pearson', compare_method=\"pair\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (2,2)})\n",
    "sns.set_theme(style='white', palette='pastel', font_scale=0.5)\n",
    "\n",
    "# Create the scatter plot\n",
    "g = sns.kdeplot(x=Conf_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), y=Pearson_Exp_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), fill=True, cmap='Spectral_r', cbar=True)\n",
    "g.set_xlim(-0.1,1)\n",
    "g.set_ylim(-0.5,1)\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Configuration Similarity')\n",
    "plt.ylabel('Expression Correlation')\n",
    "plt.title('Xenium Breast Cancer Dataset\\n(Top 20 genes)')\n",
    "\n",
    "\n",
    "# Show the legend\n",
    "# plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "#result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[sample] = adata\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### conf similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topgene = 20\n",
    "Conf_DstMx_lung5 = RMSD_distance_matrix(adatas, ngene=topgene)\n",
    "Conf_SimMx_lung5 = np.divide((np.max(Conf_DstMx_lung5) - Conf_DstMx_lung5), np.max(Conf_DstMx_lung5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### exp correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Pearson_Exp_SimMx_lung5 = expression_similarity(adatas, ngene=topgene, method='pearson', compare_method=\"pair\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_data(Pearson_Exp_SimMx_lung5,'Pearson_Exp_SimMx_lung5_thresh0.5.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Pearson_Exp_SimMx_lung5 = load_data('Pearson_Exp_SimMx_lung5_thresh0.5.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "\n",
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (2,2)})\n",
    "sns.set_theme(style='white', palette='pastel', font_scale=0.5)\n",
    "\n",
    "# Create the scatter plot\n",
    "g = sns.kdeplot(x=Conf_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), y=Pearson_Exp_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), fill=True, cmap='Spectral_r', cbar=True)\n",
    "\n",
    "g.set_xlim(-0.1,1)\n",
    "g.set_ylim(-0.6,1)\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Configuration Similarity')\n",
    "plt.ylabel('Expression Correlation')\n",
    "plt.title('Stereo-seq MOSTA Dataset\\n(Top 20 genes)')\n",
    "\n",
    "\n",
    "# Show the legend\n",
    "# plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ileum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/merfish_ileum/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    ctype = info[-1][:-5]\n",
    "    seed = info[-1][-4:]\n",
    "    if 'adata.h5ad' in files and seed=='FQ1Z' and not ctype in ['Removed','ICC','Tuft','B_Follicular_Circulating','Myenteric_Plexus']: \n",
    "        adatas[ctype] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[ctype] = adatas[ctype].uns['F']\n",
    "        GeneLists[ctype] = adatas[ctype].uns['F'].index\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topgene = 15\n",
    "# configuration similarity\n",
    "Conf_DstMx_lung5 = RMSD_distance_matrix(adatas, ngene=topgene)\n",
    "Conf_SimMx_lung5 = np.divide((np.max(Conf_DstMx_lung5) - Conf_DstMx_lung5), np.max(Conf_DstMx_lung5))\n",
    "# expression similarity\n",
    "Pearson_Exp_SimMx_lung5 = expression_similarity(adatas, ngene=topgene, method='pearson', compare_method=\"pair\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (2,2)})\n",
    "sns.set_theme(style='white', palette='pastel', font_scale=0.5)\n",
    "\n",
    "# Create the scatter plot\n",
    "g = sns.kdeplot(x=Conf_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), y=Pearson_Exp_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), fill=True, cmap='Spectral_r', cbar=True)\n",
    "\n",
    "g.set_xlim(0,1)\n",
    "g.set_ylim(-1,1.1)\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Configuration Similarity')\n",
    "plt.ylabel('Expression Correlation')\n",
    "plt.title('MERFISH Ileum Dataset')\n",
    "\n",
    "\n",
    "# Show the legend\n",
    "# plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Axolotls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and sample[0].isalpha(): \n",
    "        adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topgene = 20\n",
    "# configuration similarity\n",
    "Conf_DstMx_lung5 = RMSD_distance_matrix(adatas, ngene=topgene)\n",
    "Conf_SimMx_lung5 = np.divide((np.max(Conf_DstMx_lung5) - Conf_DstMx_lung5), np.max(Conf_DstMx_lung5))\n",
    "# expression similarity\n",
    "Pearson_Exp_SimMx_lung5 = expression_similarity(adatas, ngene=topgene, method='pearson', compare_method=\"pair\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\": 300, \"figure.figsize\": (2,2)})\n",
    "sns.set_theme(style='white', palette='pastel', font_scale=0.5)\n",
    "\n",
    "# Create the scatter plot\n",
    "g = sns.kdeplot(x=Conf_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), y=Pearson_Exp_SimMx_lung5[np.triu_indices(Conf_SimMx_lung5.shape[0],k=1)].flatten(), fill=True, cmap='Spectral_r', cbar=True)\n",
    "\n",
    "g.set_xlim(-0.1,1)\n",
    "g.set_ylim(-0.5,1.1)\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Configuration Similarity')\n",
    "plt.ylabel('Expression Correlation')\n",
    "plt.title('Stereo-seq Axolotls Dataset\\n(Top 20 genes)')\n",
    "\n",
    "\n",
    "# Show the legend\n",
    "# plt.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 13. Relation of gene distance and gene expression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expression_similarity_gene(adata):\n",
    "    GeneList = list(adata.uns['X'].index)\n",
    "    similarity_matrix = np.zeros((len(GeneList), len(GeneList)))  # Initialize similarity_matrix with the correct shape\n",
    "    for i, gene1 in enumerate(tqdm(GeneList)):\n",
    "        for j, gene2 in enumerate(GeneList[: i + 1]):\n",
    "            idx1 = adata.var_names.get_loc(gene1)\n",
    "            idx2 = adata.var_names.get_loc(gene2)\n",
    "            similarity = np.corrcoef(adata.X[:, idx1].todense(), adata.X[:, idx2].todense())[0, 1]\n",
    "            similarity_matrix[i, j] = similarity_matrix[j, i] = similarity  # Update assignment of similarity values \n",
    "\n",
    "    return similarity_matrix\n",
    "\n",
    "def fast_expression_similarity_gene(adata):\n",
    "    from scipy.sparse import csr_matrix\n",
    "    \n",
    "    gene_index = {gene: idx for idx, gene in enumerate(adata.var_names)}\n",
    "    \n",
    "    X = adata.X.todense() if isinstance(adata.X, csr_matrix) else adata.X\n",
    "    X = (X - np.mean(X, axis=0)) / np.std(X, axis=0)   # Standardize the columns (genes)\n",
    "    \n",
    "    similarity_matrix = X.T @ X / (X.shape[0] - 1)   # Compute the correlation matrix\n",
    "    \n",
    "    GeneList = adata.uns['X'].index.tolist()\n",
    "    similarity_matrix = similarity_matrix[[gene_index[gene] for gene in GeneList], :]  # Reorder rows\n",
    "    similarity_matrix = similarity_matrix[:, [gene_index[gene] for gene in GeneList]]  # Reorder columns\n",
    "    \n",
    "    return similarity_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "adatas={}\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Lung5' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = chip+' '+ctype\n",
    "        if 'adata.h5ad' in files : \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            if len(adata.uns['X'].index) > 150:\n",
    "                adatas[sample] = adata\n",
    "                Xs[sample] = adata.uns['X']\n",
    "                GeneLists[sample] = adata.uns['X'].index\n",
    "        else:\n",
    "            print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Calculate the number of rows and columns for the subplots grid\n",
    "num_plots = len(adatas)\n",
    "num_rows = int(num_plots / 7) + (num_plots % 7 > 0)\n",
    "num_cols = min(num_plots, 7)\n",
    "\n",
    "# Create the subplots grid\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(20, 20))\n",
    "\n",
    "# Iterate over the adatas dictionary and plot each adata object\n",
    "for i, (key, adata) in enumerate(adatas.items()):\n",
    "    row = i // num_cols\n",
    "    col = i % num_cols\n",
    "    ax = axes[row, col] if num_plots > 1 else axes\n",
    "    dist_mtx = gene_gene_distance_matrix(adata.uns['X'])\n",
    "    exp_sim = fast_expression_similarity_gene(adata)\n",
    "    ax.scatter(dist_mtx[np.triu_indices(dist_mtx.shape[0], k=1)].flatten(), exp_sim[np.triu_indices(exp_sim.shape[0], k=1)].flatten().A1, s=1)\n",
    "    ax.set_xlabel('Gene Distance')\n",
    "    ax.set_ylabel('Expression Correlation')\n",
    "    ax.set_title(key)\n",
    "\n",
    "# Adjust the spacing between subplots\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the combined plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Mice/\"\n",
    "#result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            #print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[sample] = adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Calculate the number of rows and columns for the subplots grid\n",
    "num_plots = len(adatas)\n",
    "num_rows = int(num_plots / 5) + (num_plots % 5 > 0)\n",
    "num_cols = min(num_plots, 5)\n",
    "\n",
    "# Create the subplots grid\n",
    "fig, axes = plt.subplots(num_rows, num_cols, figsize=(20, 20))\n",
    "\n",
    "# Iterate over the adatas dictionary and plot each adata object\n",
    "for i, (key, adata) in enumerate(adatas.items()):\n",
    "    row = i // num_cols\n",
    "    col = i % num_cols\n",
    "    ax = axes[row, col] if num_plots > 1 else axes\n",
    "    dist_mtx = gene_gene_distance_matrix(adata.uns['X'])\n",
    "    exp_sim = fast_expression_similarity_gene(adata)\n",
    "    ax.scatter(dist_mtx[np.triu_indices(dist_mtx.shape[0], k=1)].flatten(), exp_sim[np.triu_indices(exp_sim.shape[0], k=1)].flatten().A1, s=1)\n",
    "    ax.set_xlabel('Gene Distance')\n",
    "    ax.set_ylabel('Expression Correlation')\n",
    "    ax.set_title(key)\n",
    "\n",
    "# Adjust the spacing between subplots\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the combined plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 14.Genie3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lung"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "\n",
    "adatas={}\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Lung/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Lung5' in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[-1][:-5]\n",
    "        sample = chip+' '+ctype\n",
    "        if 'adata.h5ad' in files : \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            if len(adata.uns['X'].index) > 150:\n",
    "                adatas[sample] = adata\n",
    "                Xs[sample] = adata.uns['X']\n",
    "                GeneLists[sample] = adata.uns['X'].index\n",
    "                print(sample)\n",
    "                print(adata.n_obs)\n",
    "#        else:\n",
    "#            print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "samples=list(Xs.keys())\n",
    "for sample in samples:\n",
    "    adata=adatas[sample]\n",
    "    genes=list(adata.uns['X'].index)\n",
    "    subset_adata = adata[:,genes]\n",
    "    sample=sample.replace(' ','_')\n",
    "    np.savetxt('./Genie3/'+str(sample)+'.exp.csv',subset_adata.X.todense(), delimiter=',',header=','.join(genes), fmt='%f', comments='', )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### save matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample='Lung5_Rep1 T_CD4_memory'\n",
    "adata=adatas[sample]\n",
    "genes=list(adata.uns['X'].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_adata = adata[:,genes]\n",
    "np.savetxt(sample+'.exp.csv',subset_adata.X.todense(),delimiter=',',header=','.join(genes), fmt='%f', comments='', )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Genie3\n",
    "df = pd.DataFrame(columns=['sample','gene1', 'gene2', 'distance', 'source'])\n",
    "samples=list(Xs.keys())\n",
    "for sample in samples:\n",
    "    adata=adatas[sample]\n",
    "    genes=list(adata.uns['X'].index)\n",
    "    subset_adata = adata[:,genes]\n",
    "    sample_legal=sample.replace(' ','_')\n",
    "    links = pd.read_csv('./Genie3/Lung/'+sample_legal+'.exp.csv.LinkList.csv')\n",
    "\n",
    "    # remove the '.' in gene names\n",
    "    pair_ct = links[['regulatoryGene','targetGene']]\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('.', '-')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('.', '-')\n",
    "\n",
    "    # statistics\n",
    "    import itertools \n",
    "    # Genie3 pairs\n",
    "    gene1 = adata.uns['X'].loc[pair_ct.regulatoryGene].to_numpy() # convert to numpy array\n",
    "    gene2 = adata.uns['X'].loc[pair_ct.targetGene].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':pair_ct.regulatoryGene,'gene2':pair_ct.targetGene,'distance':dist,'source':'Genie3 Pairs'})\n",
    "    df = df.append(s1, ignore_index=True)\n",
    "    # all regulatory pairs\n",
    "    pathways = pd.read_csv(\"./pathways.tsv\", sep='\\t', index_col=0)\n",
    "    pairs = pathways[pathways.species==\"Human\"][['src','dest']]\n",
    "    delete_idx = [idx for idx, row in pairs.iterrows() if row[0] not in genes or row[1] not in genes]\n",
    "    pairs.drop(delete_idx, inplace=True)\n",
    "    del delete_idx\n",
    "    print(sample+' '+str(len(pairs)))\n",
    "    gene1 = adata.uns['X'].loc[pairs.src].to_numpy() # convert to numpy array\n",
    "    gene2 = adata.uns['X'].loc[pairs.dest].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':pairs.src,'gene2':pairs.dest,'distance':dist,'source':'Regularity Pairs'})\n",
    "    df = df.append(s1, ignore_index=True)\n",
    "    # all pairs\n",
    "    all_pairs = np.array(list(itertools.combinations(genes, 2))) # convert to numpy array\n",
    "    gene1 = adata.uns['X'].loc[all_pairs[:,0]].to_numpy() # convert to numpy array\n",
    "    gene2 = adata.uns['X'].loc[all_pairs[:,1]].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':all_pairs[:,0],'gene2':all_pairs[:,1],'distance':dist,'source':'All Pairs'})\n",
    "    df = df.append(s1, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_pairs = []\n",
    "for s in samples:\n",
    "    ls_pairs.append(((s, 'Regularity Pairs'), (s, 'Genie3 Pairs')))\n",
    "    ls_pairs.append(((s, 'Regularity Pairs'), (s, 'All Pairs')))  \n",
    "    ls_pairs.append(((s, 'Genie3 Pairs'), (s, 'All Pairs')))  \n",
    "\n",
    "import seaborn as sns \n",
    "from statannot import add_stat_annotation\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(40, 5)},font_scale=2)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "ax = sns.boxplot(data=df, x='sample', y='distance', hue='source', showfliers=False, width=0.7, saturation=1)\n",
    "add_stat_annotation(ax, data=df, x='sample', y='distance', hue='source', box_pairs=ls_pairs, loc='outside', test=\"Mann-Whitney\", text_format=\"star\", verbose=2)\n",
    "\n",
    "#ax.set_xticklabels([\"Regulatory Pairs\",\"All Pairs\"])\n",
    "#ax.legend(title=' ', loc='upper right', labels=['regulatory','all'])\n",
    "\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"Euclidean Distance\", fontsize=16)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xticklabels(labels=df['sample'].drop_duplicates(), rotation=90, fontsize=16)\n",
    "\n",
    "#handles, labels = ax.get_legend_handles_labels()\n",
    "#new_labels = ['Regulatory Gene Pairs', 'All Gene Pairs']\n",
    "#plt.legend(handles, new_labels, fontsize=16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import copy\n",
    "import itertools \n",
    "samples = list(Xs.keys())\n",
    "E = pd.DataFrame(columns=['sample','gene1', 'gene2', 'distance', 'source'])\n",
    "pair_ct={}\n",
    "for sample in samples:\n",
    "    # Pairs in celltype\n",
    "    GeneList = GeneLists[sample]\n",
    "    pair_ct[sample] = copy.deepcopy(pairs)\n",
    "    delete_idx = [idx for idx, row in pair_ct[sample].iterrows() if row[0] not in GeneList or row[1] not in GeneList]\n",
    "    pair_ct[sample].drop(delete_idx, inplace=True)\n",
    "    del delete_idx\n",
    "    print(sample+' '+str(len(pair_ct[sample])))\n",
    "    # Physical Distance\n",
    "    # regulatory pairs\n",
    "    gene1 = Xs[sample].loc[pair_ct[sample].src].to_numpy() # convert to numpy array\n",
    "    gene2 = Xs[sample].loc[pair_ct[sample].dest].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':pair_ct[sample].src,'gene2':pair_ct[sample].dest,'distance':dist,'source':'rgl'})\n",
    "    E = E.append(s1, ignore_index=True)\n",
    "    # all pairs\n",
    "    all_pairs = np.array(list(itertools.combinations(GeneList, 2))) # convert to numpy array\n",
    "    gene1 = Xs[sample].loc[all_pairs[:,0]].to_numpy() # convert to numpy array\n",
    "    gene2 = Xs[sample].loc[all_pairs[:,1]].to_numpy() # convert to numpy array\n",
    "    dist = np.linalg.norm(gene1-gene2, axis=1) # calculate distance vector\n",
    "    s1 = pd.DataFrame({'sample':sample,'gene1':all_pairs[:,0],'gene2':all_pairs[:,1],'distance':dist,'source':'all'})\n",
    "    E = E.append(s1, ignore_index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mice"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### save matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import RMSD_distance_matrix\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "Xs={}\n",
    "result_path=\"./results/Mice_onesample/\"\n",
    "#result_path=\"./results/Mice/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if 'Brain' not in sample:\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        sample = sample.split('.')[1]\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        if 'adata.h5ad' in files: \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[sample] = adata\n",
    "            Xs[sample]=adata.uns['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "samples=list(Xs.keys())\n",
    "for sample in samples:\n",
    "    adata=adatas[sample]\n",
    "    genes=list(adata.uns['X'].index)\n",
    "    subset_adata = adata[:,genes]\n",
    "    #sample=sample.replace(' ','_')\n",
    "    np.savetxt('./Genie3/Mice/'+str(sample)+'.exp.csv',subset_adata.X.todense(), delimiter=',',header=','.join(genes), fmt='%f', comments='', )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig.3j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run Genie3\n",
    "df = pd.DataFrame(columns=['sample','gene1', 'gene2', 'distance', 'source'])\n",
    "samples = list(Xs.keys())\n",
    "for sample in samples:\n",
    "    adata = adatas[sample]\n",
    "    genes = list(adata.uns['X'].index)\n",
    "    subset_adata = adata[:, genes]\n",
    "    sample_legal = sample.replace(' ', '_')\n",
    "    links = pd.read_csv('./Genie3/Mice/' + sample_legal + '.exp.csv.LinkList.csv')\n",
    "\n",
    "    # remove the '.' in gene names\n",
    "    pair_ct = links[['regulatoryGene', 'targetGene']]\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('.', '-')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('.', '-')\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('AC154200-1', 'AC154200.1')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('AC154200-1', 'AC154200.1')\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('X4930481A15Rik', '4930481A15Rik')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('X4930481A15Rik', '4930481A15Rik')\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('X9330179D12Rik', '9330179D12Rik')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('X9330179D12Rik', '9330179D12Rik')\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('X2600014E21Rik', '2600014E21Rik')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('X2600014E21Rik', '2600014E21Rik')\n",
    "    pair_ct['regulatoryGene'] = pair_ct['regulatoryGene'].str.replace('X1700007G11Rik', '1700007G11Rik')\n",
    "    pair_ct['targetGene'] = pair_ct['targetGene'].str.replace('X1700007G11Rik', '1700007G11Rik')\n",
    "\n",
    "    # statistics\n",
    "    import itertools\n",
    "    # Genie3 pairs\n",
    "    gene1 = adata.uns['X'].loc[pair_ct.regulatoryGene].to_numpy()\n",
    "    gene2 = adata.uns['X'].loc[pair_ct.targetGene].to_numpy()\n",
    "    dist = np.linalg.norm(gene1 - gene2, axis=1)\n",
    "    s1 = pd.DataFrame({'sample': sample, 'gene1': pair_ct.regulatoryGene, 'gene2': pair_ct.targetGene, 'distance': dist, 'source': 'Genie3 Pairs'})\n",
    "    df = pd.concat([df, s1], ignore_index=True)\n",
    "    # all regulatory pairs\n",
    "    pathways = pd.read_csv(\"./pathways.tsv\", sep='\\t', index_col=0)\n",
    "    pairs = pathways[pathways.species == \"Mouse\"][['src', 'dest']]\n",
    "    delete_idx = [idx for idx, row in pairs.iterrows() if row[0] not in genes or row[1] not in genes]\n",
    "    pairs.drop(delete_idx, inplace=True)\n",
    "    del delete_idx\n",
    "    print(sample + ' ' + str(len(pairs)))\n",
    "    gene1 = adata.uns['X'].loc[pairs.src].to_numpy()\n",
    "    gene2 = adata.uns['X'].loc[pairs.dest].to_numpy()\n",
    "    dist = np.linalg.norm(gene1 - gene2, axis=1)\n",
    "    s1 = pd.DataFrame({'sample': sample, 'gene1': pairs.src, 'gene2': pairs.dest, 'distance': dist, 'source': 'Regularity Pairs'})\n",
    "    df = pd.concat([df, s1], ignore_index=True)\n",
    "    # all pairs\n",
    "    all_pairs = np.array(list(itertools.combinations(genes, 2)))\n",
    "    gene1 = adata.uns['X'].loc[all_pairs[:, 0]].to_numpy()\n",
    "    gene2 = adata.uns['X'].loc[all_pairs[:, 1]].to_numpy()\n",
    "    dist = np.linalg.norm(gene1 - gene2, axis=1)\n",
    "    s1 = pd.DataFrame({'sample': sample, 'gene1': all_pairs[:, 0], 'gene2': all_pairs[:, 1], 'distance': dist, 'source': 'All Pairs'})\n",
    "    df = pd.concat([df, s1], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls_pairs = []\n",
    "for s in samples:\n",
    "    ls_pairs.append(((s, 'Regularity Pairs'), (s, 'Genie3 Pairs')))\n",
    "    ls_pairs.append(((s, 'Regularity Pairs'), (s, 'All Pairs')))  \n",
    "    ls_pairs.append(((s, 'Genie3 Pairs'), (s, 'All Pairs')))  \n",
    "\n",
    "import seaborn as sns \n",
    "from statannot import add_stat_annotation\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(16, 4.65)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "ax = sns.boxplot(data=df, x='sample', y='distance', hue='source', showfliers=False, width=0.7, saturation=1,\n",
    "    palette=[\"#0fcfc0\",\"#9cded6\", \"#ef9708\"])\n",
    "    \n",
    "add_stat_annotation(ax, data=df, x='sample', y='distance', hue='source', box_pairs=ls_pairs, loc='outside', test=\"Mann-Whitney\", text_format=\"star\", verbose=2)\n",
    "\n",
    "#ax.set_xticklabels([\"Regulatory Pairs\",\"All Pairs\"])\n",
    "#ax.legend(title=' ', loc='upper right', labels=['regulatory','all'])\n",
    "\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"Euclidean Distance\", fontsize=16)\n",
    "plt.yticks(fontsize=16)\n",
    "ax.set_xticklabels(labels=['Cardiomyocyte',\n",
    "'Chondrocyte',\n",
    "'Choroid plexus',\n",
    "'Dorsal Mb neuron',\n",
    "'Endothelial cell',\n",
    "'Epithelial cell',\n",
    "'Erythrocyte',\n",
    "'Facial fibroblast',\n",
    "'Fibroblast',\n",
    "'Fb neuron',\n",
    "'Fb radial glia cell',\n",
    "'Ganglion',\n",
    "'Hepatocyte',\n",
    "'Immune cell',\n",
    "'Keratinocyte',\n",
    "'Limb fibroblast',\n",
    "'Macrophage',\n",
    "'Meninges cell',\n",
    "'Mb/Hb/SpC neuron',\n",
    "'Myoblast',\n",
    "'Olfactory epithelial cell',\n",
    "'Radial glia cell',\n",
    "'Smooth muscle cell',\n",
    "'SpC neuron',\n",
    "'Thalamus neuron'], rotation=90, fontsize=16)\n",
    "\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "new_labels = ['Predicted Regulatory Gene Pairs', 'Known Regulatory Gene Pairs', 'Random Gene Pairs']\n",
    "plt.legend(handles, new_labels, fontsize=14)\n",
    "plt.savefig('Fig3j.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 15. prove cell-type configuration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fig.3b convergence rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Define the directory paths\n",
    "directories = ['./results/Mice_onesample/', './results/Mice_rand1k_sh/']\n",
    "df = pd.DataFrame(columns=['Sample Name', 'Source', 'Cell Number', 'Gene Number', 'Gene Number for Rotation Derivatives', 'loop', 'RMSD'])\n",
    "\n",
    "# Iterate over each directory\n",
    "for directory in directories:\n",
    "    if 'Mice_onesample' in directory:\n",
    "        for root, dirs, files in os.walk(directory):\n",
    "            # Iterate over each file in the current directory\n",
    "            for file in files:\n",
    "                # Check if the file is a log file\n",
    "                if file.endswith('.log'):\n",
    "                    # Create the file path\n",
    "                    file_path = os.path.join(root, file)\n",
    "                    # Read the contents of the log file\n",
    "                    with open(file_path, 'r') as f:\n",
    "                        log_contents = f.read()\n",
    "                        # Extract relevant information from log_contents\n",
    "                        dict = {}\n",
    "                        loop = 0\n",
    "                        for line in log_contents.strip().split('\\n'):\n",
    "                            if ':' in line:\n",
    "                                key, value = line.split(':')\n",
    "                                dict[key.strip()] = value.strip()\n",
    "                                # Calculate RMSD\n",
    "                                if 'Distance' in line and loop <= 9:\n",
    "                                    loop += 1\n",
    "                                    RMSD = float(line.split(':')[1].strip())\n",
    "                                    entry = pd.DataFrame([{\n",
    "                                        'Sample Name': dict.get('Sample Name').split('.')[1],\n",
    "                                        'Source': 'Cell type',\n",
    "                                        'Cell Number': dict.get('Cell Number'),\n",
    "                                        'Gene Number': dict.get('Gene Number'),\n",
    "                                        'Gene Number for Rotation Derivatives': dict.get('Number of genes used for Rotation Derivation is'),\n",
    "                                        'loop': loop,\n",
    "                                        'RMSD': RMSD\n",
    "                                    }])\n",
    "                                    df = pd.concat([df, entry], ignore_index=True)\n",
    "    elif 'Mice_rand1k_sh' in directory:\n",
    "        for random in os.listdir(directory):\n",
    "            rt = os.path.join(directory, random)\n",
    "            entries = os.listdir(rt)\n",
    "            for entry in entries:\n",
    "                dirs = os.path.join(rt, entry)\n",
    "                for root, _, files in os.walk(dirs):\n",
    "                    if 'adata.h5ad' in files:\n",
    "                        # Iterate over each file in the current directory\n",
    "                        for file in files:\n",
    "                            # Check if the file is a log file\n",
    "                            if file.endswith('.log'):\n",
    "                                # Create the file path\n",
    "                                file_path = os.path.join(root, file)\n",
    "                                # Read the contents of the log file\n",
    "                                with open(file_path, 'r') as f:\n",
    "                                    log_contents = f.read()\n",
    "                                    # Extract relevant information from log_contents\n",
    "                                    dict = {}\n",
    "                                    loop = 0\n",
    "                                    for line in log_contents.strip().split('\\n'):\n",
    "                                        try:\n",
    "                                            key, value = line.split(': ')\n",
    "                                        except ValueError:\n",
    "                                            pass\n",
    "                                        dict[key.strip()] = value.strip()\n",
    "                                        # Calculate RMSD\n",
    "                                        if 'RMSD between New Configuration and Old Configuration' in line and loop <= 9:\n",
    "                                            loop += 1\n",
    "                                            RMSD = float(line.split('is')[1].split()[0])\n",
    "                                        \n",
    "                                            entry = pd.DataFrame([{\n",
    "                                                'Sample Name': root.split('/')[-2],\n",
    "                                                'Source': 'Random',\n",
    "                                                'Cell Number': dict.get('Cell Number'),\n",
    "                                                'Gene Number': dict.get('Gene Number'),\n",
    "                                                'Gene Number for Rotation Derivatives': dict.get('Anchor Gene Number is'),\n",
    "                                                'loop': loop,\n",
    "                                                'RMSD': RMSD\n",
    "                                            }])\n",
    "                                            df = pd.concat([df, entry], ignore_index=True)\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df\n",
    "random_count = df[df['Source'] == 'Random'].shape[0]\n",
    "print(random_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from statannot import add_stat_annotation\n",
    "\n",
    "# Create the plot using seaborn\n",
    "sns.set(style=\"whitegrid\", rc={\"figure.dpi\":300,'figure.figsize':(8, 2)}, font_scale=1)\n",
    "sns.set_theme(style='ticks', palette='pastel')\n",
    "# Draw the boxplot\n",
    "ax = sns.boxplot(data=df, x=\"loop\", y=\"RMSD\", hue=\"Source\", palette=\"Set1\", fliersize=0.5, linewidth=0.5, dodge=True)\n",
    "\n",
    "# Define the pairs for comparison\n",
    "pairs = [((1, 'Cell type'), (1, 'Random')),\n",
    "         ((2, 'Cell type'), (2, 'Random')),\n",
    "         ((3, 'Cell type'), (3, 'Random')),\n",
    "         ((4, 'Cell type'), (4, 'Random')),\n",
    "         ((5, 'Cell type'), (5, 'Random')),\n",
    "         ((6, 'Cell type'), (6, 'Random')),\n",
    "         ((7, 'Cell type'), (7, 'Random')),\n",
    "         ((8, 'Cell type'), (8, 'Random')),\n",
    "         ((9, 'Cell type'), (9, 'Random')),\n",
    "         ((10, 'Cell type'), (10, 'Random'))]\n",
    "\n",
    "# Add statistical annotations\n",
    "add_stat_annotation(ax, data=df, x=\"loop\", y=\"RMSD\", hue=\"Source\", box_pairs=pairs, \n",
    "                    test='Mann-Whitney', text_format='star', loc='outside', verbose=2)\n",
    "\n",
    "plt.xlabel(\"Loop\")\n",
    "plt.ylabel(\"RMSD\")\n",
    "plt.title(\"Convergence Curve\", pad=30)\n",
    "plt.legend(loc=\"upper right\", title='Cell Origin')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fig.3c variance of cell error "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "# Define the directory paths\n",
    "directories = ['./results/Mice_onesample/', './results/Mice_rand10k/']\n",
    "adatas = {}\n",
    "\n",
    "# Iterate over each directory\n",
    "for directory in directories:\n",
    "    if 'Mice_onesample' in directory:\n",
    "        samples = [ name for name in sorted(os.listdir(directory)) if os.path.isdir(os.path.join(directory, name)) ]\n",
    "        for sample in samples:\n",
    "            result_dir = os.path.join(directory,sample)\n",
    "            sample = sample.split('.')[1]\n",
    "            files = os.listdir(os.path.join(result_dir))\n",
    "            if 'adata.h5ad' in files: \n",
    "                adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "                adatas[sample] = adata\n",
    "                #Xs[sample] = adata.uns['X']\n",
    "                #GeneLists[sample] = adata.uns['X'].index\n",
    "    elif 'Mice_rand10k' in directory:\n",
    "        for random in os.listdir(directory):\n",
    "            rt = os.path.join(directory, random)\n",
    "            entries = os.listdir(rt)\n",
    "            for entry in entries:\n",
    "                dirs = os.path.join(rt, entry)\n",
    "                for root, _, files in os.walk(dirs):\n",
    "                    if 'adata.h5ad' in files:\n",
    "                        file_path = os.path.join(root, 'adata.h5ad')\n",
    "                        adata = sc.read_h5ad(file_path)\n",
    "                        adatas[random]=adata\n",
    "                                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(adatas.keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### tc count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata=adatas[sample]\n",
    "    try:\n",
    "        W=adata.uns['W']\n",
    "        X=adata.uns['X']\n",
    "    except:\n",
    "        W=adata.uns['Z']\n",
    "        X=adata.uns['F']\n",
    "    n_tc = X.shape[0]\n",
    "\n",
    "    if 'rand1k' in sample:\n",
    "        data.append([sample, n_tc, 'Random'])\n",
    "    else:\n",
    "        data.append([sample, n_tc, 'Cell type'])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "df_tcn = pd.DataFrame(data, columns=['Sample', 'tc_count', 'Source'])\n",
    "\n",
    "# rectangular box plot\n",
    "# group_means=diffs.groupby(['celltype'])['error'].median().sort_values(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from statannot import add_stat_annotation\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"whitegrid\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(3, 4)})\n",
    "\n",
    "# Create the box plot\n",
    "ax = sns.boxplot(data=df_tcn, x='Source', y='tc_count')\n",
    "\n",
    "# Add statistical annotations\n",
    "add_stat_annotation(ax, data=df_tcn, x='Source', y='tc_count',\n",
    "                    box_pairs=[(\"Cell type\", \"Random\")],\n",
    "                    test='Mann-Whitney', text_format='star', loc='outside', verbose=2)\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_ylabel(\"Transcription Center Count\")\n",
    "#ax.set_title(\"Transcription Center Count\")\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sample errors - top 10 of slice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cytocraft.craft as cc\n",
    "gem_allcells = cc.read_gem_as_csv('./data/Mice/E16.5_E1S3/SS200000108BR_A3A4_scgem.csv',sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted_genes = gem_allcells.groupby('geneID')['MIDCount'].sum().sort_values(ascending=False)\n",
    "top_genes = sorted_genes.head(50).index.tolist()\n",
    "print(top_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas['rand1k_11']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata=adatas[sample]\n",
    "    try:\n",
    "        W=adata.uns['W']\n",
    "        X=adata.uns['X']\n",
    "    except:\n",
    "        W=adata.uns['Z']\n",
    "        X=adata.uns['F']\n",
    "    R = adata.obsm['Rotation']\n",
    "    ncell = int(W.shape[0] / 2)\n",
    "    ngene = int(W.shape[1])\n",
    "\n",
    "    for c in range(ncell):\n",
    "        Z = np.dot(X, R[c, :, :])[:, 0:2].T\n",
    "        E = W[c*2:c*2+2,:10] - Z[:,:10]\n",
    "        E_nan = E[:, ~np.isnan(E).any(axis=0)]\n",
    "        errors = np.linalg.norm(E_nan, axis=0)\n",
    "\n",
    "        for e in errors:\n",
    "            if 'rand1k' in sample:\n",
    "                data.append([sample, e, 'Random'])\n",
    "            else:\n",
    "                data.append([sample, e, 'Cell type'])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['Sample', 'error', 'Source'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sample MSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data = []\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata = adatas[sample]\n",
    "    try:\n",
    "        W = adata.uns['W']\n",
    "        X = adata.uns['X']\n",
    "    except:\n",
    "        W = adata.uns['Z']\n",
    "        X = adata.uns['F']\n",
    "    R = adata.obsm['Rotation']\n",
    "    ncell = int(W.shape[0] / 2)\n",
    "    ngene = int(W.shape[1])\n",
    "\n",
    "    for c in range(ncell):\n",
    "        Z = np.dot(X, R[c, :, :])[:, 0:2].T\n",
    "        for top in [10, 20, 30, 40, 50]:\n",
    "            W_top = W[c*2:c*2+2, :top]\n",
    "            Z_top = Z[:, :top]\n",
    "            # Remove columns with NaN values\n",
    "            valid_columns = ~np.isnan(W_top).any(axis=0) & ~np.isnan(Z_top).any(axis=0)\n",
    "            W_top = W_top[:, valid_columns]\n",
    "            Z_top = Z_top[:, valid_columns]\n",
    "            mse = np.mean((cc.normalizeZ(W_top) - cc.normalizeZ(Z_top))**2)\n",
    "            if 'rand10k' in sample:\n",
    "                data.append([sample, mse, top, 'Random'])\n",
    "            else:\n",
    "                data.append([sample, mse, top, 'Cell type'])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['Sample', 'MSE', 'top_n', 'Source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statannot import add_stat_annotation\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"whitegrid\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(4, 3)})\n",
    "\n",
    "# Create the box plot with narrower box lines\n",
    "ax = sns.boxplot(data=diffs, x='top_n', y='MSE', hue='Source', showfliers=False, linewidth=0.3)\n",
    "\n",
    "# Calculate p-values for each top_n\n",
    "p_values = []\n",
    "box_pairs = []\n",
    "for top_n in diffs['top_n'].unique():\n",
    "    data1 = diffs[(diffs['top_n'] == top_n) & (diffs['Source'] == 'Cell type')]['MSE']\n",
    "    data2 = diffs[(diffs['top_n'] == top_n) & (diffs['Source'] == 'Random')]['MSE']\n",
    "    stat, p = mannwhitneyu(data1, data2)\n",
    "    p_values.append(p)\n",
    "    box_pairs.append(((top_n, 'Cell type'), (top_n, 'Random')))\n",
    "\n",
    "# Adjust p-values for multiple comparisons\n",
    "adjusted_p_values = multipletests(p_values, method='fdr_bh')[1]\n",
    "\n",
    "# Add statistical annotations\n",
    "for i, (p, pair) in enumerate(zip(adjusted_p_values, box_pairs)):\n",
    "    add_stat_annotation(ax, data=diffs, x='top_n', y='MSE', hue='Source', box_pairs=[pair], \n",
    "                        perform_stat_test=False, pvalues=[p], text_format='star', loc='outside', verbose=2)\n",
    "\n",
    "# Set the axis labels\n",
    "ax.set_xlabel(\"Top $N$ Expressing Genes\")\n",
    "ax.set_ylabel(\"MSE\")\n",
    "\n",
    "# Define the legend\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "new_labels = ['Same Type Cells', 'Random Cells']\n",
    "ax.legend(handles, new_labels, title='Source')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### sample errors - top10 of sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata=adatas[sample]\n",
    "    try:\n",
    "        W=adata.uns['W']\n",
    "        X=adata.uns['X']\n",
    "    except:\n",
    "        W=adata.uns['Z']\n",
    "        X=adata.uns['F']\n",
    "    R = adata.obsm['Rotation']\n",
    "    ncell = int(W.shape[0] / 2)\n",
    "    ngene = int(W.shape[1])\n",
    "\n",
    "    for c in range(ncell):\n",
    "        Z = np.dot(X, R[c, :, :])[:, 0:2].T\n",
    "        for top in [10, 20, 30, 40, 50]:\n",
    "            E = cc.normalizeZ(W[c*2:c*2+2,:top] - Z[:,:top])\n",
    "            E_nan = E[:, ~np.isnan(E).any(axis=0)]\n",
    "            errors = np.linalg.norm(E_nan, axis=0)\n",
    "            for e in errors:\n",
    "                if 'rand10k' in sample:\n",
    "                    data.append([sample, e, top, 'Random'])\n",
    "                else:\n",
    "                    data.append([sample, e, top, 'Cell type'])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['Sample', 'error', 'top_n', 'Source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata=adatas[sample]\n",
    "    try:\n",
    "        W=adata.uns['W']\n",
    "        X=adata.uns['X']\n",
    "    except:\n",
    "        W=adata.uns['Z']\n",
    "        X=adata.uns['F']\n",
    "    R = adata.obsm['Rotation']\n",
    "    ncell = int(W.shape[0] / 2)\n",
    "    ngene = int(W.shape[1])\n",
    "\n",
    "    for c in range(ncell):\n",
    "        Z = np.dot(X, R[c, :, :])[:, 0:2].T\n",
    "        for top in [10, 20, 30, 40, 50]:\n",
    "            E = cc.normalizeZ(W[c*2:c*2+2,:top] - Z[:,:top])\n",
    "            E_nan = E[:, ~np.isnan(E).any(axis=0)]\n",
    "            errors = np.linalg.norm(E_nan, axis=0)\n",
    "            if errors.size>0:\n",
    "                max_error = np.max(errors)\n",
    "                nan_errors = np.full(E.shape[1], max_error)\n",
    "                nan_errors[~np.isnan(E).any(axis=0)] = errors\n",
    "                errors = nan_errors\n",
    "                for e in errors:\n",
    "                    if 'rand1k' in sample:\n",
    "                        data.append([sample, e, top, 'Random'])\n",
    "                    else:\n",
    "                        data.append([sample, e, top, 'Cell type'])\n",
    "            else:\n",
    "                #print(sample)\n",
    "                pass\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['Sample', 'error', 'top_n', 'Source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Calculate the mean error for each sample\n",
    "median_errors = diffs.groupby('Sample')['error'].median().sort_values()\n",
    "\n",
    "# Set the order of samples based on mean error\n",
    "sample_order = median_errors.index\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"whitegrid\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(20, 4)})\n",
    "\n",
    "# Create the box plot with hue by Source\n",
    "ax = sns.boxplot(data=diffs, x='Sample', y='error', hue='Source', showfliers=False, order=sample_order, dodge=False)\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"Sample\")\n",
    "ax.set_ylabel(\"Error\")\n",
    "ax.set_title(\"Error Distribution of Genes by Sample\")\n",
    "\n",
    "# Split x tick labels by '_' and keep the first element\n",
    "ax.set_xticklabels([' '.join(label.get_text().split('_')[:-1]) for label in ax.get_xticklabels()], rotation=90)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate mean and variance for errors of each sample in each top_n\n",
    "mean_errors = diffs.groupby(['Sample', 'top_n'])['error'].mean()\n",
    "variance_errors = diffs.groupby(['Sample', 'top_n'])['error'].var()\n",
    "source = diffs.groupby(['Sample', 'top_n'])['Source'].first()\n",
    "\n",
    "# Create a DataFrame to store the results\n",
    "df_errors = pd.DataFrame({\n",
    "    'Sample': mean_errors.index.get_level_values(0),\n",
    "    'Top N': mean_errors.index.get_level_values(1),\n",
    "    'Mean Error': mean_errors.values,\n",
    "    'Variance Error': variance_errors.values,\n",
    "    'Source': source.values\n",
    "})\n",
    "\n",
    "# Display the DataFrame\n",
    "print(df_errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import mannwhitneyu\n",
    "from statannot import add_stat_annotation\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"whitegrid\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(4, 3)})\n",
    "\n",
    "# Create the box plot with narrower box lines\n",
    "ax = sns.boxplot(data=df_errors, x='Top N', y='Mean Error', hue='Source', showfliers=False, linewidth=0.3)\n",
    "\n",
    "# Calculate p-values for each Top N\n",
    "p_values = []\n",
    "box_pairs = []\n",
    "for top_n in df_errors['Top N'].unique():\n",
    "    data1 = df_errors[(df_errors['Top N'] == top_n) & (df_errors['Source'] == 'Cell type')]['Mean Error']\n",
    "    data2 = df_errors[(df_errors['Top N'] == top_n) & (df_errors['Source'] == 'Random')]['Mean Error']\n",
    "    stat, p = mannwhitneyu(data1, data2)\n",
    "    p_values.append(p)\n",
    "    box_pairs.append(((top_n, 'Cell type'), (top_n, 'Random')))\n",
    "\n",
    "# Adjust p-values for multiple comparisons\n",
    "adjusted_p_values = multipletests(p_values, method='fdr_bh')[1]\n",
    "\n",
    "# Add statistical annotations\n",
    "for i, (p, pair) in enumerate(zip(adjusted_p_values, box_pairs)):\n",
    "    add_stat_annotation(ax, data=df_errors, x='Top N', y='Mean Error', hue='Source', box_pairs=[pair], \n",
    "                        perform_stat_test=False, pvalues=[p], text_format='star', loc='outside', verbose=2)\n",
    "\n",
    "# Set the axis labels\n",
    "ax.set_xlabel(\"Top $N$ Expressing Genes\")\n",
    "ax.set_ylabel(\"Mean Error\")\n",
    "\n",
    "# Define the legend\n",
    "handles, labels = ax.get_legend_handles_labels()\n",
    "new_labels = ['Same Type Cells', 'Random Cells']\n",
    "ax.legend(handles, new_labels, title='Source')\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import norm\n",
    "import numpy as np\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"white\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(6, 4)})\n",
    "\n",
    "# Create the histogram plot\n",
    "df_errors=df_errors[df_errors['Sample']!='Hepatocyte_WQV6']\n",
    "ax = sns.histplot(data=df_errors, x='Mean Error', hue='Source', element='step', stat='density', common_norm=False)\n",
    "\n",
    "# Fit a Gaussian curve for each source\n",
    "for source in df_errors['Source'].unique():\n",
    "    data = df_errors[df_errors['Source'] == source]['Mean Error']\n",
    "    mu, std = norm.fit(data)\n",
    "    xmin, xmax = ax.get_xlim()\n",
    "    x = np.linspace(xmin, xmax, 100)\n",
    "    p = norm.pdf(x, mu, std)\n",
    "    ax.plot(x, p, label=f'{source} fit: $\\mu={mu:.2f}$, $\\sigma={std:.2f}$')\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"Variance of Error\")\n",
    "ax.set_ylabel(\"Density\")\n",
    "ax.set_title(\"Distribution of Error Variance of top 10 Expressed Gene\")\n",
    "ax.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata=adatas[sample]\n",
    "    try:\n",
    "        W=adata.uns['W']\n",
    "        X=adata.uns['X']\n",
    "    except:\n",
    "        W=adata.uns['Z']\n",
    "        X=adata.uns['F']\n",
    "    R = adata.obsm['Rotation']\n",
    "    ncell = int(W.shape[0] / 2)\n",
    "    ngene = int(W.shape[1])\n",
    "\n",
    "    errors = []\n",
    "    for c in range(ncell):\n",
    "        Z = np.dot(X, R[c, :, :])[:, 0:2].T\n",
    "        E = W[c*2:c*2+2,:10] - Z[:,:10]\n",
    "        E_nan = E[:, ~np.isnan(E).any(axis=0)]\n",
    "        e = np.linalg.norm(E_nan, axis=0)\n",
    "        errors.extend(e)\n",
    "    \n",
    "    errors = np.array(errors)\n",
    "    mean_error = errors.mean()\n",
    "    variance_error = errors.var()\n",
    "    if 'rand1k' in sample:\n",
    "        data.append([sample, mean_error, variance_error, 'Random'])\n",
    "    else:\n",
    "        data.append([sample, mean_error, variance_error, 'Cell type'])\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['Sample', 'mean_error', 'var_error', 'Source'])\n",
    "\n",
    "# rectangular box plot\n",
    "# group_means=diffs.groupby(['celltype'])['error'].median().sort_values(ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import norm\n",
    "import numpy as np\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"white\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(6, 4)})\n",
    "\n",
    "# Create the histogram plot\n",
    "ax = sns.histplot(data=diffs, x='var_error', hue='Source', element='step', stat='density', common_norm=False)\n",
    "\n",
    "# Fit a Gaussian curve for each source\n",
    "for source in diffs['Source'].unique():\n",
    "    data = diffs[diffs['Source'] == source]['var_error']\n",
    "    mu, std = norm.fit(data)\n",
    "    xmin, xmax = ax.get_xlim()\n",
    "    x = np.linspace(xmin, xmax, 100)\n",
    "    p = norm.pdf(x, mu, std)\n",
    "    ax.plot(x, p, label=f'{source} fit: $\\mu={mu:.2f}$, $\\sigma={std:.2f}$')\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"Variance of Error\")\n",
    "ax.set_ylabel(\"Density\")\n",
    "ax.set_title(\"Distribution of Error Variance of top 10 Expressed Gene\")\n",
    "ax.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### all errors - top10 of sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import norm\n",
    "import numpy as np\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"white\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(8, 6)})\n",
    "\n",
    "# Create the histogram plot\n",
    "ax = sns.histplot(data=diffs, x='Error', hue='Sample', element='step', stat='density', common_norm=False, legend=False)\n",
    "\n",
    "# Fit and plot Gaussian distribution for each sample\n",
    "for sample in diffs['Sample'].unique():\n",
    "    sample_data = diffs[diffs['Sample'] == sample]['Error']\n",
    "    mu, std = norm.fit(sample_data)\n",
    "    xmin, xmax = ax.get_xlim()\n",
    "    x = np.linspace(xmin, xmax, 100)\n",
    "    p = norm.pdf(x, mu, std)\n",
    "    #ax.plot(x, p, label=f'{sample} fit')\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"Error\")\n",
    "ax.set_ylabel(\"Density\")\n",
    "ax.set_title(\"Distribution of Error of top 10 Expressed Gene\")\n",
    "\n",
    "# Show the legend\n",
    "#ax.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "\n",
    "data=[]\n",
    "for sample in tqdm(list(adatas.keys())):\n",
    "    adata=adatas[sample]\n",
    "    try:\n",
    "        W=adata.uns['W']\n",
    "        X=adata.uns['X']\n",
    "    except:\n",
    "        W=adata.uns['Z']\n",
    "        X=adata.uns['F']\n",
    "    R = adata.obsm['Rotation']\n",
    "    ncell = int(W.shape[0] / 2)\n",
    "    ngene = int(W.shape[1])\n",
    "\n",
    "    errors = []\n",
    "    for c in range(ncell):\n",
    "        Z = np.dot(X, R[c, :, :])[:, 0:2].T\n",
    "        E = W[c*2:c*2+2,:] - Z\n",
    "        #E = W[c*2:c*2+2,:100] - Z[:,:100]\n",
    "        E_nan = E[:, ~np.isnan(E).any(axis=0)]\n",
    "        errors = np.linalg.norm(E_nan, axis=0)\n",
    "\n",
    "    if 'rand1k' in sample:\n",
    "        for e in errors:\n",
    "            data.append([sample, e, 'Random'])\n",
    "    else:\n",
    "        for e in errors:\n",
    "            data.append([sample, e, 'Cell type'])\n",
    "\n",
    "\n",
    "# Build the final DataFrame in one operation.\n",
    "diffs = pd.DataFrame(data, columns=['sample', 'error', 'Source'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diffs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import poisson\n",
    "import numpy as np\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"white\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(8, 6)})\n",
    "\n",
    "# Create the histogram plot\n",
    "ax = sns.histplot(data=diffs, x='Error_var', hue='Source', element='step', stat='density', common_norm=False)\n",
    "\n",
    "# Fit a Poisson curve for each source\n",
    "for source in diffs['Source'].unique():\n",
    "    data = diffs[diffs['Source'] == source]['Error_var']\n",
    "    mu = np.mean(data)\n",
    "    xmin, xmax = ax.get_xlim()\n",
    "    x = np.arange(xmin, xmax)\n",
    "    p = poisson.pmf(x, mu)\n",
    "    ax.plot(x, p, label=f'{source} fit: $\\lambda={mu:.2f}$')\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"Error\")\n",
    "ax.set_ylabel(\"Density\")\n",
    "ax.set_title(\"Distribution of Error of top 100 Expressed Gene\")\n",
    "ax.legend()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "# Set the style and color palette\n",
    "sns.set(style=\"whitegrid\", palette=\"Set2\", rc={\"figure.dpi\":300,'figure.figsize':(4, 4)})\n",
    "\n",
    "# Create the box plot\n",
    "ax = sns.boxplot(data=diffs, x='Source', y='var_error')\n",
    "\n",
    "# Calculate the p-value\n",
    "celltype_var_error = diffs[diffs['Source'] == 'Cell type']['var_error']\n",
    "random_var_error = diffs[diffs['Source'] == 'Random']['var_error']\n",
    "stat, p_value = mannwhitneyu(celltype_var_error, random_var_error)\n",
    "\n",
    "# Annotate the plot with the p-value\n",
    "ax.text(0.5, max(diffs['var_error']), f'p-value: {p_value:.2e}', ha='center', va='bottom')\n",
    "\n",
    "# Set the axis labels and title\n",
    "ax.set_xlabel(\"Source\")\n",
    "ax.set_ylabel(\"Variance of Error\")\n",
    "ax.set_title(\"Variance of Error by Source\")\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "st",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
