{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrap_labels(label, max_length=20, ignore=[]):\n",
    "    if len(label) > max_length:\n",
    "        for i in ignore:\n",
    "            if i in label:\n",
    "                return label\n",
    "        last_space = label.rfind(' ', 0, max_length)\n",
    "        if last_space > 0:\n",
    "            if len(label[last_space+1:]) > max_length:\n",
    "                label = label[:last_space] + '<br>' + wrap_labels(label[last_space+1:], max_length=max_length)\n",
    "            else:\n",
    "                label = label[:last_space] + '<br>' + label[last_space+1:]\n",
    "    return label\n",
    "\n",
    "def remove_rare_values(df, column_names):\n",
    "    for column in column_names:\n",
    "        counts = df[column].value_counts()\n",
    "        to_remove = counts[counts < 3].index\n",
    "        df = df[~df[column].isin(to_remove)]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# check completeness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in os.listdir(result_path) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    sample = sample[:-5] \n",
    "    chip = sample.split(\".\")[0]\n",
    "    if 'adata.h5ad' not in files: \n",
    "        print(sample)\n",
    "        #if sample[0].isdigit():\n",
    "        #    !python ./CRUST.py ./data/Axolotls/regeneration/{chip}/{sample}.csv ./results/Axolotls/ Axolotls\n",
    "        #else:\n",
    "        #    !python ./CRUST.py ./data/Axolotls/development/{chip}/{sample}.csv ./results/Axolotls/ Axolotls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dataview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## gem check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gems={}\n",
    "gems['Stage44'] = pd.read_csv(\"./data/Axolotls/development/Stage44_telencephalon_rep2_FP200000239BL_E4_scgem.csv\")\n",
    "gems['Stage54'] = pd.read_csv(\"./data/Axolotls/development/Stage54_telencephalon_rep2_DP8400015649BRD6_2_scgem.csv\")\n",
    "gems['Stage57'] = pd.read_csv(\"./data/Axolotls/development/Stage57_telencephalon_rep2_DP8400015649BRD5_1_scgem.csv\")\n",
    "gems['Juv'] = pd.read_csv(\"./data/Axolotls/development/Injury_control_FP200000239BL_E3_scgem.csv\")\n",
    "gems['Adult'] = pd.read_csv(\"./data/Axolotls/development/Adult_telencephalon_rep2_DP8400015234BLA3_1_scgem.csv\")\n",
    "gems['Meta'] = pd.read_csv(\"./data/Axolotls/development/Meta_telencephalon_rep1_DP8400015234BLB2_1_scgem.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Set the style of the visualization\n",
    "sns.set(style=\"whitegrid\")\n",
    "\n",
    "# Create a figure and axis\n",
    "plt.figure(figsize=(12, 8))\n",
    "\n",
    "# Define a color palette\n",
    "palette = sns.color_palette(\"husl\", len(gems))\n",
    "\n",
    "# Loop through each sample in gems and plot the histogram\n",
    "for (sample_name, gem_data), color in zip(gems.items(), palette):\n",
    "    gem_total_counts = gem_data.groupby('geneID')['MIDCounts'].sum()\n",
    "    sns.histplot(gem_total_counts, bins=1000, kde=True, label=sample_name, color=color, alpha=0.6)\n",
    "\n",
    "# Add titles and labels\n",
    "plt.title('Distribution of Total Counts per Gene for Each Sample', fontsize=16)\n",
    "plt.xlabel('Total Counts', fontsize=14)\n",
    "plt.ylabel('Frequency', fontsize=14)\n",
    "\n",
    "# Set x-axis limits\n",
    "plt.xlim(0, 10000)\n",
    "\n",
    "# Add a legend\n",
    "plt.legend(title='Sample', fontsize=12)\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## development"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata_dev = sc.read_h5ad(\"./data/Axolotls/Development.h5ad\")\n",
    "adata_dev.obs_names_make_unique()\n",
    "adata_reg = sc.read_h5ad(\"./data/Axolotls/Regeneration.h5ad\")\n",
    "adata_reg.obs_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dev.obs['x']=adata_dev.obsm['spatial'][:,0]\n",
    "adata_dev.obs['y']=4500-adata_dev.obsm['spatial'][:,1]\n",
    "adata_reg.obs['x']=adata_reg.obsm['spatial'][:,0]\n",
    "adata_reg.obs['y']=13000-adata_reg.obsm['spatial'][:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(dpi=300, figsize=[15,3], fontsize=12)\n",
    "sc.pl.scatter(adata_dev,x=\"x\",y=\"y\",color='Annotation', size=9, legend_loc = 'right margin', legend_fontsize='small', show=False, frameon=False, title='')\n",
    "\n",
    "#legend = ax.get_legend()\n",
    "#legend.set_bbox_to_anchor((1, 0.5)) # set legend position\n",
    "#plt.setp(legend.get_texts(), fontsize='x-small') # set legend font size\n",
    "#legend._legend_box.align = \"left\" # set legend alignment\n",
    "#legend.set_title('Legend Title') # set legend title\n",
    "#legend._legend_box.ncol = 0 # set number of columns\n",
    "ax = plt.gca()\n",
    "# To remove the background frame\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "# To remove x-y axis\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "ax.axes.get_yaxis().set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sc.set_figure_params(dpi=300, figsize=[15,3], fontsize=12)\n",
    "sc.pl.scatter(adata_dev,x=\"x\",y=\"y\",color='Annotation', size=9, legend_loc = 'upper right', legend_fontsize='large', show=False, \n",
    "    frameon=False, title='', groups=['VLMC'], palette=sns.color_palette('viridis'))\n",
    "\n",
    "ax = plt.gca()\n",
    "# To remove the background frame\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "# To remove x-y axis\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "for label, x in zip(['St.44', 'St.54', 'St.57', 'Juv.', 'Adult', 'Meta.'], [760, 3900, 7700, 13700, 20400, 27000]):\n",
    "    ax.text(x, -200, label, ha='center', va='top', fontsize=14)\n",
    "ax.axes.get_yaxis().set_visible(False)\n",
    "# Add annotation for developmental stages\n",
    "ax.annotate('Developmental stages', xy=(4230, 400), xytext=(4230, 4000),\n",
    "            fontsize=16, color='black', ha='center')\n",
    "\n",
    "plt.savefig(\"Fig5a_u.pdf\", format=\"pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_dev.obs['Annotation'] = adata_dev.obs['Annotation'].replace({'CP  ':'CP'})\n",
    "sc.set_figure_params(dpi=300, figsize=[15,3], fontsize=12)\n",
    "sc.pl.scatter(adata_dev,x=\"x\",y=\"y\",color='Annotation', size=9, legend_loc = 'upper right', legend_fontsize='large', show=False, \n",
    "    frameon=False, title='', groups=['CP'], palette=sns.color_palette('viridis'))\n",
    "\n",
    "ax = plt.gca()\n",
    "# To remove the background frame\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "# To remove x-y axis\n",
    "ax.axes.get_xaxis().set_visible(False)\n",
    "for label, x in zip(['St.44', 'St.54', 'St.57', 'Juv.', 'Adult', 'Meta.'], [760, 3900, 7700, 13700, 20400, 27000]):\n",
    "    ax.text(x, -200, label, ha='center', va='top', fontsize=14)\n",
    "ax.axes.get_yaxis().set_visible(False)\n",
    "\n",
    "plt.savefig(\"Fig5a_d.pdf\", format=\"pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# downstram analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import scanpy as sc\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in os.listdir(result_path) if os.path.isdir(os.path.join(result_path, name))]\n",
    "adata={}\n",
    "Xs={}\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    chip = sample.split(\".\")[0]\n",
    "    if 'adata.h5ad' in files: \n",
    "        adata[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata[sample].uns['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in sample:\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.netplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from CRUST import normalizeX, numpy_svd_rmsd_rot\n",
    "\n",
    "def mirror(X):\n",
    "    mirrorX = np.copy(X)\n",
    "    mirrorX[:, 2] = -mirrorX[:, 2]\n",
    "    return mirrorX\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "ctypes=[]\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    result_dir = os.path.join(result_path,sample)\n",
    "    files = os.listdir(os.path.join(result_dir))\n",
    "    info = sample.split(\".\")\n",
    "    chip = info[0]\n",
    "    ctype = info[-1][:-5]\n",
    "    ctypes.append(ctype)\n",
    "    sample = chip+' '+ctype\n",
    "    if 'adata.h5ad' in files and sample[0].isalpha(): \n",
    "        print(sample)\n",
    "        adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "        Xs[sample] = adata.uns['X']\n",
    "        GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "# Build distance matrix\n",
    "utypes = np.unique(ctypes)\n",
    "keys = list(Xs.keys())\n",
    "N = len(keys)\n",
    "D_axo_dev = np.zeros((N,N))\n",
    "\n",
    "for n, key_n in enumerate(keys):\n",
    "    for m, key_m in enumerate(keys[:n+1]):\n",
    "        intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])[:100]\n",
    "        boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "        boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "        normalized_x_n = normalizeX(Xs[key_n][boolean_arrays_n], method='mean')\n",
    "        normalized_x_m = normalizeX(Xs[key_m][boolean_arrays_m], method='mean')\n",
    "        d1, _, _ = numpy_svd_rmsd_rot(normalized_x_n, normalized_x_m)\n",
    "        d2, _, _ = numpy_svd_rmsd_rot(mirror(normalized_x_n), normalized_x_m)\n",
    "        D_axo_dev[n,m] = D_axo_dev[m,n] = min(d1,d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "godsnot_102 = [\n",
    "    #\"#000000\",  # remove the black, as often, we have black colored annotation\n",
    "    \"#FFFF00\",\n",
    "    \"#1CE6FF\",\n",
    "    \"#FF34FF\",\n",
    "    \"#FF4A46\",\n",
    "    \"#008941\",\n",
    "    \"#006FA6\",\n",
    "    \"#A30059\",\n",
    "    \"#FFDBE5\",\n",
    "    \"#7A4900\",\n",
    "    \"#0000A6\",\n",
    "    \"#63FFAC\",\n",
    "    \"#B79762\",\n",
    "    \"#004D43\",\n",
    "    \"#8FB0FF\",\n",
    "    \"#997D87\",\n",
    "    \"#5A0007\",\n",
    "    \"#809693\",\n",
    "    \"#6A3A4C\",\n",
    "    \"#1B4400\",\n",
    "    \"#4FC601\",\n",
    "    \"#3B5DFF\",\n",
    "    \"#4A3B53\",\n",
    "    \"#FF2F80\",\n",
    "    \"#61615A\",\n",
    "    \"#BA0900\",\n",
    "    \"#6B7900\",\n",
    "    \"#00C2A0\",\n",
    "    \"#FFAA92\",\n",
    "    \"#FF90C9\",\n",
    "    \"#B903AA\",\n",
    "    \"#D16100\",\n",
    "    \"#DDEFFF\",\n",
    "    \"#000035\",\n",
    "    \"#7B4F4B\",\n",
    "    \"#A1C299\",\n",
    "    \"#300018\",\n",
    "    \"#0AA6D8\",\n",
    "    \"#013349\",\n",
    "    \"#00846F\",\n",
    "    \"#372101\",\n",
    "    \"#FFB500\",\n",
    "    \"#C2FFED\",\n",
    "    \"#A079BF\",\n",
    "    \"#CC0744\",\n",
    "    \"#C0B9B2\",\n",
    "    \"#C2FF99\",\n",
    "    \"#001E09\",\n",
    "    \"#00489C\",\n",
    "    \"#6F0062\",\n",
    "    \"#0CBD66\",\n",
    "    \"#EEC3FF\",\n",
    "    \"#456D75\",\n",
    "    \"#B77B68\",\n",
    "    \"#7A87A1\",\n",
    "    \"#788D66\",\n",
    "    \"#885578\",\n",
    "    \"#FAD09F\",\n",
    "    \"#FF8A9A\",\n",
    "    \"#D157A0\",\n",
    "    \"#BEC459\",\n",
    "    \"#456648\",\n",
    "    \"#0086ED\",\n",
    "    \"#886F4C\",\n",
    "    \"#34362D\",\n",
    "    \"#B4A8BD\",\n",
    "    \"#00A6AA\",\n",
    "    \"#452C2C\",\n",
    "    \"#636375\",\n",
    "    \"#A3C8C9\",\n",
    "    \"#FF913F\",\n",
    "    \"#938A81\",\n",
    "    \"#575329\",\n",
    "    \"#00FECF\",\n",
    "    \"#B05B6F\",\n",
    "    \"#8CD0FF\",\n",
    "    \"#3B9700\",\n",
    "    \"#04F757\",\n",
    "    \"#C8A1A1\",\n",
    "    \"#1E6E00\",\n",
    "    \"#7900D7\",\n",
    "    \"#A77500\",\n",
    "    \"#6367A9\",\n",
    "    \"#A05837\",\n",
    "    \"#6B002C\",\n",
    "    \"#772600\",\n",
    "    \"#D790FF\",\n",
    "    \"#9B9700\",\n",
    "    \"#549E79\",\n",
    "    \"#FFF69F\",\n",
    "    \"#201625\",\n",
    "    \"#72418F\",\n",
    "    \"#BC23FF\",\n",
    "    \"#99ADC0\",\n",
    "    \"#3A2465\",\n",
    "    \"#922329\",\n",
    "    \"#5B4534\",\n",
    "    \"#FDE8DC\",\n",
    "    \"#404E55\",\n",
    "    \"#0089A3\",\n",
    "    \"#CB7E98\",\n",
    "    \"#A4E804\",\n",
    "    \"#324E72\",\n",
    "]\n",
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#color_map_ct =  dict(zip(utypes, godsnot_102))\n",
    "\n",
    "color_map_ct =  {\n",
    "'CP': '#1CE6FF',\n",
    "'IMN': '#FF34FF',\n",
    "'Immature_CMPN': '#FF4A46',\n",
    "'CMPN': '#FFFF00',\n",
    "'Immature_MSN': '#006FA6',\n",
    "'MCG': '#63FFAC',\n",
    "'MSN': '#B79762',\n",
    "'Oligo': '#004D43',\n",
    "'VLMC': '#997D87',\n",
    "'WSN': '#5A0007',\n",
    "'dNBL1': '#1B4400',\n",
    "'dNBL2': '#4FC601',\n",
    "'dNBL3': '#3B5DFF',\n",
    "'dNBL4': '#4A3B53',\n",
    "'dNBL5': '#FF2F80',\n",
    "'obNBL': '#B903AA',\n",
    "'tlNBL': '#372101',\n",
    "'Immature_dpEX': '#FFDBE5',\n",
    "'Immature_mpEX': '#7A4900',\n",
    "'Immature_nptxEX': '#0000A6',\n",
    "'dpEX': '#61615A',\n",
    "'mpEX': '#BA0900',\n",
    "'nptxEX': '#00C2A0',\n",
    "'mpIN': '#6B7900',\n",
    "'cckIN': '#809693',\n",
    "'npyIN': '#FFAA92',\n",
    "'Immature_DMIN': '#008941',\n",
    "'Immature_cckIN': '#A30059',\n",
    "'ntng1IN': '#FF90C9',\n",
    "'scgnIN': '#0AA6D8',\n",
    "'sstIN': '#00846F',\n",
    "'rIPC1': '#D16100',\n",
    "'rIPC2': '#DDEFFF',\n",
    "'rIPC3': '#000035',\n",
    "'rIPC4': '#7B4F4B',\n",
    "'dEGC': '#6A3A4C',\n",
    "'reaEGC': '#A1C299',\n",
    "'ribEGC': '#300018',\n",
    "'sfrpEGC': '#013349',\n",
    "'wntEGC': '#FFB500',\n",
    "'Unknown': '#8FB0FF',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_map_ct =  {\n",
    "'CP': '#1CE6FF',\n",
    "'IMN': '#FF34FF',\n",
    "'Immature_CMPN': '#FF4A46',\n",
    "'CMPN': '#FF4A46',\n",
    "'Immature_MSN': '#006FA6',\n",
    "'MSN': '#006FA6',\n",
    "'MCG': '#63FFAC',\n",
    "'Oligo': '#004D43',\n",
    "'VLMC': '#997D87',\n",
    "'WSN': '#5A0007',\n",
    "'dNBL1': '#1B4400',\n",
    "'dNBL2': '#1B4400',\n",
    "'dNBL3': '#1B4400',\n",
    "'dNBL4': '#1B4400',\n",
    "'dNBL5': '#1B4400',\n",
    "'obNBL': '#B903AA',\n",
    "'tlNBL': '#372101',\n",
    "\n",
    "'Immature_dpEX': '#FFDBE5',\n",
    "'dpEX': '#FFDBE5',\n",
    "'Immature_mpEX': '#7A4900',\n",
    "'mpEX': '#7A4900',\n",
    "'Immature_nptxEX': '#0000A6',\n",
    "'nptxEX': '#0000A6',\n",
    "'Immature_cckIN': '#A30059',\n",
    "'cckIN': '#A30059',\n",
    "\n",
    "'mpIN': '#6B7900',\n",
    "'npyIN': '#FFAA92',\n",
    "'Immature_DMIN': '#008941',\n",
    "'ntng1IN': '#FF90C9',\n",
    "'scgnIN': '#0AA6D8',\n",
    "'sstIN': '#00846F',\n",
    "\n",
    "'rIPC1': '#D16100',\n",
    "'rIPC2': '#D16100',\n",
    "'rIPC3': '#D16100',\n",
    "'rIPC4': '#D16100',\n",
    "'dEGC': '#FFB500',\n",
    "'reaEGC': '#FFB500',\n",
    "'ribEGC': '#FFB500',\n",
    "'sfrpEGC': '#FFB500',\n",
    "'wntEGC': '#FFB500',\n",
    "'Unknown': '#8FB0FF',\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyvis.network import Network\n",
    "from IPython.display import HTML\n",
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "import plotly.io as pio\n",
    "import kaleido, random\n",
    "random.seed(997)\n",
    "np.random.seed(997)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_axo_dev.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i, label=keys[i])\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_axo_dev[i,:]\n",
    "    min_indices = np.argpartition(arr, 4)[:4]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,int(j),{'weight':float(1/arr[j]*100-60)}))\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "\n",
    "# Draw pyvis\n",
    "net = Network(width=2000, height=2000, notebook=True, cdn_resources='remote')\n",
    "net.from_nx(G)\n",
    "\n",
    "# set node color based on label\n",
    "for node in net.nodes:\n",
    "    node['group'] = node['label'].split(\" \")[1]\n",
    "    node['color'] = color_map_ct[node['group']]\n",
    "\n",
    "# show\n",
    "net.show_buttons()\n",
    "net.set_options(\"\"\"\n",
    "const options = {\n",
    "  \"configure\": {\n",
    "    \"enabled\": true,\n",
    "    \"filter\": [\"nodes\",\"edges\",\"physics\"]\n",
    "  },\n",
    "  \"nodes\": {\n",
    "    \"borderWidth\": 3,\n",
    "    \"opacity\": 1,\n",
    "    \"font\": {\n",
    "      \"size\": 15,\n",
    "      \"strokeWidth\": 5\n",
    "    },\n",
    "    \"size\": 0\n",
    "  },\n",
    "  \"edges\": {\n",
    "    \"color\": {\n",
    "      \"opacity\": 0.7\n",
    "    },\n",
    "    \"selfReferenceSize\": 0,\n",
    "    \"selfReference\": {\n",
    "      \"size\": 0,\n",
    "      \"angle\": 0.7853981633974483\n",
    "    },\n",
    "    \"smooth\": {\n",
    "      \"forceDirection\": \"vertical\"\n",
    "    }\n",
    "  },\n",
    "  \"physics\": {\n",
    "    \"minVelocity\": 0.75\n",
    "  }\n",
    "}\n",
    "\"\"\")\n",
    "\n",
    "net.show('networkplot_Axo_dev_neighbor3.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.sub-conformation by clustering on variance matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1][:-5]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='CP': \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[sample] = adata\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(lambda x, y: x.intersection(y), GeneLists.values())\n",
    "M = intersected_genes.size\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build gene-gene distance matrix\n",
    "keys = ['Stage44_CP',\n",
    "        'Stage54_CP',\n",
    "        'Stage57_CP',\n",
    "        'Control_Juv_CP',\n",
    "        'Adult_CP',\n",
    "        'Meta_CP']\n",
    "N = len(keys)\n",
    "GG_dev = np.zeros((N,M,M))\n",
    "for n in tqdm(range(N)):\n",
    "    index_k_intersection = [list(GeneLists[keys[n]]).index(gene) for gene in intersected_genes]\n",
    "    for i, idx_n in enumerate(index_k_intersection):\n",
    "        for j, idx_m in enumerate(index_k_intersection[:i+1]):\n",
    "            g1_cord = Xs[keys[n]].iloc[idx_n,:].to_numpy()\n",
    "            g2_cord = Xs[keys[n]].iloc[idx_m,:].to_numpy()\n",
    "            d = np.linalg.norm(g1_cord - g2_cord)\n",
    "            GG_dev[n,i,j] = GG_dev[n,j,i] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "V_dev_CP = np.zeros((M,M))\n",
    "for i in range(M):\n",
    "    for j in range(M):\n",
    "        v = statistics.variance(GG_dev[:,i,j])\n",
    "        V_dev_CP[i,j] = V_dev_CP[j,i] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate gene\n",
    "import pandas as pd\n",
    "anno = pd.read_csv(\"./Axo_Summary_Gene_Annotation_0325.txt\", sep='\\t')\n",
    "annodict = {row['Axolotl_ID']: row['hs_gene'] for _, row in anno.iterrows() if row['hs_gene'] != '-'}\n",
    "\n",
    "intersected_symbols=[]\n",
    "for i in intersected_genes:\n",
    "    intersected_symbols.append(annodict.get(i, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "V_dev_sqrt_CP = np.sqrt(V_dev_CP)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hierarchy\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "clustergrid = sns.clustermap(V_dev_sqrt_CP, xticklabels=intersected_symbols, yticklabels=intersected_symbols, method='ward')\n",
    "plt.close()\n",
    "\n",
    "# set\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(10, 10)},font_scale=1)\n",
    "\n",
    "ax = sns.clustermap(V_dev_sqrt_CP, xticklabels=[], yticklabels=[], cmap=\"viridis\", method='ward', cbar_kws={'ticks': [0,5,10],'orientation':'horizontal'}, vmax=10)\n",
    "ax.ax_cbar.set_title(\"Standard Deviation\", fontdict={'fontsize': 15})\n",
    "ax.ax_cbar.set_position((0.03, .89, .20, .05))\n",
    "# Remove the dendrograms\n",
    "#ax.ax_row_dendrogram.set_visible(False)\n",
    "#ax.ax_col_dendrogram.set_visible(False)\n",
    "\n",
    "## subconformation 1\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 45\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "# add patches\n",
    "row_order = ax.dendrogram_row.reordered_ind\n",
    "# start\n",
    "x = y = 0\n",
    "# width\n",
    "w = h = 0\n",
    "formal_c = 1\n",
    "count = 1\n",
    "for c in np.roll(clusters[row_order],-1):\n",
    "    if c == formal_c:\n",
    "        formal_c = c\n",
    "        count += 1\n",
    "    else:\n",
    "        w = h = count\n",
    "        if w>=3:\n",
    "            ax.ax_heatmap.add_patch(patches.Rectangle((x, y), w, h, fill=False, edgecolor='white', lw=1.5, clip_on=False))\n",
    "        x += count\n",
    "        y = x\n",
    "        formal_c = c\n",
    "        count = 1\n",
    "\n",
    "## subconformation 2\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 20\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "# add patches\n",
    "row_order = ax.dendrogram_row.reordered_ind\n",
    "# start\n",
    "x = y = 0\n",
    "# width\n",
    "w = h = 0\n",
    "formal_c = 1\n",
    "count = 1\n",
    "for c in np.roll(clusters[row_order],-1):\n",
    "    if c == formal_c:\n",
    "        formal_c = c\n",
    "        count += 1\n",
    "    else:\n",
    "        w = h = count\n",
    "        if w>=3:\n",
    "            ax.ax_heatmap.add_patch(patches.Rectangle((x, y), w, h, fill=False, edgecolor='orange', lw=1.5, clip_on=False))\n",
    "        x += count\n",
    "        y = x\n",
    "        formal_c = c\n",
    "        count = 1\n",
    "\n",
    "plt.savefig(\"figures/Fig5f.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "# mask\n",
    "#mask = np.tril(np.ones_like(V_dev_sqrt_CP),k=-1)\n",
    "#values = ax.ax_heatmap.collections[0].get_array().reshape(V_dev_sqrt_CP.shape)\n",
    "#new_values = np.ma.array(values, mask=mask)\n",
    "#ax.ax_heatmap.collections[0].set_array(new_values)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sunburst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# Generates Seaborn clustermap and dendrograms\n",
    "df = pd.DataFrame(V_dev_sqrt_CP)\n",
    "col_name = 'class'\n",
    "sns_cluster = sns.clustermap(df, method='ward', metric='euclidean')\n",
    "plt.close()\n",
    "row_linkage = sns_cluster.dendrogram_row.linkage\n",
    "col_linkage = sns_cluster.dendrogram_col.linkage\n",
    "\n",
    "# Assigns clusters to the data\n",
    "df[col_name + '1'] = fcluster(row_linkage, t=20, criterion='distance')\n",
    "df[col_name + '2'] = fcluster(row_linkage, t=45, criterion='distance')\n",
    "df['weight']=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "    enr = pd.DataFrame(columns=['Class','Sub-conformation','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "\n",
    "    # Enricr API\n",
    "    for clusterclass in ['class1','class2']:\n",
    "        unique_values, counts = np.unique(df[clusterclass], return_counts=True)\n",
    "        for value, count in zip(unique_values, counts):\n",
    "            if count>=3:\n",
    "                try:\n",
    "                    genelist = np.array(intersected_symbols)[df[clusterclass]==value].tolist()\n",
    "                    filterlist = list(filter(lambda x: 'AMEX60DD' not in x, genelist))\n",
    "                    enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                                        organism='Human',\n",
    "                                        gene_sets=gs,\n",
    "                                        outdir=None, \n",
    "                                        cutoff=0.05,\n",
    "                                        no_plot=True)\n",
    "\n",
    "                    # trim (go:...)\n",
    "                    enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "                    enr_up.res2d['Class'] = clusterclass\n",
    "                    enr_up.res2d['Sub-conformation'] = value\n",
    "                    enr = pd.concat([enr, enr_up.res2d[enr_up.res2d['Adjusted P-value']<0.05]], ignore_index=True)\n",
    "                    # dotplot\n",
    "                    # gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(value), size=10, cmap = plt.cm.viridis_r)\n",
    "                except ValueError or FileNotFoundError or TypeError as e:\n",
    "                    print(value)\n",
    "                    print(e)\n",
    "                    continue\n",
    "\n",
    "    enr.to_csv('AXO_dev_CP_subconformation_hierachy_clustering_'+gs+'.pjust.05.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['KEGG_label1']=df['GO_BP_label1']=df['GO_MF_label1']=df['class1']\n",
    "df['KEGG_label2']=df['GO_BP_label2']=df['GO_MF_label2']=df['class2']\n",
    "\n",
    "# GO_BP\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_CP_subconformation_hierachy_clustering_GO_Biological_Process_2023.pjust.05.csv')\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['GO_BP_label1', 'GO_BP_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(df[clusterclass]):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "        except:\n",
    "            enr_class_top1_label = 'unknown'\n",
    "        df[label] = df[label].replace(cluster, str(cluster)+' '+enr_class_top1_label)\n",
    "\n",
    "# GO_MF\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_CP_subconformation_hierachy_clustering_GO_Molecular_Function_2023.pjust.05.csv')\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['GO_MF_label1', 'GO_MF_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(df[clusterclass]):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "        except:\n",
    "            enr_class_top1_label = 'unknown'\n",
    "        df[label] = df[label].replace(cluster, str(cluster)+' '+enr_class_top1_label)\n",
    "\n",
    "# KEGG\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_CP_subconformation_hierachy_clustering_KEGG_2021_Human.pjust.05.csv')\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['KEGG_label1', 'KEGG_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(df[clusterclass]):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "        except:\n",
    "            enr_class_top1_label = 'unknown'\n",
    "        df[label] = df[label].replace(cluster, str(cluster)+' '+enr_class_top1_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "labels=[]\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_CP_subconformation_hierachy_clustering_GO_Molecular_Function_2023.pjust.05.csv', index_col=0)\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['GO_BP_label1', 'GO_BP_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(enr_class['Sub-conformation']):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "            labels.append(enr_class_top1_label)\n",
    "        except:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### draw and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.io as pio\n",
    "import plotly.graph_objs as go\n",
    "from cytocraft.plot import get_cmap\n",
    "# naming a layout theme for future reference\n",
    "pio.templates[\"viridis\"] = go.layout.Template(\n",
    "    layout_colorway=get_cmap('viridis', 8)\n",
    ")\n",
    "pio.templates[\"google\"] = go.layout.Template(\n",
    "    layout_colorway=['#4285F4', '#DB4437', '#F4B400', '#0F9D58',\n",
    "                     '#185ABC', '#B31412', '#EA8600', '#137333',\n",
    "                     '#d2e3fc', '#ceead6']\n",
    ")\n",
    "# setting Google color palette as default\n",
    "pio.templates.default = \"viridis\"\n",
    "\n",
    "df=remove_rare_values(df, column_names=['class1','class2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=df['KEGG_label1'].unique()\n",
    "dict(zip(labels,labels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['KEGG_label1'] = df['KEGG_label1'].map({'8 Ribosome': '8 Ribosome',\n",
    " '4 unknown': '4 unknown',\n",
    " '1 unknown': '1 unknown',\n",
    " '5 Oxidative phosphorylation': '5 Oxidative phosphorylation',\n",
    " '12 Ribosome': '12 Ribosome',\n",
    " '10 unknown': '10 unknown',\n",
    " '7 Ribosome': '7 Ribosome',\n",
    " '2 Endocytosis': '2 Endocytosis',\n",
    " '19 unknown': '18 unknown',\n",
    " '6 Ribosome': '6 Ribosome',\n",
    " '23 Synaptic vesicle cycle': '22 Synaptic vesicle cycle',\n",
    " '9 Ribosome': '9 Ribosome',\n",
    " '16 unknown': '15 unknown',\n",
    " '11 unknown': '11 unknown',\n",
    " '3 unknown': '3 unknown',\n",
    " '14 unknown': '13 unknown',\n",
    " '18 Amino sugar and nucleotide sugar metabolism': '17 Amino sugar and nucleotide sugar metabolism',\n",
    " '17 unknown': '16 unknown',\n",
    " '15 Cell adhesion molecules': '14 Cell adhesion molecules',\n",
    " '20 Type I diabetes mellitus': '19 Type I diabetes mellitus',\n",
    " '21 unknown': '20 unknown',\n",
    " '22 unknown': '21 unknown'})\n",
    "\n",
    "df['GO_MF_label1'] = df['GO_MF_label1'].map({'8 RNA Binding': '8 RNA Binding',\n",
    " '4 MHC Class II Protein Binding': '4 MHC Class II Protein Binding',\n",
    " '1 Protein Serine/Threonine Phosphatase Activity': '1 Protein Serine/Threonine Phosphatase Activity',\n",
    " '5 Oxidoreduction-Driven Active Transmembrane Transporter Activity': '5 Oxidoreduction-Driven Active Transmembrane Transporter Activity',\n",
    " '12 RNA Binding': '12 RNA Binding',\n",
    " '10 RNA Binding': '10 RNA Binding',\n",
    " '7 RNA Binding': '7 RNA Binding',\n",
    " '2 Purine Ribonucleoside Triphosphate Binding': '2 Purine Ribonucleoside Triphosphate Binding',\n",
    " '19 Calcium Channel Inhibitor Activity': '18 Calcium Channel Inhibitor Activity',\n",
    " '6 RNA Binding': '6 RNA Binding',\n",
    " '23 Acetyltransferase Activator Activity': '22 Acetyltransferase Activator Activity',\n",
    " \"9 5'-3' Exonuclease Activity\": \"9 5'-3' Exonuclease Activity\",\n",
    " '16 Metal-Dependent Deubiquitinase Activity': '15 Metal-Dependent Deubiquitinase Activity',\n",
    " '11 mRNA Binding': '11 mRNA Binding',\n",
    " '3 RNA Binding': '3 RNA Binding',\n",
    " '14 unknown': '13 unknown',\n",
    " '18 Cytidylyltransferase Activity': '17 Cytidylyltransferase Activity',\n",
    " '17 Aminophospholipid Flippase Activity': '16 Aminophospholipid Flippase Activity',\n",
    " '15 Axon Guidance Receptor Activity': '14 Axon Guidance Receptor Activity',\n",
    " '20 Transmembrane Receptor Protein Phosphatase Activity': '19 Transmembrane Receptor Protein Phosphatase Activity',\n",
    " '21 Transferrin Receptor Binding': '20 Transferrin Receptor Binding',\n",
    " '22 Clathrin Light Chain Binding': '21 Clathrin Light Chain Binding'})\n",
    "\n",
    "df['GO_BP_label1'] = df['GO_BP_label1'].map({'8 Cytoplasmic Translation': '8 Cytoplasmic Translation',\n",
    " '4 Humoral Immune Response Mediated By Circulating Immunoglobulin': '4 Humoral Immune Response Mediated By Circulating Immunoglobulin',\n",
    " '1 Peptidyl-Serine Dephosphorylation': '1 Peptidyl-Serine Dephosphorylation',\n",
    " '5 Negative Regulation Of Protein Polyubiquitination': '5 Negative Regulation Of Protein Polyubiquitination',\n",
    " '12 Macromolecule Biosynthetic Process': '12 Macromolecule Biosynthetic Process',\n",
    " '10 Cytoplasmic Translation': '10 Cytoplasmic Translation',\n",
    " '7 Macromolecule Biosynthetic Process': '7 Macromolecule Biosynthetic Process',\n",
    " '2 Vesicle-Mediated Transport': '2 Vesicle-Mediated Transport',\n",
    " '19 Negative Regulation Of Heart Rate': '18 Negative Regulation Of Heart Rate',\n",
    " '6 Cytoplasmic Translation': '6 Cytoplasmic Translation',\n",
    " '23 Synaptic Vesicle Docking': '22 Synaptic Vesicle Docking',\n",
    " '9 Cytoplasmic Translation': '9 Cytoplasmic Translation',\n",
    " '16 Regulation Of Wnt Signaling Pathway': '15 Regulation Of Wnt Signaling Pathway',\n",
    " '11 Cytoplasmic Translation': '11 Cytoplasmic Translation',\n",
    " '3 unknown': '3 unknown',\n",
    " '14 Regulation Of Dendrite Development': '13 Regulation Of Dendrite Development',\n",
    " '18 Nucleotide-Sugar Biosynthetic Process': '17 Nucleotide-Sugar Biosynthetic Process',\n",
    " '17 DNA-templated DNA Replication Maintenance Of Fidelity': '16 DNA-templated DNA Replication Maintenance Of Fidelity',\n",
    " '15 Positive Regulation Of Axon Extension': '14 Positive Regulation Of Axon Extension',\n",
    " '20 Regulation Of Protein Depolymerization': '19 Regulation Of Protein Depolymerization',\n",
    " '21 Positive Regulation Of Dendritic Spine Morphogenesis': '20 Positive Regulation Of Dendritic Spine Morphogenesis',\n",
    " '22 Regulation Of Hyaluronan Biosynthetic Process': '21 Regulation Of Hyaluronan Biosynthetic Process'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG\n",
    "import plotly.express as px\n",
    "from plotly.io import write_image\n",
    "df['KEGG_label1'] = df['KEGG_label1'].apply(lambda x: wrap_labels(x,28,ignore=['14']))\n",
    "df['KEGG_label2'] = df['KEGG_label2'].apply(lambda x: wrap_labels(x,28,ignore=['5','7']))\n",
    "fig = px.sunburst(df, path=['KEGG_label2','KEGG_label1'], values='weight')\n",
    "fig.update_layout(autosize=False, width=600, height=600, margin=dict(l=0, r=0, t=0, b=0))\n",
    "fig.update_traces(insidetextorientation='radial')\n",
    "fig.show()\n",
    "# export figure to pdf image\n",
    "write_image(fig, 'figures/Sunburst_KEGG_AXO_dev_CP.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO MF\n",
    "import plotly.express as px\n",
    "from plotly.io import write_image\n",
    "df['GO_MF_label1'] = df['GO_MF_label1'].apply(lambda x: wrap_labels(x,32,ignore=['15','18','19','21','22','23']))\n",
    "df['GO_MF_label2'] = df['GO_MF_label2'].apply(lambda x: wrap_labels(x,32))\n",
    "fig = px.sunburst(df, path=['GO_MF_label2','GO_MF_label1'], values='weight')\n",
    "fig.update_layout(autosize=False, width=600, height=600, margin=dict(l=0, r=0, t=0, b=0))\n",
    "fig.update_traces(insidetextorientation='radial')\n",
    "fig.show()\n",
    "# export figure to png image\n",
    "write_image(fig, 'figures/Sunburst_GO_MF_AXO_dev_CP.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO BP\n",
    "import plotly.express as px\n",
    "from plotly.io import write_image\n",
    "df['GO_BP_label1'] = df['GO_BP_label1'].apply(lambda x: wrap_labels(x,32,ignore=['15','14']))\n",
    "df['GO_BP_label2'] = df['GO_BP_label2'].apply(lambda x: wrap_labels(x,32,ignore=['6']))\n",
    "fig = px.sunburst(df, path=['GO_BP_label2','GO_BP_label1'], values='weight')\n",
    "fig.update_layout(autosize=False, width=600, height=600,margin=dict(l=0, r=0, t=0, b=0))\n",
    "fig.update_traces(insidetextorientation='radial')\n",
    "fig.show()\n",
    "# export figure to png image\n",
    "write_image(fig, 'figures/Sunburst_GO_BP_AXO_dev_CP.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VLMC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Preprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "adatas={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1][:-5]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='VLMC': \n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            adatas[sample] = adata\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(lambda x, y: x.intersection(y), GeneLists.values())\n",
    "M = intersected_genes.size\n",
    "print(M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build gene-gene distance matrix\n",
    "keys = ['Stage44_VLMC',\n",
    "        'Stage54_VLMC',\n",
    "        'Stage57_VLMC',\n",
    "        'Control_Juv_VLMC',\n",
    "        'Adult_VLMC',\n",
    "        'Meta_VLMC']\n",
    "N = len(keys)\n",
    "GG_dev = np.zeros((N,M,M))\n",
    "for n in tqdm(range(N)):\n",
    "    index_k_intersection = [list(GeneLists[keys[n]]).index(gene) for gene in intersected_genes]\n",
    "    for i, idx_n in enumerate(index_k_intersection):\n",
    "        for j, idx_m in enumerate(index_k_intersection[:i+1]):\n",
    "            g1_cord = Xs[keys[n]].iloc[idx_n,:].to_numpy()\n",
    "            g2_cord = Xs[keys[n]].iloc[idx_m,:].to_numpy()\n",
    "            d = np.linalg.norm(g1_cord - g2_cord)\n",
    "            GG_dev[n,i,j] = GG_dev[n,j,i] = d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "V_dev_VLMC = np.zeros((M,M))\n",
    "for i in range(M):\n",
    "    for j in range(M):\n",
    "        v = statistics.variance(GG_dev[:,i,j])\n",
    "        V_dev_VLMC[i,j] = V_dev_VLMC[j,i] = v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate gene\n",
    "import pandas as pd\n",
    "anno = pd.read_csv(\"./Axo_Summary_Gene_Annotation_0325.txt\", sep='\\t')\n",
    "annodict = {row['Axolotl_ID']: row['hs_gene'] for _, row in anno.iterrows() if row['hs_gene'] != '-'}\n",
    "\n",
    "intersected_symbols=[]\n",
    "for i in intersected_genes:\n",
    "    intersected_symbols.append(annodict.get(i, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "V_dev_sqrt_VLMC = np.sqrt(V_dev_VLMC)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hierarchy\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "\n",
    "clustergrid = sns.clustermap(V_dev_sqrt_VLMC, xticklabels=intersected_symbols, yticklabels=intersected_symbols, method='ward')\n",
    "plt.close()\n",
    "\n",
    "# set\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(10, 10)},font_scale=1)\n",
    "\n",
    "ax = sns.clustermap(V_dev_sqrt_VLMC, xticklabels=[], yticklabels=[], cmap=\"viridis\", method='ward', cbar_kws={'ticks': [0,5,10],'orientation':'horizontal'}, vmax=10)\n",
    "ax.ax_cbar.set_title(\"Standard Deviation\", fontdict={'fontsize': 15})\n",
    "ax.ax_cbar.set_position((0.03, .89, .20, .05))\n",
    "# Remove the dendrograms\n",
    "#ax.ax_row_dendrogram.set_visible(False)\n",
    "#ax.ax_col_dendrogram.set_visible(False)\n",
    "\n",
    "## subconformation 1\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 60\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "# add patches\n",
    "row_order = ax.dendrogram_row.reordered_ind\n",
    "# start\n",
    "x = y = 0\n",
    "# width\n",
    "w = h = 0\n",
    "formal_c = 1\n",
    "count = 1\n",
    "for c in np.roll(clusters[row_order],-1):\n",
    "    if c == formal_c:\n",
    "        formal_c = c\n",
    "        count += 1\n",
    "    else:\n",
    "        w = h = count\n",
    "        if w>=3:\n",
    "            ax.ax_heatmap.add_patch(patches.Rectangle((x, y), w, h, fill=False, edgecolor='white', lw=1.5, clip_on=False))\n",
    "        x += count\n",
    "        y = x\n",
    "        formal_c = c\n",
    "        count = 1\n",
    "\n",
    "## subconformation 2\n",
    "# Get clusters with threshold\n",
    "linkage = clustergrid.dendrogram_row.linkage\n",
    "threshold = 30\n",
    "clusters = fcluster(linkage, threshold, criterion='distance')\n",
    "# add patches\n",
    "row_order = ax.dendrogram_row.reordered_ind\n",
    "# start\n",
    "x = y = 0\n",
    "# width\n",
    "w = h = 0\n",
    "formal_c = 1\n",
    "count = 1\n",
    "for c in np.roll(clusters[row_order],-1):\n",
    "    if c == formal_c:\n",
    "        formal_c = c\n",
    "        count += 1\n",
    "    else:\n",
    "        w = h = count\n",
    "        if w>=3:\n",
    "            ax.ax_heatmap.add_patch(patches.Rectangle((x, y), w, h, fill=False, edgecolor='orange', lw=1.5, clip_on=False))\n",
    "        x += count\n",
    "        y = x\n",
    "        formal_c = c\n",
    "        count = 1\n",
    "\n",
    "plt.savefig(\"figures/Fig5f_l.pdf\", format=\"pdf\", bbox_inches=\"tight\")\n",
    "# mask\n",
    "#mask = np.tril(np.ones_like(V_dev_sqrt_VLMC),k=-1)\n",
    "#values = ax.ax_heatmap.collections[0].get_array().reshape(V_dev_sqrt_VLMC.shape)\n",
    "#new_values = np.ma.array(values, mask=mask)\n",
    "#ax.ax_heatmap.collections[0].set_array(new_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import the Python Image \n",
    "# processing Library \n",
    "from PIL import Image \n",
    "\n",
    "# Giving The Original image Directory \n",
    "# Specified \n",
    "Original_Image = Image.open(\"./Clustermap_subconf_Axo_dev_sqrt_VLMC.png\") \n",
    "\n",
    "# Rotate Image By 180 Degree \n",
    "rotated_image1 = Original_Image.rotate(180) \n",
    "\n",
    "# This is Alternative Syntax To Rotate \n",
    "# The Image \n",
    "rotated_image2 = Original_Image.transpose(Transpose.ROTATE_90)\n",
    "\n",
    "# This Will Rotate Image By 60 Degree \n",
    "rotated_image3 = Original_Image.rotate(60) \n",
    "\n",
    "rotated_image1.show() \n",
    "rotated_image2.show() \n",
    "rotated_image3.show() \n",
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cv2  # importing cv \n",
    "import imutils \n",
    "\n",
    "# read an image as input using OpenCV \n",
    "image = cv2.imread(\"./Clustermap_subconf_Axo_dev_sqrt_VLMC.jpg\") \n",
    "#Rotated_image = imutils.rotate(image, angle=45) \n",
    "#Rotated1_image = imutils.rotate(image, angle=90) \n",
    "  \n",
    "# display the image using OpenCV of \n",
    "# angle 45 \n",
    "#cv2.imshow(\"Rotated\", Rotated_image) \n",
    "  \n",
    "# display the image using OpenCV of\n",
    "# angle 90 \n",
    "#cv2.imshow(\"Rotated\", Rotated1_image)\n",
    "# This is used for To Keep On Displaying \n",
    "# The Image Until Any Key is Pressed \n",
    "#cv2.waitKey(0) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sunburst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generates Seaborn clustermap and dendrograms\n",
    "df = pd.DataFrame(V_dev_sqrt_VLMC)\n",
    "col_name = 'class'\n",
    "sns_cluster = sns.clustermap(df, method='ward', metric='euclidean')\n",
    "plt.close()\n",
    "row_linkage = sns_cluster.dendrogram_row.linkage\n",
    "col_linkage = sns_cluster.dendrogram_col.linkage\n",
    "\n",
    "# Assigns clusters to the data\n",
    "df[col_name + '1'] = fcluster(row_linkage, t=30, criterion='distance')\n",
    "df[col_name + '2'] = fcluster(row_linkage, t=60, criterion='distance')\n",
    "df['weight']=1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "    enr = pd.DataFrame(columns=['Class','Sub-conformation','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "\n",
    "    # Enricr API\n",
    "    for clusterclass in ['class1','class2']:\n",
    "        unique_values, counts = np.unique(df[clusterclass], return_counts=True)\n",
    "        for value, count in zip(unique_values, counts):\n",
    "            if count >= 3:\n",
    "                try:\n",
    "                    genelist = np.array(intersected_symbols)[df[clusterclass]==value].tolist()\n",
    "                    filterlist = list(filter(lambda x: 'AMEX60DD' not in x, genelist))\n",
    "                    enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                                        organism='Human',\n",
    "                                        gene_sets=gs,\n",
    "                                        outdir=None, \n",
    "                                        cutoff=0.05,\n",
    "                                        no_plot=True)\n",
    "\n",
    "                    # trim (go:...)\n",
    "                    enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "                    enr_up.res2d['Class'] = clusterclass\n",
    "                    enr_up.res2d['Sub-conformation'] = value\n",
    "                    enr = pd.concat([enr, enr_up.res2d[enr_up.res2d['Adjusted P-value']<0.05]], ignore_index=True)\n",
    "                    # dotplot\n",
    "                    # gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(value), size=10, cmap = plt.cm.viridis_r)\n",
    "                except ValueError or FileNotFoundError or TypeError as e:\n",
    "                    print(value)\n",
    "                    print(e)\n",
    "                    continue\n",
    "\n",
    "    enr.to_csv('AXO_dev_VLMC_subconformation_hierachy_clustering_'+gs+'.pjust.05.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write df\n",
    "df['KEGG_label1']=df['GO_BP_label1']=df['GO_MF_label1']=df['class1']\n",
    "df['KEGG_label2']=df['GO_BP_label2']=df['GO_MF_label2']=df['class2']\n",
    "\n",
    "# GO_BP\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_VLMC_subconformation_hierachy_clustering_GO_Biological_Process_2023.pjust.05.csv')\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['GO_BP_label1', 'GO_BP_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(df[clusterclass]):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "        except:\n",
    "            enr_class_top1_label = 'unknown'\n",
    "        df[label] = df[label].replace(cluster, str(cluster)+' '+enr_class_top1_label)\n",
    "\n",
    "# GO_MF\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_VLMC_subconformation_hierachy_clustering_GO_Molecular_Function_2023.pjust.05.csv')\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['GO_MF_label1', 'GO_MF_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(df[clusterclass]):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "        except:\n",
    "            enr_class_top1_label = 'unknown'\n",
    "        df[label] = df[label].replace(cluster, str(cluster)+' '+enr_class_top1_label)\n",
    "\n",
    "# KEGG\n",
    "enr = pd.read_csv('subconfiguration/AXO_dev_VLMC_subconformation_hierachy_clustering_KEGG_2021_Human.pjust.05.csv')\n",
    "for clusterclass, label in zip(['class1', 'class2'] ,['KEGG_label1', 'KEGG_label2']):\n",
    "    enr_class = enr[enr[\"Class\"]==clusterclass]\n",
    "    for cluster in np.unique(df[clusterclass]):\n",
    "        try:\n",
    "            enr_class_top1_label = enr_class[enr_class['Sub-conformation']==cluster].iloc[0]['Term']\n",
    "        except:\n",
    "            enr_class_top1_label = 'unknown'\n",
    "        df[label] = df[label].replace(cluster, str(cluster)+' '+enr_class_top1_label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### draw and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.io as pio\n",
    "import plotly.graph_objs as go\n",
    "from cytocraft.plot import get_cmap\n",
    "# naming a layout theme for future reference\n",
    "pio.templates[\"viridis\"] = go.layout.Template(\n",
    "    layout_colorway=get_cmap('viridis', 8)\n",
    ")\n",
    "pio.templates[\"google\"] = go.layout.Template(\n",
    "    layout_colorway=['#4285F4', '#DB4437', '#F4B400', '#0F9D58',\n",
    "                     '#185ABC', '#B31412', '#EA8600', '#137333',\n",
    "                     '#d2e3fc', '#ceead6']\n",
    ")\n",
    "# setting Google color palette as default\n",
    "pio.templates.default = \"viridis\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df=remove_rare_values(df, column_names=['class1','class2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels=df['KEGG_label1'].unique()\n",
    "dict(zip(labels,labels))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['KEGG_label1'] = df['KEGG_label1'].map({'7 Ribosome': '7 Ribosome',\n",
    "    '12 unknown': '12 unknown',\n",
    "    '15 Ribosome': '15 Ribosome',\n",
    "    '9 Ribosome': '9 Ribosome',\n",
    "    '6 Ribosome': '6 Ribosome',\n",
    "    '10 Diabetic cardiomyopathy': '10 Diabetic cardiomyopathy',\n",
    "    '2 unknown': '2 unknown',\n",
    "    '3 Ribosome': '3 Ribosome',\n",
    "    '5 Ribosome': '5 Ribosome',\n",
    "    '14 Pentose phosphate pathway': '14 Pentose phosphate pathway',\n",
    "    '8 Amyotrophic lateral sclerosis': '8 Amyotrophic lateral sclerosis',\n",
    "    '4 unknown': '4 unknown',\n",
    "    '1 Adrenergic signaling in cardiomyocytes': '1 Adrenergic signaling in cardiomyocytes',\n",
    "    '22 unknown': '21 unknown',\n",
    "    '25 Human papillomavirus infection': '23 Human papillomavirus infection',\n",
    "    '11 unknown': '11 unknown',\n",
    "    '20 unknown': '19 unknown',\n",
    "    '16 unknown': '16 unknown',\n",
    "    '26 unknown': '24 unknown',\n",
    "    '13 unknown': '13 unknown',\n",
    "    '24 Legionellosis': '22 Legionellosis',\n",
    "    '18 Type I diabetes mellitus': '18 Type I diabetes mellitus',\n",
    "    '21 Vasopressin-regulated water reabsorption': '20 Vasopressin-regulated water reabsorption',\n",
    "    '17 Collecting duct acid secretion': '17 Collecting duct acid secretion'})\n",
    "\n",
    "df['GO_MF_label1'] = df['GO_MF_label1'].map({'7 RNA Binding': '7 RNA Binding',\n",
    "    '12 unknown': '12 unknown',\n",
    "    '15 Translation Initiation Factor Activity': '15 Translation Initiation Factor Activity',\n",
    "    '9 RNA Binding': '9 RNA Binding',\n",
    "    '6 RNA Binding': '6 RNA Binding',\n",
    "    '10 RNA Binding': '10 RNA Binding',\n",
    "    '2 Cadherin Binding': '2 Cadherin Binding',\n",
    "    '3 Serine-Type Endopeptidase Inhibitor Activity': '3 Serine-Type Endopeptidase Inhibitor Activity',\n",
    "    '5 RNA Binding': '5 RNA Binding',\n",
    "    '14 Dynactin Binding': '14 Dynactin Binding',\n",
    "    \"8 pre-mRNA 3'-Splice Site Binding\": \"8 pre-mRNA 3'-Splice Site Binding\",\n",
    "    '4 P-type Calcium Transporter Activity': '4 P-type Calcium Transporter Activity',\n",
    "    '1 RNA Binding': '1 RNA Binding',\n",
    "    '22 High-Density Lipoprotein Particle Binding': '21 High-Density Lipoprotein Particle Binding',\n",
    "    '25 Platelet-Derived Growth Factor Binding': '23 Platelet-Derived Growth Factor Binding',\n",
    "    '11 pre-mRNA Binding': '11 pre-mRNA Binding',\n",
    "    '20 Arginine N-methyltransferase Activity': '19 Arginine N-methyltransferase Activity',\n",
    "    '16 acetyl-CoA C-acyltransferase Activity': '16 acetyl-CoA C-acyltransferase Activity',\n",
    "    '26 Axon Guidance Receptor Activity': '24 Axon Guidance Receptor Activity',\n",
    "    '13 unknown': '13 unknown',\n",
    "    '24 Disulfide Oxidoreductase Activity': '22 Disulfide Oxidoreductase Activity',\n",
    "    '18 Transmembrane Receptor Protein Phosphatase Activity': '18 Transmembrane Receptor Protein Phosphatase Activity',\n",
    "    '21 GDP-dissociation Inhibitor Activity': '20 GDP-dissociation Inhibitor Activity',\n",
    "    '17 Proton-Transporting ATPase Activity, Rotational Mechanism': '17 Proton-Transporting ATPase Activity, Rotational Mechanism'})\n",
    "\n",
    "df['GO_BP_label1'] = df['GO_BP_label1'].map({'7 Cytoplasmic Translation': '7 Cytoplasmic Translation',\n",
    "    '12 Cytoplasmic Translation': '12 Cytoplasmic Translation',\n",
    "    '15 Cytoplasmic Translation': '15 Cytoplasmic Translation',\n",
    "    '9 Cytoplasmic Translation': '9 Cytoplasmic Translation',\n",
    "    '6 Cytoplasmic Translation': '6 Cytoplasmic Translation',\n",
    "    '10 Alternative mRNA Splicing, Via Spliceosome': '10 Alternative mRNA Splicing, Via Spliceosome',\n",
    "    '2 Glucose Metabolic Process': '2 Glucose Metabolic Process',\n",
    "    '3 Regulation Of Translation': '3 Regulation Of Translation',\n",
    "    '5 Cytoplasmic Translation': '5 Cytoplasmic Translation',\n",
    "    '14 Protein Stabilization': '14 Protein Stabilization',\n",
    "    '8 unknown': '8 unknown',\n",
    "    '4 Regulation Of Cellular Response To Insulin Stimulus': '4 Regulation Of Cellular Response To Insulin Stimulus',\n",
    "    '1 Cytoplasmic Translation': '1 Cytoplasmic Translation',\n",
    "    '22 Regulation Of Cytokinesis': '21 Regulation Of Cytokinesis',\n",
    "    '25 Regulation Of Microtubule Binding': '23 Regulation Of Microtubule Binding',\n",
    "    '11 unknown': '11 unknown',\n",
    "    '20 Negative Regulation Of Programmed Cell Death': '19 Negative Regulation Of Programmed Cell Death',\n",
    "    '16 Hypotonic Response': '16 Hypotonic Response',\n",
    "    '26 Neuron Projection Guidance': '24 Neuron Projection Guidance',\n",
    "    '13 SNARE Complex Assembly': '13 SNARE Complex Assembly',\n",
    "    '24 Regulation Of Gene Expression': '22 Regulation Of Gene Expression',\n",
    "    '18 Regulation Of Transport': '18 Regulation Of Transport',\n",
    "    '21 Negative Regulation Of Trophoblast Cell Migration': '20 Negative Regulation Of Trophoblast Cell Migration',\n",
    "    '17 ncRNA Transcription': '17 ncRNA Transcription'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG\n",
    "import plotly.express as px\n",
    "from plotly.io import write_image\n",
    "df['KEGG_label1'] = df['KEGG_label1'].apply(lambda x: wrap_labels(x,35,ignore=['18']))\n",
    "df['KEGG_label2'] = df['KEGG_label2'].apply(lambda x: wrap_labels(x,32))\n",
    "fig = px.sunburst(df, path=['KEGG_label2','KEGG_label1'], values='weight')\n",
    "fig.update_layout(autosize=False, width=600, height=600, margin=dict(l=0, r=0, t=0, b=0))\n",
    "fig.update_traces(insidetextorientation='radial')\n",
    "fig.show()\n",
    "# export figure to pdf image\n",
    "write_image(fig, 'figures/Sunburst_KEGG_AXO_dev_VLMC.pdf', scale=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO MF\n",
    "import plotly.express as px\n",
    "from plotly.io import write_image\n",
    "df['GO_MF_label1'] = df['GO_MF_label1'].apply(lambda x: wrap_labels(x,32,ignore=['18','21','25', '20']))\n",
    "df['GO_MF_label2'] = df['GO_MF_label2'].apply(lambda x: wrap_labels(x,35, ignore=['10']))\n",
    "fig = px.sunburst(df, path=['GO_MF_label2','GO_MF_label1'], values='weight')\n",
    "fig.update_layout(autosize=True, width=600, height=600,margin=dict(l=0, r=0, t=0, b=0))\n",
    "fig.update_traces(insidetextorientation='radial')\n",
    "fig.update_layout(title='CP')\n",
    "fig.show()\n",
    "# export figure to png image\n",
    "write_image(fig, 'figures/Sunburst_GO_MF_AXO_dev_VLMC.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO BP\n",
    "import plotly.express as px\n",
    "from plotly.io import write_image\n",
    "\n",
    "df['GO_BP_label1'] = df['GO_BP_label1'].apply(lambda x: wrap_labels(x,35,ignore=['23']))\n",
    "df['GO_BP_label2'] = df['GO_BP_label2'].apply(lambda x: wrap_labels(x,30,ignore=['6']))\n",
    "fig = px.sunburst(df, path=['GO_BP_label2','GO_BP_label1'], values='weight', color_discrete_map='Set2')\n",
    "fig.update_layout(autosize=False, width=600, height=600,margin=dict(l=0, r=0, t=0, b=0))\n",
    "fig.update_traces(insidetextorientation='radial')\n",
    "fig.show()\n",
    "# export figure to pdf image\n",
    "write_image(fig, 'figures/Sunburst_GO_BP_AXO_dev_VLMC.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.similarity matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(5, 2)}, font_scale=1)\n",
    "fig, axes = plt.subplots(1, 3, gridspec_kw={'width_ratios':[1,1,0.1]})\n",
    "\n",
    "# VLMC heatmap\n",
    "sns.heatmap(D_Axo_dev_VLMC, xticklabels=labels_dev, yticklabels=labels_dev, cmap=\"viridis\", cbar=False, ax=axes[0])\n",
    "axes[0].set_title('VLMC', fontdict={'fontsize': 10, 'fontweight': 'bold'})\n",
    "\n",
    "# CP heatmap\n",
    "sns.heatmap(D_Axo_dev_CP, xticklabels=labels_dev, yticklabels=labels_dev, cmap=\"viridis\", cbar_ax=axes[2], ax=axes[1])\n",
    "axes[1].set_title('CP', fontdict={'fontsize': 10, 'fontweight': 'bold'})\n",
    "\n",
    "axes[2].set_ylabel('RMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)\n",
    "plt.tight_layout()\n",
    "plt.savefig('figures/Fig5c_combined.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.show()\n",
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import *\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "adatas={}\n",
    "GeneLists={}\n",
    "GeneCounts={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='CP':\n",
    "                adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "                adatas[sample] = adata\n",
    "                GeneLists[sample] = adata.uns['X'].index.values\n",
    "                GeneCounts[sample] = adata.n_vars  \n",
    "\n",
    "keys = ['Stage44_CP',\n",
    "        'Stage54_CP',\n",
    "        'Stage57_CP',\n",
    "        'Control_Juv_CP',\n",
    "        'Adult_CP',\n",
    "        'Meta_CP']\n",
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig5c heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D_Axo_dev_CP = RMSD_distance_matrix(adatas, order=keys, ngene=100)\n",
    "\n",
    "labels_dev = ['St.44',\n",
    "        'St.54',\n",
    "        'St.57',\n",
    "        'Juv.',\n",
    "        'Adult',\n",
    "        'Meta.']\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1.8, 1.5)},font_scale=1)\n",
    "f,(ax,axcb) = plt.subplots(1,2, gridspec_kw={'width_ratios':[1,0.1]})\n",
    "g = sns.heatmap(D_Axo_dev_CP, xticklabels=labels_dev, yticklabels=labels_dev, cmap=\"viridis\", cbar_ax=axcb, ax=ax, vmax=D_Axo_dev_VLMC.max())\n",
    "g.set_title('CP', fontdict={'fontsize': 10, 'fontweight': 'bold'})\n",
    "axcb.set_ylabel('RMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)\n",
    "plt.savefig('figures/Fig5c_r.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig.5b Lineplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "count=[]\n",
    "file_1 = './data/Axolotls/development/Stage44/Stage44.CP.csv'\n",
    "file_2 = './data/Axolotls/development/Stage54/Stage54.CP.csv'\n",
    "file_3 = './data/Axolotls/development/Stage57/Stage57.CP.csv'\n",
    "file_4 = './data/Axolotls/development/Control_Juv/Control_Juv.CP.csv'\n",
    "file_5 = './data/Axolotls/development/Adult/Adult.CP.csv'\n",
    "file_6 = './data/Axolotls/development/Meta/Meta.CP.csv'\n",
    "\n",
    "# Read the file and calculate the row count\n",
    "for file in [file_5, file_4, file_6, file_1, file_2, file_3]:\n",
    "    df = pd.read_csv(file, sep='\\t')\n",
    "    count.append(df['MIDCounts'].count())\n",
    "\n",
    "# Converting Dictionary into DataFrame\n",
    "df = pd.DataFrame(list(GeneLists.items()), columns=['Stage', 'Genes'])\n",
    "df['Count'] = df['Genes'].str.len()\n",
    "df['Stage'] = ['Adult','Juv.','Meta.','St.44','St.54','St.57']\n",
    "df['Genes'] = GeneCounts.values()\n",
    "df['Transcripts'] = count\n",
    "\n",
    "df.set_index('Stage', inplace=True)\n",
    "df=df.reindex(['St.44','St.54','St.57','Juv.','Adult','Meta.'])\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "transcription_center_count = df['Count']\n",
    "transcript_count = df['Transcripts']\n",
    "\n",
    "correlation, p_value = pearsonr(transcription_center_count, transcript_count)\n",
    "\n",
    "# Draw Barplot\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(3, 2.2)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='Set2')\n",
    "ax=sns.lineplot(data=df, x='Stage', y='Transcripts', color='#023fa5')\n",
    "ax.set_ylabel('Total transcript count', color='#023fa5')\n",
    "plt.xticks(rotation=90, fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "\n",
    "ax2 = ax.twinx()\n",
    "sns.lineplot(data=df, x='Stage', y='Count', color='#8e063b', ax=ax2)\n",
    "ax2.set_ylabel('Transcription center count', color='#8e063b')\n",
    "ax.set_xlabel('')\n",
    "plt.title('CP')\n",
    "\n",
    "ax.set_ylim(0,2500000)\n",
    "ax2.set_ylim(0,10000)\n",
    "ax.text(0.05, 0.98, f'Pearson r={correlation:.4f}\\n$P={p_value:.4f}$', transform=ax.transAxes, fontsize=11, color='black', va='top')\n",
    "\n",
    "plt.savefig('figures/Fig5b_d.pdf', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn import manifold\n",
    "\n",
    "mds = manifold.MDS(n_components=2, dissimilarity=\"precomputed\", random_state=6)\n",
    "results = mds.fit(D_Axo_dev_CP)\n",
    "coords = results.embedding_\n",
    "mds_CP=pd.DataFrame(coords, columns=['x','y'], index=labels_dev)\n",
    "\n",
    "mds = manifold.MDS(n_components=2, dissimilarity=\"precomputed\", random_state=6)\n",
    "results = mds.fit(D_Axo_dev_VLMC)\n",
    "coords = results.embedding_\n",
    "mds_VLMC=pd.DataFrame(coords, columns=['x','y'], index=labels_dev)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mds_CP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mds_VLMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(2, 2)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "ax = sns.scatterplot(data=mds, x='x', y='y', s=5)\n",
    "\n",
    "for i in mds.index.values:\n",
    "    plt.text(mds.loc[i]['x'],mds.loc[i]['y'], i, fontdict=dict(color='black', alpha=0.5, size=6))\n",
    "\n",
    "ax.set_ylabel('y')\n",
    "ax.set_xlabel('x')\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(2, 2)}, font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "\n",
    "ax = sns.scatterplot(data=mds_VLMC, x='x', y='y', s=5)\n",
    "\n",
    "for i in mds_VLMC.index.values:\n",
    "    plt.text(mds_VLMC.loc[i]['x'],mds_VLMC.loc[i]['y'], i, fontdict=dict(color='black', alpha=0.5, size=6))\n",
    "\n",
    "ax.set_ylabel('y')\n",
    "ax.set_xlabel('x')\n",
    "ax.set_yticks([])\n",
    "ax.set_xticks([])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### netplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objs as go\n",
    "import itertools\n",
    "import networkx as nx\n",
    "random.seed(996)\n",
    "np.random.seed(996)\n",
    "\n",
    "# create an empty graph\n",
    "G = nx.Graph()\n",
    "\n",
    "# add nodes\n",
    "N = D_Axo_dev_CP.shape[0]\n",
    "for i in range(N):\n",
    "    G.add_node(i)\n",
    "\n",
    "# add edges\n",
    "edges = []\n",
    "\n",
    "# known edges:\n",
    "#for i,j in [(3,4),(4,5),(5,1),(1,0),(0,2)]:\n",
    "#    edges.append((i,j,{'weight':1/D_Axo_dev_CP[i,j]}))\n",
    "\n",
    "for i in range(N):\n",
    "    arr = D_Axo_dev_CP[i,:]\n",
    "    min_indices = np.argpartition(arr, 3)[:3]\n",
    "    min_values = arr[min_indices]\n",
    "    for j in min_indices:\n",
    "        if not i==j:\n",
    "            edges.append((i,j,{'weight':1/arr[j]}))\n",
    "\n",
    "#G.add_weighted_edges_from(edges)\n",
    "\n",
    "G.add_edges_from(edges)\n",
    "labels = nx.get_edge_attributes(G,'weight')\n",
    "final = dict()\n",
    "for key in labels:\n",
    "    final[key] = round(labels[key], 2)\n",
    "\n",
    "widths = nx.get_edge_attributes(G,'weight')\n",
    "widths.update((key, value * 6) for key, value in widths.items())\n",
    "nodelist = G.nodes()\n",
    "\n",
    "plt.figure(figsize=(25,25))\n",
    "\n",
    "#pos = nx.shell_layout(G)\n",
    "pos = nx.spring_layout(G, k=0.5, iterations=100)\n",
    "for n, p in pos.items():\n",
    "    G.nodes[n]['pos'] = p\n",
    "\n",
    "# edges\n",
    "# edge_trace setup\n",
    "edge_trace = []\n",
    "for edge in G.edges():\n",
    "    weight = G.edges[edge]['weight'] * 150 - 95\n",
    "    edge_trace.append(go.Scatter(\n",
    "       x=[],\n",
    "       y=[],\n",
    "       line=dict(width=weight, color='#888'),\n",
    "       hoverinfo='none',\n",
    "       mode='lines'))\n",
    "\n",
    "for trace, edge in zip(edge_trace, G.edges()):\n",
    "   x0, y0 = G.nodes[edge[0]]['pos']\n",
    "   x1, y1 = G.nodes[edge[1]]['pos']\n",
    "   trace['x'] += tuple([x0, x1, None])\n",
    "   trace['y'] += tuple([y0, y1, None])\n",
    "\n",
    "# nodes\n",
    "node_trace = go.Scatter(\n",
    "    x=[],\n",
    "    y=[],\n",
    "    text=[],\n",
    "    mode='markers+text',\n",
    "    hoverinfo='text',\n",
    "    marker=dict(\n",
    "        showscale=True,\n",
    "        colorscale='YlOrRd',\n",
    "        reversescale=False,\n",
    "        color=[],\n",
    "        size=50,\n",
    "        colorbar=dict(\n",
    "            thickness=7,\n",
    "            title='Node Connections',\n",
    "            xanchor='left',\n",
    "            titleside='right',\n",
    "            tickmode='array',\n",
    "            tickvals=[1, 2, 3, 4, 5],\n",
    "        ),\n",
    "        line=dict(width=0)))\n",
    "\n",
    "for node in G.nodes():\n",
    "    x, y = G.nodes[node]['pos']\n",
    "    node_trace['x'] += tuple([x])\n",
    "    node_trace['y'] += tuple([y])\n",
    "\n",
    "node_adjacencies = []\n",
    "for node, adjacencies in enumerate(G.adjacency()):\n",
    "    node_adjacencies.append(len(adjacencies[1]))\n",
    "    \n",
    "node_trace.marker.color = node_adjacencies +[4,5]+[1]\n",
    "node_trace.text = labels_dev\n",
    "\n",
    "fig = go.Figure(data = edge_trace+ [node_trace],\n",
    "            layout=go.Layout(\n",
    "                title=\"\",\n",
    "                titlefont=dict(size=16),\n",
    "                showlegend=False,\n",
    "                hovermode='closest',\n",
    "                margin=dict(b=0, l=0, r=0, t=0),\n",
    "                plot_bgcolor='#fff',\n",
    "                annotations=[dict(\n",
    "                    text=\"Development - CP\",\n",
    "                    showarrow=False,\n",
    "                    xref=\"paper\", yref=\"paper\",\n",
    "                    x=0.98, y=0.98, font=dict(size=16))],\n",
    "                xaxis=dict(showgrid=False, zeroline=False,\n",
    "                        showticklabels=False, mirror=True),\n",
    "                yaxis=dict(showgrid=False, zeroline=False,\n",
    "                showticklabels=False, mirror=True)))\n",
    "\n",
    "fig.update_layout(\n",
    "    autosize=False,\n",
    "    width=400,\n",
    "    height=300)\n",
    "\n",
    "fig.show()\n",
    "#save a figure of 300dpi, with 1.5 inches, and  height 0.75inches\n",
    "pio.write_image(fig, \"networkplot_axo_dev_CP.png\", width=1.3*300, height=0.8*300, scale=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_cmap('Blues',10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cytocraft.plot import *\n",
    "color_map_sample = {\n",
    "'St.44': '#b8d5ea',\n",
    "'St.54': '#97c6df',\n",
    "'St.57': '#4e9acb',\n",
    "'Juv.': '#3282be',\n",
    "'Adult': '#1a68ae',\n",
    "'Meta.': '#084f99',\n",
    "}\n",
    "plot_network(D_Axo_dev_CP, labels=labels_dev, cmap=color_map_sample, solver='forceAtlas2Based', corder=0, html='network_Axo_dev_CP_neighbor2.html', edge_scale=10, edge_adjust=0, N_neighbor=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VLMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def RMSD_distance_matrix_v1(\n",
    "    Confs, GeneLists, order=None, ngene=100, compare_method=\"pair\", norm_method=None\n",
    "):\n",
    "    # check if Confs and GeneLists have the same keys\n",
    "    if order==None:\n",
    "        keys = list(Confs.keys())\n",
    "    else:\n",
    "        keys = order\n",
    "    for i in keys:\n",
    "        if not i in GeneLists.keys():\n",
    "            print(f\"Warning: {i} in Confs but not in GeneLists\")\n",
    "            return\n",
    "\n",
    "    # calculate the distance matrix\n",
    "    N = len(keys)\n",
    "    DM = np.zeros((N, N))\n",
    "    if compare_method == \"complete\":\n",
    "        from functools import reduce\n",
    "\n",
    "        intersected_values = reduce(np.intersect1d, GeneLists.values())\n",
    "        intersected_values = intersected_values[:ngene]\n",
    "        if len(intersected_values) < ngene:\n",
    "            print(\n",
    "                f\"Warning: {len(intersected_values)} common genes between samples are less than {ngene}\"\n",
    "            )\n",
    "    from tqdm import tqdm\n",
    "\n",
    "    for n, key_n in enumerate(tqdm(keys)):\n",
    "        for m, key_m in enumerate(keys[: n + 1]):\n",
    "            if compare_method == \"pair\":\n",
    "                intersected_values = np.intersect1d(GeneLists[key_n], GeneLists[key_m])\n",
    "                intersected_values = intersected_values[:ngene]\n",
    "                if len(intersected_values) < ngene:\n",
    "                    print(\n",
    "                        f\"Warning: {len(intersected_values)} common genes between {key_n} and {key_m} are less than {ngene}\"\n",
    "                    )\n",
    "            boolean_arrays_n = np.in1d(GeneLists[key_n], intersected_values)\n",
    "            boolean_arrays_m = np.in1d(GeneLists[key_m], intersected_values)\n",
    "            Conf_n = Confs[key_n][boolean_arrays_n]\n",
    "            Conf_m = Confs[key_m][boolean_arrays_m]\n",
    "            if norm_method:\n",
    "                Conf_n = normalizeF(Conf_n, method=norm_method)\n",
    "                Conf_m = normalizeF(Conf_m, method=norm_method)\n",
    "            d1, _, _ = numpy_svd_rmsd_rot(Conf_n, Conf_m)\n",
    "            d2, _, _ = numpy_svd_rmsd_rot(mirror(Conf_n), Conf_m)\n",
    "            DM[n, m] = DM[m, n] = min(d1, d2)\n",
    "    return DM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from cytocraft.craft import *\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "adatas={}\n",
    "GeneLists={}\n",
    "GeneCounts={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='VLMC':\n",
    "                adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "                adatas[sample] = adata\n",
    "                GeneLists[sample] = adata.uns['X'].index.values\n",
    "                GeneCounts[sample] = adata.n_vars  \n",
    "\n",
    "keys = ['Stage44_VLMC',\n",
    "        'Stage54_VLMC',\n",
    "        'Stage57_VLMC',\n",
    "        'Control_Juv_VLMC',\n",
    "        'Adult_VLMC',\n",
    "        'Meta_VLMC']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig5c heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D_Axo_dev_VLMC.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D_Axo_dev_VLMC = RMSD_distance_matrix(adatas, order=keys, ngene=100)\n",
    "\n",
    "labels_dev = ['St.44',\n",
    "        'St.54',\n",
    "        'St.57',\n",
    "        'Juv.',\n",
    "        'Adult',\n",
    "        'Meta.']  \n",
    "   \n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1.8, 1.5)},font_scale=1)\n",
    "f,(ax,axcb) = plt.subplots(1,2, gridspec_kw={'width_ratios':[1,0.1]})\n",
    "g = sns.heatmap(D_Axo_dev_VLMC, xticklabels=labels_dev, yticklabels=labels_dev, cmap=\"viridis\", cbar_ax=axcb, ax=ax)\n",
    "g.set_title('VLMC', fontdict={'fontsize': 10, 'fontweight': 'bold'})\n",
    "axcb.set_ylabel('RMSD', rotation=90, va=\"bottom\", loc=\"center\", size=10, labelpad=10)\n",
    "plt.savefig('figures/Fig5c_l.pdf', bbox_inches='tight', dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fig.5b lineplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "count=[]\n",
    "file_1 = './data/Axolotls/development/Stage44/Stage44.VLMC.csv'\n",
    "file_2 = './data/Axolotls/development/Stage54/Stage54.VLMC.csv'\n",
    "file_3 = './data/Axolotls/development/Stage57/Stage57.VLMC.csv'\n",
    "file_4 = './data/Axolotls/development/Control_Juv/Control_Juv.VLMC.csv'\n",
    "file_5 = './data/Axolotls/development/Adult/Adult.VLMC.csv'\n",
    "file_6 = './data/Axolotls/development/Meta/Meta.VLMC.csv'\n",
    "\n",
    "# Read the file and calculate the row count\n",
    "for file in [file_5, file_4, file_6, file_1, file_2, file_3]:\n",
    "    df = pd.read_csv(file, sep='\\t')\n",
    "    count.append(df['MIDCounts'].count())\n",
    "\n",
    "# Converting Dictionary into DataFrame\n",
    "df = pd.DataFrame(list(GeneLists.items()), columns=['Stage', 'Genes'])\n",
    "df['Count'] = df['Genes'].str.len()\n",
    "df['Stage'] = ['Adult','Juv.','Meta.','St.44','St.54','St.57']\n",
    "df['Genes'] = GeneCounts.values()\n",
    "df['Transcripts'] = count\n",
    "\n",
    "df.set_index('Stage', inplace=True)\n",
    "df=df.reindex(['St.44','St.54','St.57','Juv.','Adult','Meta.'])\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "\n",
    "transcription_center_count = df['Count']\n",
    "transcript_count = df['Transcripts']\n",
    "\n",
    "correlation, p_value = pearsonr(transcription_center_count, transcript_count)\n",
    "\n",
    "# Draw Barplot\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(3, 2.2)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='Set2')\n",
    "ax=sns.lineplot(data=df, x='Stage', y='Transcripts', color='#023fa5')\n",
    "ax.set_ylabel('Total transcript count', color='#023fa5')\n",
    "plt.xticks(rotation=90, fontsize=12)\n",
    "plt.yticks(fontsize=12)\n",
    "\n",
    "ax2 = ax.twinx()\n",
    "sns.lineplot(data=df, x='Stage', y='Count', color='#8e063b', ax=ax2)\n",
    "ax2.set_ylabel('Transcription Center Count', color='#8e063b')\n",
    "ax.set_xlabel('')\n",
    "\n",
    "ax.set_ylim(0,2500000)\n",
    "ax2.set_ylim(0,10000)\n",
    "ax.text(0.05, 0.98, f'Pearson r={correlation:.4f}\\n$P={p_value:.4f}$', transform=ax.transAxes, fontsize=11, color='black', va='top')\n",
    "\n",
    "plt.title('VLMC')\n",
    "plt.savefig('figures/Fig5b_u.pdf', bbox_inches='tight')\n",
    "#ax.set_ylim(0,10000)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import spearmanr\n",
    "\n",
    "transcription_center_count = df['Count']\n",
    "transcript_count = df['Transcripts']\n",
    "\n",
    "correlation, p_value = spearmanr(transcription_center_count, transcript_count)\n",
    "\n",
    "correlation, p_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import pearsonr\n",
    "\n",
    "transcription_center_count = df['Count']\n",
    "transcript_count = df['Transcripts']\n",
    "\n",
    "correlation, p_value = pearsonr(transcription_center_count, transcript_count)\n",
    "\n",
    "correlation, p_value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### netplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cytocraft.plot import *\n",
    "color_map_sample = {\n",
    "'St.44': '#b8d5ea',\n",
    "'St.54': '#97c6df',\n",
    "'St.57': '#4e9acb',\n",
    "'Juv.': '#3282be',\n",
    "'Adult': '#1a68ae',\n",
    "'Meta.': '#084f99',\n",
    "}\n",
    "plot_network(D_Axo_dev_VLMC, labels=labels_dev, cmap=color_map_sample, solver='forceAtlas2Based', corder=0, html='network_Axo_dev_VLMC_neighbor2.html', edge_scale=10, edge_adjust=0, N_neighbor=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4.gene pairwise distance pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='CP': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(np.intersect1d, GeneLists.values())\n",
    "boolean_arrays = [np.in1d(a, intersected_genes) for a in GeneLists.values()]\n",
    "\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "D_pattern = pd.DataFrame()\n",
    "order=0\n",
    "for s, df in Xs.items():\n",
    "    print(s)\n",
    "    distances = pdist(df[boolean_arrays[order]].values, metric='euclidean')\n",
    "    df_dist = pd.DataFrame(squareform(distances), index=intersected_genes, columns=intersected_genes).stack()\n",
    "    df_dist = df_dist.reset_index()\n",
    "    df_dist.columns = ['gene1', 'gene2', 'distance']\n",
    "    inf = s.split('_')\n",
    "    df_dist['stage'] = inf[0]\n",
    "    df_dist['celltype'] = inf[1]\n",
    "    df_dist['genepair'] = df_dist['gene1'] + \",\" + df_dist['gene2']\n",
    "    D_pattern = D_pattern.append(df_dist)\n",
    "    order+=1\n",
    "\n",
    "D_pattern.reset_index(drop=True, inplace=True)\n",
    "matrix_pattern = D_pattern.pivot_table(index='genepair', columns='stage', values='distance')\n",
    "matrix_pattern.to_csv('Axo_dev_CP_pattern.mx.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "Mfuzz_tumor=pd.read_csv('Mfuzz_AXO_dev_CP.xls', sep='\\t')\n",
    "Mfuzz_tumor_sub=Mfuzz_tumor[Mfuzz_tumor['max'] >0.95]\n",
    "\n",
    "# unique gene list in each cluster\n",
    "unique_gs_cluster={}\n",
    "cluster_names = np.unique(Mfuzz_tumor_sub['Cluster'])\n",
    "for i in cluster_names:\n",
    "    Mfuzz_tumor_sub_cluster = Mfuzz_tumor_sub[Mfuzz_tumor_sub['Cluster'] == i]\n",
    "    gene_list = Mfuzz_tumor_sub_cluster['gene'].apply(lambda x: x.split(','))\n",
    "    genes = [gene for sublist in gene_list for gene in sublist]\n",
    "    unique_gs_cluster[i] = set(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate gene\n",
    "new_unique_gs_cluster={}\n",
    "anno = pd.read_csv(\"./Axo_Summary_Gene_Annotation_0325.txt\", sep='\\t')\n",
    "annodict = {row['Axolotl_ID']: row['hs_gene'] for _, row in anno.iterrows() if row['hs_gene'] != '-'}\n",
    "for cluster in cluster_names:\n",
    "    new_unique_gs_cluster[cluster] = {annodict.get(i, i) for i in unique_gs_cluster[cluster]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# enrichment\n",
    "import gseapy as gp\n",
    "enr = pd.DataFrame(columns=['Cluster','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "\n",
    "# Enricr API\n",
    "for key, value in new_unique_gs_cluster.items():\n",
    "    try:\n",
    "        for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "            filterlist = list(filter(lambda x: 'AMEX60DD' not in x, list(value)))\n",
    "            enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                                organism='Human',\n",
    "                                gene_sets=gs,\n",
    "                                outdir=None, \n",
    "                                cutoff=0.05,\n",
    "                                no_plot= True)\n",
    "\n",
    "            # trim (go:...)\n",
    "            enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "            enr_up.res2d['Cluster'] = key\n",
    "            enr = pd.concat([enr, enr_up.res2d[enr_up.res2d['Adjusted P-value']<0.05]], ignore_index=True)\n",
    "            # dotplot\n",
    "            # gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(value), size=10, cmap = plt.cm.viridis_r)\n",
    "    except ValueError or FileNotFoundError or TypeError as e:\n",
    "        print(value)\n",
    "        print(e)\n",
    "        continue\n",
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG_2021_Human\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "f,axs = plt.subplots(len(cluster_names),1)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 3)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.colormaps['viridis_r'] # create a color map object\n",
    "enr_gs = enr[enr[\"Gene_set\"]=='KEGG_2021_Human']\n",
    "\n",
    "for i,cluster in enumerate(cluster_names):\n",
    "    enr_top5_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "    log_adj_pvalue = -np.log10(enr_top5_gs_cluster[\"Adjusted P-value\"])\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=30)\n",
    "    colors = [cmap(norm(value)) for value in log_adj_pvalue]\n",
    "    g = sns.barplot(data=enr_top5_gs_cluster, x=\"Combined Score\", y=enr_top5_gs_cluster[\"Term\"], hue=enr_top5_gs_cluster[\"Term\"], fill=\"Adjusted P-value\", ax=axs[i], palette=colors, legend=False)\n",
    "    g.set_xlabel(\"\")\n",
    "    axs[i].yaxis.set_label_position(\"right\")\n",
    "    g.set_ylabel(cluster,rotation=270,labelpad=12)\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "g.set_xlabel(\"Combined Score\")\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='-log10(Adjusted P-value)') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-0.5, -0.4, 1.5, 0.5]) # xmin, ymin, dx, and dy\n",
    "plt.show()\n",
    "#f.savefig('Mfuzz_KEGG_Axo_dev_CP.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO_Molecular_Function_2023\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "f,axs = plt.subplots(len(cluster_names),1)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 3)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.colormaps['viridis_r'] # create a color map object\n",
    "enr_gs = enr[enr[\"Gene_set\"]=='GO_Molecular_Function_2023']\n",
    "\n",
    "for i,cluster in enumerate(cluster_names):\n",
    "    enr_top5_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "    log_adj_pvalue = -np.log10(enr_top5_gs_cluster[\"Adjusted P-value\"])\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=30)\n",
    "    colors = [cmap(norm(value)) for value in log_adj_pvalue]\n",
    "    g = sns.barplot(data=enr_top5_gs_cluster, x=\"Combined Score\", y=enr_top5_gs_cluster[\"Term\"], hue=enr_top5_gs_cluster[\"Term\"], fill=\"Adjusted P-value\", ax=axs[i], palette=colors, legend=False)\n",
    "    g.set_xlabel(\"\")\n",
    "    axs[i].yaxis.set_label_position(\"right\")\n",
    "    g.set_ylabel(cluster,rotation=270,labelpad=12)\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "g.set_xlabel(\"Combined Score\")\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='-log10(Adjusted P-value)') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-0.5, -0.4, 1.5, 0.5]) # xmin, ymin, dx, and dy\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO_Biological_Process_2023\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "f,axs = plt.subplots(len(cluster_names),1)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 3)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.colormaps['viridis_r'] # create a color map object\n",
    "enr_gs = enr[enr[\"Gene_set\"]=='GO_Biological_Process_2023']\n",
    "\n",
    "for i,cluster in enumerate(cluster_names):\n",
    "    enr_top5_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "    log_adj_pvalue = -np.log10(enr_top5_gs_cluster[\"Adjusted P-value\"])\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=30)\n",
    "    colors = [cmap(norm(value)) for value in log_adj_pvalue]\n",
    "    g = sns.barplot(data=enr_top5_gs_cluster, x=\"Combined Score\", y=enr_top5_gs_cluster[\"Term\"], hue=enr_top5_gs_cluster[\"Term\"], fill=\"Adjusted P-value\", ax=axs[i], palette=colors, legend=False)\n",
    "    g.set_xlabel(\"\")\n",
    "    axs[i].yaxis.set_label_position(\"right\")\n",
    "    g.set_ylabel(cluster,rotation=270,labelpad=12)\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "g.set_xlabel(\"Combined Score\")\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='-log10(Adjusted P-value)') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-0.5, -0.4, 1.5, 0.5]) # xmin, ymin, dx, and dy\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### VLMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "from functools import reduce\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "Xs={}\n",
    "GeneLists={}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+'_'+ctype\n",
    "        if 'adata.h5ad' in files and ctype=='VLMC': \n",
    "            print(sample)\n",
    "            adata = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            Xs[sample] = adata.uns['X']\n",
    "            GeneLists[sample] = adata.uns['X'].index\n",
    "\n",
    "intersected_genes = reduce(np.intersect1d, GeneLists.values())\n",
    "boolean_arrays = [np.in1d(a, intersected_genes) for a in GeneLists.values()]\n",
    "\n",
    "D_pattern = pd.DataFrame()\n",
    "order=0\n",
    "for s, df in Xs.items():\n",
    "    print(s)\n",
    "    distances = pdist(df[boolean_arrays[order]].values, metric='euclidean')\n",
    "    df_dist = pd.DataFrame(squareform(distances), index=intersected_genes, columns=intersected_genes).stack()\n",
    "    df_dist = df_dist.reset_index()\n",
    "    df_dist.columns = ['gene1', 'gene2', 'distance']\n",
    "    inf = s.split('_')\n",
    "    df_dist['stage'] = inf[0]\n",
    "    df_dist['celltype'] = inf[1]\n",
    "    df_dist['genepair'] = df_dist['gene1'] + \",\" + df_dist['gene2']\n",
    "    D_pattern = D_pattern.append(df_dist)\n",
    "    order+=1\n",
    "\n",
    "D_pattern.reset_index(drop=True, inplace=True)\n",
    "matrix_pattern = D_pattern.pivot_table(index='genepair', columns='stage', values='distance')\n",
    "matrix_pattern.to_csv('Axo_dev_VLMC_pattern.mx.csv')\n",
    "#### enrichment\n",
    "Mfuzz_VLMC=pd.read_csv('Mfuzz_membership_AXO_dev_VLMC.xls', sep='\\t')\n",
    ""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "Mfuzz_tumor=pd.read_csv('Mfuzz_AXO_dev_VLMC.xls', sep='\\t')\n",
    "Mfuzz_tumor_sub=Mfuzz_tumor[Mfuzz_tumor['max'] >0.95]\n",
    "\n",
    "# unique gene list in each cluster\n",
    "unique_gs_cluster={}\n",
    "cluster_names = np.unique(Mfuzz_tumor_sub['Cluster'])\n",
    "for i in cluster_names:\n",
    "    Mfuzz_tumor_sub_cluster = Mfuzz_tumor_sub[Mfuzz_tumor_sub['Cluster'] == i]\n",
    "    gene_list = Mfuzz_tumor_sub_cluster['gene'].apply(lambda x: x.split(','))\n",
    "    genes = [gene for sublist in gene_list for gene in sublist]\n",
    "    unique_gs_cluster[i] = set(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# translate gene\n",
    "new_unique_gs_cluster={}\n",
    "anno = pd.read_csv(\"./Axo_Summary_Gene_Annotation_0325.txt\", sep='\\t')\n",
    "annodict = {row['Axolotl_ID']: row['hs_gene'] for _, row in anno.iterrows() if row['hs_gene'] != '-'}\n",
    "for cluster in cluster_names:\n",
    "    new_unique_gs_cluster[cluster] = {annodict.get(i, i) for i in unique_gs_cluster[cluster]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# enrichment\n",
    "import gseapy as gp\n",
    "enr = pd.DataFrame(columns=['Cluster','Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\t'Old Adjusted P-value',\t'Odds Ratio', 'Combined Score', 'Genes'])\n",
    "\n",
    "# Enricr API\n",
    "for key, value in new_unique_gs_cluster.items():\n",
    "    try:\n",
    "        for gs in ['KEGG_2021_Human','GO_Molecular_Function_2023','GO_Biological_Process_2023']:\n",
    "            filterlist = list(filter(lambda x: 'AMEX60DD' not in x, list(value)))\n",
    "            enr_up = gp.enrichr(gene_list=filterlist,\n",
    "                                organism='Human',\n",
    "                                gene_sets=gs,\n",
    "                                outdir=None, \n",
    "                                cutoff=0.05,\n",
    "                                no_plot= True)\n",
    "\n",
    "            # trim (go:...)\n",
    "            enr_up.res2d.Term = enr_up.res2d.Term.str.split(\" \\(GO\").str[0]\n",
    "            enr_up.res2d['Cluster'] = key\n",
    "            enr = pd.concat([enr, enr_up.res2d[enr_up.res2d['Adjusted P-value']<0.05]], ignore_index=True)\n",
    "            # dotplot\n",
    "            # gp.dotplot(enr_up.res2d, title=\"Cluster \"+str(value), size=10, cmap = plt.cm.viridis_r)\n",
    "    except ValueError or FileNotFoundError or TypeError as e:\n",
    "        print(value)\n",
    "        print(e)\n",
    "        continue\n",
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# KEGG_2021_Human\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "f,axs = plt.subplots(len(cluster_names),1)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 3)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.colormaps['viridis_r'] # create a color map object\n",
    "enr_gs = enr[enr[\"Gene_set\"]=='KEGG_2021_Human']\n",
    "\n",
    "for i,cluster in enumerate(cluster_names):\n",
    "    enr_top5_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "    log_adj_pvalue = -np.log10(enr_top5_gs_cluster[\"Adjusted P-value\"])\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=30)\n",
    "    colors = [cmap(norm(value)) for value in log_adj_pvalue]\n",
    "    g = sns.barplot(data=enr_top5_gs_cluster, x=\"Combined Score\", y=enr_top5_gs_cluster[\"Term\"], hue=enr_top5_gs_cluster[\"Term\"], fill=\"Adjusted P-value\", ax=axs[i], palette=colors, legend=False)\n",
    "    g.set_xlabel(\"\")\n",
    "    axs[i].yaxis.set_label_position(\"right\")\n",
    "    g.set_ylabel(cluster,rotation=270,labelpad=12)\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "g.set_xlabel(\"Combined Score\")\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='-log10(Adjusted P-value)') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-0.5, -0.4, 1.5, 0.5]) # xmin, ymin, dx, and dy\n",
    "plt.show()\n",
    "#f.savefig('Mfuzz_KEGG_Axo_dev_CP.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO_Molecular_Function_2023\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "f,axs = plt.subplots(len(cluster_names),1)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 3)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.colormaps['viridis_r'] # create a color map object\n",
    "enr_gs = enr[enr[\"Gene_set\"]=='GO_Molecular_Function_2023']\n",
    "\n",
    "for i,cluster in enumerate(cluster_names):\n",
    "    enr_top5_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "    log_adj_pvalue = -np.log10(enr_top5_gs_cluster[\"Adjusted P-value\"])\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=30)\n",
    "    colors = [cmap(norm(value)) for value in log_adj_pvalue]\n",
    "    g = sns.barplot(data=enr_top5_gs_cluster, x=\"Combined Score\", y=enr_top5_gs_cluster[\"Term\"], hue=enr_top5_gs_cluster[\"Term\"], fill=\"Adjusted P-value\", ax=axs[i], palette=colors, legend=False)\n",
    "    g.set_xlabel(\"\")\n",
    "    axs[i].yaxis.set_label_position(\"right\")\n",
    "    g.set_ylabel(cluster,rotation=270,labelpad=12)\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "g.set_xlabel(\"Combined Score\")\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='-log10(Adjusted P-value)') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-0.5, -0.4, 1.5, 0.5]) # xmin, ymin, dx, and dy\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# GO_Biological_Process_2023\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "f,axs = plt.subplots(len(cluster_names),1)\n",
    "sns.set(rc={\"figure.dpi\":300,'figure.figsize':(1, 3)},font_scale=1)\n",
    "sns.set_theme(style='white', palette='pastel')\n",
    "plt.subplots_adjust(left=0, bottom=-.1, right=1, top=1.1)\n",
    "cmap = mpl.colormaps['viridis_r'] # create a color map object\n",
    "enr_gs = enr[enr[\"Gene_set\"]=='GO_Biological_Process_2023']\n",
    "\n",
    "for i,cluster in enumerate(cluster_names):\n",
    "    enr_top5_gs_cluster = enr_gs[enr_gs[\"Cluster\"]==cluster][:5]\n",
    "    log_adj_pvalue = -np.log10(enr_top5_gs_cluster[\"Adjusted P-value\"])\n",
    "    norm = mpl.colors.Normalize(vmin=0, vmax=30)\n",
    "    colors = [cmap(norm(value)) for value in log_adj_pvalue]\n",
    "    g = sns.barplot(data=enr_top5_gs_cluster, x=\"Combined Score\", y=enr_top5_gs_cluster[\"Term\"], hue=enr_top5_gs_cluster[\"Term\"], fill=\"Adjusted P-value\", ax=axs[i], palette=colors, legend=False)\n",
    "    g.set_xlabel(\"\")\n",
    "    axs[i].yaxis.set_label_position(\"right\")\n",
    "    g.set_ylabel(cluster,rotation=270,labelpad=12)\n",
    "    axs[i].tick_params(axis='x', which='major', pad=-4)\n",
    "\n",
    "g.set_xlabel(\"Combined Score\")\n",
    "mappable = mpl.cm.ScalarMappable(cmap=cmap, norm=norm) # create a scalar mappable object\n",
    "cbar = f.colorbar(mappable, ax=axs, orientation='horizontal',label='-log10(Adjusted P-value)') # add a colorbar to the figure\n",
    "cbar.ax.set_position([-0.5, -0.4, 1.5, 0.5]) # xmin, ymin, dx, and dy\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "Mfuzz_tumor=pd.read_csv('Mfuzz_AXO_dev_CP.xls', sep='\\t')\n",
    "\n",
    "# unique gene list in each cluster\n",
    "Mfuzz_genepairs_CP={}\n",
    "cluster_names = np.unique(Mfuzz_tumor['Pattern'])\n",
    "for i in cluster_names:\n",
    "    gp = set(Mfuzz_tumor[Mfuzz_tumor['Pattern'] == i].gene)\n",
    "    Mfuzz_genepairs_CP[i] = gp\n",
    "    print(len(gp))\n",
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "Mfuzz_tumor=pd.read_csv('Mfuzz_AXO_dev_VLMC.xls', sep='\\t')\n",
    "\n",
    "\n",
    "# unique gene list in each cluster\n",
    "Mfuzz_genepairs_VLMC={}\n",
    "cluster_names = np.unique(Mfuzz_tumor['Pattern'])\n",
    "for i in cluster_names:\n",
    "    gp = set(Mfuzz_tumor[Mfuzz_tumor['Pattern'] == i].gene)\n",
    "    Mfuzz_genepairs_VLMC[i] = gp\n",
    "    print(len(gp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chi-square test\n",
    "from scipy.stats import chi2_contingency\n",
    "from scipy.stats import chi2\n",
    "\n",
    "# Define the patterns to compare\n",
    "patterns_to_compare = [\n",
    "    ('Pattern 3', 'Pattern 1'),\n",
    "    ('Pattern 2', 'Pattern 2'),\n",
    "    ('Pattern 1', 'Pattern 3')\n",
    "]\n",
    "\n",
    "# Perform chi-square test for each pair of patterns\n",
    "for cp_pattern, vlmc_pattern in patterns_to_compare:\n",
    "    # Create a contingency table\n",
    "    overlap = len(Mfuzz_genepairs_CP[cp_pattern].intersection(Mfuzz_genepairs_VLMC[vlmc_pattern]))\n",
    "    only_cp = len(Mfuzz_genepairs_CP[cp_pattern]) - overlap\n",
    "    only_vlmc = len(Mfuzz_genepairs_VLMC[vlmc_pattern]) - overlap\n",
    "    neither = len(set.union(Mfuzz_genepairs_CP[cp_pattern], Mfuzz_genepairs_VLMC[vlmc_pattern])) - overlap - only_cp - only_vlmc\n",
    "\n",
    "    contingency_table = [[overlap, only_cp], [only_vlmc, neither]]\n",
    "\n",
    "    print(contingency_table)\n",
    "    # Perform chi-square test\n",
    "    chi2_stat, p_val, dof, ex = chi2_contingency(contingency_table)\n",
    "\n",
    "    print(f\"Chi-square test for CP {cp_pattern} and VLMC {vlmc_pattern}:\")\n",
    "    print(f\"Chi2 Stat: {chi2_stat}, P-value: {p_val}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import openchord as ocd\n",
    "\n",
    "# Find the overlaps\n",
    "overlap_11 = len(Mfuzz_genepairs_CP['Pattern 1'].intersection(Mfuzz_genepairs_VLMC['Pattern 1']))\n",
    "overlap_12 = len(Mfuzz_genepairs_CP['Pattern 1'].intersection(Mfuzz_genepairs_VLMC['Pattern 2']))\n",
    "overlap_13 = len(Mfuzz_genepairs_CP['Pattern 1'].intersection(Mfuzz_genepairs_VLMC['Pattern 3']))\n",
    "overlap_21 = len(Mfuzz_genepairs_CP['Pattern 2'].intersection(Mfuzz_genepairs_VLMC['Pattern 1']))\n",
    "overlap_22 = len(Mfuzz_genepairs_CP['Pattern 2'].intersection(Mfuzz_genepairs_VLMC['Pattern 2']))\n",
    "overlap_23 = len(Mfuzz_genepairs_CP['Pattern 2'].intersection(Mfuzz_genepairs_VLMC['Pattern 3']))\n",
    "overlap_31 = len(Mfuzz_genepairs_CP['Pattern 3'].intersection(Mfuzz_genepairs_VLMC['Pattern 1']))\n",
    "overlap_32 = len(Mfuzz_genepairs_CP['Pattern 3'].intersection(Mfuzz_genepairs_VLMC['Pattern 2']))\n",
    "overlap_33 = len(Mfuzz_genepairs_CP['Pattern 3'].intersection(Mfuzz_genepairs_VLMC['Pattern 3']))\n",
    "\n",
    "# The input should be a matrix showing the relation (overlap) between each gene list\n",
    "matrix = [\n",
    "  [0, 0, 0, overlap_11, overlap_12, overlap_13],\n",
    "  [0, 0, 0, overlap_21, overlap_22, overlap_23],\n",
    "  [0, 0, 0, overlap_31, overlap_32, overlap_33],\n",
    "  [overlap_11, overlap_21, overlap_31, 0, 0, 0],\n",
    "  [overlap_12, overlap_22, overlap_32, 0, 0, 0],\n",
    "  [overlap_13, overlap_23, overlap_33, 0, 0, 0]\n",
    "]\n",
    "print(matrix)\n",
    "names = ['CP Pattern 1', 'CP Pattern 2', 'CP Pattern 3', 'VLMC Pattern 1', 'VLMC Pattern 2', 'VLMC Pattern 3']\n",
    "fig = ocd.Chord(matrix, names)\n",
    "fig.colormap = [\"#FF6B6B\", \"#F9844A\", \"#F9C74F\", \"#90BE6D\", \"#43AA8B\", \"#4D908E\", \"#577590\", \"#277DA1\"]\n",
    "fig.radius = 80\n",
    "fig.padding = 80\n",
    "bg_transparancy=1\n",
    "\n",
    "fig.save_svg(\"figures/chord_patterns_vlmc&cp.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Corr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adatas={}\n",
    "celltypes = {}\n",
    "\n",
    "result_path=\"./results/Axolotls/\"\n",
    "samples = [ name for name in sorted(os.listdir(result_path)) if os.path.isdir(os.path.join(result_path, name)) ]\n",
    "for sample in samples:\n",
    "    if sample[0].isalpha():\n",
    "        result_dir = os.path.join(result_path,sample)\n",
    "        files = os.listdir(os.path.join(result_dir))\n",
    "        info = sample.split(\".\")\n",
    "        chip = info[0]\n",
    "        ctype = info[1].split('_')[0]\n",
    "        sample = chip+' '+ctype\n",
    "        if chip not in celltypes:\n",
    "            celltypes[chip] = []\n",
    "        celltypes[chip].append(ctype)\n",
    "        if 'adata.h5ad' in files and ('CP' in sample or 'VLMC' in sample):\n",
    "            adatas[sample] = sc.read_h5ad(os.path.join(result_dir+\"/adata.h5ad\"))\n",
    "            #GeneLists[sample] = adatas[sample].uns['X'].index\n",
    "            #Xs[sample] = adatas[sample].uns['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas[sample].uns['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf = pd.read_csv(\"./notebook_yf/ReST3D/main/src/cytocraft/gtf/AmexT_v47-AmexG_v6.0-DD.gene.gtf\", sep='\\t', comment='#', header=None)\n",
    "gtf = gtf.rename(columns=lambda x: 'col_' + str(x))\n",
    "# get position of genes\n",
    "gene_chr={}\n",
    "gene_length={}\n",
    "gene_length_cbrt={}\n",
    "for _, row in gtf.query(\"col_2 == 'gene'\").iterrows():\n",
    "    key = row['col_8'].split(';')[1].split()[1].strip('\"')\n",
    "    gene_chr[key] = row['col_0']\n",
    "    gene_length[key] = int(row['col_4'])-int(row['col_3'])\n",
    "    gene_length_cbrt[key] = np.cbrt(gene_length[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_gene_distance_matrix(F):\n",
    "    GeneList = list(F.index)\n",
    "    N = len(GeneList)\n",
    "    DM = np.zeros((N, N))\n",
    "    F_values = F.values  # Convert DataFrame to numpy array for faster access\n",
    "    for n in range(N):\n",
    "        for m in range(n + 1):\n",
    "            d = np.linalg.norm(F_values[n] - F_values[m])\n",
    "            DM[n, m] = DM[m, n] = d\n",
    "    return DM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "thresholds = [0.2, 0.3, 0.5, 1, 2]\n",
    "# Initialize an empty dataframe to store the results\n",
    "combined_df = pd.DataFrame(columns=['sample', 'celltype', 'gene', 'thresh', 'num_neighbors', 'length'])\n",
    "\n",
    "# Precompute the distance matrices for all cell types\n",
    "distance_matrices = {cell_type: gene_gene_distance_matrix(adatas[cell_type].uns['X']) for cell_type in tqdm(adatas.keys())}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    ""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "#thresholds = [0.2, 0.3, 0.5, 1, 2]\n",
    "# Initialize an empty dataframe to store the results\n",
    "#combined_df = pd.DataFrame(columns=['sample', 'celltype', 'gene', 'thresh', 'num_neighbors', 'length'])\n",
    "\n",
    "# Precompute the distance matrices for all cell types\n",
    "#distance_matrices = {cell_type: gene_gene_distance_matrix(adatas[cell_type].uns['X']) for cell_type in tqdm(adatas.keys())}\n",
    "\n",
    "# Iterate over each cell type in GeneLists\n",
    "for sample in tqdm(adatas.keys()):\n",
    "    #cell_type = sample+' '+ct\n",
    "    # Get the gene list and distance matrix for the current cell type\n",
    "    gene_list = adatas[sample].uns['X'].index\n",
    "    distance_matrix = distance_matrices[sample]\n",
    "    \n",
    "    # Filter the gene list based on the available gene lengths\n",
    "    filtered_gene_list = [gene for gene in gene_list if gene in gene_length_cbrt.keys()]\n",
    "    filtered_gene_indices = [gene_list.get_loc(gene) for gene in filtered_gene_list]\n",
    "    \n",
    "    # Calculate the number of neighbors for each gene using different thresholds\n",
    "    for threshold in thresholds:\n",
    "        num_neighbors = np.sum(distance_matrix[filtered_gene_indices][:, filtered_gene_indices] < threshold, axis=1) - 1\n",
    "        \n",
    "        # Create a dataframe with the current results\n",
    "        \n",
    "        df = pd.DataFrame({\n",
    "            'sample': sample,\n",
    "            #'celltype': ct,\n",
    "            'gene': filtered_gene_list,\n",
    "            'thresh': [threshold] * len(filtered_gene_list),\n",
    "            'num_neighbors': num_neighbors,\n",
    "            'length': [gene_length_cbrt[gene] for gene in filtered_gene_list]\n",
    "        })\n",
    "        \n",
    "        # Append the current dataframe to the combined dataframe\n",
    "        combined_df = pd.concat([combined_df, df], ignore_index=True)\n",
    "\n",
    "# Print the combined dataframe\n",
    "print(combined_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresholds = [0.2, 0.3, 0.5, 1, 2]\n",
    "#combined_df['celltype'] = combined_df['celltype'].replace({\n",
    "#    'Cardiomyocyte': 'Cardiomyocyte', 'Chondrocyte': 'Chondrocyte', 'Choroid_plexus': 'Choroid plexus', \n",
    "#    'Dorsal_midbrain_neuron': 'Dorsal Mb neuron', 'Endothelial_cell': 'Endothelial cell', \n",
    "#    'Epithelial_cell': 'Epithelial cell', 'Erythrocyte': 'Erythrocyte', 'Facial_fibroblast': 'Facial fibroblast', \n",
    "#    'Fibroblast': 'Fibroblast', 'Forebrain_neuron': 'Fb neuron', 'Forebrain_radial_glia_cell': 'Fb radial glia cell', \n",
    "#    'Ganglion': 'Ganglion', 'Hepatocyte': 'Hepatocyte', 'Immune_cell': 'Immune cell', 'Keratinocyte': 'Keratinocyte', \n",
    "#    'Limb_fibroblast': 'Limb fibroblast', 'Macrophage': 'Macrophage', 'Meninges_cell': 'Meninges cell', \n",
    "#    'Mid-hindbrain_and_spinal_cord_neuron': 'Mb/Hb/SpC neuron', 'Myoblast': 'Myoblast', \n",
    "#    'Olfactory_epithelial_cell': 'Olfactory epithelial cell', 'Radial_glia_cell': 'Radial glia cell', \n",
    "#    'Smooth_muscle_cell': 'Smooth muscle cell', 'Spinal_cord_neuron': 'SpC neuron', 'Thalamus_neuron': 'Thalamus neuron'\n",
    "#})\n",
    "#\n",
    "#palette = {\n",
    "#    'Limb fibroblast': '#809693', 'Keratinocyte': '#A30059', 'Endothelial cell': '#006FA6', \n",
    "#    'Facial fibroblast': '#63FFAC', 'Meninges cell': '#4FC601', 'Mb/Hb/SpC neuron': '#3B5DFF', \n",
    "#    'SpC neuron': '#FFAA92', 'Myoblast': '#FF2F80', 'Choroid plexus': '#FF34FF', 'Cardiomyocyte': '#8B0000', \n",
    "#    'Chondrocyte': '#1CE6FF', 'Dorsal Mb neuron': '#FF92CC', 'Fibroblast': '#FF4A46', 'Thalamus neuron': '#B79762', \n",
    "#    'Erythrocyte': '#FFE4E1', 'Smooth muscle cell': '#00C2A0', 'Macrophage': '#1B4400', 'Fb neuron': '#004D43', \n",
    "#    'Hepatocyte': '#997D87', 'Ganglion': '#800080', 'Olfactory epithelial cell': '#008941', 'Immune cell': '#5A0007', \n",
    "#    'Radial glia cell': '#6B7900', 'Epithelial cell': '#0000A6', 'Fb radial glia cell': '#8FB0FF'\n",
    "#}\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "# Calculate the correlation between gene length and number of proximal genes for each threshold\n",
    "correlations = []\n",
    "for threshold in thresholds:\n",
    "    subset = combined_df[combined_df['thresh'] == threshold]\n",
    "    correlation = subset.groupby('sample').apply(lambda x: spearmanr(x['length'], x['num_neighbors'].astype(float))[0])\n",
    "    correlations.append(correlation)\n",
    "\n",
    "# Create a DataFrame for plotting\n",
    "correlation_df = pd.concat(correlations, axis=1).T\n",
    "\n",
    "# Ensure columns have more than one level before dropping a level\n",
    "if correlation_df.columns.nlevels > 1:\n",
    "    correlation_df.columns = correlation_df.columns.droplevel(0)\n",
    "\n",
    "correlation_df['Threshold'] = np.array(thresholds) * 0.5\n",
    "\n",
    "# Melt the DataFrame for seaborn\n",
    "correlation_df_melted = correlation_df.melt(id_vars='Threshold', var_name='Celltype', value_name='Correlation')\n",
    "\n",
    "# Plot the relation of threshold and the correlation\n",
    "plt.figure(figsize=(4, 4))\n",
    "sns.set_theme(style='ticks')\n",
    "sns.lineplot(data=correlation_df_melted, x='Threshold', y='Correlation', hue='Celltype', marker='o', #palette=palette,\n",
    " legend=True)\n",
    "plt.axhline(0, color='black', linestyle='--')  # Highlight the y=0 axis\n",
    "plt.xlabel('Radius (\u03bcm)')\n",
    "plt.ylabel('Correlation')\n",
    "#plt.ylabel('Correlation between Cube Root of Gene Length\\nand Number of Proximal Genes')\n",
    "plt.title('')\n",
    "plt.grid(True)\n",
    "plt.show()\n",
    ""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "st",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}